// bitfab-plugin-lib/dist/flows/commandWrappers.js
import { defineCommandCatalog } from "bitfab-flow";
const PLUGIN_DIRS = {
    claude: "bitfab-claude-plugin",
    cursor: "bitfab-cursor-plugin",
    codex: "bitfab-codex-plugin",
};
const COMMAND_WRAPPERS = [
    {
        name: "clearStudioSession",
        sharedImports: ["runClearStudioSession"],
        body: "runClearStudioSession()",
    },
    {
        name: "closeStudio",
        sharedImports: ["runCloseStudio"],
        body: `runCloseStudio().catch((err) => {
  console.error("Failed to close Studio:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "detectCapabilities",
        sharedImports: ["runDetectCapabilities"],
        body: `runDetectCapabilities(process.argv[2]).catch((err) => {
  console.error("Capability detection failed:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "login",
        sharedImports: ["runLogin"],
        localImports: ["platform", "version"],
        body: `runLogin(platform, getVersion()).catch((err) => {
  console.error("Login failed:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "logout",
        sharedImports: ["runLogout"],
        body: "runLogout()",
    },
    {
        name: "openDatasetExperiments",
        sharedImports: ["runOpenDatasetExperiments"],
        localImports: ["platform"],
        body: `runOpenDatasetExperiments(platform).catch((err) => {
  console.error("Failed to open dataset experiments page:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "openExperiments",
        sharedImports: ["runOpenExperiments"],
        localImports: ["platform"],
        body: `runOpenExperiments(platform).catch((err) => {
  console.error("Failed to open experiments page:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "openStudioTo",
        sharedImports: ["runOpenStudioTo"],
        localImports: ["platform"],
        body: `runOpenStudioTo(platform).catch((err) => {
  console.error("Failed to open Studio:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "openTracePlan",
        sharedImports: ["runOpenTracePlan"],
        localImports: ["platform", "version"],
        body: `runOpenTracePlan(platform, getVersion()).catch((err) => {
  console.error("Failed to open trace plan:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "persistReplayLabels",
        sharedImports: ["runPersistReplayLabels"],
        localImports: ["platform", "version"],
        body: `runPersistReplayLabels(platform, getVersion()).catch((err) => {
  console.error("Persist replay labels failed:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "pushActivity",
        sharedImports: ["runPushActivity"],
        body: `runPushActivity().catch((err) => {
  console.error("Push activity failed:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "readReplayStatus",
        sharedImports: ["runReadReplayStatus"],
        localImports: ["platform"],
        body: `runReadReplayStatus(platform).catch((err) => {
  console.error("Read replay status failed:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "readTracesBatched",
        sharedImports: ["runReadTracesBatched"],
        localImports: ["platform", "version"],
        body: `runReadTracesBatched(platform, getVersion()).catch((err) => {
  console.error("Read traces (batched) failed:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "reapDaemon",
        sharedImports: ["runReapDaemon"],
        body: `runReapDaemon().catch((err) => {
  console.error("Failed to reap daemon:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "replayProgress",
        sharedImports: ["runReplayProgress"],
        body: `runReplayProgress().catch((err) => {
  console.error("Replay progress failed:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "sessionLogConsent",
        sharedImports: ["runSessionLogConsent"],
        body: `try {
  runSessionLogConsent(process.argv.slice(2))
} catch (err) {
  console.error(
    "Session log consent failed:",
    err instanceof Error ? err.message : String(err),
  )
  process.exit(1)
}`,
    },
    {
        name: "startDataset",
        sharedImports: ["runStartDataset"],
        localImports: ["platform", "version"],
        body: `runStartDataset(platform, getVersion()).catch((err) => {
  console.error("Failed to open dataset page:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "startTemplatePreview",
        sharedImports: ["runStartTemplatePreview"],
        localImports: ["platform", "version"],
        body: `runStartTemplatePreview(platform, getVersion()).catch((err) => {
  console.error("Failed to open template preview page:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "status",
        sharedImports: ["runStatus"],
        localImports: ["platform", "version", "pluginRoot"],
        body: `runStatus(getVersion(), platform, PLUGIN_ROOT, import.meta.url).catch((err) => {
  console.error("Status check failed:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "switchOrg",
        sharedImports: ["runSwitchOrg"],
        localImports: ["platform"],
        body: `runSwitchOrg(platform).catch((err) => {
  console.error("Org switch failed:", err.message)
  process.exit(1)
})`,
    },
    {
        name: "update",
        sharedImports: {
            claude: [
                "detectLegacyInstall",
                "legacyMigrationMessage",
                "parseUpdateMode",
                "runUpdate",
            ],
            default: ["parseUpdateMode", "runUpdate"],
        },
        localImports: ["platform", "version"],
        body: {
            claude: `async function main() {
  if (detectLegacyInstall(platform)) {
    console.log(legacyMigrationMessage(platform))
    return
  }

  const mode = parseUpdateMode(process.argv[2])
  await runUpdate(getVersion(), platform, mode)
}

main().catch((err) => {
  console.error("Update failed:", err.message)
  process.exit(1)
})`,
            default: `const mode = parseUpdateMode(process.argv[2])
runUpdate(getVersion(), platform, mode).catch((err) => {
  console.error("Update failed:", err.message)
  process.exit(1)
})`,
        },
    },
    {
        name: "waitForTrace",
        sharedImports: ["runWaitForTrace"],
        localImports: ["platform", "version"],
        body: `runWaitForTrace(platform, getVersion()).catch((err) => {
  console.error("Wait for trace failed:", err.message)
  process.exit(1)
})`,
    },
];
function commandFilesFromWrappers(wrappers) {
    const out = {};
    for (const wrapper of wrappers) {
        out[wrapper.name] = {
            file: `${wrapper.name}.js`,
        };
    }
    return out;
}
const COMMAND_FILES = commandFilesFromWrappers(COMMAND_WRAPPERS);
export const BITFAB_PLUGIN_COMMANDS = defineCommandCatalog(COMMAND_FILES);
export function listCommandCatalogNames() {
    return Object.keys(COMMAND_FILES);
}
export function listCommandWrapperNames() {
    return COMMAND_WRAPPERS.map((wrapper) => wrapper.name);
}
function resolveBody(wrapper, target) {
    if (typeof wrapper.body === "string") {
        return wrapper.body;
    }
    const body = wrapper.body[target] ?? wrapper.body.default;
    if (body === undefined) {
        throw new Error(`Command wrapper "${wrapper.name}" has no body for ${target}`);
    }
    return body;
}
function resolveSharedImports(wrapper, target) {
    if (isSharedImportList(wrapper.sharedImports)) {
        return wrapper.sharedImports;
    }
    const imports = wrapper.sharedImports[target] ?? wrapper.sharedImports.default;
    if (imports === undefined) {
        throw new Error(`Command wrapper "${wrapper.name}" has no shared imports for ${target}`);
    }
    return imports;
}
function isSharedImportList(value) {
    return Array.isArray(value);
}
function renderSharedImport(imports) {
    if (imports.length === 1) {
        return `import { ${imports[0]} } from "bitfab-plugin-lib"`;
    }
    if (imports.length > 3) {
        return [
            "import {",
            ...imports.map((importName) => `  ${importName},`),
            '} from "bitfab-plugin-lib"',
        ].join("\n");
    }
    return `import { ${imports.join(", ")} } from "bitfab-plugin-lib"`;
}
function renderLocalImports(imports) {
    return [...(imports ?? [])].sort().map((localImport) => {
        if (localImport === "platform") {
            return `import { platform } from "../platform.js"`;
        }
        if (localImport === "version") {
            return `import { getVersion } from "../version.js"`;
        }
        return `import { PLUGIN_ROOT } from "../pluginRoot.js"`;
    });
}
function renderCommandWrapper(wrapper, target) {
    const lines = [
        "// Generated by bitfab-plugin-lib/src/flows/commandWrappers.ts. Do not edit directly.",
        renderSharedImport(resolveSharedImports(wrapper, target)),
        ...renderLocalImports(wrapper.localImports),
        "",
        resolveBody(wrapper, target),
        "",
    ];
    return lines.join("\n");
}
export function renderAllCommandWrappers() {
    const out = [];
    for (const [target, pluginDir] of Object.entries(PLUGIN_DIRS)) {
        for (const wrapper of COMMAND_WRAPPERS) {
            out.push({
                target,
                outputPath: `${pluginDir}/src/commands/${wrapper.name}.ts`,
                content: renderCommandWrapper(wrapper, target),
            });
        }
    }
    return out;
}
