// bitfab-plugin-lib/dist/flows/frameworks.js
export const FRAMEWORKS = [
    {
        id: "langgraph",
        label: "LangGraph / LangChain",
        docsPath: "/frameworks/langgraph",
        handlerKind: "callbackHandler",
        handlerNoun: "callback handler",
        handler: {
            ts: "getLangGraphCallbackHandler",
            py: "get_langgraph_callback_handler",
        },
        aliasHandler: {
            ts: "getLangChainCallbackHandler",
            py: "get_langchain_callback_handler",
        },
        signals: {
            ts: "`@langchain/langgraph`, `@langchain/core`",
            py: "`langgraph`, `langchain`, `langchain_core`",
        },
    },
    {
        id: "openai-agents",
        label: "OpenAI Agents SDK",
        docsPath: "/frameworks/openai-agents",
        handlerKind: "traceProcessor",
        handlerNoun: "tracing processor",
        handler: {
            ts: "getOpenAiTracingProcessor",
            py: "get_openai_tracing_processor",
        },
        signals: {
            ts: "`@openai/agents`, `setTraceProcessors`",
            py: "`agents` (`from agents import ...`)",
        },
    },
    {
        id: "claude-agent-sdk",
        label: "Claude Agent SDK",
        docsPath: "/frameworks/claude-agent-sdk",
        handlerKind: "callbackHandler",
        handlerNoun: "handler",
        handler: {
            ts: "getClaudeAgentHandler",
            py: "get_claude_agent_handler",
        },
        signals: {
            ts: "`@anthropic-ai/claude-agent-sdk`, `query(`",
            py: "`claude_agent_sdk`, `ClaudeSDKClient`, `query(`",
        },
    },
    {
        id: "baml",
        label: "BAML",
        docsPath: "/frameworks/baml",
        handlerKind: "wrapper",
        handlerNoun: "wrapper",
        handler: { ts: "wrapBAML", py: "wrap_baml" },
        signals: {
            ts: "`@boundaryml/baml`, `baml_client` import",
            py: "`baml-py`, `from baml_client import b`",
        },
    },
];
const code = (s) => `\`${s}\``;
// Docs are served as plain markdown at the `.md` suffix; rendered URLs use it,
// while the suffix-free `docsPath` maps to the `.mdx` source file in tests.
const docsUrl = (docsPath) => `${docsPath}.md`;
/** Frameworks whose handler records a root span (everything but wrappers). */
const handlerRoots = FRAMEWORKS.filter((f) => f.handlerKind !== "wrapper");
/** Frameworks whose handler takes a trace-function key (replay by explicit key). */
const keyedHandlers = FRAMEWORKS.filter((f) => f.handlerKind === "callbackHandler");
/** Join with an Oxford "or": "a, b, or c". */
function joinOr(parts) {
    if (parts.length <= 1) {
        return parts.join("");
    }
    if (parts.length === 2) {
        return `${parts[0]} or ${parts[1]}`;
    }
    return `${parts.slice(0, -1).join(", ")}, or ${parts[parts.length - 1]}`;
}
/** Comma list of every framework docs path. */
export function frameworkDocsPaths() {
    return FRAMEWORKS.map((f) => code(docsUrl(f.docsPath))).join(", ");
}
/** The per-framework detection bullets for Instrument's detect-language step. */
export function frameworkDetectionBullets() {
    return FRAMEWORKS.map((f) => `- **${f.label}**: TS: ${f.signals.ts}; Python: ${f.signals.py}`).join("\n");
}
/** The framework -> docs-page + handler-name routing for read-sdk-reference. */
export function frameworkReferenceList() {
    return FRAMEWORKS.map((f) => {
        const methods = `${code(f.handler.ts)} / ${code(f.handler.py)}`;
        const alias = f.aliasHandler
            ? `; in a LangChain-only project, prefer the identical aliases ${code(f.aliasHandler.ts)} / ${code(f.aliasHandler.py)} so the code reads naturally`
            : "";
        return `${f.label} → ${code(docsUrl(f.docsPath))} (${methods}${alias})`;
    }).join("; ");
}
/** "<label> via `tsHandler` / `pyHandler`" for each root-handler framework. */
export function handlerFrameworksWithMethods() {
    return handlerRoots
        .map((f) => `${f.label} via ${code(f.handler.ts)} / ${code(f.handler.py)}`)
        .join(", ");
}
/** Oxford list of root-handler framework labels. */
export function handlerFrameworkNames() {
    return joinOr(handlerRoots.map((f) => f.label));
}
/** Keyed-handler call refs for replay prose: '`py("key")`, ..., or the TS equivalents'. */
export function keyedHandlerKeyRefs() {
    const refs = keyedHandlers.map((f) => code(`${f.handler.py}("key")`));
    return `${refs.join(", ")}, or the TS equivalents`;
}
/** "<label> `pyHandler`" for each keyed-handler framework. */
export function keyedHandlerExamples() {
    return keyedHandlers.map((f) => `${f.label} ${code(f.handler.py)}`).join(", ");
}
/** "<label> <noun>" list for each root-handler framework. */
function handlerNounList() {
    return handlerRoots.map((f) => `${f.label} ${f.handlerNoun}`).join(", ");
}
/** The SDK grep patterns used to find existing instrumented trace functions. */
export function buildSdkPatterns() {
    const base = '`getFunction("key")`, `get_function("key")`, `bitfab_function "key"`, `WithFunctionName("key")`';
    const keyed = keyedHandlers
        .map((f) => {
        const pair = `${code(`${f.handler.ts}("key")`)} / ${code(`${f.handler.py}("key")`)}`;
        return f.aliasHandler ? `${pair} (or the LangChain-named aliases)` : pair;
    })
        .join(" and ");
    return `${base}, plus keyed framework handlers: ${keyed}`;
}
/**
 * The shared "handler-instrumented workflows" replay guidance. Identical across
 * Instrument's write-instrumentation step (both the inline and codex branches)
 * apart from the verb that introduces the briefing list, so it is parameterized
 * on that verb and surfaced as a Targetable variable.
 */
function handlerReplayBlock(verb) {
    return `- **Handler-instrumented workflows (no decorated root)**: when this cycle's instrumentation is a framework handler (${handlerNounList()}) rather than a decorated root function, replace the "Trace function root" item with key-based replay: the replay pipeline passes the handler's key plus a plain callable to \`replay()\` (Python: \`client.replay("<key>", fn, ...)\`; TypeScript: \`bitfab.replay("<key>", fn, opts)\`), and the callable re-invokes the framework entrypoint with the recorded root input. The SDK wraps the callable internally; on SDKs that predate explicit-key replay, wrap it under the same key yourself (Python \`@bitfab.span("<key>")\`, TS \`getFunction(key).withSpan(...)\`). ${verb} the framework entrypoint + import path (e.g. the compiled graph's \`invoke\`/\`ainvoke\`, the agent run call), the recorded root-input shape (a dict root input like a LangGraph state arrives as a single positional argument on the explicit-key path; on the older same-key-wrapper path it splats into kwargs, so legacy Python wrappers take \`(**state)\`), and the environment the wrapper must construct fresh (framework config, dependency objects), using **safe no-op substitutes for side-effectful wiring** (billing/credit callbacks, notification senders) so replay never charges or notifies anyone. The handler-recorded production traces and the wrapper share the key, which is all \`replay()\` needs; never report a handler-instrumented key as not replayable.`;
}
/**
 * Flow variables derived from the framework registry. Spread into a flow's
 * `variables` map so the registry single-sources every per-framework string.
 */
export const FRAMEWORK_VARIABLES = {
    sdkPatterns: buildSdkPatterns(),
    frameworkDocsPaths: frameworkDocsPaths(),
    frameworkDetectionBullets: frameworkDetectionBullets(),
    frameworkReferenceList: frameworkReferenceList(),
    handlerFrameworksWithMethods: handlerFrameworksWithMethods(),
    handlerFrameworkNames: handlerFrameworkNames(),
    keyedHandlerKeyRefs: keyedHandlerKeyRefs(),
    keyedHandlerExamples: keyedHandlerExamples(),
    handlerReplayBlock: {
        default: handlerReplayBlock("Brief the subagent on:"),
        codex: handlerReplayBlock("Work out:"),
    },
};
