// bitfab-plugin-lib/dist/flows/frameworks.js
export const FRAMEWORKS = [
    {
        id: "langgraph",
        label: "LangGraph / LangChain",
        docsPath: "/frameworks/langgraph",
        handlerKind: "callbackHandler",
        handlerNoun: "callback handler",
        handler: {
            ts: "getLangGraphCallbackHandler",
            py: "get_langgraph_callback_handler",
        },
        aliasHandler: {
            ts: "getLangChainCallbackHandler",
            py: "get_langchain_callback_handler",
        },
        signals: {
            ts: "`@langchain/langgraph`, `@langchain/core`",
            py: "`langgraph`, `langchain`, `langchain_core`",
        },
    },
    {
        id: "openai-agents",
        label: "OpenAI Agents SDK",
        docsPath: "/frameworks/openai-agents",
        handlerKind: "traceProcessor",
        handlerNoun: "tracing processor",
        handler: {
            ts: "getOpenAiTracingProcessor",
            py: "get_openai_tracing_processor",
        },
        rootWrapper: {
            ts: "getOpenAiAgentHandler",
            py: "get_openai_agent_handler",
        },
        signals: {
            ts: "`@openai/agents`, `setTraceProcessors`",
            py: "`agents` (`from agents import ...`)",
        },
    },
    {
        id: "claude-agent-sdk",
        label: "Claude Agent SDK",
        docsPath: "/frameworks/claude-agent-sdk",
        handlerKind: "callbackHandler",
        handlerNoun: "handler",
        handler: {
            ts: "getClaudeAgentHandler",
            py: "get_claude_agent_handler",
        },
        signals: {
            ts: "`@anthropic-ai/claude-agent-sdk`, `query(`",
            py: "`claude_agent_sdk`, `ClaudeSDKClient`, `query(`",
        },
    },
    {
        id: "baml",
        label: "BAML",
        docsPath: "/frameworks/baml",
        handlerKind: "wrapper",
        handlerNoun: "wrapper",
        handler: { ts: "wrapBAML", py: "wrap_baml" },
        signals: {
            ts: "`@boundaryml/baml`, `baml_client` import",
            py: "`baml-py`, `from baml_client import b`",
        },
    },
    {
        // TypeScript only: the Vercel AI SDK (`ai`) has no Python equivalent, so
        // there is no `py` handler or signal. The integration is a language-model
        // middleware (`wrapLanguageModel`) that records each model call as a keyed
        // `llm` span carrying the prompt as input, so it replays by key like a
        // callback handler.
        id: "vercel-ai-sdk",
        label: "Vercel AI SDK",
        docsPath: "/frameworks/vercel-ai-sdk",
        handlerKind: "callbackHandler",
        handlerNoun: "middleware",
        handler: { ts: "getVercelAiMiddleware" },
        signals: {
            ts: "`ai`, `wrapLanguageModel`, `streamText`, `generateText`",
        },
    },
];
const code = (s) => `\`${s}\``;
// Docs are served as plain markdown at the `.md` suffix; rendered URLs use it,
// while the suffix-free `docsPath` maps to the `.mdx` source file in tests.
const docsUrl = (docsPath) => `${docsPath}.md`;
/** Frameworks whose Bitfab integration records a root span (everything but wrappers). */
const handlerRoots = FRAMEWORKS.filter((f) => f.handlerKind !== "wrapper");
/**
 * Frameworks that replay by explicit key with no hand-written `withSpan`/`@span`
 * root, paired with the keyed method that records that root: callback handlers
 * via their `handler`, and trace processors via their `rootWrapper` (e.g. OpenAI
 * Agents `getOpenAiAgentHandler(...).wrapRun`). A bare trace processor with no
 * `rootWrapper` is excluded - it records an empty-input root and is not
 * replayable on its own.
 */
const keyedReplayHandlers = FRAMEWORKS.flatMap((f) => {
    if (f.rootWrapper) {
        return [{ framework: f, method: f.rootWrapper, noun: "run wrapper" }];
    }
    if (f.handlerKind === "callbackHandler") {
        return [{ framework: f, method: f.handler, noun: f.handlerNoun }];
    }
    return [];
});
/** Join with an Oxford "or": "a, b, or c". */
function joinOr(parts) {
    if (parts.length <= 1) {
        return parts.join("");
    }
    if (parts.length === 2) {
        return `${parts[0]} or ${parts[1]}`;
    }
    return `${parts.slice(0, -1).join(", ")}, or ${parts[parts.length - 1]}`;
}
/** Comma list of every framework docs path. */
export function frameworkDocsPaths() {
    return FRAMEWORKS.map((f) => code(docsUrl(f.docsPath))).join(", ");
}
/** The per-framework detection bullets for Instrument's detect-language step. */
export function frameworkDetectionBullets() {
    return FRAMEWORKS.map((f) => {
        const py = f.signals.py ? `; Python: ${f.signals.py}` : " (TypeScript only)";
        return `- **${f.label}**: TS: ${f.signals.ts}${py}`;
    }).join("\n");
}
/** The framework -> docs-page + handler-name routing for read-sdk-reference. */
export function frameworkReferenceList() {
    return FRAMEWORKS.map((f) => {
        const methods = f.handler.py
            ? `${code(f.handler.ts)} / ${code(f.handler.py)}`
            : code(f.handler.ts);
        const alias = f.aliasHandler
            ? `; in a LangChain-only project, prefer the identical aliases ${code(f.aliasHandler.ts)} / ${code(f.aliasHandler.py ?? f.aliasHandler.ts)} so the code reads naturally`
            : "";
        // Trace processors expose a separate keyed run wrapper that records the
        // replayable root; name it here so write-instrumentation calls it instead
        // of leaving a bare, non-replayable processor root.
        const wrapper = f.rootWrapper
            ? `, plus the replayable run wrapper ${code(f.rootWrapper.ts)} / ${code(f.rootWrapper.py ?? f.rootWrapper.ts)} (drop-in for the run call)`
            : "";
        return `${f.label} → ${code(docsUrl(f.docsPath))} (${methods}${alias}${wrapper})`;
    }).join("; ");
}
/** "<label> via `tsHandler` / `pyHandler`" for each root-handler framework. */
export function handlerFrameworksWithMethods() {
    return handlerRoots
        .map((f) => {
        const methods = f.handler.py
            ? `${code(f.handler.ts)} / ${code(f.handler.py)}`
            : code(f.handler.ts);
        return `${f.label} via ${methods}`;
    })
        .join(", ");
}
/** Oxford list of root-handler framework labels. */
export function handlerFrameworkNames() {
    return joinOr(handlerRoots.map((f) => f.label));
}
/** Keyed-handler call refs for replay prose: '`py("key")`, ..., or the TS equivalents'. */
export function keyedHandlerKeyRefs() {
    // Prefer the Python form (the prose says "or the TS equivalents"); fall back
    // to the TS method for TypeScript-only frameworks that have no Python form.
    const refs = keyedReplayHandlers.map((k) => code(`${k.method.py ?? k.method.ts}("key")`));
    return `${refs.join(", ")}, or the TS equivalents`;
}
/** "<label> `handler`" for each keyed-replay framework (Python form, else TS). */
export function keyedHandlerExamples() {
    return keyedReplayHandlers
        .map((k) => `${k.framework.label} ${code(k.method.py ?? k.method.ts)}`)
        .join(", ");
}
/**
 * Oxford list of the framework integrations that record a replayable root with
 * no hand-written `withSpan`/`@span`: callback handlers (which record the
 * framework's own input as the root) and trace processors paired with their
 * root wrapper (OpenAI Agents `wrapRun`, which records the run input). These are
 * replayable as-is by key. A bare trace processor with no root wrapper is the
 * only root integration that is NOT replayable on its own (empty-input root).
 */
export function replayableRootHandlerNames() {
    return joinOr(keyedReplayHandlers.map((k) => k.framework.label));
}
/**
 * Oxford list of the callback-handler frameworks only (the subset of replayable
 * roots whose own handler records the root, as opposed to trace processors that
 * need a run wrapper). Used to contrast the two replayable mechanisms in prose.
 */
export function callbackHandlerNames() {
    return joinOr(FRAMEWORKS.filter((f) => f.handlerKind === "callbackHandler").map((f) => f.label));
}
/**
 * Oxford list of trace-processor framework labels. The processor captures a
 * run's internals but records an empty-input root, so it is replayable only when
 * paired with its root wrapper (`wrapRun` / `wrap_run`) or a hand-written
 * `withSpan`/`@span` root; a bare processor-only trace is not replayable.
 */
export function traceProcessorNames() {
    return joinOr(FRAMEWORKS.filter((f) => f.handlerKind === "traceProcessor").map((f) => f.label));
}
/** Comma list of trace-processor root-wrapper methods: '`tsWrap` / `pyWrap`'. */
export function traceProcessorRootWrappers() {
    return FRAMEWORKS.filter((f) => f.handlerKind === "traceProcessor" && f.rootWrapper)
        .map((f) => {
        const w = f.rootWrapper;
        return w.py ? `${code(w.ts)} / ${code(w.py)}` : code(w.ts);
    })
        .join(", ");
}
/**
 * "<label> <noun>" list for each keyed-replay framework: callback handlers by
 * their handler noun, trace processors by their run wrapper. These are the
 * integrations with no hand-written root that replay by key.
 */
function handlerNounList() {
    return keyedReplayHandlers
        .map((k) => `${k.framework.label} ${k.noun}`)
        .join(", ");
}
/** The SDK grep patterns used to find existing instrumented trace functions. */
export function buildSdkPatterns() {
    const base = '`getFunction("key")`, `get_function("key")`, `bitfab_function "key"`, `WithFunctionName("key")`';
    const keyed = keyedReplayHandlers
        .map(({ framework, method }) => {
        const pair = method.py
            ? `${code(`${method.ts}("key")`)} / ${code(`${method.py}("key")`)}`
            : code(`${method.ts}("key")`);
        return framework.aliasHandler
            ? `${pair} (or the LangChain-named aliases)`
            : pair;
    })
        .join(" and ");
    const processors = FRAMEWORKS.filter((f) => f.handlerKind === "traceProcessor")
        .map((f) => `${code(`${f.handler.ts}()`)} / ${code(`${f.handler.py}()`)}`)
        .join(" and ");
    return `${base}, plus keyed framework handlers: ${keyed}; plus trace-processor registrations (unkeyed in code, the key is derived server-side from the workflow name): ${processors}`;
}
/**
 * The shared "handler-instrumented workflows" replay guidance. Identical across
 * Instrument's write-instrumentation step (both the inline and codex branches)
 * apart from the verb that introduces the briefing list, so it is parameterized
 * on that verb and surfaced as a Targetable variable.
 */
function handlerReplayBlock(verb) {
    return `- **Handler-instrumented workflows (no decorated root)**: when this cycle's instrumentation is a framework handler (${handlerNounList()}) rather than a decorated root function, replace the "Trace function root" item with key-based replay: the replay pipeline passes the handler's key plus a plain callable to \`replay()\` (Python: \`client.replay("<key>", fn, ...)\`; TypeScript: \`bitfab.replay("<key>", fn, opts)\`), and the callable re-invokes the framework entrypoint with the recorded root input. The SDK wraps the callable internally; on SDKs that predate explicit-key replay, wrap it under the same key yourself (Python \`@bitfab.span("<key>")\`, TS \`getFunction(key).withSpan(...)\`). ${verb} the framework entrypoint + import path (e.g. the compiled graph's \`invoke\`/\`ainvoke\`, the agent run call), the recorded root-input shape (a dict root input like a LangGraph state arrives as a single positional argument on the explicit-key path; on the older same-key-wrapper path it splats into kwargs, so legacy Python wrappers take \`(**state)\`), and the environment the wrapper must construct fresh (framework config, dependency objects), using **safe no-op substitutes for side-effectful wiring** (billing/credit callbacks, notification senders) so replay never charges or notifies anyone. The handler-recorded production traces and the wrapper share the key, which is all \`replay()\` needs; never report a handler-instrumented key as not replayable.`;
}
/**
 * Flow variables derived from the framework registry. Spread into a flow's
 * `variables` map so the registry single-sources every per-framework string.
 */
export const FRAMEWORK_VARIABLES = {
    sdkPatterns: buildSdkPatterns(),
    frameworkDocsPaths: frameworkDocsPaths(),
    frameworkDetectionBullets: frameworkDetectionBullets(),
    frameworkReferenceList: frameworkReferenceList(),
    handlerFrameworksWithMethods: handlerFrameworksWithMethods(),
    handlerFrameworkNames: handlerFrameworkNames(),
    replayableRootHandlerNames: replayableRootHandlerNames(),
    callbackHandlerNames: callbackHandlerNames(),
    traceProcessorNames: traceProcessorNames(),
    traceProcessorRootWrappers: traceProcessorRootWrappers(),
    keyedHandlerKeyRefs: keyedHandlerKeyRefs(),
    keyedHandlerExamples: keyedHandlerExamples(),
    handlerReplayBlock: {
        default: handlerReplayBlock("Brief the subagent on:"),
        codex: handlerReplayBlock("Work out:"),
    },
};
