// bitfab-plugin-lib/dist/flows/index.js
import { renderMarkdown, TARGETS, } from "bitfab-flow";
import { assistantFlow } from "./assistant.js";
import { setupFlow } from "./setup.js";
import { updateFlow } from "./update.js";
export { applyMustacheSections, computeModesPerStep, DEV_TARGETS, describeFlow, Flow, nextRefsForMode, renderMarkdown, renderMermaid, SHARED_VARIABLES, TARGETS, } from "bitfab-flow";
export { assistantFlow } from "./assistant.js";
export { setupFlow } from "./setup.js";
export { updateFlow } from "./update.js";
/**
 * Every flow that should be rendered into per-target plugin skill files at
 * build time. Add new flows here as they are migrated from hand-written
 * markdown to the schema-driven representation.
 */
export const ALL_FLOWS = {
    update: updateFlow,
    setup: setupFlow,
    assistant: assistantFlow,
};
/**
 * Output paths (relative to the monorepo root) for each (flow, target) pair.
 * The build-skills script writes here.
 *
 * If a flow shouldn't be rendered for a target, omit it from the inner map.
 */
export const FLOW_OUTPUT_PATHS = {
    update: {
        claude: "bitfab-claude-plugin/commands/update.md",
        cursor: "bitfab-cursor-plugin/skills/bitfab-update/SKILL.md",
        codex: "bitfab-codex-plugin/skills/update/SKILL.md",
    },
    setup: {
        claude: "bitfab-claude-plugin/commands/setup.md",
        cursor: "bitfab-cursor-plugin/skills/bitfab-setup/SKILL.md",
        codex: "bitfab-codex-plugin/skills/setup/SKILL.md",
    },
    assistant: {
        claude: "bitfab-claude-plugin/commands/assistant.md",
        cursor: "bitfab-cursor-plugin/skills/bitfab-assistant/SKILL.md",
        codex: "bitfab-codex-plugin/skills/assistant/SKILL.md",
    },
};
/**
 * Produce every (flow, target) rendering as in-memory strings. Used by both
 * the write script and the check-only script so they share rendering logic.
 */
export function renderAllSkills() {
    const out = [];
    for (const [flowId, flow] of Object.entries(ALL_FLOWS)) {
        const paths = FLOW_OUTPUT_PATHS[flowId] ?? {};
        for (const [targetName, outputPath] of Object.entries(paths)) {
            const target = TARGETS[targetName];
            out.push({
                flowId,
                target: targetName,
                outputPath,
                content: renderMarkdown(flow, target),
            });
        }
    }
    return out;
}
