// bitfab-plugin-lib/dist/flows/index.js
import { renderFlow, TARGETS } from "bitfab-flow";
import { assistantFlow } from "./assistant.js";
import { setupFlow } from "./setup.js";
import { updateFlow } from "./update.js";
export { applyMustacheSections, computeModesPerStep, DEV_TARGETS, describeFlow, Flow, nextRefsForMode, renderFlow, renderMarkdown, renderMermaid, SHARED_VARIABLES, TARGETS, } from "bitfab-flow";
export { assistantFlow } from "./assistant.js";
export { renderAllCommandWrappers } from "./commandWrappers.js";
export { setupFlow } from "./setup.js";
export { updateFlow } from "./update.js";
/**
 * Every flow that should be rendered into per-target plugin skill files at
 * build time. Add new flows here as they are migrated from hand-written
 * markdown to the schema-driven representation.
 */
export const ALL_FLOWS = {
    update: updateFlow,
    setup: setupFlow,
    assistant: assistantFlow,
};
/**
 * Output paths (relative to the monorepo root) for each (flow, target) pair.
 * The build-skills script writes here.
 *
 * If a flow shouldn't be rendered for a target, omit it from the inner map.
 */
export const FLOW_OUTPUT_PATHS = {
    update: {
        // Claude: a skill orchestrator that dispatches to per-phase Agent Skills.
        // (Main migrated Claude flows from commands/ to skills/; the orchestrator
        // follows that convention.) Cursor and Codex stay monolith for now (gated
        // by per-target compile.mode).
        claude: {
            orchestrator: "bitfab-claude-plugin/skills/update/SKILL.md",
            phase: "bitfab-claude-plugin/skills/update-{phase}/SKILL.md",
        },
        cursor: "bitfab-cursor-plugin/skills/bitfab-update/SKILL.md",
        codex: "bitfab-codex-plugin/skills/update/SKILL.md",
    },
    setup: {
        // Claude: a skill orchestrator that dispatches to per-phase Agent Skills.
        // Cursor and Codex stay monolith (gated by per-target compile.mode).
        claude: {
            orchestrator: "bitfab-claude-plugin/skills/setup/SKILL.md",
            phase: "bitfab-claude-plugin/skills/setup-{phase}/SKILL.md",
        },
        cursor: "bitfab-cursor-plugin/skills/bitfab-setup/SKILL.md",
        codex: "bitfab-codex-plugin/skills/setup/SKILL.md",
    },
    assistant: {
        // Claude: a skill orchestrator that dispatches to per-phase Agent Skills.
        // Cursor and Codex stay monolith (gated by per-target compile.mode).
        claude: {
            orchestrator: "bitfab-claude-plugin/skills/assistant/SKILL.md",
            phase: "bitfab-claude-plugin/skills/assistant-{phase}/SKILL.md",
        },
        cursor: "bitfab-cursor-plugin/skills/bitfab-assistant/SKILL.md",
        codex: "bitfab-codex-plugin/skills/assistant/SKILL.md",
    },
};
/**
 * Resolve a rendered file's `slug` against a target's output config to its
 * final path. Monolith and the split orchestrator use slug `""`; sub-skills
 * use their phase id, substituted into the split config's `phase` template.
 */
function resolveOutputPath(config, slug) {
    if (typeof config === "string") {
        if (slug === "") {
            return config;
        }
        throw new Error(`Sub-skill slug "${slug}" emitted for a monolith output config "${config}". Give this flow a split output config.`);
    }
    if (slug === "") {
        return config.orchestrator;
    }
    return config.phase.replace(/\{phase\}/g, slug);
}
/**
 * Produce every (flow, target) rendering as in-memory strings. Used by both
 * the write script and the check-only script so they share rendering logic.
 * A split flow expands to multiple `RenderedSkill`s (orchestrator + sub-skills).
 */
export function renderAllSkills() {
    const out = [];
    for (const [flowId, flow] of Object.entries(ALL_FLOWS)) {
        const paths = FLOW_OUTPUT_PATHS[flowId] ?? {};
        for (const [targetName, outputConfig] of Object.entries(paths)) {
            const target = TARGETS[targetName];
            for (const file of renderFlow(flow, target)) {
                out.push({
                    flowId,
                    target: targetName,
                    outputPath: resolveOutputPath(outputConfig, file.slug),
                    content: file.content,
                });
            }
        }
    }
    return out;
}
