// bitfab-plugin-lib/dist/flows/setup.test.js
import { computeModesPerStep, computeReachableSteps, nextRefsForMode, renderFlow, renderMarkdown, TARGETS, } from "bitfab-flow";
import { describe, expect, it } from "vitest";
import { setupFlow } from "./setup.js";
const TARGET_NAMES = ["claude", "cursor", "codex"];
function findStep(phaseId, stepId) {
    const phase = setupFlow.phases.find((p) => p.id === phaseId);
    if (phase === undefined) {
        throw new Error(`phase "${phaseId}" not found`);
    }
    const step = phase.steps.find((s) => s.id === stepId);
    if (step === undefined) {
        throw new Error(`step "${phaseId}/${stepId}" not found`);
    }
    return step;
}
describe("setupFlow schema invariants", () => {
    it("declares all twelve invocation modes with `wizard` as defaultMode", () => {
        expect(Object.keys(setupFlow.entries).sort()).toEqual([
            "analyze-repo",
            "db-snapshot",
            "explain",
            "inspect",
            "instrument",
            "login",
            "modify",
            "replay",
            "session-logs",
            "switch-org",
            "templates",
            "view",
            "wizard",
        ]);
        expect(setupFlow.defaultMode).toBe("wizard");
    });
    it("analyze-repo mode enters at preflight and is a self-contained non-interactive phase", () => {
        expect(setupFlow.entries["analyze-repo"]).toBe("analyze-repo/preflight");
        const phase = setupFlow.phases.find((p) => p.id === "analyze-repo");
        expect(phase).toBeDefined();
        const analyzePhase = phase;
        // Non-interactive: no step in the phase may call AskUserQuestion, and none
        // may open Studio via the blocking openTracePlan CLI.
        for (const step of analyzePhase.steps) {
            expect(step.toolCalls ?? []).not.toContain("ask");
            expect(step.commandCalls ?? []).not.toContain("openTracePlan");
        }
        // preflight gates on auth via get_bitfab_api_key.
        expect(findStep("analyze-repo", "preflight").toolCalls).toContain("apiKey");
        // build-and-upload uploads plans (create_trace_plan) but never opens the
        // browser confirmation UI.
        const build = findStep("analyze-repo", "build-and-upload");
        expect(build.toolCalls).toContain("createTracePlan");
        // The body must explicitly instruct skipping the blocking browser
        // confirmation (the phase-wide loop above already asserts openTracePlan is
        // not in commandCalls).
        expect(build.body).toMatch(/do NOT run `openTracePlan/);
        // report converges on the shared cleanup step like every other mode.
        const report = findStep("analyze-repo", "report");
        if (report.kind === "action") {
            expect(report.next).toBe("cleanup/close-studio");
        }
    });
    it("analyze-repo detects existing instrumentation and uploads a per-node diff for already-traced workflows", () => {
        // scan gathers the existing-instrumentation map (greps for SDK usage).
        const scan = findStep("analyze-repo", "scan");
        expect(scan.body).toMatch(/already instrumented/i);
        expect(scan.body).toContain("withSpan");
        expect(scan.body).toContain("alreadyTraced");
        // build-and-upload sets the diff fields per node and reuses the existing key.
        const build = findStep("analyze-repo", "build-and-upload");
        expect(build.body).toContain("alreadyTraced");
        expect(build.body).toContain("alreadyMocked");
        // The diff must be unambiguous: every node carries alreadyTraced on a diff
        // plan, and greenfield plans omit it entirely.
        expect(build.body).toMatch(/every.*node/i);
        expect(build.body).toMatch(/existing.*traceFunctionKey/i);
        expect(build.body).toMatch(/[Oo]mit both fields entirely on greenfield/);
        // report surfaces the delta (new spans vs dropped spans) per candidate.
        const report = findStep("analyze-repo", "report");
        expect(report.body).toMatch(/new span/i);
        expect(report.body).toMatch(/drop/i);
    });
    it("the claude analyze-repo sub-skill is NOT granted AskUserQuestion (non-interactivity is enforced, not just instructed)", () => {
        const files = renderFlow(setupFlow, TARGETS.claude);
        const subSkill = files.find((f) => f.slug === "analyze-repo");
        expect(subSkill).toBeDefined();
        // The allowed-tools frontmatter line must not grant AskUserQuestion, so the
        // Claude permission layer denies any attempt to prompt the user.
        const allowedToolsLine = subSkill.content
            .split("\n")
            .find((line) => line.startsWith("allowed-tools:"));
        expect(allowedToolsLine).toBeDefined();
        expect(allowedToolsLine).not.toContain("AskUserQuestion");
        // Sanity: it still grants the tools the phase actually needs.
        expect(allowedToolsLine).toContain("create_trace_plan");
    });
    it("the analyze-repo phase is reachable only in `analyze-repo` mode", () => {
        for (const targetName of TARGET_NAMES) {
            const modesPerStep = computeModesPerStep(setupFlow, targetName);
            for (const stepId of [
                "analyze-repo/preflight",
                "analyze-repo/scan",
                "analyze-repo/select",
                "analyze-repo/build-and-upload",
                "analyze-repo/report",
            ]) {
                expect([...(modesPerStep.get(stepId) ?? [])]).toEqual(["analyze-repo"]);
            }
        }
    });
    it("preamble is reachable only in `wizard` mode for every target", () => {
        for (const targetName of TARGET_NAMES) {
            const modesPerStep = computeModesPerStep(setupFlow, targetName);
            const preambleModes = modesPerStep.get("preamble/render-intro");
            expect(preambleModes).toBeDefined();
            expect([...(preambleModes ?? [])]).toEqual(["wizard"]);
        }
    });
    it("login phase gates `wizard`, `login`, and `instrument` modes", () => {
        for (const targetName of TARGET_NAMES) {
            const modesPerStep = computeModesPerStep(setupFlow, targetName);
            const loginStartModes = modesPerStep.get("login/status-check");
            expect(loginStartModes).toBeDefined();
            expect([...(loginStartModes ?? [])].sort()).toEqual([
                "instrument",
                "login",
                "wizard",
            ]);
        }
    });
    it("login/run-login failure path routes to cleanup", () => {
        const step = findStep("login", "run-login");
        expect(step.kind).toBe("branch");
        if (step.kind !== "branch") {
            throw new Error("unreachable");
        }
        const failureBranch = step.branches.find((b) => b.when === "login fails, errors, or times out" &&
            b.next === "cleanup/close-studio");
        expect(failureBranch).toBeDefined();
    });
    it.each(TARGET_NAMES)("tells %s to surface fresh Studio URLs from window-open-requested events", (targetName) => {
        const output = renderMarkdown(setupFlow, TARGETS[targetName]);
        expect(output).toContain('{"event":"window-open-requested","url":"..."}');
        expect(output).toContain("Opening Studio: <url>");
        expect(output).toContain("click it if a window doesn't appear");
    });
    it("instrument/search-existing 'Modify' option cross-phases to modify", () => {
        const step = findStep("instrument", "search-existing");
        expect(step.kind).toBe("branch");
        if (step.kind !== "branch") {
            throw new Error("unreachable");
        }
        const modifyBranch = step.branches.find((b) => b.option?.label === "Modify an existing trace setup");
        expect(modifyBranch).toBeDefined();
        expect(modifyBranch?.next).toBe("modify/gather-existing");
    });
    it("instrument/next-workflow offers another workflow or cleanup", () => {
        const step = findStep("instrument", "next-workflow");
        expect(step.kind).toBe("branch");
        if (step.kind !== "branch") {
            throw new Error("unreachable");
        }
        const aBranch = step.branches.find((b) => b.option?.letter === "A");
        expect(aBranch?.next).toBe("instrument/identify-workflows");
        const bBranch = step.branches.find((b) => b.option?.letter === "B");
        expect(bBranch?.next).toBe("instrument/choose-discovery");
        const cBranch = step.branches.find((b) => b.option?.letter === "C");
        expect(cBranch?.next).toBe("cleanup/close-studio");
    });
    it("replay/safety-net falls through to cleanup when every function is invocable", () => {
        const step = findStep("replay", "safety-net");
        expect(step.kind).toBe("branch");
        if (step.kind !== "branch") {
            throw new Error("unreachable");
        }
        // The A/B/C options only fire when a function is un-invocable; the common
        // "everything callable" path must be an explicit `when` arm to cleanup, or
        // the split-chain dead-ends without running closeStudio.js.
        const fallThrough = step.branches.find((b) => b.option === undefined && b.when !== undefined);
        expect(fallThrough).toBeDefined();
        expect(fallThrough?.next).toBe("cleanup/close-studio");
    });
    it("switch-org/perform-switch routes not-member/error to cleanup, not a prose stop", () => {
        const step = findStep("switch-org", "perform-switch");
        expect(step.kind).toBe("branch");
        if (step.kind !== "branch") {
            throw new Error("unreachable");
        }
        const failureArm = step.branches.find((b) => b.when?.includes("not-member"));
        expect(failureArm).toBeDefined();
        expect(failureArm?.next).toBe("cleanup/close-studio");
        const successArm = step.branches.find((b) => b.when?.includes("switched"));
        expect(successArm?.next).toBe("switch-org/update-local-env");
    });
    it("templates/check-traces routes an existing trace through preview-mode, never straight to open-preview", () => {
        const step = findStep("templates", "check-traces");
        expect(step.kind).toBe("branch");
        if (step.kind !== "branch") {
            throw new Error("unreachable");
        }
        const traceExists = step.branches.find((b) => b.when?.includes("trace exists"));
        expect(traceExists?.next).toBe("templates/preview-mode");
    });
    it("templates/preview-mode lets the user edit inline (skip preview) or open the live preview", () => {
        const step = findStep("templates", "preview-mode");
        expect(step.kind).toBe("branch");
        if (step.kind !== "branch") {
            throw new Error("unreachable");
        }
        // Inline arm skips the preview entirely and drops straight into the edit loop,
        // so a user already viewing a trace is never yanked onto the preview page.
        const inlineArm = step.branches.find((b) => b.when?.includes("inline"));
        expect(inlineArm?.next).toBe("templates/edit-loop");
        // Preview arm opens the chromeless preview first, then enters the same edit loop.
        const previewArm = step.branches.find((b) => b.when?.includes("live preview"));
        expect(previewArm?.next).toBe("templates/open-preview");
    });
    it("replay phase is reachable only through explicit replay mode", () => {
        for (const targetName of TARGET_NAMES) {
            const modesPerStep = computeModesPerStep(setupFlow, targetName);
            const replayStartModes = modesPerStep.get("replay/gather-keys");
            expect(replayStartModes).toBeDefined();
            expect([...(replayStartModes ?? [])].sort()).toEqual(["replay"]);
        }
    });
    it("modify phase is reachable in `wizard`, `instrument`, and `modify` modes", () => {
        for (const targetName of TARGET_NAMES) {
            const modesPerStep = computeModesPerStep(setupFlow, targetName);
            const modifyStartModes = modesPerStep.get("modify/gather-existing");
            expect(modifyStartModes).toBeDefined();
            expect([...(modifyStartModes ?? [])].sort()).toEqual([
                "instrument",
                "modify",
                "wizard",
            ]);
        }
    });
    it("explain phase is reachable only in `explain` mode", () => {
        for (const targetName of TARGET_NAMES) {
            const modesPerStep = computeModesPerStep(setupFlow, targetName);
            const explainModes = modesPerStep.get("explain/overview");
            expect(explainModes).toBeDefined();
            expect([...(explainModes ?? [])].sort()).toEqual(["explain"]);
        }
    });
    it("inspect phase is reachable only in `inspect` mode", () => {
        for (const targetName of TARGET_NAMES) {
            const modesPerStep = computeModesPerStep(setupFlow, targetName);
            const inspectModes = modesPerStep.get("inspect/status-check");
            expect(inspectModes).toBeDefined();
            expect([...(inspectModes ?? [])].sort()).toEqual(["inspect"]);
        }
    });
    it("db-snapshot phase is reachable only in `db-snapshot` mode", () => {
        for (const targetName of TARGET_NAMES) {
            const modesPerStep = computeModesPerStep(setupFlow, targetName);
            const startModes = modesPerStep.get("db-snapshot/preflight");
            expect(startModes).toBeDefined();
            expect([...(startModes ?? [])].sort()).toEqual(["db-snapshot"]);
        }
    });
    it("registers the get_database_connection_status MCP tool and polls it in await-connection", () => {
        expect(setupFlow.tools.dbConnectionStatus).toEqual({
            kind: "mcp",
            name: "get_database_connection_status",
        });
        const connect = findStep("db-snapshot", "connect-db");
        expect(connect.toolCalls).toContain("dbConnectionStatus");
        const await_ = findStep("db-snapshot", "await-connection");
        expect(await_.toolCalls).toContain("dbConnectionStatus");
    });
    it("db-snapshot/await-connection loops on `checking` and exits to wire-replay on `connected`", () => {
        const step = findStep("db-snapshot", "await-connection");
        expect(step.kind).toBe("branch");
        if (step.kind !== "branch") {
            throw new Error("unreachable");
        }
        const connected = step.branches.find((b) => b.when === "status is connected");
        expect(connected?.next).toBe("db-snapshot/wire-replay");
        const checking = step.branches.find((b) => b.when === "status is checking");
        expect(checking?.next).toBe("db-snapshot/await-connection");
    });
    it("db-snapshot wires ReplayEnvironment per language and has no capture-config step (capture is always-on)", () => {
        const phase = setupFlow.phases.find((p) => p.id === "db-snapshot");
        expect(phase?.steps.some((s) => s.id === "enable-snapshot")).toBe(false);
        const wire = findStep("db-snapshot", "wire-replay");
        expect(wire.body).toMatch(/new ReplayEnvironment\(\)/); // TypeScript
        expect(wire.body).toMatch(/environment=env/); // Python
        expect(wire.body).toMatch(/Bitfab::ReplayEnvironment\.new/); // Ruby
        const smoke = findStep("db-snapshot", "smoke-test");
        expect(smoke.kind).toBe("action");
        if (smoke.kind !== "action") {
            throw new Error("unreachable");
        }
        expect(smoke.next).toBe("cleanup/close-studio");
    });
    it("modify/reconstruct-current-plan calls getTracePlan to bootstrap from a prior plan", () => {
        const step = findStep("modify", "reconstruct-current-plan");
        expect(step.toolCalls).toContain("getTracePlan");
        expect(step.body).toMatch(/traceFunctionKey/);
    });
    it("modify/build-modified-plan requires surrounding callers + callees as pure context", () => {
        const step = findStep("modify", "build-modified-plan");
        expect(step.body).toMatch(/~10 callers above/);
        expect(step.body).toMatch(/~10 callees below/);
        expect(step.body).toMatch(/`pure`/);
    });
    it("modify/present-diff routes Trace plan cancelled to the modifications branch (not terminate)", () => {
        const step = findStep("modify", "present-diff");
        expect(step.kind).toBe("branch");
        if (step.kind !== "branch") {
            throw new Error("unreachable");
        }
        expect(step.toolCalls).toContain("createTracePlan");
        expect(step.toolCalls).toContain("getTracePlan");
        const proceed = step.branches.find((b) => b.option?.letter === "A");
        expect(proceed?.next).toBe("modify/apply-changes");
        const modifications = step.branches.find((b) => b.option?.letter === "C");
        expect(modifications?.option?.label).toBe("Modifications");
        expect(modifications?.next).toBe("modify/build-modified-plan");
        const abort = step.branches.find((b) => b.option?.letter === "D");
        expect(abort?.option?.label).toBe("Abort entirely");
        expect(abort?.next).toBe("cleanup/close-studio");
        expect(step.body).toMatch(/"event":"cancelled".*branch \*\*C\*\*/s);
    });
    it("instrument/build-trace-plan passes traceFunctionKey to create_trace_plan", () => {
        const step = findStep("instrument", "build-trace-plan");
        expect(step.body).toMatch(/traceFunctionKey/);
        expect(step.body).toMatch(/~10 callers above/);
        expect(step.body).toMatch(/~10 callees below/);
    });
    it("trace-plan delivery renders inline ASCII first, then offers browser or continue", () => {
        // Instrument: renders the plan inline, asks View in browser vs Continue,
        // and the Continue path proceeds without a Studio round-trip.
        const build = findStep("instrument", "build-trace-plan");
        expect(build.body).toMatch(/[Rr]ender the trace plan inline as ASCII/);
        expect(build.body).toMatch(/View in browser/);
        expect(build.body).toMatch(/If they pick Continue/);
        // Continue must persist the plan as confirmed (else view/modify can't find
        // it by key), so it calls confirmTracePlan, not just getTracePlan.
        expect(build.body).toMatch(/\{\{tool:confirmTracePlan\}\}/);
        expect(build.toolCalls ?? []).toContain("confirmTracePlan");
        // The old "delivered ONE way only / forbidden to render inline" invariant
        // must be gone, inline ASCII is now the default surface.
        expect(build.body).not.toMatch(/delivered ONE way only/);
        expect(build.body).not.toMatch(/is forbidden/);
        // Modify: same render-then-ask shape; Continue maps to the apply branch.
        const modify = findStep("modify", "present-diff");
        expect(modify.body).toMatch(/[Rr]ender the modified plan inline as ASCII/);
        expect(modify.body).toMatch(/View in browser/);
        // Substep 1 posts the plan without opening Studio (render + ask come after).
        expect(modify.body).not.toMatch(/Post the modified plan and open the UI/);
        // "View in browser" is an explicit AskUserQuestion branch, not prose-only.
        if (modify.kind === "branch") {
            const browserBranch = modify.branches.find((b) => b.option?.label === "View in browser");
            expect(browserBranch).toBeDefined();
            expect(browserBranch?.next).toBe("modify/apply-changes");
        }
        // Modify Continue also persists via confirmTracePlan.
        expect(modify.body).toMatch(/\{\{tool:confirmTracePlan\}\}/);
        expect(modify.toolCalls ?? []).toContain("confirmTracePlan");
        // View: renders inline, then offers the browser as an optional read-only look.
        const view = findStep("view", "open-in-ui");
        expect(view.body).toMatch(/[Rr]ender the trace plan inline as ASCII/);
        expect(view.body).toMatch(/View in browser/);
    });
    it("trace-plan steps require analysis on every node, including context", () => {
        const buildTracePlan = findStep("instrument", "build-trace-plan");
        const buildModifiedPlan = findStep("modify", "build-modified-plan");
        const presentDiff = findStep("modify", "present-diff");
        expect(buildTracePlan.body).toMatch(/Every node in the `TracePlanTree`/);
        expect(buildTracePlan.body).toMatch(/including uncaptured surrounding context nodes/);
        expect(buildTracePlan.body).toMatch(/mockable `external_read` and `side_effect` nodes serve their recorded output by default/);
        expect(buildTracePlan.body).toMatch(/server defaults it to `true` for mockable `external_read` and `side_effect` nodes/);
        expect(buildTracePlan.body).toMatch(/`false` for `model_call` and `pure` nodes/);
        expect(buildTracePlan.body).toMatch(/smallest span that actually crosses/);
        expect(buildTracePlan.body).toMatch(/broad external parents with live descendants/);
        expect(buildTracePlan.body).not.toMatch(/uncaptured context nodes omit it/);
        expect(buildTracePlan.body).not.toMatch(/Omit `analysis`/);
        expect(buildTracePlan.body).not.toMatch(/replay re-runs every span by default/);
        expect(buildModifiedPlan.body).toMatch(/Every node in the modified `TracePlanTree`/);
        expect(buildModifiedPlan.body).toMatch(/Mockable `external_read` and `side_effect` nodes default to mocked replay/);
        expect(buildModifiedPlan.body).toMatch(/smallest external boundary/);
        expect(buildModifiedPlan.body).toMatch(/broad external parents with live descendants/);
        expect(buildModifiedPlan.body).toMatch(/older prior plans that lack `analysis`/);
        expect(buildModifiedPlan.body).toMatch(/backfill `analysis`/);
        expect(buildModifiedPlan.body).toMatch(/still attach `analysis`/);
        expect(buildModifiedPlan.body).not.toMatch(/Omit `analysis`/);
        expect(presentDiff.body).toMatch(/Every node carries the `analysis` you set/);
        expect(presentDiff.body).not.toMatch(/set its `mockOnReplay`/);
    });
    it("build-trace-plan gates plan construction on a replayable root", () => {
        const buildTracePlan = findStep("instrument", "build-trace-plan");
        // The gate must fire BEFORE the tree is constructed, not after the plan is posted.
        expect(buildTracePlan.body).toMatch(/Replayability gate/);
        expect(buildTracePlan.body).toMatch(/verify the root is replayable BEFORE you construct the tree/);
        // Both clearing cases are spelled out.
        expect(buildTracePlan.body).toMatch(/Serializable input/);
        expect(buildTracePlan.body).toMatch(/Replayable-root handler\/processor/);
        // A live-client argument does not clear the gate on its own.
        expect(buildTracePlan.body).toMatch(/single live-client argument does NOT clear the gate/);
        // If it clears neither, do not build or post the plan.
        expect(buildTracePlan.body).toMatch(/do not build or post the plan/);
        // The three resolutions route back through present-workflows.
        expect(buildTracePlan.body).toMatch(/Return to step \{\{step:instrument\/present-workflows\}\}/);
    });
    it("replay script instructions use marked mocks for captured external calls", () => {
        const writeInstrumentation = findStep("instrument", "write-instrumentation");
        expect(writeInstrumentation.body).toMatch(/mock: "marked"/);
        expect(writeInstrumentation.body).toMatch(/captured external DB\/HTTP calls are isolated by their trace-plan span's `mockOnReplay: true`/);
        expect(writeInstrumentation.body).toMatch(/no hand-mocked DB clients/);
        expect(writeInstrumentation.body).not.toMatch(/leave the DB unmocked/);
        expect(writeInstrumentation.body).not.toMatch(/reads today's data/);
    });
    it("declares the `reference` appendix consulted by build-trace-plan and build-modified-plan", () => {
        expect(setupFlow.appendices.find((a) => a.id === "reference")).toBeDefined();
        const buildTracePlan = findStep("instrument", "build-trace-plan");
        const buildModifiedPlan = findStep("modify", "build-modified-plan");
        expect(buildTracePlan.kind).toBe("reference");
        expect(buildModifiedPlan.kind).toBe("reference");
        if (buildTracePlan.kind === "reference") {
            expect(buildTracePlan.refs).toEqual(["reference"]);
        }
        if (buildModifiedPlan.kind === "reference") {
            expect(buildModifiedPlan.refs).toEqual(["reference"]);
        }
    });
    it("every step is reachable in some (mode, target) combination", () => {
        const allStepIds = new Set();
        for (const phase of setupFlow.phases) {
            for (const step of phase.steps) {
                allStepIds.add(`${phase.id}/${step.id}`);
            }
        }
        const reachableUnion = new Set();
        for (const targetName of TARGET_NAMES) {
            for (const id of computeReachableSteps(setupFlow, targetName)) {
                reachableUnion.add(id);
            }
        }
        expect([...reachableUnion].sort()).toEqual([...allStepIds].sort());
    });
});
describe("renderMarkdown(setupFlow)", () => {
    it("renders without throwing for every target", () => {
        for (const targetName of TARGET_NAMES) {
            const target = TARGETS[targetName];
            expect(() => renderMarkdown(setupFlow, target)).not.toThrow();
        }
    });
    it("preamble ASCII diagram appears in every target's rendered output", () => {
        for (const targetName of TARGET_NAMES) {
            const rendered = renderMarkdown(setupFlow, TARGETS[targetName]);
            expect(rendered).toContain("CODE");
            expect(rendered).toContain("TRACES");
            expect(rendered).toContain("DATASETS");
            expect(rendered).toContain("IMPROVE");
        }
    });
    it("codex output contains the BITFAB_PLUGIN_DIR resolution block; non-codex does not", () => {
        const codex = renderMarkdown(setupFlow, TARGETS.codex);
        expect(codex).toContain("BITFAB_PLUGIN_DIR=$(");
        expect(codex).toContain("CODEX_HOME");
        const claude = renderMarkdown(setupFlow, TARGETS.claude);
        expect(claude).not.toContain("CODEX_HOME");
        const cursor = renderMarkdown(setupFlow, TARGETS.cursor);
        expect(cursor).not.toContain("CODEX_HOME");
    });
    it("codex output uses 'ask the user' framing where claude/cursor use AskUserQuestion", () => {
        const codex = renderMarkdown(setupFlow, TARGETS.codex);
        const claude = renderMarkdown(setupFlow, TARGETS.claude);
        expect(claude).toContain("AskUserQuestion");
        expect(codex).not.toContain("Use `AskUserQuestion`");
        expect(codex).toContain("Ask the user");
    });
    it("apiKey MCP tool name reflects the target prefix", () => {
        expect(renderMarkdown(setupFlow, TARGETS.claude)).toContain("mcp__plugin_bitfab_Bitfab__get_bitfab_api_key");
        expect(renderMarkdown(setupFlow, TARGETS.cursor)).toContain("mcp__Bitfab__get_bitfab_api_key");
        expect(renderMarkdown(setupFlow, TARGETS.codex)).toContain("mcp__Bitfab__get_bitfab_api_key");
    });
    it("instrumentation replay guidance enforces root parity in every target", () => {
        for (const targetName of TARGET_NAMES) {
            const rendered = renderMarkdown(setupFlow, TARGETS[targetName]);
            expect(rendered).toContain("Replay root parity (hard rule)");
            expect(rendered).toContain("exact same exported top-level traced wrapper that production/runtime calls to create the root span");
            expect(rendered).toContain("You must do this unless it is genuinely impossible in the host app");
            expect(rendered).toContain("inconvenience, extra refactoring, an inline wrapper, or needing to move code is not impossible");
            expect(rendered).toContain("extract it into the nearest appropriate service/module, export it, and update both production and replay to import and call that same symbol");
            expect(rendered).toContain("Do not replay a convenient inner helper unless that exact helper is also the production root traced wrapper");
            expect(rendered).toContain("Avoid duplicate semantic wrappers split across production and replay with names like `runX`, `processX`, or `generateX`");
            expect(rendered).toContain("re-invokes the same framework entrypoint production calls");
            expect(rendered).not.toContain("re-invokes the framework entrypoint with the recorded root input plus");
            expect(rendered).not.toContain("the handler-recorded root stays replayable via a same-key wrapper in the replay script");
            expect(rendered).toContain("Replay root parity:");
            expect(rendered).toContain("- Production root symbol:");
            expect(rendered).toContain("- Production import/path:");
            expect(rendered).toContain("- Replay symbol:");
            expect(rendered).toContain("- Replay import/path:");
            expect(rendered).toContain("- Same symbol? yes/no");
            expect(rendered).toContain("- If no, why is this impossible?");
        }
    });
    it("standalone replay script creation also enforces root parity", () => {
        const step = findStep("replay", "create-scripts");
        expect(step.kind).toBe("action");
        if (step.kind !== "action") {
            throw new Error("unreachable");
        }
        expect(step.body).toContain("Replay root parity (hard rule)");
        expect(step.body).toContain("exact same exported top-level traced wrapper that production/runtime calls to create the root span");
        expect(step.body).toContain("You must do this unless it is genuinely impossible in the host app");
        expect(step.body).toContain("For handler-instrumented keys with explicit-key replay, verify parity against the production keyed handler/run-wrapper entrypoint");
        expect(step.body).toContain("Replay root parity verification");
        expect(step.body).toContain("Production root symbol:");
        expect(step.body).toContain("If no, why is this impossible?");
    });
});
describe("split-chain dispatch integrity", () => {
    const STEP_BY_ID = new Map();
    for (const phase of setupFlow.phases) {
        for (const step of phase.steps) {
            STEP_BY_ID.set(`${phase.id}/${step.id}`, step);
        }
    }
    function entryStart(entry) {
        if (typeof entry === "string") {
            return entry;
        }
        const resolved = entry.claude ?? entry.default;
        if (resolved === undefined) {
            throw new Error("entry has no resolvable start step");
        }
        return resolved;
    }
    function walkChain(mode, start) {
        const visited = new Set();
        const terminals = new Set();
        const queue = [start];
        while (queue.length > 0) {
            const id = queue.shift();
            if (id === null || id === undefined || visited.has(id)) {
                continue;
            }
            visited.add(id);
            const step = STEP_BY_ID.get(id);
            if (step === undefined) {
                throw new Error(`mode "${mode}": Next ref "${id}" has no matching step`);
            }
            const refs = nextRefsForMode(step, mode);
            if (refs.every((ref) => ref === null)) {
                terminals.add(id);
            }
            for (const ref of refs) {
                queue.push(ref);
            }
        }
        return { visited, terminals };
    }
    it("every Next ref in every mode resolves to a real step or null", () => {
        const validIds = new Set(STEP_BY_ID.keys());
        for (const mode of Object.keys(setupFlow.entries)) {
            for (const [id, step] of STEP_BY_ID) {
                for (const ref of nextRefsForMode(step, mode)) {
                    if (ref !== null) {
                        expect(validIds.has(ref), `mode "${mode}" step "${id}" routes to missing step "${ref}"`).toBe(true);
                    }
                }
            }
        }
    });
    it("every mode's dispatch chain converges on cleanup/close-studio as its only terminal", () => {
        for (const [mode, entry] of Object.entries(setupFlow.entries)) {
            const { visited, terminals } = walkChain(mode, entryStart(entry));
            expect(visited.has("cleanup/close-studio"), `mode "${mode}" never reaches cleanup/close-studio`).toBe(true);
            expect([...terminals].sort(), `mode "${mode}" has unexpected terminal steps`).toEqual(["cleanup/close-studio"]);
        }
    });
    it("wizard chains preamble -> login -> instrument -> cleanup", () => {
        const wizard = walkChain("wizard", entryStart(setupFlow.entries.wizard)).visited;
        for (const id of [
            "preamble/render-intro",
            "login/status-check",
            "instrument/detect-language",
            "cleanup/close-studio",
        ]) {
            expect(wizard.has(id), `wizard chain missing "${id}"`).toBe(true);
        }
        const instrument = walkChain("instrument", entryStart(setupFlow.entries.instrument)).visited;
        expect(instrument.has("instrument/detect-language")).toBe(true);
        expect(wizard.has("replay/gather-keys")).toBe(false);
        expect(instrument.has("replay/gather-keys"), "instrument mode must NOT fall through into the replay phase").toBe(false);
    });
});
describe("rendered output hygiene", () => {
    const EXPECTED_FILE_COUNT = {
        claude: 15,
        cursor: 1,
        codex: 1,
    };
    it("renderFlow leaves no unresolved {{...}} template tokens in any file", () => {
        for (const targetName of TARGET_NAMES) {
            const files = renderFlow(setupFlow, TARGETS[targetName]);
            for (const file of files) {
                const leftover = file.content.match(/\{\{[^}]*\}\}/g);
                expect(leftover, `target "${targetName}" file "${file.slug || file.kind}" has unresolved tokens: ${leftover?.join(", ")}`).toBeNull();
            }
        }
    });
    it("compiles claude to an orchestrator + per-phase sub-skills, cursor/codex to a monolith", () => {
        for (const targetName of TARGET_NAMES) {
            const files = renderFlow(setupFlow, TARGETS[targetName]);
            expect(files.length, `target "${targetName}" produced an unexpected number of files`).toBe(EXPECTED_FILE_COUNT[targetName]);
        }
        const claudeKinds = renderFlow(setupFlow, TARGETS.claude).map((f) => f.kind);
        expect(claudeKinds.filter((kind) => kind === "orchestrator")).toHaveLength(1);
        expect(claudeKinds.filter((kind) => kind === "phase")).toHaveLength(14);
        expect(renderFlow(setupFlow, TARGETS.cursor)[0].kind).toBe("monolith");
        expect(renderFlow(setupFlow, TARGETS.codex)[0].kind).toBe("monolith");
    });
});
