// bitfab-plugin-lib/dist/chatSessions/arming.js
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
function sessionsDir() {
    return path.join(os.homedir(), ".config", "bitfab", "sessions");
}
export function ensureSessionsDir() {
    fs.mkdirSync(sessionsDir(), { recursive: true });
}
function armedMarkerPath(sessionId) {
    return path.join(sessionsDir(), `${sessionId}.armed`);
}
export function localOffsetPath(sessionId) {
    return path.join(sessionsDir(), `${sessionId}.offset`);
}
const BITFAB_MCP_PREFIXES = [
    "mcp__plugin_bitfab_Bitfab__",
    "mcp__Bitfab__",
];
export function isBitfabMcpToolName(toolName) {
    return BITFAB_MCP_PREFIXES.some((prefix) => toolName.startsWith(prefix));
}
export function isArmed(sessionId) {
    return fs.existsSync(armedMarkerPath(sessionId));
}
export function arm(sessionId, metadata) {
    ensureSessionsDir();
    const marker = armedMarkerPath(sessionId);
    if (!fs.existsSync(marker)) {
        fs.writeFileSync(marker, JSON.stringify(metadata));
    }
}
export function readArmedMetadata(sessionId) {
    try {
        const raw = fs.readFileSync(armedMarkerPath(sessionId), "utf-8").trim();
        if (!raw) {
            return null;
        }
        const parsed = JSON.parse(raw);
        if (typeof parsed.transcriptPath !== "string") {
            return null;
        }
        const VALID_AGENTS = ["claude-code", "cursor", "codex"];
        const agent = typeof parsed.agent === "string" &&
            VALID_AGENTS.includes(parsed.agent)
            ? parsed.agent
            : "claude-code";
        return {
            transcriptPath: parsed.transcriptPath,
            cwd: typeof parsed.cwd === "string" ? parsed.cwd : process.cwd(),
            agent,
        };
    }
    catch {
        return null;
    }
}
export function listArmedSessionIds() {
    try {
        return fs
            .readdirSync(sessionsDir())
            .filter((f) => f.endsWith(".armed"))
            .map((f) => f.replace(/\.armed$/, ""));
    }
    catch {
        return [];
    }
}
export function disarm(sessionId) {
    try {
        fs.unlinkSync(armedMarkerPath(sessionId));
    }
    catch { }
}
function studioMappingPath(agentSessionId) {
    return path.join(sessionsDir(), `${agentSessionId}.studio`);
}
export function writeStudioMapping(agentSessionId, studioSessionId) {
    ensureSessionsDir();
    const data = {
        studioSessionId,
        createdAt: new Date().toISOString(),
    };
    fs.writeFileSync(studioMappingPath(agentSessionId), JSON.stringify(data));
}
export function readStudioMapping(agentSessionId) {
    try {
        const raw = fs
            .readFileSync(studioMappingPath(agentSessionId), "utf-8")
            .trim();
        if (!raw) {
            return null;
        }
        const parsed = JSON.parse(raw);
        if (typeof parsed.studioSessionId !== "string") {
            return null;
        }
        return {
            studioSessionId: parsed.studioSessionId,
            createdAt: typeof parsed.createdAt === "string"
                ? parsed.createdAt
                : new Date().toISOString(),
        };
    }
    catch {
        return null;
    }
}
export function clearStudioMapping(agentSessionId) {
    try {
        fs.unlinkSync(studioMappingPath(agentSessionId));
    }
    catch { }
}
