// bitfab-plugin-lib/dist/chatSessions/arming.js
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
function sessionsDir() {
    return path.join(os.homedir(), ".config", "bitfab", "sessions");
}
export function ensureSessionsDir() {
    fs.mkdirSync(sessionsDir(), { recursive: true });
}
function armedMarkerPath(sessionId) {
    return path.join(sessionsDir(), `${sessionId}.armed`);
}
function agentRunIdPath(sessionId) {
    return path.join(sessionsDir(), `${sessionId}.agent-run-id`);
}
export function localOffsetPath(sessionId) {
    return path.join(sessionsDir(), `${sessionId}.offset`);
}
const BITFAB_MCP_PREFIXES = [
    "mcp__plugin_bitfab_Bitfab__",
    "mcp__Bitfab__",
];
export function isBitfabMcpToolName(toolName) {
    return BITFAB_MCP_PREFIXES.some((prefix) => toolName.startsWith(prefix));
}
export function isArmed(sessionId) {
    return fs.existsSync(armedMarkerPath(sessionId));
}
export function arm(sessionId, metadata) {
    ensureSessionsDir();
    fs.writeFileSync(armedMarkerPath(sessionId), JSON.stringify(metadata));
}
export function writeAgentRunId(sessionId, agentRunId) {
    if (!sessionId || !agentRunId.trim()) {
        return;
    }
    ensureSessionsDir();
    fs.writeFileSync(agentRunIdPath(sessionId), agentRunId);
}
export function readAgentRunId(sessionId) {
    try {
        const value = fs.readFileSync(agentRunIdPath(sessionId), "utf-8").trim();
        return value === "" ? undefined : value;
    }
    catch {
        return undefined;
    }
}
export function clearAgentRunId(sessionId) {
    try {
        fs.unlinkSync(agentRunIdPath(sessionId));
    }
    catch { }
}
export function readArmedMetadata(sessionId) {
    try {
        const raw = fs.readFileSync(armedMarkerPath(sessionId), "utf-8").trim();
        if (!raw) {
            return null;
        }
        const parsed = JSON.parse(raw);
        if (typeof parsed.transcriptPath !== "string") {
            return null;
        }
        const VALID_AGENTS = ["claude-code", "cursor", "codex"];
        const agent = typeof parsed.agent === "string" &&
            VALID_AGENTS.includes(parsed.agent)
            ? parsed.agent
            : "claude-code";
        return {
            transcriptPath: parsed.transcriptPath,
            cwd: typeof parsed.cwd === "string" ? parsed.cwd : process.cwd(),
            agent,
            agentRunId: typeof parsed.agentRunId === "string" ? parsed.agentRunId : undefined,
            pluginVersion: typeof parsed.pluginVersion === "string"
                ? parsed.pluginVersion
                : undefined,
        };
    }
    catch {
        return null;
    }
}
export function listArmedSessionIds() {
    try {
        return fs
            .readdirSync(sessionsDir())
            .filter((f) => f.endsWith(".armed"))
            .map((f) => f.replace(/\.armed$/, ""));
    }
    catch {
        return [];
    }
}
export function disarm(sessionId) {
    try {
        fs.unlinkSync(armedMarkerPath(sessionId));
    }
    catch { }
}
