// bitfab-plugin-lib/dist/chatSessions/transcriptReader.js
import fs from "node:fs";
/**
 * Reads the transcript file from `fromOffset` (in bytes) to EOF. Each newline-
 * terminated JSON line is parsed and normalised. The returned `newOffset` is
 * the byte length of the file after reading, suitable for passing as the next
 * `fromOffset`.
 *
 * Lines that fail to parse or don't carry a recognised role are skipped and
 * counted in `malformedLines`; sequence numbers stay aligned with line indices
 * so the server-side ordering is preserved across imports.
 */
export function readTranscriptDelta(transcriptPath, fromOffset, sequenceBaseOverride) {
    const stat = fs.statSync(transcriptPath);
    const fileSize = stat.size;
    if (fromOffset > fileSize) {
        // File shrank — most likely compaction. The caller is expected to detect
        // this and flush a reset rather than continue appending.
        return { turns: [], newOffset: fileSize, malformedLines: 0 };
    }
    if (fromOffset === fileSize) {
        return { turns: [], newOffset: fileSize, malformedLines: 0 };
    }
    const fd = fs.openSync(transcriptPath, "r");
    let buffer;
    try {
        const bytesToRead = fileSize - fromOffset;
        buffer = Buffer.alloc(bytesToRead);
        fs.readSync(fd, buffer, 0, bytesToRead, fromOffset);
    }
    finally {
        fs.closeSync(fd);
    }
    const text = buffer.toString("utf-8");
    const lines = text.split("\n");
    // The slice may end mid-line if the writer hasn't flushed a newline yet.
    // Drop a trailing partial line and back the offset off so we re-read it on
    // the next call.
    const lastLineComplete = text.endsWith("\n");
    const completeLines = lines.slice(0, -1);
    const trailingPartialBytes = lastLineComplete
        ? 0
        : Buffer.byteLength(lines[lines.length - 1] ?? "", "utf-8");
    // The line-index base for sequence numbers is the count of newline-
    // terminated lines that precede `fromOffset`. Compute by reading the file
    // up to the offset (cheap; transcripts stay reasonably small).
    const sequenceBase = sequenceBaseOverride ?? countLinesBefore(transcriptPath, fromOffset);
    const turns = [];
    let malformedLines = 0;
    for (let i = 0; i < completeLines.length; i++) {
        const line = completeLines[i];
        if (!line.trim()) {
            continue;
        }
        const turn = parseLine(line, sequenceBase + i);
        if (turn === null) {
            malformedLines++;
            continue;
        }
        turns.push(turn);
    }
    const newOffset = fileSize - trailingPartialBytes;
    return { turns, newOffset, malformedLines };
}
function countLinesBefore(transcriptPath, offset) {
    if (offset === 0) {
        return 0;
    }
    const fd = fs.openSync(transcriptPath, "r");
    try {
        const buffer = Buffer.alloc(offset);
        fs.readSync(fd, buffer, 0, offset, 0);
        let count = 0;
        for (let i = 0; i < buffer.length; i++) {
            if (buffer[i] === 0x0a) {
                count++;
            }
        }
        return count;
    }
    finally {
        fs.closeSync(fd);
    }
}
function parseLine(line, sequence) {
    let entry;
    try {
        entry = JSON.parse(line);
    }
    catch {
        return null;
    }
    const role = entry.message?.role;
    if (!role) {
        return null;
    }
    const turnStartedAt = entry.timestamp ?? new Date().toISOString();
    const rawContent = entry.message?.content;
    if (role === "user") {
        const toolResultBlocks = extractToolResultBlocks(rawContent);
        if (toolResultBlocks !== null) {
            return {
                sequence,
                role: "tool_result",
                content: toolResultBlocks,
                turnStartedAt,
            };
        }
        return {
            sequence,
            role: "user",
            content: rawContent ?? null,
            turnStartedAt,
        };
    }
    if (role === "assistant") {
        const toolCalls = extractToolCalls(rawContent);
        return {
            sequence,
            role: "assistant",
            content: rawContent ?? null,
            ...(toolCalls.length > 0 ? { toolCalls } : {}),
            ...(entry.message?.usage !== undefined
                ? { usage: entry.message.usage }
                : {}),
            turnStartedAt,
        };
    }
    if (role === "system") {
        return {
            sequence,
            role: "system",
            content: rawContent ?? null,
            turnStartedAt,
        };
    }
    return null;
}
function extractToolCalls(content) {
    if (!Array.isArray(content)) {
        return [];
    }
    return content.filter((block) => block.type === "tool_use");
}
function extractToolResultBlocks(content) {
    if (!Array.isArray(content)) {
        return null;
    }
    const blocks = content.filter((block) => block.type === "tool_result");
    if (blocks.length === content.length && blocks.length > 0) {
        return blocks;
    }
    return null;
}
