// bitfab-flow/dist/render.js
import { collectCalledToolIds, computeModesPerStep, computeReachableSteps, MUSTACHE_SECTION_NAMES, resolvePerTarget, resolveTargetableString, } from "./schema.js";
/** Shorthand: resolve a flow string field for the current target. */
function s(value, target, label) {
    return resolveTargetableString(value, target, label);
}
/**
 * Render a flow to one or more skill files for a target.
 *
 * - `monolith` (default): a single file from `renderMarkdown` (unchanged).
 * - `split`: an orchestrator (slug `""`) plus one sub-skill per non-inline
 *   phase (slug = phase id). Phases listed in `compile.inlinePhases` stay
 *   inside the orchestrator and emit no sub-skill of their own.
 *
 * `compile.mode` is resolved per target, so a flow can split one target while
 * the rest stay monolith (e.g. `{ default: "monolith", claude: "split" }`).
 */
export function renderFlow(flow, target) {
    const mode = resolvePerTarget(flow.compile.mode, target.name);
    if (mode === "split") {
        return renderSplit(flow, target);
    }
    return [{ slug: "", kind: "monolith", content: renderMarkdown(flow, target) }];
}
function makeRenderContext(flow, target) {
    return {
        allModes: Object.keys(flow.entries),
        modesPerStep: computeModesPerStep(flow, target.name),
        reachable: computeReachableSteps(flow, target.name),
        emitTemplate: flow.emitTemplate !== undefined
            ? s(flow.emitTemplate, target.name, "flow.emitTemplate")
            : undefined,
    };
}
function phaseHasReachableSteps(phase, reachable) {
    return phase.steps.some((step) => reachable.has(`${phase.id}/${step.id}`));
}
/**
 * Expand a split flow into an orchestrator plus one sub-skill per non-inline
 * phase that has at least one reachable step for this target. Reachability is
 * per-target, so a phase only invokable in (say) Codex modes emits no sub-skill
 * and no dispatch entry for the other targets.
 */
function renderSplit(flow, target) {
    const ctx = makeRenderContext(flow, target);
    const inlineSet = new Set(flow.compile.inlinePhases);
    const subPhases = flow.phases.filter((phase) => !inlineSet.has(phase.id) && phaseHasReachableSteps(phase, ctx.reachable));
    const files = [
        {
            slug: "",
            kind: "orchestrator",
            content: renderOrchestrator(flow, target, ctx, inlineSet, subPhases),
        },
    ];
    for (const phase of subPhases) {
        files.push({
            slug: phase.id,
            kind: "phase",
            content: renderSubSkill(flow, target, ctx, phase),
        });
    }
    return files;
}
/**
 * The orchestrator carries the flow's shared chrome (intro, command table,
 * invariants, appendices), renders any inline phases in place, and for each
 * non-inline phase emits a short dispatch section pointing at its sub-skill.
 */
function renderOrchestrator(flow, target, ctx, inlineSet, subPhases) {
    const inlinePhaseIds = flow.phases
        .filter((phase) => inlineSet.has(phase.id))
        .map((phase) => phase.id);
    const sections = [
        target.frontmatter({
            flowId: flow.id,
            description: s(flow.frontmatter.description, target.name, "flow.frontmatter.description"),
            argumentHint: resolveArgumentHint(flow, target),
            model: flow.frontmatter.model,
            allowedTools: orchestratorAllowedTools(flow, target, inlinePhaseIds, subPhases.length > 0),
        }),
        `# ${s(flow.title, target.name, "flow.title")}`,
    ];
    pushIntro(sections, flow, target);
    if (Object.keys(flow.commands).length > 0) {
        sections.push(renderCommandsTable(flow));
    }
    pushInvariants(sections, flow);
    const subPhaseIds = new Set(subPhases.map((phase) => phase.id));
    for (const phase of flow.phases) {
        if (inlineSet.has(phase.id)) {
            const rendered = renderPhase(phase, target.name, ctx.allModes, ctx.modesPerStep, ctx.reachable, ctx.emitTemplate);
            if (rendered.length > 0) {
                sections.push(rendered);
            }
        }
        else if (subPhaseIds.has(phase.id)) {
            sections.push(renderDispatch(flow, target, ctx, phase));
        }
    }
    pushAppendices(sections, flow, target);
    return finalize(sections, flow, target);
}
/**
 * One dispatch section in the orchestrator pointing at a phase sub-skill. It
 * carries the phase's mode gate (so the agent only invokes the sub-skill in
 * the right modes) and tells the agent to invoke the sub-skill by name.
 */
function renderDispatch(flow, target, ctx, phase) {
    const title = s(phase.title, target.name, `phase[${phase.id}].title`);
    const stepKeys = phase.steps
        .map((step) => `${phase.id}/${step.id}`)
        .filter((key) => ctx.reachable.has(key));
    const { phaseGate } = precomputePhaseAndStepGates(stepKeys, ctx.modesPerStep, ctx.allModes);
    const blocks = [`## ${title}`];
    if (phaseGate !== null) {
        blocks.push(phaseGate);
    }
    blocks.push(`Invoke the \`${flow.id}-${phase.id}\` skill, then return here.`);
    return blocks.join("\n\n");
}
/**
 * A phase sub-skill: its own frontmatter plus the phase body, rendered exactly
 * as the monolith would render that phase. Step numbering already resets per
 * phase, so the sub-skill reads as a standalone 1-indexed sequence.
 */
function renderSubSkill(flow, target, ctx, phase) {
    const phaseTitle = s(phase.title, target.name, `phase[${phase.id}].title`);
    const flowTitle = s(flow.title, target.name, "flow.title");
    // No trailing period: the cursor/codex frontmatter appends a `. Usage…` /
    // `. Invoke with…` clause, which would otherwise read as a double period.
    const description = `${phaseTitle} phase of the ${flowTitle} flow. Invoked by the ${flow.id} orchestrator; not run directly`;
    const sections = [
        target.subSkillFrontmatter({
            flowId: `${flow.id}-${phase.id}`,
            description,
            argumentHint: resolveArgumentHint(flow, target),
            model: flow.frontmatter.model,
            allowedTools: resolveAllowedToolsForPhases(flow, target, [phase.id]),
        }),
        `# ${flowTitle}: ${phaseTitle}`,
    ];
    // The H1 above already names the phase, so the body skips the `## {title}`
    // that the monolith renders (it would duplicate the heading).
    const rendered = renderPhase(phase, target.name, ctx.allModes, ctx.modesPerStep, ctx.reachable, ctx.emitTemplate, false);
    if (rendered.length > 0) {
        sections.push(rendered);
    }
    return finalize(sections, flow, target);
}
function resolveArgumentHint(flow, target) {
    return flow.frontmatter.argumentHint !== undefined
        ? s(flow.frontmatter.argumentHint, target.name, "frontmatter.argumentHint")
        : undefined;
}
function pushIntro(sections, flow, target) {
    if (flow.intro === undefined) {
        return;
    }
    const introStr = s(flow.intro, target.name, "flow.intro");
    if (introStr.length > 0) {
        sections.push(introStr);
    }
}
function pushInvariants(sections, flow) {
    if (flow.invariants.length > 0) {
        const items = flow.invariants.map((inv) => `- ${inv}`).join("\n");
        sections.push(`## Invariants\n\n${items}`);
    }
}
function pushAppendices(sections, flow, target) {
    for (const appendix of flow.appendices) {
        const title = s(appendix.title, target.name, `appendix[${appendix.id}].title`);
        const body = s(appendix.body, target.name, `appendix[${appendix.id}].body`);
        sections.push(`## ${title}\n\n${body}`);
    }
}
/**
 * `allowed-tools` for a split orchestrator: the inline phases' tools plus, when
 * the orchestrator dispatches to sub-skills, the target's skill-invocation tool
 * (Claude's `Skill`) so a scoped command can still hand off. When the inline
 * tools resolve to `undefined` (omit → full access) the orchestrator can
 * already invoke anything, so nothing is appended.
 */
function orchestratorAllowedTools(flow, target, inlinePhaseIds, hasSubSkills) {
    const tools = resolveAllowedToolsForPhases(flow, target, inlinePhaseIds);
    if (tools === undefined) {
        return undefined;
    }
    if (hasSubSkills &&
        target.subSkillTool &&
        !tools.includes(target.subSkillTool)) {
        return [...tools, target.subSkillTool];
    }
    return tools;
}
/**
 * Auto-derive `allowed-tools` from the union of the named phases' tool calls,
 * in registry order. Returns `undefined` (omit the field) when the flow
 * declares no tools and no manual override, or when the selected phases call
 * no tools — an empty `allowed-tools: []` would needlessly handcuff a pure
 * dispatcher orchestrator.
 */
function resolveAllowedToolsForPhases(flow, target, phaseIds) {
    if (Object.keys(flow.tools).length === 0) {
        return flow.frontmatter.allowedTools;
    }
    const ids = collectPhaseToolIds(flow, target, phaseIds);
    if (ids.length === 0) {
        return undefined;
    }
    return ids.map((id) => fullToolName(flow, id, target));
}
const TOOL_PLACEHOLDER_RE = /\{\{tool:([a-zA-Z][a-zA-Z0-9_-]*)\}\}/g;
/** Tool ids called within the named phases, in flow registry order. */
function collectPhaseToolIds(flow, target, phaseIds) {
    const wanted = new Set(phaseIds);
    const called = new Set();
    const scan = (text) => {
        if (text === undefined) {
            return;
        }
        for (const match of text.matchAll(TOOL_PLACEHOLDER_RE)) {
            called.add(match[1]);
        }
    };
    for (const phase of flow.phases) {
        if (!wanted.has(phase.id)) {
            continue;
        }
        if (phase.intro !== undefined) {
            scan(s(phase.intro, target.name, `phase[${phase.id}].intro`));
        }
        for (const step of phase.steps) {
            for (const id of step.toolCalls ?? []) {
                called.add(id);
            }
            scan(s(step.body, target.name, `step[${step.id}].body`));
        }
    }
    return Object.keys(flow.tools).filter((id) => called.has(id));
}
export function renderMarkdown(flow, target) {
    const sections = [];
    const description = s(flow.frontmatter.description, target.name, `flow.frontmatter.description`);
    const argumentHint = flow.frontmatter.argumentHint !== undefined
        ? s(flow.frontmatter.argumentHint, target.name, "frontmatter.argumentHint")
        : undefined;
    sections.push(target.frontmatter({
        flowId: flow.id,
        description,
        argumentHint,
        model: flow.frontmatter.model,
        allowedTools: resolveAllowedTools(flow, target),
    }));
    sections.push(`# ${s(flow.title, target.name, "flow.title")}`);
    if (flow.intro !== undefined) {
        const introStr = s(flow.intro, target.name, "flow.intro");
        if (introStr.length > 0) {
            sections.push(introStr);
        }
    }
    if (Object.keys(flow.commands).length > 0) {
        sections.push(renderCommandsTable(flow));
    }
    // Flow-wide invariants — rules that apply throughout, rendered once near
    // the top so the agent reads them before working through the phases. Each
    // invariant is a one-line bullet under `## Invariants`. Standard
    // placeholders (`{{tool:…}}`, `{{cmd}}`, Mustache sections) work because
    // the substitution pipeline runs over the joined output below.
    if (flow.invariants.length > 0) {
        const items = flow.invariants.map((inv) => `- ${inv}`).join("\n");
        sections.push(`## Invariants\n\n${items}`);
    }
    const allModes = Object.keys(flow.entries);
    const modesPerStep = computeModesPerStep(flow, target.name);
    const reachable = computeReachableSteps(flow, target.name);
    const emitTemplate = flow.emitTemplate !== undefined
        ? s(flow.emitTemplate, target.name, "flow.emitTemplate")
        : undefined;
    for (const phase of flow.phases) {
        const rendered = renderPhase(phase, target.name, allModes, modesPerStep, reachable, emitTemplate);
        // Skip phases with no reachable steps — pushing an empty string here would
        // create a double `\n\n` separator (triple newline) between adjacent
        // visible sections, which shows up as an extra blank line in the output.
        if (rendered.length > 0) {
            sections.push(rendered);
        }
    }
    if (flow.appendices.length > 0) {
        for (const appendix of flow.appendices) {
            const title = s(appendix.title, target.name, `appendix[${appendix.id}].title`);
            const body = s(appendix.body, target.name, `appendix[${appendix.id}].body`);
            sections.push(`## ${title}\n\n${body}`);
        }
    }
    return finalize(sections, flow, target);
}
/**
 * Join rendered sections and run the substitution pipeline. Shared by the
 * monolith renderer and the split orchestrator / sub-skill renderers so every
 * generated file resolves placeholders identically.
 *
 * Order matters: strip non-matching Mustache sections FIRST. Otherwise
 * placeholders inside (e.g.) `{{#codex}}{{step:codex-only-step}}{{/codex}}`
 * get resolved against the wrong target's reachability set and throw.
 */
function finalize(sections, flow, target) {
    const joined = sections.join("\n\n");
    const afterSections = applyMustacheSections(joined, target.name);
    const withFlowData = applyFlowPlaceholders(afterSections, flow, target);
    const substituted = applyTokens(withFlowData, target, flow);
    return ensureTrailingNewline(substituted);
}
/**
 * Substitute flow-level data placeholders.
 *
 * - `{{step:phase-id/step-id}}` — renders the step's number for the current
 *   target (preamble steps render as "0"). Catches stale step references
 *   when the flow's steps are reordered or split.
 * - `{{tool:id}}` — backtick-wrapped per-target tool name from `flow.tools`
 *   (with the target's `mcpPrefix` for MCP tools)
 * - `{{languages:FIELD}}` — table with the named field (and `name`)
 * - `{{languages:F1,F2,...}}` — table with multiple fields (and `name`)
 * - `{{languages:imports}}` is special-cased as a bullet list because it's
 *   the only field rendered as `displayName: value` rather than a table row.
 *
 * Output still contains target tokens (`{{cmd}}`, `{{editor}}`); those are
 * resolved by `applyTemplate` after this pass.
 */
function applyFlowPlaceholders(text, flow, target) {
    let result = text;
    const stepNumbers = computeStepNumbers(flow, target.name);
    result = result.replace(/\{\{step:([a-z0-9-]+\/[a-z0-9-]+)\}\}/g, (_match, ref) => {
        const num = stepNumbers.get(ref);
        if (num === undefined) {
            throw new Error(`Step reference {{step:${ref}}} not found in flow "${flow.id}" for target "${target.name}"`);
        }
        return String(num);
    });
    result = result.replace(/\{\{tool:([a-zA-Z][a-zA-Z0-9_-]*)\}\}/g, (_match, toolId) => `\`${fullToolName(flow, toolId, target)}\``);
    result = result.replace(/\{\{command:([a-zA-Z][a-zA-Z0-9_-]*)\}\}/g, (_match, commandId) => {
        const def = flow.commands[commandId];
        if (def === undefined) {
            throw new Error(`Command "${commandId}" not declared in flow.commands`);
        }
        return `node "{{pluginRoot}}/dist/commands/${def.file}"`;
    });
    result = result.replace(/\{\{languages:([a-z,]+)\}\}/g, (_match, columnsRaw) => renderLanguagesBlock(flow, columnsRaw.split(",")));
    return result;
}
/**
 * Map of `phase-id/step-id` → step number for the given target. Mirrors the
 * numbering logic in `renderPhaseAsHeaders` / `renderPhaseAsList`: preamble
 * steps render as 0, everything else 1-indexed among visible steps.
 */
function computeStepNumbers(flow, target) {
    const result = new Map();
    const reachable = computeReachableSteps(flow, target);
    for (const phase of flow.phases) {
        let counter = 1;
        for (const step of phase.steps) {
            if (!reachable.has(`${phase.id}/${step.id}`)) {
                continue;
            }
            const num = step.preamble === true ? 0 : counter;
            if (step.preamble !== true) {
                counter += 1;
            }
            result.set(`${phase.id}/${step.id}`, num);
        }
    }
    return result;
}
/**
 * Auto-derive `allowed-tools` from the union of every step's `toolCalls`,
 * using registry order. When the flow declares no tools the explicit
 * `frontmatter.allowedTools` is used as-is.
 */
function resolveAllowedTools(flow, target) {
    if (Object.keys(flow.tools).length === 0) {
        return flow.frontmatter.allowedTools;
    }
    const calledIds = collectCalledToolIds(flow);
    return calledIds.map((id) => fullToolName(flow, id, target));
}
function fullToolName(flow, toolId, target) {
    const def = flow.tools[toolId];
    if (def === undefined) {
        throw new Error(`Tool "${toolId}" not declared in flow.tools`);
    }
    const name = resolvePerTarget(def.name, target.name);
    if (name === undefined) {
        throw new Error(`Tool "${toolId}" has no name for target "${target.name}" (and no default)`);
    }
    return def.kind === "mcp" ? `${target.mcpPrefix}${name}` : name;
}
function renderCommandsTable(flow) {
    const rows = [];
    for (const def of Object.values(flow.commands)) {
        const sig = def.args !== undefined ? `${def.file} ${def.args}` : def.file;
        rows.push(`| \`${sig}\` | ${def.description} |`);
    }
    return [
        "**CLI commands** available via Bash (all paths relative to `{{pluginRoot}}/dist/commands/`):",
        "",
        "| Command | Description |",
        "|---------|-------------|",
        ...rows,
    ].join("\n");
}
function renderLanguagesBlock(flow, columns) {
    if (flow.languages.length === 0) {
        return "";
    }
    // `imports` renders as a bullet list using displayName, the only field
    // styled this way today. Everything else is a table.
    if (columns.length === 1 && columns[0] === "imports") {
        const lines = [];
        for (const lang of flow.languages) {
            if (lang.imports !== undefined) {
                lines.push(`  - ${lang.displayName}: ${lang.imports}`);
            }
        }
        return lines.join("\n");
    }
    const headers = ["Language", ...columns.map((c) => columnHeader(flow, c))];
    const lines = [
        `| ${headers.join(" | ")} |`,
        `|${headers.map(() => "---").join("|")}|`,
    ];
    for (const lang of flow.languages) {
        const cells = columns.map((col) => readLanguageField(lang, col));
        if (cells.every((c) => c !== undefined)) {
            lines.push(`| ${lang.name} | ${cells.join(" | ")} |`);
        }
    }
    return lines.join("\n");
}
function columnHeader(flow, column) {
    return flow.languageColumnHeaders[column] ?? titleCase(column);
}
function titleCase(s) {
    return s.charAt(0).toUpperCase() + s.slice(1);
}
function readLanguageField(lang, column) {
    switch (column) {
        case "imports":
            return lang.imports;
        case "detection":
            return lang.detection;
        case "outdated":
            return lang.outdated;
        case "update":
            return lang.update;
        default:
            throw new Error(`Unknown languages field "${column}"`);
    }
}
function renderPhase(phase, target, allModes, modesPerStep, reachable, emitTemplate, 
// List-style phases lead with a `## {title}` heading. Sub-skills pass false
// because they already carry the phase name in their own H1. Headers-style
// phases never emit the title, so the flag is a no-op there.
includePhaseTitle = true) {
    const visibleSteps = phase.steps
        .filter((step) => reachable.has(`${phase.id}/${step.id}`))
        .map((step) => ({
        step,
        stepKey: `${phase.id}/${step.id}`,
    }));
    if (visibleSteps.length === 0) {
        return "";
    }
    if (phase.stepStyle === "headers") {
        return renderPhaseAsHeaders(phase, visibleSteps, allModes, modesPerStep, target, emitTemplate);
    }
    return renderPhaseAsList(phase, visibleSteps, allModes, modesPerStep, target, emitTemplate, includePhaseTitle);
}
function modesForStep(stepKey, modesPerStep, allModes) {
    return modesPerStep.get(stepKey) ?? new Set(allModes);
}
/**
 * Format a positive mode-gate sentence: "Run only when mode is `X`." for one
 * mode, "Run only when mode is `X` or `Y`." for two, "Run only when mode is
 * `X`, `Y` or `Z`." for three or more.
 */
function formatPositiveGate(inModes) {
    const phrases = inModes.map((m) => `\`${m}\``);
    const list = phrases.length === 1
        ? phrases[0]
        : `${phrases.slice(0, -1).join(", ")} or ${phrases[phrases.length - 1]}`;
    return `**Run only when mode is ${list}.**`;
}
/**
 * Build mode-gating prose for a phase, in the positive form: "Run only when
 * mode is `X` [or `Y`]". Two output channels:
 *
 *   - `phaseGate`: a single line emitted just under the phase heading when
 *     every visible step in the phase is reachable in the same proper subset
 *     of modes. One line gates the whole phase; no per-step prose needed.
 *
 *   - `perStepGates`: per-step gates emitted in the step body when the
 *     phase has mixed reachability (different steps reachable in different
 *     mode subsets — e.g. update.ts where step 2 runs in `all`/`plugin` but
 *     step 3 runs in `all`/`sdk`). Each affected step gets its own positive
 *     gate; gates do not cascade or inherit.
 *
 * Steps reachable in every declared mode get no gate. Phases where every
 * visible step is reachable in every mode also get no `phaseGate`.
 *
 * The positive form is unambiguous on scope (it states when the step runs,
 * full stop) and is usually shorter than the negative "Skip if X" form when
 * there are more excluded modes than included modes.
 */
function precomputePhaseAndStepGates(stepKeys, modesPerStep, allModes) {
    if (stepKeys.length === 0) {
        return { phaseGate: null, perStepGates: new Map() };
    }
    const stepModeSets = stepKeys.map((key) => modesForStep(key, modesPerStep, allModes));
    const allModesCount = allModes.length;
    const firstSet = stepModeSets[0];
    const uniform = stepModeSets.every((s) => s.size === firstSet.size && [...s].every((mode) => firstSet.has(mode)));
    if (uniform) {
        if (firstSet.size === allModesCount) {
            // Phase reachable in every mode: no gate at all.
            return { phaseGate: null, perStepGates: new Map() };
        }
        const inModes = allModes.filter((m) => firstSet.has(m));
        return {
            phaseGate: formatPositiveGate(inModes),
            perStepGates: new Map(),
        };
    }
    // Mixed reachability: each affected step gets its own positive gate.
    const perStepGates = new Map();
    for (let i = 0; i < stepKeys.length; i++) {
        const set = stepModeSets[i];
        if (set.size === allModesCount) {
            continue;
        }
        const inModes = allModes.filter((m) => set.has(m));
        perStepGates.set(stepKeys[i], formatPositiveGate(inModes));
    }
    return { phaseGate: null, perStepGates };
}
function renderEmit(step, template) {
    if (step.emit === undefined || template === undefined) {
        return null;
    }
    let displayName;
    let action;
    if (typeof step.emit === "string") {
        displayName = step.emit;
        action = "started";
    }
    else {
        displayName = step.emit.displayName;
        action = step.emit.action;
    }
    return template
        .replace(/\{action\}/g, action)
        .replace(/\{displayName\}/g, displayName);
}
function bodyWithModeGate(step, stepKey, perStepGates, target, emitTemplate) {
    const withBranches = expandBranchesPlaceholders(step, target);
    const parts = [];
    const gate = perStepGates.get(stepKey);
    if (gate !== undefined) {
        parts.push(gate);
    }
    const emitted = renderEmit(step, emitTemplate);
    if (emitted !== null) {
        parts.push(emitted);
    }
    parts.push(withBranches);
    return parts.join("\n\n");
}
/**
 * Replace branch-related placeholders in a branch step's body:
 *
 * - `{{branches}}` / `{{branches:prompt-name}}` — formatted AskUserQuestion
 *   option block. Only branches with an `option` field are rendered.
 * - `{{whens}}` — formatted runtime-branch routing block. Only branches
 *   with a `when` field (no `option`) are rendered. Each line shows the
 *   runtime condition and the action (the branch's `description`, falling
 *   back to "stop the flow" for `next: null` or "continue" otherwise).
 *
 * Runtime `when` and user-facing `option` branches are intentionally
 * separate placeholders — the agent presents `{{branches}}` to the user
 * via AskUserQuestion, and reads `{{whens}}` to itself when routing on
 * observed state.
 */
function expandBranchesPlaceholders(step, target) {
    const body = s(step.body, target, `step[${step.id}].body`);
    if (step.kind !== "branch") {
        return body;
    }
    return body
        .replace(/\{\{branches(?::([a-z0-9-]+))?\}\}/g, (_match, promptFilter) => {
        const lines = [];
        for (const branch of step.branches) {
            if (branch.option === undefined) {
                continue;
            }
            if (promptFilter !== undefined && branch.prompt !== promptFilter) {
                continue;
            }
            const { letter, label, description } = branch.option;
            const recommended = branch.recommended === true ? " *(recommended)*" : "";
            const desc = description !== undefined ? ` — ${description}` : "";
            lines.push(`> ${letter}) **${label}**${desc}${recommended}`);
        }
        return lines.join("\n");
    })
        .replace(/\{\{whens\}\}/g, () => {
        const lines = [];
        for (const branch of step.branches) {
            if (branch.when === undefined || branch.option !== undefined) {
                continue;
            }
            const action = branch.description ??
                (branch.next === null ? "stop the flow" : "continue");
            lines.push(`- **${branch.when}** — ${action}`);
        }
        return lines.join("\n");
    });
}
function renderPhaseAsHeaders(phase, visibleSteps, allModes, modesPerStep, target, emitTemplate) {
    const blocks = [];
    const stepKeys = visibleSteps.map((v) => v.stepKey);
    const { phaseGate, perStepGates } = precomputePhaseAndStepGates(stepKeys, modesPerStep, allModes);
    if (phaseGate !== null) {
        blocks.push(phaseGate);
    }
    if (phase.intro !== undefined) {
        const intro = s(phase.intro, target, `phase[${phase.id}].intro`);
        if (intro.length > 0) {
            blocks.push(intro);
        }
    }
    let counter = 1;
    for (const { step, stepKey } of visibleSteps) {
        const number = step.preamble === true ? 0 : counter;
        if (step.preamble !== true) {
            counter += 1;
        }
        const title = s(step.title, target, `step[${step.id}].title`);
        const body = bodyWithModeGate(step, stepKey, perStepGates, target, emitTemplate);
        blocks.push(`## ${number}. ${title}\n\n${body}`);
    }
    if (phase.outro !== undefined) {
        const outro = s(phase.outro, target, `phase[${phase.id}].outro`);
        if (outro.length > 0) {
            blocks.push(outro);
        }
    }
    return blocks.join("\n\n");
}
function renderPhaseAsList(phase, visibleSteps, allModes, modesPerStep, target, emitTemplate, includePhaseTitle) {
    const blocks = [];
    if (includePhaseTitle) {
        blocks.push(`## ${s(phase.title, target, `phase[${phase.id}].title`)}`);
    }
    const stepKeys = visibleSteps.map((v) => v.stepKey);
    const { phaseGate, perStepGates } = precomputePhaseAndStepGates(stepKeys, modesPerStep, allModes);
    if (phaseGate !== null) {
        blocks.push(phaseGate);
    }
    if (phase.intro !== undefined) {
        const intro = s(phase.intro, target, `phase[${phase.id}].intro`);
        if (intro.length > 0) {
            blocks.push(intro);
        }
    }
    const items = [];
    let counter = 1;
    for (const { step, stepKey } of visibleSteps) {
        const number = step.preamble === true ? 0 : counter;
        if (step.preamble !== true) {
            counter += 1;
        }
        const body = bodyWithModeGate(step, stepKey, perStepGates, target, emitTemplate);
        items.push(`${number}. ${indentBody(body)}`);
    }
    blocks.push(items.join("\n"));
    if (phase.outro !== undefined) {
        const outro = s(phase.outro, target, `phase[${phase.id}].outro`);
        if (outro.length > 0) {
            blocks.push(outro);
        }
    }
    return blocks.join("\n\n");
}
function indentBody(body) {
    const lines = body.split("\n");
    return lines
        .map((line, i) => (i === 0 ? line : line.length > 0 ? `   ${line}` : line))
        .join("\n");
}
/**
 * Strip Mustache conditional sections (`{{#target}}...{{/target}}` and the
 * inverse `{{^target}}...{{/target}}`) leaving only content that matches the
 * current target. Exported so the dev-server payload builder can apply the
 * same stripping to step / appendix bodies before they reach the UI; without
 * it the StepDetail "Body" pane would still show `{{#codex}}…{{/codex}}`
 * blocks even when target=claude.
 *
 * In `renderMarkdown` this runs BEFORE flow-data placeholder resolution so
 * that placeholders inside non-matching sections (which would throw if
 * evaluated against the wrong target's reachability set) never get evaluated.
 */
export function applyMustacheSections(text, targetName) {
    const knownSectionNames = new Set(MUSTACHE_SECTION_NAMES);
    const sectionRegex = /\{\{([#^])([a-z]+)\}\}([\s\S]*?)\{\{\/\2\}\}/g;
    return text.replace(sectionRegex, (_match, kind, name, inner) => {
        if (!knownSectionNames.has(name)) {
            throw new Error(`Unknown target name "${name}" in Mustache section. Valid names: ${[...knownSectionNames].join(", ")}.`);
        }
        const isMatch = name === targetName;
        const include = kind === "#" ? isMatch : !isMatch;
        return include ? inner : "";
    });
}
/**
 * Resolve target tokens (cmd, editor, etc.) and flow variables, then sanity-
 * check that no `{{...}}` placeholder slipped through. Runs AFTER Mustache
 * sections and flow-data placeholders.
 */
function applyTokens(text, target, flow) {
    // Build a unified substitution pool: target tokens + flow variables.
    // Variables that collide with target tokens are caught at parse time by
    // the schema's superRefine, so this loop never throws in practice.
    const pool = { ...target.tokens };
    for (const [name, value] of Object.entries(flow.variables)) {
        pool[name] = s(value, target.name, `flow.variables.${name}`);
    }
    const tokenRegex = /\{\{([a-zA-Z][a-zA-Z0-9_]*)\}\}/g;
    const result = text.replace(tokenRegex, (_match, key) => {
        if (!Object.hasOwn(pool, key)) {
            throw new Error(`Unresolved placeholder {{${key}}} — not a target token or flow variable. Define it in flow.variables, fix the typo, or remove the placeholder.`);
        }
        return pool[key];
    });
    // Final sanity scan — any leftover `{{...}}` is a malformed or unknown
    // namespaced placeholder that no earlier pass recognized.
    const leftovers = result.match(/\{\{[^}]+\}\}/g);
    if (leftovers) {
        throw new Error(`Unresolved placeholder(s): ${leftovers.join(", ")}. Check spelling and namespace (step:, tool:, languages:, branches, branches:).`);
    }
    return result;
}
function ensureTrailingNewline(text) {
    return text.endsWith("\n") ? text : `${text}\n`;
}
