// bitfab-flow/dist/targets.js
const renderClaudeFrontmatter = ({ description, argumentHint, model, allowedTools, }) => {
    const lines = ["---", `description: ${description}`];
    if (argumentHint !== undefined) {
        // JSON.stringify produces a double-quoted YAML scalar so values containing
        // `[...]` aren't parsed as flow sequences (e.g. "[all|dataset] [<key>]").
        lines.push(`argument-hint: ${JSON.stringify(argumentHint)}`);
    }
    if (model !== undefined) {
        lines.push(`model: ${model}`);
    }
    if (allowedTools !== undefined) {
        const formatted = `[${allowedTools.map((t) => JSON.stringify(t)).join(", ")}]`;
        lines.push(`allowed-tools: ${formatted}`);
    }
    lines.push("---");
    return lines.join("\n");
};
/**
 * Claude Agent Skill frontmatter (for split phase sub-skills). Unlike a Claude
 * command, a skill needs a `name:` (matching its directory). Skills are
 * model-invoked by `description`, so no `argument-hint`.
 *
 * `user-invocable: false` hides the sub-skill from the `/` menu (it's only
 * meant to be dispatched by the orchestrator) while leaving it invocable by the
 * model via the Skill tool. (`disable-model-invocation` would do the opposite,
 * and is also ignored for plugin skills per anthropics/claude-code#22345.)
 */
const renderClaudeSkillFrontmatter = ({ flowId, description, model, allowedTools, }) => {
    const lines = [
        "---",
        `name: ${flowId}`,
        `description: ${description}`,
        "user-invocable: false",
    ];
    if (model !== undefined) {
        lines.push(`model: ${model}`);
    }
    if (allowedTools !== undefined) {
        const formatted = `[${allowedTools.map((t) => JSON.stringify(t)).join(", ")}]`;
        lines.push(`allowed-tools: ${formatted}`);
    }
    lines.push("---");
    return lines.join("\n");
};
const renderCursorFrontmatter = ({ flowId, description, argumentHint, }) => {
    const usage = argumentHint !== undefined
        ? `. Usage: /bitfab-${flowId} ${argumentHint}`
        : "";
    return [
        "---",
        `name: bitfab-${flowId}`,
        `description: "${description}${usage}"`,
        "---",
    ].join("\n");
};
const renderCodexFrontmatter = ({ flowId, description, argumentHint, }) => {
    const usage = argumentHint !== undefined
        ? `. Invoke with $bitfab:${flowId} ${argumentHint}.`
        : "";
    return [
        "---",
        `name: ${flowId}`,
        `description: "${description}${usage}"`,
        "---",
    ].join("\n");
};
const renderCursorDevFrontmatter = ({ flowId, description, argumentHint, }) => {
    const usage = argumentHint !== undefined
        ? `. Usage: /bitfab-dev-${flowId} ${argumentHint}`
        : "";
    return [
        "---",
        `name: bitfab-dev-${flowId}`,
        `description: "${description}${usage}"`,
        "---",
    ].join("\n");
};
const renderCodexDevFrontmatter = ({ flowId, description, argumentHint, }) => {
    const usage = argumentHint !== undefined
        ? `. Invoke with $bitfab-dev:${flowId} ${argumentHint}.`
        : "";
    // Codex scopes a skill's handle as `<plugin-name>:<skill-name>`. The plugin is
    // named `bitfab-dev`, so the skill name is the bare flowId (e.g. `sync`),
    // yielding `bitfab-dev:sync`. Prefixing the name with `dev-` here would double
    // it to `bitfab-dev:dev-sync`. (Cursor needs the full `bitfab-dev-` prefix
    // because its skills share one flat namespace with no plugin scoping.)
    return [
        "---",
        `name: ${flowId}`,
        `description: "${description}${usage}"`,
        "---",
    ].join("\n");
};
export const TARGETS = {
    claude: {
        name: "claude",
        tokens: {
            cmd: "/bitfab:",
            devCmd: "/bitfab-dev:",
            editor: "Claude Code",
            // biome-ignore lint/suspicious/noTemplateCurlyInString: shell variable reference rendered into skill markdown, not a JS template
            pluginRoot: "${CLAUDE_PLUGIN_ROOT}",
            askUser: "with `AskUserQuestion` ",
            askVerb: "use `AskUserQuestion`",
            askVerbCap: "Use `AskUserQuestion`",
            askNoun: "AskUserQuestion",
        },
        mcpPrefix: "mcp__plugin_bitfab_Bitfab__",
        argsToken: "$ARGUMENTS",
        frontmatter: renderClaudeFrontmatter,
        subSkillFrontmatter: renderClaudeSkillFrontmatter,
        subSkillTool: "Skill",
    },
    cursor: {
        name: "cursor",
        tokens: {
            cmd: "/bitfab-",
            devCmd: "/bitfab-dev-",
            editor: "Cursor",
            // biome-ignore lint/suspicious/noTemplateCurlyInString: shell variable reference rendered into skill markdown, not a JS template
            pluginRoot: "${CURSOR_PLUGIN_ROOT:-${CLAUDE_PLUGIN_ROOT}}",
            askUser: "with `AskUserQuestion` ",
            askVerb: "use `AskUserQuestion`",
            askVerbCap: "Use `AskUserQuestion`",
            askNoun: "AskUserQuestion",
        },
        mcpPrefix: "mcp__Bitfab__",
        argsToken: "$ARGUMENTS",
        frontmatter: renderCursorFrontmatter,
        subSkillFrontmatter: renderCursorFrontmatter,
    },
    codex: {
        name: "codex",
        tokens: {
            cmd: "$bitfab:",
            devCmd: "$bitfab-dev:",
            editor: "Codex",
            // biome-ignore lint/suspicious/noTemplateCurlyInString: shell variable reference rendered into skill markdown, not a JS template
            pluginRoot: "${BITFAB_PLUGIN_DIR}",
            askUser: "",
            askVerb: "ask the user",
            askVerbCap: "Ask the user",
            askNoun: "choice prompt",
        },
        mcpPrefix: "mcp__Bitfab__",
        argsToken: "$ARGUMENTS",
        frontmatter: renderCodexFrontmatter,
        subSkillFrontmatter: renderCodexFrontmatter,
        // `bitfab` plugin scope: a sub-skill `sync-local` is invoked as
        // `$bitfab:sync-local` (matches renderCodexFrontmatter's `$bitfab:` usage).
        dispatchRef: (skillName) => `\`$bitfab:${skillName}\``,
    },
};
export const DEV_TARGETS = {
    claude: {
        ...TARGETS.claude,
        frontmatter: renderClaudeFrontmatter,
        subSkillFrontmatter: renderClaudeSkillFrontmatter,
    },
    cursor: {
        ...TARGETS.cursor,
        frontmatter: renderCursorDevFrontmatter,
        subSkillFrontmatter: renderCursorDevFrontmatter,
    },
    codex: {
        ...TARGETS.codex,
        frontmatter: renderCodexDevFrontmatter,
        subSkillFrontmatter: renderCodexDevFrontmatter,
        // `bitfab-dev` plugin scope: a sub-skill `sync-local` is invoked as
        // `$bitfab-dev:sync-local` (matches renderCodexDevFrontmatter's `$bitfab-dev:`
        // usage). Overrides the `$bitfab:` inherited from TARGETS.codex.
        dispatchRef: (skillName) => `\`$bitfab-dev:${skillName}\``,
    },
};
