// bitfab-flow/dist/targets.js
const renderClaudeFrontmatter = ({ description, argumentHint, model, allowedTools, }) => {
    const lines = ["---", `description: ${description}`];
    if (argumentHint !== undefined) {
        // JSON.stringify produces a double-quoted YAML scalar so values containing
        // `[...]` aren't parsed as flow sequences (e.g. "[all|dataset] [<key>]").
        lines.push(`argument-hint: ${JSON.stringify(argumentHint)}`);
    }
    if (model !== undefined) {
        lines.push(`model: ${model}`);
    }
    if (allowedTools !== undefined) {
        const formatted = `[${allowedTools.map((t) => JSON.stringify(t)).join(", ")}]`;
        lines.push(`allowed-tools: ${formatted}`);
    }
    lines.push("---");
    return lines.join("\n");
};
const renderCursorFrontmatter = ({ flowId, description, argumentHint, }) => {
    const usage = argumentHint !== undefined
        ? `. Usage: /bitfab-${flowId} ${argumentHint}`
        : "";
    return [
        "---",
        `name: bitfab-${flowId}`,
        `description: "${description}${usage}"`,
        "---",
    ].join("\n");
};
const renderCodexFrontmatter = ({ flowId, description, argumentHint, }) => {
    const usage = argumentHint !== undefined
        ? `. Invoke with $bitfab:${flowId} ${argumentHint}.`
        : "";
    return [
        "---",
        `name: ${flowId}`,
        `description: "${description}${usage}"`,
        "---",
    ].join("\n");
};
const renderCursorDevFrontmatter = ({ flowId, description, argumentHint, }) => {
    const usage = argumentHint !== undefined
        ? `. Usage: /bitfab-dev-${flowId} ${argumentHint}`
        : "";
    return [
        "---",
        `name: bitfab-dev-${flowId}`,
        `description: "${description}${usage}"`,
        "---",
    ].join("\n");
};
const renderCodexDevFrontmatter = ({ flowId, description, argumentHint, }) => {
    const usage = argumentHint !== undefined
        ? `. Invoke with $bitfab-dev:${flowId} ${argumentHint}.`
        : "";
    return [
        "---",
        `name: dev-${flowId}`,
        `description: "${description}${usage}"`,
        "---",
    ].join("\n");
};
export const TARGETS = {
    claude: {
        name: "claude",
        tokens: {
            cmd: "/bitfab:",
            devCmd: "/bitfab-dev:",
            editor: "Claude Code",
            // biome-ignore lint/suspicious/noTemplateCurlyInString: shell variable reference rendered into skill markdown, not a JS template
            pluginRoot: "${CLAUDE_PLUGIN_ROOT}",
            askUser: "with `AskUserQuestion` ",
            askVerb: "use `AskUserQuestion`",
            askVerbCap: "Use `AskUserQuestion`",
            askNoun: "AskUserQuestion",
        },
        mcpPrefix: "mcp__plugin_bitfab_Bitfab__",
        frontmatter: renderClaudeFrontmatter,
    },
    cursor: {
        name: "cursor",
        tokens: {
            cmd: "/bitfab-",
            devCmd: "/bitfab-dev-",
            editor: "Cursor",
            // biome-ignore lint/suspicious/noTemplateCurlyInString: shell variable reference rendered into skill markdown, not a JS template
            pluginRoot: "${CURSOR_PLUGIN_ROOT:-${CLAUDE_PLUGIN_ROOT}}",
            askUser: "with `AskUserQuestion` ",
            askVerb: "use `AskUserQuestion`",
            askVerbCap: "Use `AskUserQuestion`",
            askNoun: "AskUserQuestion",
        },
        mcpPrefix: "mcp__Bitfab__",
        frontmatter: renderCursorFrontmatter,
    },
    codex: {
        name: "codex",
        tokens: {
            cmd: "$bitfab:",
            devCmd: "$bitfab-dev:",
            editor: "Codex",
            // biome-ignore lint/suspicious/noTemplateCurlyInString: shell variable reference rendered into skill markdown, not a JS template
            pluginRoot: "${BITFAB_PLUGIN_DIR}",
            askUser: "",
            askVerb: "ask the user",
            askVerbCap: "Ask the user",
            askNoun: "choice prompt",
        },
        mcpPrefix: "mcp__Bitfab__",
        frontmatter: renderCodexFrontmatter,
    },
};
export const DEV_TARGETS = {
    claude: {
        ...TARGETS.claude,
        frontmatter: renderClaudeFrontmatter,
    },
    cursor: {
        ...TARGETS.cursor,
        frontmatter: renderCursorDevFrontmatter,
    },
    codex: {
        ...TARGETS.codex,
        frontmatter: renderCodexDevFrontmatter,
    },
};
