// bitfab-plugin-lib/dist/daemon/client.js
import net from "node:net";
import { DaemonOpenTimeoutError } from "../errors.js";
import { ensureDaemon, reapAllDaemons } from "./ensure.js";
import { COMMAND_TIMEOUT_MS, isPush, isResponse, OPEN_TIMEOUT_MS, SINGLETON_DAEMON_PATHS, serialize, } from "./protocol.js";
export class DaemonClient {
    socket = null;
    buffer = "";
    responseQueue = [];
    eventHandler = null;
    disconnectHandler = null;
    paths;
    socketPath;
    openTimeoutMs;
    commandTimeoutMs;
    // The single in-flight wedged-daemon reap, shared by every timeout path.
    // Sharing it is load-bearing: `open`'s retry must AWAIT the same reap the
    // command timeout fire-and-forgot, or a trailing reap's file cleanup could
    // finish after the respawned daemon bound its socket and wrote its pid
    // file, unlinking the new daemon's files and orphaning it.
    reapInFlight = null;
    constructor(paths = SINGLETON_DAEMON_PATHS, timeouts = {}) {
        this.paths = paths;
        this.socketPath = paths.socketPath;
        this.openTimeoutMs = timeouts.openTimeoutMs ?? OPEN_TIMEOUT_MS;
        this.commandTimeoutMs = timeouts.commandTimeoutMs ?? COMMAND_TIMEOUT_MS;
    }
    async connect() {
        await ensureDaemon(this.paths);
        return new Promise((resolve, reject) => {
            const socket = net.createConnection(this.socketPath, () => {
                this.socket = socket;
                resolve();
            });
            socket.on("data", (chunk) => {
                this.buffer += chunk.toString();
                this.drain();
            });
            socket.on("error", (err) => {
                if (!this.socket) {
                    reject(err);
                    return;
                }
                this.socket = null;
                this.disconnectHandler?.(err);
            });
            socket.on("close", () => {
                // Only an unexpected drop: a deliberate destroy() nulls the socket
                // first, so `this.socket` is already null and we stay quiet.
                const wasConnected = this.socket !== null;
                this.socket = null;
                if (wasConnected) {
                    this.disconnectHandler?.(new Error("daemon connection closed"));
                }
            });
        });
    }
    drain() {
        let idx = this.buffer.indexOf("\n");
        while (idx !== -1) {
            const line = this.buffer.slice(0, idx).trim();
            this.buffer = this.buffer.slice(idx + 1);
            if (line) {
                try {
                    const msg = JSON.parse(line);
                    if (isResponse(msg) && this.responseQueue.length > 0) {
                        const resolver = this.responseQueue.shift();
                        resolver(msg);
                    }
                    else if (isPush(msg) && this.eventHandler) {
                        this.eventHandler(msg);
                    }
                }
                catch {
                    // malformed line, skip
                }
            }
            idx = this.buffer.indexOf("\n");
        }
    }
    // Every command runs under a timeout so a wedged daemon can never hang a
    // caller indefinitely: a never-resolving promise is not a rejection, so it
    // slips straight past callers' .catch() guards. `open` passes its own tighter
    // timeout + typed error to drive the reap-and-retry path; everything else
    // gets the generous default ceiling.
    sendCommand(cmd, opts = {}) {
        if (!this.socket) {
            return Promise.reject(new Error("not connected"));
        }
        const timeoutMs = opts.timeoutMs ?? this.commandTimeoutMs;
        const cmdName = cmd.cmd ?? "command";
        return new Promise((resolve, reject) => {
            let settled = false;
            const timer = setTimeout(() => {
                if (settled) {
                    return;
                }
                settled = true;
                // A timed-out command leaves the FIFO pairing indeterminate: if the
                // daemon answers late, its reply would be shifted onto the NEXT
                // command's resolver and every subsequent call would read the wrong
                // response. The connection is no longer trustworthy, so tear it down:
                // fail any other in-flight commands explicitly (their replies can
                // never be paired safely), then destroy the socket. Callers reconnect
                // (open's reap-and-retry path) or fall back to the direct channel.
                const pending = this.responseQueue;
                this.responseQueue = [];
                for (const other of pending) {
                    if (other !== resolver) {
                        other({
                            ok: false,
                            error: `daemon connection reset after ${cmdName} timed out`,
                        });
                    }
                }
                this.destroy();
                // The daemon is wedged (it accepted the socket but never answered), so
                // reap it: leaving it running would create dual ownership when the
                // caller falls back to the direct channel - the daemon keeps polling
                // the same session (duplicate durable-log events) and keeps answering
                // ping so later ensures reuse it. Fire-and-forget here; `open`'s retry
                // path awaits the SAME in-flight reap (see reapWedgedDaemon) before
                // respawning, and a session the reap tears down is rehydrated from
                // disk on the next daemon start.
                void this.reapWedgedDaemon();
                reject(opts.onTimeout?.() ??
                    new Error(`daemon did not respond to ${cmdName} within ${timeoutMs}ms`));
            }, timeoutMs);
            const resolver = (msg) => {
                if (settled) {
                    return;
                }
                settled = true;
                clearTimeout(timer);
                resolve(msg);
            };
            this.responseQueue.push(resolver);
            try {
                this.socket.write(serialize(cmd));
            }
            catch (err) {
                settled = true;
                clearTimeout(timer);
                const idx = this.responseQueue.indexOf(resolver);
                if (idx !== -1) {
                    this.responseQueue.splice(idx, 1);
                }
                reject(err);
            }
        });
    }
    async open(opts) {
        try {
            return await this.openOnce(opts);
        }
        catch (err) {
            if (!(err instanceof DaemonOpenTimeoutError)) {
                throw err;
            }
            // A wedged daemon answered ensureDaemon's ping (so connect() succeeded)
            // but never acked the open. Await the reap (joining the one the command
            // timeout already started, so no reap can outlive the respawn), then
            // respawn via connect() and retry the open exactly once.
            this.destroy();
            await this.reapWedgedDaemon();
            await this.connect();
            return await this.openOnce(opts);
        }
    }
    async openOnce(opts) {
        const cmd = { cmd: "open", ...opts };
        const res = await this.sendCommand(cmd, {
            timeoutMs: this.openTimeoutMs,
            onTimeout: () => new DaemonOpenTimeoutError(),
        });
        if (!res.ok) {
            throw new Error(res.error);
        }
        return res;
    }
    // Start (or join) the wedged-daemon reap. All timeout paths funnel through
    // this so at most one reap runs at a time and `open`'s retry can await the
    // in-flight one instead of racing it - a reap that finished AFTER the retry
    // respawned would unlink the new daemon's socket/pid files and orphan it.
    reapWedgedDaemon() {
        if (!this.reapInFlight) {
            this.reapInFlight = reapAllDaemons(this.paths)
                .catch(() => { })
                .finally(() => {
                this.reapInFlight = null;
            });
        }
        return this.reapInFlight;
    }
    async navigate(key, navPath) {
        const res = await this.sendCommand({ cmd: "navigate", key, path: navPath });
        if (!res.ok) {
            throw new Error(res.error);
        }
        return res;
    }
    async close(key) {
        const res = await this.sendCommand({ cmd: "close", key });
        if (!res.ok) {
            throw new Error(res.error);
        }
    }
    /**
     * Forget a session from the daemon's in-memory map (the recovery path after
     * a `not-responding` gate). Matched daemon-side on the studio sessionId, so a
     * session re-bound under the same key is never touched. Returns how many
     * entries were dropped (0 or 1).
     */
    async clear(sessionId) {
        const res = await this.sendCommand({ cmd: "clear", sessionId });
        if (!res.ok) {
            throw new Error(res.error);
        }
        return res.cleared;
    }
    async detach(key) {
        const res = await this.sendCommand({ cmd: "detach", key });
        if (!res.ok) {
            throw new Error(res.error);
        }
    }
    async status() {
        const res = await this.sendCommand({ cmd: "status" });
        if (!res.ok) {
            throw new Error(res.error);
        }
        return res;
    }
    async ping() {
        const res = await this.sendCommand({ cmd: "ping" });
        if (!res.ok) {
            throw new Error(res.error);
        }
    }
    onEvent(handler) {
        this.eventHandler = handler;
    }
    /** Fires once when the daemon connection drops unexpectedly (not on destroy). */
    onDisconnect(handler) {
        this.disconnectHandler = handler;
    }
    destroy() {
        this.eventHandler = null;
        this.disconnectHandler = null;
        this.responseQueue = [];
        if (this.socket) {
            this.socket.destroy();
            this.socket = null;
        }
    }
}
export { resolveAgentSessionKey } from "../agentSessionKey.js";
