// bitfab-plugin-lib/dist/daemon/client.js
import net from "node:net";
import { ensureDaemon } from "./ensure.js";
import { isPush, isResponse, SOCKET_PATH, serialize } from "./protocol.js";
export class DaemonClient {
    socket = null;
    buffer = "";
    responseQueue = [];
    eventHandler = null;
    disconnectHandler = null;
    socketPath;
    constructor(socketPath) {
        this.socketPath = socketPath ?? SOCKET_PATH;
    }
    async connect() {
        await ensureDaemon(this.socketPath);
        return new Promise((resolve, reject) => {
            const socket = net.createConnection(this.socketPath, () => {
                this.socket = socket;
                resolve();
            });
            socket.on("data", (chunk) => {
                this.buffer += chunk.toString();
                this.drain();
            });
            socket.on("error", (err) => {
                if (!this.socket) {
                    reject(err);
                    return;
                }
                this.socket = null;
                this.disconnectHandler?.(err);
            });
            socket.on("close", () => {
                // Only an unexpected drop: a deliberate destroy() nulls the socket
                // first, so `this.socket` is already null and we stay quiet.
                const wasConnected = this.socket !== null;
                this.socket = null;
                if (wasConnected) {
                    this.disconnectHandler?.(new Error("daemon connection closed"));
                }
            });
        });
    }
    drain() {
        let idx = this.buffer.indexOf("\n");
        while (idx !== -1) {
            const line = this.buffer.slice(0, idx).trim();
            this.buffer = this.buffer.slice(idx + 1);
            if (line) {
                try {
                    const msg = JSON.parse(line);
                    if (isResponse(msg) && this.responseQueue.length > 0) {
                        const resolver = this.responseQueue.shift();
                        resolver(msg);
                    }
                    else if (isPush(msg) && this.eventHandler) {
                        this.eventHandler(msg);
                    }
                }
                catch {
                    // malformed line, skip
                }
            }
            idx = this.buffer.indexOf("\n");
        }
    }
    sendCommand(cmd) {
        if (!this.socket) {
            return Promise.reject(new Error("not connected"));
        }
        return new Promise((resolve, reject) => {
            this.responseQueue.push(resolve);
            try {
                this.socket.write(serialize(cmd));
            }
            catch (err) {
                this.responseQueue.pop();
                reject(err);
            }
        });
    }
    async open(opts) {
        const cmd = { cmd: "open", ...opts };
        const res = await this.sendCommand(cmd);
        if (!res.ok) {
            throw new Error(res.error);
        }
        return res;
    }
    async navigate(key, navPath) {
        const res = await this.sendCommand({ cmd: "navigate", key, path: navPath });
        if (!res.ok) {
            throw new Error(res.error);
        }
        return res;
    }
    async close(key) {
        const res = await this.sendCommand({ cmd: "close", key });
        if (!res.ok) {
            throw new Error(res.error);
        }
    }
    /**
     * Forget a session from the daemon's in-memory map (the recovery path after
     * a `not-responding` gate). Matched daemon-side on the studio sessionId, so a
     * session re-bound under the same key is never touched. Returns how many
     * entries were dropped (0 or 1).
     */
    async clear(sessionId) {
        const res = await this.sendCommand({ cmd: "clear", sessionId });
        if (!res.ok) {
            throw new Error(res.error);
        }
        return res.cleared;
    }
    async detach(key) {
        const res = await this.sendCommand({ cmd: "detach", key });
        if (!res.ok) {
            throw new Error(res.error);
        }
    }
    async status() {
        const res = await this.sendCommand({ cmd: "status" });
        if (!res.ok) {
            throw new Error(res.error);
        }
        return res;
    }
    async ping() {
        const res = await this.sendCommand({ cmd: "ping" });
        if (!res.ok) {
            throw new Error(res.error);
        }
    }
    onEvent(handler) {
        this.eventHandler = handler;
    }
    /** Fires once when the daemon connection drops unexpectedly (not on destroy). */
    onDisconnect(handler) {
        this.disconnectHandler = handler;
    }
    destroy() {
        this.eventHandler = null;
        this.disconnectHandler = null;
        this.responseQueue = [];
        if (this.socket) {
            this.socket.destroy();
            this.socket = null;
        }
    }
}
export { resolveAgentSessionKey } from "../agentSessionKey.js";
