// bitfab-plugin-lib/dist/daemon/ensure.js
import { execFileSync, fork } from "node:child_process";
import crypto from "node:crypto";
import fs from "node:fs";
import net from "node:net";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { DaemonUnavailableError } from "../errors.js";
import { ENSURE_POLL_INTERVAL_MS, ENSURE_TIMEOUT_MS, PING_TIMEOUT_MS, SINGLETON_DAEMON_PATHS, } from "./protocol.js";
import { sweepZombieDaemons } from "./scope.js";
const LOCK_STALE_MS = 10_000;
function processIsAlive(pid) {
    try {
        process.kill(pid, 0);
        return true;
    }
    catch (err) {
        return err.code !== "ESRCH";
    }
}
function readPid(pidPath) {
    try {
        const raw = fs.readFileSync(pidPath, "utf-8").trim();
        const pid = Number.parseInt(raw, 10);
        return Number.isNaN(pid) ? null : pid;
    }
    catch {
        return null;
    }
}
async function pingSocket(socketPath) {
    return new Promise((resolve) => {
        const timeout = setTimeout(() => {
            socket.destroy();
            resolve({ alive: false, build: null });
        }, PING_TIMEOUT_MS);
        const socket = net.createConnection(socketPath, () => {
            socket.write('{"cmd":"ping"}\n');
        });
        let buffer = "";
        socket.on("data", (chunk) => {
            buffer += chunk.toString();
            if (buffer.includes("\n")) {
                clearTimeout(timeout);
                socket.destroy();
                try {
                    const msg = JSON.parse(buffer.split("\n")[0]);
                    resolve({ alive: msg.ok === true, build: msg.build ?? null });
                }
                catch {
                    resolve({ alive: false, build: null });
                }
            }
        });
        socket.on("error", () => {
            clearTimeout(timeout);
            resolve({ alive: false, build: null });
        });
    });
}
async function waitForSocket(socketPath, timeoutMs) {
    const deadline = Date.now() + timeoutMs;
    while (Date.now() < deadline) {
        const { alive } = await pingSocket(socketPath);
        if (alive) {
            return;
        }
        await new Promise((r) => setTimeout(r, ENSURE_POLL_INTERVAL_MS));
    }
    throw new Error(`daemon did not become ready within ${timeoutMs}ms`);
}
/**
 * Wait for the daemon the lock holder is bringing up. We reach here only after
 * losing the spawn lock, so another ensureDaemon is (re)spawning; prefer to
 * return once the answering daemon reports OUR local build, so a build-mismatch
 * respawn can't hand us the stale daemon that is about to be reaped (the window
 * Bugbot flagged). If the local build is unknowable, the daemon doesn't report
 * one, or a matching build never appears within the timeout (a rare cross-install
 * race where the holder is spawning a different build), fall back to the prior
 * waitForSocket contract: return only if a daemon is responding RIGHT NOW. We
 * re-ping at the deadline rather than trusting an earlier success, because the
 * holder may have reaped the old daemon and not yet bound the replacement, and
 * returning on a dead socket would falsely claim readiness.
 */
export async function waitForLockHolderDaemon(socketPath, timeoutMs, localBuild = localBuildFingerprint()) {
    const deadline = Date.now() + timeoutMs;
    while (Date.now() < deadline) {
        const ping = await pingSocket(socketPath);
        if (ping.alive && buildMatches(localBuild, ping.build)) {
            return;
        }
        await new Promise((r) => setTimeout(r, ENSURE_POLL_INTERVAL_MS));
    }
    // Timed out waiting for a build match. Return only if a daemon is currently
    // responding (the cross-install race), never on a dead socket.
    const finalPing = await pingSocket(socketPath);
    if (finalPing.alive) {
        return;
    }
    throw new Error(`daemon did not become ready within ${timeoutMs}ms`);
}
/**
 * Whether a running daemon's reported build is safe to reuse. A daemon that
 * reports OUR build matches. A daemon that reports a DIFFERENT build, or reports
 * NONE at all (too old to answer `build` on ping - the exact stale daemon that
 * kept serving old code), does NOT match and must be reaped + respawned. When we
 * can't fingerprint our own build (`unknown`), we can't judge, so we accept
 * whatever is running rather than respawn-loop.
 */
export function buildMatches(localBuild, remoteBuild) {
    if (localBuild === "unknown") {
        return true;
    }
    return remoteBuild === localBuild;
}
function localBuildFingerprint() {
    try {
        const thisDir = path.dirname(fileURLToPath(import.meta.url));
        const serverPath = path.join(thisDir, "server.js");
        const content = fs.readFileSync(serverPath);
        return crypto
            .createHash("sha256")
            .update(content)
            .digest("hex")
            .slice(0, 16);
    }
    catch {
        return "unknown";
    }
}
/**
 * Find every running studio daemon that targets this socket path, by command
 * line, not just the one named in the pidfile. The pidfile only ever names the
 * newest daemon; older ones orphaned by a socket-steal (a different plugin
 * install, a build-mismatch respawn, or a lock race) keep running and are
 * invisible to a pidfile-only kill, which is how they pile up over time. The
 * daemon is forked as `node .../daemon/main.js <socketPath> <pidPath>`, so its
 * command line contains both `main.js` and the socket path; we match on both so
 * the scan is scoped to our socket (and, in tests, to the temp socket) and can
 * never reap an unrelated process. Best-effort: an unavailable `ps` yields [].
 */
function findDaemonPids(socketPath) {
    let out;
    try {
        out = execFileSync("ps", ["-axww", "-o", "pid=,command="], {
            encoding: "utf-8",
        });
    }
    catch {
        return [];
    }
    const pids = [];
    for (const line of out.split("\n")) {
        const match = line.trim().match(/^(\d+)\s+(.*)$/);
        if (!match) {
            continue;
        }
        const pid = Number.parseInt(match[1], 10);
        const command = match[2];
        if (pid === process.pid || Number.isNaN(pid)) {
            continue;
        }
        if (command.includes("daemon/main.js") &&
            commandTargetsSocket(command, socketPath)) {
            pids.push(pid);
        }
    }
    return pids;
}
/**
 * True when `socketPath` appears in the command line as a whole argv token, not
 * just a substring. The daemon is forked with the socket path as a discrete
 * argument, so the character after it is an arg separator (space) or end of
 * string; requiring that boundary stops a longer path that merely starts with
 * the same string (e.g. `studio-daemon.sock2`) from matching `studio-daemon.sock`.
 */
function commandTargetsSocket(command, socketPath) {
    let idx = command.indexOf(socketPath);
    while (idx !== -1) {
        const after = command[idx + socketPath.length];
        if (after === undefined || after === " ") {
            return true;
        }
        idx = command.indexOf(socketPath, idx + 1);
    }
    return false;
}
/**
 * Reap ALL daemons bound to this socket, not just the pidfile's. Called once,
 * under the spawn lock, immediately before spawning a fresh daemon, so every
 * respawn path (no pidfile, stale pidfile, build mismatch, unresponsive socket)
 * converges on a single live daemon instead of orphaning the previous one.
 */
export async function reapAllDaemons(paths) {
    const pids = new Set(findDaemonPids(paths.socketPath));
    const pidfilePid = readPid(paths.pidPath);
    if (pidfilePid !== null) {
        pids.add(pidfilePid);
    }
    for (const pid of pids) {
        try {
            process.kill(pid, "SIGTERM");
        }
        catch { }
    }
    const deadline = Date.now() + 5_000;
    while (Date.now() < deadline) {
        if (![...pids].some(processIsAlive)) {
            break;
        }
        await new Promise((r) => setTimeout(r, 50));
    }
    for (const pid of pids) {
        if (processIsAlive(pid)) {
            try {
                process.kill(pid, "SIGKILL");
            }
            catch { }
        }
    }
    cleanupStaleFiles(paths);
}
function spawnDaemon(paths) {
    const thisDir = path.dirname(fileURLToPath(import.meta.url));
    const mainPath = path.join(thisDir, "main.js");
    if (!fs.existsSync(mainPath)) {
        throw new DaemonUnavailableError(`daemon entry point not found: ${mainPath}`);
    }
    // argv: socketPath, pidPath, metaPath, worktreeRoot ("" = absent for prod).
    const child = fork(mainPath, [
        paths.socketPath,
        paths.pidPath,
        paths.metaPath ?? "",
        paths.worktreeRoot ?? "",
    ], {
        detached: true,
        stdio: "ignore",
        execArgv: [],
    });
    child.unref();
}
function cleanupStaleFiles(paths) {
    try {
        fs.unlinkSync(paths.socketPath);
    }
    catch { }
    try {
        fs.unlinkSync(paths.pidPath);
    }
    catch { }
    if (paths.metaPath) {
        try {
            fs.unlinkSync(paths.metaPath);
        }
        catch { }
    }
}
function lockPath(socketPath) {
    return `${socketPath}.lock`;
}
function acquireLock(socketPath) {
    const lock = lockPath(socketPath);
    try {
        fs.mkdirSync(path.dirname(lock), { recursive: true });
        fs.writeFileSync(lock, `${process.pid}\n`, { flag: "wx" });
        return true;
    }
    catch (err) {
        if (err.code !== "EEXIST") {
            return true;
        }
        try {
            const stat = fs.statSync(lock);
            if (Date.now() - stat.mtimeMs > LOCK_STALE_MS) {
                fs.unlinkSync(lock);
                fs.writeFileSync(lock, `${process.pid}\n`, { flag: "wx" });
                return true;
            }
        }
        catch { }
        return false;
    }
}
function releaseLock(socketPath) {
    try {
        fs.unlinkSync(lockPath(socketPath));
    }
    catch { }
}
export async function ensureDaemon(paths = SINGLETON_DAEMON_PATHS) {
    const { socketPath, pidPath } = paths;
    // Opportunistic: clear zombie dev daemons (dead pid, or worktree gone) on any
    // daemon activity, so abandoned worktrees get reaped even without a close.
    sweepZombieDaemons();
    const pid = readPid(pidPath);
    if (pid !== null && processIsAlive(pid)) {
        const ping = await pingSocket(socketPath);
        if (ping.alive) {
            const localBuild = localBuildFingerprint();
            if (buildMatches(localBuild, ping.build)) {
                return;
            }
            // Don't kill here: fall through to the locked spawn path, which reaps
            // every daemon (this stale-build one plus any orphans) in one place
            // before spawning. Killing outside the lock is what let concurrent
            // ensures race and double-spawn.
            console.error(`[daemon] build mismatch (running: ${ping.build ?? "unreported"}, local: ${localBuild}), restarting`);
        }
    }
    if (!acquireLock(socketPath)) {
        await waitForLockHolderDaemon(socketPath, ENSURE_TIMEOUT_MS);
        return;
    }
    try {
        await reapAllDaemons(paths);
        spawnDaemon(paths);
        await waitForSocket(socketPath, ENSURE_TIMEOUT_MS);
    }
    finally {
        releaseLock(socketPath);
    }
}
