// bitfab-plugin-lib/dist/daemon/ensure.js
import { fork } from "node:child_process";
import crypto from "node:crypto";
import fs from "node:fs";
import net from "node:net";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { ENSURE_POLL_INTERVAL_MS, ENSURE_TIMEOUT_MS, PID_FILE_PATH, PING_TIMEOUT_MS, SOCKET_PATH, } from "./protocol.js";
const LOCK_STALE_MS = 10_000;
function processIsAlive(pid) {
    try {
        process.kill(pid, 0);
        return true;
    }
    catch (err) {
        return err.code !== "ESRCH";
    }
}
function readPid(pidPath) {
    try {
        const raw = fs.readFileSync(pidPath, "utf-8").trim();
        const pid = Number.parseInt(raw, 10);
        return Number.isNaN(pid) ? null : pid;
    }
    catch {
        return null;
    }
}
async function pingSocket(socketPath) {
    return new Promise((resolve) => {
        const timeout = setTimeout(() => {
            socket.destroy();
            resolve({ alive: false, build: null });
        }, PING_TIMEOUT_MS);
        const socket = net.createConnection(socketPath, () => {
            socket.write('{"cmd":"ping"}\n');
        });
        let buffer = "";
        socket.on("data", (chunk) => {
            buffer += chunk.toString();
            if (buffer.includes("\n")) {
                clearTimeout(timeout);
                socket.destroy();
                try {
                    const msg = JSON.parse(buffer.split("\n")[0]);
                    resolve({ alive: msg.ok === true, build: msg.build ?? null });
                }
                catch {
                    resolve({ alive: false, build: null });
                }
            }
        });
        socket.on("error", () => {
            clearTimeout(timeout);
            resolve({ alive: false, build: null });
        });
    });
}
async function waitForSocket(socketPath, timeoutMs) {
    const deadline = Date.now() + timeoutMs;
    while (Date.now() < deadline) {
        const { alive } = await pingSocket(socketPath);
        if (alive) {
            return;
        }
        await new Promise((r) => setTimeout(r, ENSURE_POLL_INTERVAL_MS));
    }
    throw new Error(`daemon did not become ready within ${timeoutMs}ms`);
}
function localBuildFingerprint() {
    try {
        const thisDir = path.dirname(fileURLToPath(import.meta.url));
        const serverPath = path.join(thisDir, "server.js");
        const content = fs.readFileSync(serverPath);
        return crypto
            .createHash("sha256")
            .update(content)
            .digest("hex")
            .slice(0, 16);
    }
    catch {
        return "unknown";
    }
}
async function killDaemon(pidPath, socketPath) {
    const pid = readPid(pidPath);
    if (pid !== null) {
        try {
            process.kill(pid, "SIGTERM");
        }
        catch { }
        const deadline = Date.now() + 5_000;
        while (processIsAlive(pid) && Date.now() < deadline) {
            await new Promise((r) => setTimeout(r, 50));
        }
        if (processIsAlive(pid)) {
            try {
                process.kill(pid, "SIGKILL");
            }
            catch { }
        }
    }
    cleanupStaleFiles(socketPath, pidPath);
}
function spawnDaemon(socketPath, pidPath) {
    const thisDir = path.dirname(fileURLToPath(import.meta.url));
    const mainPath = path.join(thisDir, "main.js");
    if (!fs.existsSync(mainPath)) {
        throw new Error(`daemon entry point not found: ${mainPath}`);
    }
    const child = fork(mainPath, [socketPath, pidPath], {
        detached: true,
        stdio: "ignore",
        execArgv: [],
    });
    child.unref();
}
function cleanupStaleFiles(socketPath, pidPath) {
    try {
        fs.unlinkSync(socketPath);
    }
    catch { }
    try {
        fs.unlinkSync(pidPath);
    }
    catch { }
}
function lockPath(socketPath) {
    return `${socketPath}.lock`;
}
function acquireLock(socketPath) {
    const lock = lockPath(socketPath);
    try {
        fs.mkdirSync(path.dirname(lock), { recursive: true });
        fs.writeFileSync(lock, `${process.pid}\n`, { flag: "wx" });
        return true;
    }
    catch (err) {
        if (err.code !== "EEXIST") {
            return true;
        }
        try {
            const stat = fs.statSync(lock);
            if (Date.now() - stat.mtimeMs > LOCK_STALE_MS) {
                fs.unlinkSync(lock);
                fs.writeFileSync(lock, `${process.pid}\n`, { flag: "wx" });
                return true;
            }
        }
        catch { }
        return false;
    }
}
function releaseLock(socketPath) {
    try {
        fs.unlinkSync(lockPath(socketPath));
    }
    catch { }
}
export async function ensureDaemon(socketPath = SOCKET_PATH, pidPath = PID_FILE_PATH) {
    const pid = readPid(pidPath);
    if (pid !== null && processIsAlive(pid)) {
        const ping = await pingSocket(socketPath);
        if (ping.alive) {
            const localBuild = localBuildFingerprint();
            if (ping.build && ping.build !== localBuild && localBuild !== "unknown") {
                console.error(`[daemon] build mismatch (running: ${ping.build}, local: ${localBuild}), restarting`);
                await killDaemon(pidPath, socketPath);
            }
            else {
                return;
            }
        }
    }
    if (!acquireLock(socketPath)) {
        await waitForSocket(socketPath, ENSURE_TIMEOUT_MS);
        return;
    }
    try {
        cleanupStaleFiles(socketPath, pidPath);
        spawnDaemon(socketPath, pidPath);
        await waitForSocket(socketPath, ENSURE_TIMEOUT_MS);
    }
    finally {
        releaseLock(socketPath);
    }
}
