// bitfab-plugin-lib/dist/daemon/protocol.js
import crypto from "node:crypto";
import os from "node:os";
import path from "node:path";
// --- Paths ---
const BITFAB_CONFIG_DIR = path.join(os.homedir(), ".config", "bitfab");
// Prod singleton: one daemon per machine, shared by every non-dev worktree.
export const SOCKET_PATH = path.join(BITFAB_CONFIG_DIR, "studio-daemon.sock");
export const PID_FILE_PATH = path.join(BITFAB_CONFIG_DIR, "studio-daemon.pid");
// Dev daemons live here, one per worktree: dev-<worktreeHash>.{sock,pid,meta.json}.
export const DAEMON_DIR = path.join(BITFAB_CONFIG_DIR, "daemon");
export const SINGLETON_DAEMON_PATHS = {
    socketPath: SOCKET_PATH,
    pidPath: PID_FILE_PATH,
    metaPath: null,
    worktreeRoot: null,
};
export function hashWorktreeRoot(worktreeRoot) {
    return crypto
        .createHash("sha256")
        .update(worktreeRoot)
        .digest("hex")
        .slice(0, 12);
}
/** Per-worktree dev daemon paths, keyed on the worktree root (not the build). */
export function devDaemonPaths(worktreeRoot) {
    const base = path.join(DAEMON_DIR, `dev-${hashWorktreeRoot(worktreeRoot)}`);
    return {
        socketPath: `${base}.sock`,
        pidPath: `${base}.pid`,
        metaPath: `${base}.meta.json`,
        worktreeRoot,
    };
}
// --- Timing ---
export const BROWSER_HEALTH_INTERVAL_MS = 10 * 1000;
export const ENSURE_POLL_INTERVAL_MS = 200;
export const ENSURE_TIMEOUT_MS = 5000;
export const PING_TIMEOUT_MS = 2000;
// A healthy daemon acks `open` immediately (it fire-and-forget spawns the
// browser, then responds). A wedged daemon can answer `ping` yet hang here, so
// the client times out an `open` past this and reaps + respawns the daemon.
export const OPEN_TIMEOUT_MS = 10_000;
// Ceiling for every other daemon command, so a wedged daemon can never hang a
// caller indefinitely (a never-resolving promise slips straight past .catch()).
// Sized above the slowest legitimate command: `navigate` waits server-side for
// a browser ack (12s) plus a ping/pong fallback (5s), so ~17s worst case on a
// HEALTHY daemon. Timing out sooner would misread a slow-but-live navigate as a
// wedge.
export const COMMAND_TIMEOUT_MS = 30_000;
// --- Wire helpers ---
export function serialize(msg) {
    return `${JSON.stringify(msg)}\n`;
}
export function isResponse(msg) {
    return typeof msg === "object" && msg !== null && "ok" in msg;
}
export function isPush(msg) {
    return typeof msg === "object" && msg !== null && "event" in msg;
}
