// bitfab-plugin-lib/dist/daemon/scope.js
import { execFileSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { DAEMON_DIR, devDaemonPaths, SINGLETON_DAEMON_PATHS, } from "./protocol.js";
/**
 * Resolve the git worktree root from a starting directory. Returns null outside
 * a repo. Both the daemon socket and the reap key off the running module's
 * directory, so they always land on the same worktree root.
 */
export function resolveWorktreeRoot(fromDir) {
    try {
        const out = execFileSync("git", ["rev-parse", "--show-toplevel"], {
            cwd: fromDir,
            encoding: "utf-8",
            stdio: ["ignore", "pipe", "ignore"],
        }).trim();
        return out || null;
    }
    catch {
        return null;
    }
}
/** The directory the running plugin code lives in (the sidecar, in dev). */
function moduleDir() {
    return path.dirname(fileURLToPath(import.meta.url));
}
// Known host plugin-install roots: a PUBLISHED plugin/CLI lands under one of
// these, while the dev sidecar lands inside a project worktree. A path under any
// of these is prod even if it happens to sit inside a git repo (e.g. a dotfiles
// repo at $HOME, or the npx cache under ~/.npm) — so prod always shares the
// singleton and never spawns per-"worktree" daemons keyed on an unrelated repo.
const PROD_INSTALL_MARKERS = [
    `${path.sep}.claude${path.sep}plugins${path.sep}`,
    `${path.sep}.codex${path.sep}plugins${path.sep}`,
    `${path.sep}.cursor${path.sep}plugins${path.sep}`,
    `${path.sep}.npm${path.sep}`,
];
export function isProdInstallPath(dir) {
    return PROD_INSTALL_MARKERS.some((marker) => dir.includes(marker));
}
/**
 * The dev worktree this running build belongs to, or null for a prod install.
 *
 * Dev = the running plugin code lives in a git worktree (the bitfab-dev sidecar
 * at `<worktree>/.bitfab-internal-*`, or a raw monorepo checkout) AND is not
 * under a known host plugin cache. Prod (published plugin/CLI under
 * `~/.claude`/`~/.codex`/`~/.cursor` plugins or the npx cache) -> null.
 *
 * Keyed on the module dir, NOT cwd: the sidecar is stable wherever the user runs
 * from, and the reap path keys the same way so the socket the daemon binds and
 * the one close/merge reaps always match.
 */
export function resolveDevWorktreeRoot() {
    const dir = moduleDir();
    if (isProdInstallPath(dir)) {
        return null;
    }
    return resolveWorktreeRoot(dir);
}
/**
 * Which daemon this caller talks to: the shared singleton for prod, or a
 * per-worktree daemon (keyed on the worktree root) for a dev build.
 */
export function resolveDaemonPaths() {
    const worktreeRoot = resolveDevWorktreeRoot();
    return worktreeRoot ? devDaemonPaths(worktreeRoot) : SINGLETON_DAEMON_PATHS;
}
function processIsAlive(pid) {
    try {
        process.kill(pid, 0);
        return true;
    }
    catch (err) {
        return err.code !== "ESRCH";
    }
}
function unlinkQuietly(...paths) {
    for (const p of paths) {
        try {
            fs.unlinkSync(p);
        }
        catch { }
    }
}
function readMeta(metaPath) {
    try {
        const parsed = JSON.parse(fs.readFileSync(metaPath, "utf-8"));
        if (typeof parsed.pid === "number" &&
            typeof parsed.worktreeRoot === "string" &&
            typeof parsed.socketPath === "string" &&
            typeof parsed.pidPath === "string") {
            return parsed;
        }
    }
    catch { }
    return null;
}
/**
 * Reap zombie dev daemons: those whose process is dead, or whose worktree no
 * longer exists on disk. A daemon with a live process AND a live worktree is
 * healthy and never touched. Only scans dev-* metas, so the prod singleton is
 * structurally excluded. Best-effort; never throws.
 *
 * For a live-but-orphaned daemon we only SIGHUP: its handler closes its windows
 * and unlinks its own files (socket/pid/meta), so we leave the files in place
 * and let it self-clean. A wedged daemon that ignores SIGHUP is re-signalled by
 * the next sweep (pid still alive, worktree still gone, meta still present).
 */
export function sweepZombieDaemons(dir = DAEMON_DIR) {
    let entries;
    try {
        entries = fs.readdirSync(dir);
    }
    catch {
        return;
    }
    for (const name of entries) {
        if (!name.startsWith("dev-") || !name.endsWith(".meta.json")) {
            continue;
        }
        const metaPath = path.join(dir, name);
        const meta = readMeta(metaPath);
        if (!meta) {
            // Unparseable/orphaned meta — drop it.
            unlinkQuietly(metaPath);
            continue;
        }
        const pidAlive = processIsAlive(meta.pid);
        const worktreeExists = fs.existsSync(meta.worktreeRoot);
        if (pidAlive && worktreeExists) {
            continue; // healthy
        }
        if (!pidAlive) {
            // No process to kill; just clear the stale files.
            unlinkQuietly(meta.socketPath, meta.pidPath, `${meta.socketPath}.lock`, metaPath);
            continue;
        }
        // Alive but its worktree is gone: a true zombie. SIGHUP tells it to close
        // its Studio windows and self-clean (SIGTERM would leave windows orphaned,
        // since that handler is the window-preserving restart path).
        try {
            process.kill(meta.pid, "SIGHUP");
        }
        catch { }
    }
}
