// bitfab-plugin-lib/dist/daemon/server.js
import crypto from "node:crypto";
import fs from "node:fs";
import net from "node:net";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { clearSessionStateForKey, readAllDaemonSessionStates, updateSessionStateForKey, writeSessionStateForKey, } from "../activeStudioSession.js";
import { createAgentSession, navigateStudioAndAwaitAck, pollAgentSessionEvents, } from "../agentSessionChannel.js";
import { closeStudioWindowsBySession, openChromelessWindow, } from "../browser.js";
import { applyFocusTarget } from "../frontmostApp.js";
import { connectGraceMs, preAuthConnectGraceMs } from "../studioConnectGrace.js";
import { appendStudioEvent, studioEventToLogLine, truncateStudioEventLog, } from "../studioEventLog.js";
import { buildStudioSessionUrl, resolveStudioInitialPath, } from "../studioUrl.js";
import { BROWSER_HEALTH_INTERVAL_MS, PID_FILE_PATH, SOCKET_PATH, serialize, } from "./protocol.js";
const WINDOW_CLOSE_GRACE_MS = 10_000;
// A dev (per-worktree) daemon with no sessions for this long shuts itself down.
// Backstop only: close/merge/teardown reap directly, and the zombie sweep
// catches deleted worktrees. The prod singleton (no meta) never idle-exits.
const DEV_IDLE_TTL_MS = 5 * 60 * 1000;
// After the agent asks to close, give the cooperative teardown (the HTTP close
// the channel fires in parallel) this long to surface a studio:window-closed
// before the daemon force-closes the window itself. Read at close time so tests
// and operators can shorten it without a rebuild.
const CLOSE_WINDOW_FALLBACK_DELAY_MS = 2_000;
function closeWindowFallbackDelayMs() {
    const raw = process.env.BITFAB_STUDIO_CLOSE_FALLBACK_DELAY_MS;
    const parsed = raw ? Number(raw) : Number.NaN;
    return Number.isFinite(parsed) ? parsed : CLOSE_WINDOW_FALLBACK_DELAY_MS;
}
// How often the reaper sweeps closed sessions. Read at startup so tests can
// shorten it without a rebuild.
function reapIntervalMs() {
    const raw = process.env.BITFAB_DAEMON_REAP_INTERVAL_MS;
    const parsed = raw ? Number(raw) : Number.NaN;
    return Number.isFinite(parsed) ? parsed : BROWSER_HEALTH_INTERVAL_MS;
}
// Events that hand control back to the coding agent: the daemon restores the
// terminal focus the CLI captured at open time (the CLI's createEventSubscription
// used to do this in-process; with fire-and-forget that process is gone).
const FOCUS_RESTORE_EVENTS = new Set([
    "studio:return-to-agent",
    "studio:edit-with-agent",
    "studio:session-ended",
    "studio:not-member",
]);
function computeBuildFingerprint() {
    try {
        const thisFile = fileURLToPath(import.meta.url);
        const content = fs.readFileSync(thisFile);
        return crypto
            .createHash("sha256")
            .update(content)
            .digest("hex")
            .slice(0, 16);
    }
    catch {
        return "unknown";
    }
}
const sessions = new Map();
let healthTimer = null;
let buildFingerprint = null;
// The paths this process actually bound (from argv). gracefulShutdown unlinks
// THESE, never the singleton constants - a dev daemon must not stomp the
// singleton's files nor orphan its own.
let boundSocketPath = SOCKET_PATH;
let boundPidPath = PID_FILE_PATH;
let boundMetaPath = null;
// The worktree this daemon serves (null for the prod singleton). Stamped onto
// every session it writes and used to filter restore, so a per-worktree dev
// daemon never adopts or reaps another worktree's sessions.
let boundWorktreeRoot = null;
// When this dev daemon last had zero sessions (null = currently non-empty).
let idleSince = null;
// When this daemon process started serving. The never-connected reap measures
// its grace from max(browserLaunchedAt, daemonStartedAt) so a session rehydrated
// on a restart (launched long ago, adopted just now) gets a fresh window to
// reconnect before it's swept, instead of being reaped on the first sweep.
let daemonStartedAt = 0;
function buildSessionState(entry) {
    return {
        sessionId: entry.studioSessionId,
        serviceUrl: entry.serviceUrl,
        windowPid: entry.browserPid,
        pollerPid: null,
        source: "daemon",
        openedAt: new Date(entry.browserLaunchedAt).toISOString(),
        currentPath: entry.currentPath,
        authenticated: true,
        windowState: "open",
        agentIntent: "open",
        lastBrowserPingAt: null,
        browserConnected: entry.browserConnected,
        lastEventAt: null,
        lastEventCursor: entry.lastEventCursor,
        agentSessionKey: entry.key,
        worktreeRoot: boundWorktreeRoot,
    };
}
function persistSessionState(entry) {
    try {
        writeSessionStateForKey(entry.key, buildSessionState(entry));
    }
    catch { }
}
function send(socket, msg) {
    try {
        socket.write(serialize(msg));
    }
    catch {
        // socket already closed
    }
}
const STATE_CHANGING_EVENTS = new Set([
    "studio:session-started",
    "studio:window-closed",
    "studio:session-ended",
    "studio:not-member",
    "studio:authenticated",
]);
function broadcastEvent(entry, event) {
    const push = {
        id: event.id,
        event: event.type,
        sessionId: entry.studioSessionId,
        key: entry.key,
        data: event.data,
    };
    for (const client of entry.connectedClients) {
        send(client, push);
    }
    // Primary force-close trigger: the window reached the close page. Every close
    // flow (login auto-close AND agent close) lands on /studio/close, which
    // beacons studio:session-closed ("close flow completed") and then can't
    // window.close() itself, so it parks. Physically close it here. Deliberately
    // writes NO window state: session-closed does not claim the window closed;
    // window state flows only from the studio:window-closed our close emits.
    if (event.type === "studio:session-closed") {
        entry.closing = true;
        clearCloseFallbackTimer(entry);
        forceCloseWindow(entry);
    }
    // The durable buffer: append every event to the per-session log so the
    // agent's Monitor can attach at any time and replay from a cursor, whether
    // or not a client is connected right now.
    appendStudioEvent(entry.key, studioEventToLogLine(event, entry.studioSessionId));
    if (FOCUS_RESTORE_EVENTS.has(event.type)) {
        applyFocusTarget(entry.focusTarget);
    }
    if (STATE_CHANGING_EVENTS.has(event.type)) {
        const updates = {
            lastEventAt: new Date().toISOString(),
            lastEventCursor: event.id || entry.lastEventCursor,
        };
        if (event.type === "studio:window-closed") {
            // The window is confirmed gone (cooperative close, or our fallback's own
            // beforeunload beacon); cancel any still-pending fallback.
            clearCloseFallbackTimer(entry);
            entry.browserConnected = false;
            entry.browserDisconnectedAt = Date.now();
            entry.windowClosedAt = Date.now();
            updates.browserConnected = false;
            updates.windowState = "closed";
        }
        else if (event.type === "studio:not-member") {
            updates.authenticated = false;
        }
        else if (event.type === "studio:authenticated") {
            const token = typeof event.data === "object" &&
                event.data !== null &&
                "token" in event.data
                ? event.data.token
                : undefined;
            if (typeof token === "string" && !entry.apiKey) {
                entry.apiKey = token;
                entry.pollAbort.abort();
                entry.pollAbort = new AbortController();
                startPollLoop(entry);
            }
            // The sign-in page pushed this event, so the window demonstrably
            // connected. Without this flip, storing the apiKey above would silently
            // move the entry from the long pre-auth grace to the short connect grace
            // and the never-connected reaper could kill a login that took longer
            // than the short grace, right after it succeeded.
            entry.browserEverConnected = true;
            updates.authenticated = true;
        }
        else if (event.type === "studio:session-started") {
            entry.browserConnected = true;
            entry.browserEverConnected = true;
            entry.browserDisconnectedAt = null;
            entry.windowClosedAt = null;
            updates.browserConnected = true;
            updates.windowState = "open";
        }
        try {
            updateSessionStateForKey(entry.key, updates);
        }
        catch { }
    }
}
function startPollLoop(entry) {
    ;
    (async () => {
        if (!entry.lastEventCursor && entry.apiKey) {
            try {
                await createAgentSession({
                    serviceUrl: entry.serviceUrl,
                    apiKey: entry.apiKey,
                    sessionId: entry.studioSessionId,
                    extraHeaders: entry.clientHeaders,
                });
            }
            catch (err) {
                console.error(`[daemon] createAgentSession failed for ${entry.key}: ${err instanceof Error ? err.message : String(err)}`);
                return;
            }
        }
        await pollAgentSessionEvents({
            serviceUrl: entry.serviceUrl,
            apiKey: entry.apiKey || undefined,
            sessionId: entry.studioSessionId,
            startCursor: entry.lastEventCursor ?? undefined,
            abortSignal: entry.pollAbort.signal,
            onEvent: (event) => {
                entry.lastEventCursor = event.id;
                entry.lastActivity = Date.now();
                broadcastEvent(entry, event);
            },
            onBrowserStatus: (connected) => {
                const wasConnected = entry.browserConnected;
                entry.browserConnected = connected;
                if (connected) {
                    entry.browserEverConnected = true;
                    entry.browserDisconnectedAt = null;
                }
                else if (wasConnected && !connected) {
                    entry.browserDisconnectedAt = Date.now();
                }
                try {
                    updateSessionStateForKey(entry.key, {
                        browserConnected: connected,
                        lastBrowserPingAt: new Date().toISOString(),
                    });
                }
                catch { }
            },
            onError: (err) => {
                console.error(`[daemon] poll error for ${entry.key}: ${err.message}`);
            },
        });
    })().catch((err) => {
        if (!entry.pollAbort.signal.aborted) {
            console.error(`[daemon] poll terminated for ${entry.key}: ${err.message}`);
        }
    });
}
function clearCloseFallbackTimer(entry) {
    if (entry.closeFallbackTimer) {
        clearTimeout(entry.closeFallbackTimer);
        entry.closeFallbackTimer = null;
    }
}
/**
 * Best-effort kill of the spawned `--app` pid. Real teardown only when the
 * browser was not already running (otherwise this is a dead launcher pid and a
 * no-op); the AppleScript close below is what handles the handed-off case. We
 * never signal our own pid (a launcher hand-off can leave it recorded, and the
 * tests mock the spawn to return process.pid) so a fallback can't kill us.
 */
function killBrowserWindow(entry) {
    const pid = entry.browserPid;
    if (pid == null || pid === process.pid) {
        return;
    }
    try {
        process.kill(pid, "SIGTERM");
    }
    catch {
        // Already gone (or a dead launcher pid); nothing to kill.
    }
}
/**
 * Force the window closed when cooperative autoclose didn't (window.close() is
 * refused for these windows, so the page parks instead of closing). Close by
 * URL match on the session id, plus a best-effort pid kill for the rare case
 * where the spawned pid is the real window. Never marks state closed here: a
 * successful close fires the browser's beforeunload, which beacons
 * studio:window-closed and drives state through the normal path; if the close
 * was refused/denied the window really is still open, so claiming otherwise
 * would lie.
 */
function forceCloseWindow(entry) {
    killBrowserWindow(entry);
    closeStudioWindowsBySession(entry.studioSessionId);
}
function launchBrowser(entry, initialPath) {
    const studioPath = resolveStudioInitialPath(initialPath ?? entry.currentPath ?? undefined);
    const url = buildStudioSessionUrl({
        serviceUrl: entry.serviceUrl,
        path: studioPath,
        sessionId: entry.studioSessionId,
    });
    const open = openChromelessWindow(url);
    entry.browserPid = open.pid;
    entry.browserLaunched = open.launched;
    entry.browserLaunchReason = open.reason;
    entry.browserLaunchedAt = Date.now();
    entry.browserConnected = false;
    entry.browserDisconnectedAt = null;
    entry.currentPath = studioPath;
    try {
        updateSessionStateForKey(entry.key, {
            windowPid: entry.browserPid,
            browserConnected: false,
            windowState: "open",
            currentPath: studioPath,
        });
    }
    catch { }
}
async function handleOpen(socket, cmd) {
    const existing = sessions.get(cmd.key);
    if (existing) {
        if (existing.windowClosedAt != null) {
            clearCloseFallbackTimer(existing);
            existing.pollAbort.abort();
            sessions.delete(cmd.key);
        }
        else {
            existing.connectedClients.add(socket);
            existing.lastActivity = Date.now();
            // Reopening cancels an in-flight close: the agent wants the window back.
            existing.closing = false;
            clearCloseFallbackTimer(existing);
            if (cmd.apiKey && !existing.apiKey) {
                existing.apiKey = cmd.apiKey;
            }
            // Refresh the focus target: a reconnect comes from a new terminal-child
            // CLI, so its captured target is the current one to hand back to.
            if (cmd.focusTarget) {
                existing.focusTarget = cmd.focusTarget;
            }
            const RELAUNCH_COOLDOWN_MS = 10_000;
            if (!existing.browserConnected &&
                existing.browserEverConnected &&
                Date.now() - (existing.browserLaunchedAt ?? 0) > RELAUNCH_COOLDOWN_MS) {
                launchBrowser(existing);
            }
            send(socket, {
                ok: true,
                sessionId: existing.studioSessionId,
                browserConnected: existing.browserConnected,
                reconnected: true,
            });
            return;
        }
    }
    // Fresh (non-reconnect) session: reset the durable log so a resumed run never
    // sees the prior session's tail. Reconnects return early above and keep theirs.
    truncateStudioEventLog(cmd.key);
    const studioSessionId = cmd.sessionId ?? crypto.randomUUID();
    const entry = {
        key: cmd.key,
        studioSessionId,
        serviceUrl: cmd.serviceUrl,
        apiKey: cmd.apiKey ?? "",
        browserPid: null,
        browserLaunched: false,
        browserConnected: false,
        browserEverConnected: false,
        browserDisconnectedAt: null,
        browserLaunchedAt: Date.now(),
        pollAbort: new AbortController(),
        pollActive: true,
        connectedClients: new Set([socket]),
        lastActivity: Date.now(),
        clientHeaders: cmd.clientHeaders ?? {},
        lastEventCursor: null,
        currentPath: null,
        windowClosedAt: null,
        closing: false,
        closeFallbackTimer: null,
        focusTarget: cmd.focusTarget ?? null,
    };
    launchBrowser(entry, cmd.initialPath);
    sessions.set(cmd.key, entry);
    startPollLoop(entry);
    persistSessionState(entry);
    send(socket, {
        ok: true,
        sessionId: studioSessionId,
        browserConnected: false,
        reconnected: false,
        launched: entry.browserLaunched,
        launchReason: entry.browserLaunchReason,
    });
}
async function handleNavigate(socket, cmd) {
    const entry = sessions.get(cmd.key);
    if (!entry) {
        send(socket, { ok: false, error: `no session for key: ${cmd.key}` });
        return;
    }
    try {
        const result = await navigateStudioAndAwaitAck({
            serviceUrl: entry.serviceUrl,
            apiKey: entry.apiKey,
            sessionId: entry.studioSessionId,
        }, cmd.path);
        entry.lastActivity = Date.now();
        if (result.acked) {
            entry.currentPath = cmd.path;
            try {
                updateSessionStateForKey(cmd.key, {
                    currentPath: cmd.path,
                    windowState: "open",
                });
            }
            catch { }
        }
        send(socket, { ok: true, acked: result.acked, reason: result.reason });
    }
    catch (err) {
        send(socket, {
            ok: false,
            error: `navigate failed: ${err instanceof Error ? err.message : String(err)}`,
        });
    }
}
function handleClose(socket, cmd) {
    const entry = sessions.get(cmd.key);
    if (!entry) {
        send(socket, { ok: false, error: `no session for key: ${cmd.key}` });
        return;
    }
    entry.connectedClients.add(socket);
    entry.closing = true;
    try {
        updateSessionStateForKey(cmd.key, {
            agentIntent: "close",
        });
    }
    catch { }
    // Primary trigger is studio:session-closed, which the close page beacons when
    // the window reaches it (handled in broadcastEvent). This timer is only a
    // BACKSTOP for when that beacon never reaches us (SSE dropped, or the page
    // never navigated to the close page): if neither session-closed nor
    // window-closed has acted within the delay, force-close anyway. Cancelled when
    // either arrives, or if the agent reopens. The daemon outlives the CLI.
    clearCloseFallbackTimer(entry);
    entry.closeFallbackTimer = setTimeout(() => {
        entry.closeFallbackTimer = null;
        if (sessions.get(cmd.key) !== entry) {
            return;
        }
        if (!entry.closing || entry.windowClosedAt != null) {
            return;
        }
        forceCloseWindow(entry);
    }, closeWindowFallbackDelayMs());
    send(socket, { ok: true });
}
/**
 * Forget a session the agent has explicitly abandoned (the recovery path that
 * runs after openStudioTo gates `not-responding`). The on-disk pointer alone
 * could already be cleared, but the daemon's in-memory SessionEntry survives a
 * crash/sleep that emitted no `studio:window-closed`; left in place, the next
 * `open` reconnects to the dead window and gates `not-responding` forever.
 *
 * Matched on the studio `sessionId`, not the agent key: a session the agent
 * re-bound under the same key (a fresh window) keeps a different sessionId, so
 * it is never the one we drop.
 *
 * Forgets state only; it does NOT force-close the window. The window reap is the
 * caller's job (runClearStudioSession runs the AppleScript close + pid kill from
 * the CLI, mode-independent), and force-closing here would block this synchronous
 * handler on a multi-second osascript pass per Chromium app, stalling the
 * single-threaded daemon. Aborting the poll and dropping the entry is instant.
 */
function handleClear(socket, cmd) {
    let cleared = 0;
    for (const [key, entry] of sessions) {
        if (entry.studioSessionId !== cmd.sessionId) {
            continue;
        }
        clearCloseFallbackTimer(entry);
        entry.pollAbort.abort();
        sessions.delete(key);
        try {
            updateSessionStateForKey(key, {
                browserConnected: false,
                windowState: "closed",
                agentIntent: "close",
            });
        }
        catch { }
        cleared++;
    }
    send(socket, { ok: true, cleared });
}
function handleDetach(socket, cmd) {
    const entry = sessions.get(cmd.key);
    if (!entry) {
        send(socket, { ok: true });
        return;
    }
    entry.connectedClients.delete(socket);
    entry.lastActivity = Date.now();
    send(socket, { ok: true });
}
function handleStatus(socket) {
    const infos = [];
    for (const [key, entry] of sessions) {
        infos.push({
            key,
            sessionId: entry.studioSessionId,
            browserPid: entry.browserPid,
            browserConnected: entry.browserConnected,
            connectedClients: entry.connectedClients.size,
            lastActivity: entry.lastActivity,
        });
    }
    send(socket, { ok: true, sessions: infos });
}
function handlePing(socket) {
    send(socket, { ok: true, build: buildFingerprint });
}
/**
 * Reap: close every window this daemon owns, then exit. Acks first so the
 * caller's `shutdown()` resolves, then force-closes all windows and shuts down
 * on the next tick (gracefulShutdown calls process.exit).
 */
function handleShutdown(socket) {
    send(socket, { ok: true });
    setImmediate(reapShutdown);
}
/**
 * Reap teardown: force-close every window this daemon owns, then exit. Used by
 * the `shutdown` RPC and the SIGHUP handler (the signal the CLI-side reap and
 * zombie sweep send for a worktree that's going away). Distinct from
 * gracefulShutdown, which deliberately leaves windows open so a build-mismatch
 * RESTART can hand them to the respawned daemon - reap has no successor, so the
 * windows must be closed here or they orphan.
 */
function reapShutdown() {
    for (const entry of sessions.values()) {
        entry.closing = true;
        clearCloseFallbackTimer(entry);
        forceCloseWindow(entry);
    }
    gracefulShutdown();
}
function dispatch(socket, cmd) {
    switch (cmd.cmd) {
        case "open":
            handleOpen(socket, cmd).catch((err) => {
                send(socket, {
                    ok: false,
                    error: `open failed: ${err instanceof Error ? err.message : String(err)}`,
                });
            });
            break;
        case "navigate":
            handleNavigate(socket, cmd).catch((err) => {
                send(socket, {
                    ok: false,
                    error: `navigate failed: ${err instanceof Error ? err.message : String(err)}`,
                });
            });
            break;
        case "close":
            handleClose(socket, cmd);
            break;
        case "clear":
            handleClear(socket, cmd);
            break;
        case "detach":
            handleDetach(socket, cmd);
            break;
        case "status":
            handleStatus(socket);
            break;
        case "ping":
            handlePing(socket);
            break;
        case "shutdown":
            handleShutdown(socket);
            break;
        default:
            send(socket, {
                ok: false,
                error: `unknown command: ${cmd.cmd}`,
            });
    }
}
function removeClientFromAllSessions(socket) {
    for (const entry of sessions.values()) {
        entry.connectedClients.delete(socket);
    }
}
function handleConnection(socket) {
    let buffer = "";
    socket.on("data", (chunk) => {
        buffer += chunk.toString();
        let newlineIdx = buffer.indexOf("\n");
        while (newlineIdx !== -1) {
            const line = buffer.slice(0, newlineIdx).trim();
            buffer = buffer.slice(newlineIdx + 1);
            if (line) {
                try {
                    const cmd = JSON.parse(line);
                    dispatch(socket, cmd);
                }
                catch {
                    send(socket, { ok: false, error: "invalid JSON" });
                }
            }
            newlineIdx = buffer.indexOf("\n");
        }
    });
    socket.on("close", () => {
        removeClientFromAllSessions(socket);
    });
    socket.on("error", () => {
        removeClientFromAllSessions(socket);
    });
}
// The daemon NEVER relaunches a window on its own. A window opens only in
// direct response to an `open` command; an unexplained browser disconnect (the
// user closed the window) is left alone, never silently resurrected. This timer
// reaps two kinds of dead session: one whose window the browser confirmed
// closed, and one whose window never surfaced in the first place. Both are
// ended identically so the waiting command unblocks and the pointer is dropped.
function reapClosedSessions() {
    const now = Date.now();
    const connectGrace = connectGraceMs();
    const preAuthGrace = preAuthConnectGraceMs();
    for (const [key, entry] of sessions) {
        // Confirmed close: the browser beaconed window-closed and the grace elapsed.
        const closedOut = entry.windowClosedAt != null &&
            now - entry.windowClosedAt > WINDOW_CLOSE_GRACE_MS;
        // Never surfaced: a fresh window that never once reported a connection
        // within the connect grace. Gated on browserEverConnected so a window that
        // connected and then dropped (a user close with no beacon) is left alone -
        // an explicit open can still reconnect to it. Skipped while closing so a
        // teardown in flight isn't double-ended. The grace runs from the later of
        // the launch and this daemon's start, so a session rehydrated on a restart
        // gets a full window to reconnect before it's swept. Two cases get the much
        // longer manual grace instead of the short one, because a human must act by
        // hand and the auto-clear must not race them: a pre-auth (keyless login)
        // entry, where the user may be typing a password on a sign-in page that is
        // not guaranteed to report as a browser connection; and a failed launch
        // (browserLaunched false), where the open reported open-failed with a URL
        // and told the user to click it manually.
        const connectDeadlineFrom = Math.max(entry.browserLaunchedAt, daemonStartedAt);
        const graceForEntry = entry.apiKey && entry.browserLaunched ? connectGrace : preAuthGrace;
        const neverConnected = !entry.browserEverConnected &&
            !entry.closing &&
            entry.windowClosedAt == null &&
            now - connectDeadlineFrom > graceForEntry;
        if (!closedOut && !neverConnected) {
            continue;
        }
        clearCloseFallbackTimer(entry);
        // The session is really over. Send the terminal signal through
        // broadcastEvent - the daemon's single event fan-out - so it reaches every
        // consumer the same way any other event does: pushed to connected clients
        // (unblocking a waiting command), appended to the durable log (unblocking a
        // log watcher), and focus-restored. Tag the never-connected case so the
        // reason isn't lost.
        broadcastEvent(entry, {
            id: "",
            type: "studio:session-ended",
            data: neverConnected ? { reason: "never-connected" } : {},
        });
        entry.pollAbort.abort();
        sessions.delete(key);
        // The daemon owns this session's persisted state, so it clears it here on
        // reap - both in-memory and on disk in one place - instead of leaving the
        // on-disk pointer for a command process to mark closed from the outside.
        // Without this a reaped never-connected session (whose disk state never got
        // a windowState:"closed" write) would be re-adopted on the next restart.
        clearSessionStateForKey(entry.key, entry.studioSessionId);
    }
    maybeIdleShutdown(now);
}
/**
 * Backstop reaper for an abandoned-but-not-deleted worktree: a dev daemon idle
 * (no sessions) for DEV_IDLE_TTL_MS exits on its own. Only dev daemons (those
 * with a bound meta path) - the prod singleton lives indefinitely.
 */
function maybeIdleShutdown(now) {
    if (!boundMetaPath) {
        return;
    }
    if (sessions.size > 0) {
        idleSince = null;
        return;
    }
    if (idleSince == null) {
        idleSince = now;
        return;
    }
    if (now - idleSince > DEV_IDLE_TTL_MS) {
        console.error("[daemon] idle past TTL, shutting down");
        gracefulShutdown();
    }
}
function gracefulShutdown() {
    for (const entry of sessions.values()) {
        clearCloseFallbackTimer(entry);
        entry.pollAbort.abort();
        for (const client of entry.connectedClients) {
            client.destroy();
        }
    }
    sessions.clear();
    if (healthTimer) {
        clearInterval(healthTimer);
    }
    try {
        fs.unlinkSync(boundSocketPath);
    }
    catch { }
    try {
        fs.unlinkSync(boundPidPath);
    }
    catch { }
    if (boundMetaPath) {
        try {
            fs.unlinkSync(boundMetaPath);
        }
        catch { }
    }
    process.exit(0);
}
function restoreSessionsFromDisk() {
    const states = readAllDaemonSessionStates();
    for (const state of states) {
        if (!state.agentSessionKey) {
            continue;
        }
        // Only adopt sessions from THIS daemon's worktree. The session store is
        // shared across worktrees (keyed by agent-session-id), so without this a
        // second worktree's daemon would restore - and later reap - another
        // worktree's windows. `?? null` keeps the prod singleton (null) restoring
        // its own untagged sessions.
        //
        // Deliberate consequence: a DEV daemon does NOT restore legacy sessions
        // written before this field existed (they read as null). Those are
        // unattributable to any worktree, so claiming them in a dev daemon would
        // reintroduce exactly the cross-worktree adoption bug above. The cost is a
        // one-time, per-session upgrade transition: a Studio session left open
        // across the upgrade isn't reattached, so the next open may spawn a fresh
        // window beside it. After that first open the session is re-tagged and
        // behaves normally.
        if ((state.worktreeRoot ?? null) !== boundWorktreeRoot) {
            continue;
        }
        if (sessions.has(state.agentSessionKey)) {
            continue;
        }
        // `windowState` is the ONLY signal that a window is gone. A window only
        // leaves "open" when the browser confirms it closed (studio:window-closed),
        // so a non-closed state means the window is still ours to reuse. We do NOT
        // gate on `windowPid`: the pid we record is usually a short-lived launcher
        // that hands the window to an already-running Chrome and then exits, so its
        // liveness says nothing about whether the window is open. If the window did
        // die out of band without a close beacon, the poll loop flips
        // browserConnected to false and the reconnect path in handleOpen relaunches
        // it under the SAME session rather than opening a duplicate.
        if (state.windowState === "closed") {
            continue;
        }
        const entry = {
            key: state.agentSessionKey,
            studioSessionId: state.sessionId,
            serviceUrl: state.serviceUrl,
            apiKey: "",
            browserPid: state.windowPid,
            // Rehydrated from a persisted pointer: the window was launched in a prior
            // process, so treat it as launched (we have no failure to report here).
            browserLaunched: true,
            browserConnected: state.browserConnected,
            // Use only the persisted connection status, never lastBrowserPingAt: the
            // poll loop stamps that timestamp on every status poll, including
            // connected=false ones, so it is not proof the window ever connected.
            // Deriving "ever connected" from it would flip browserEverConnected true
            // for a still-open-but-disconnected window (e.g. a sleeping tab) and let
            // the next open hit the relaunch path, spawning a duplicate window.
            browserEverConnected: state.browserConnected,
            browserDisconnectedAt: null,
            browserLaunchedAt: new Date(state.openedAt).getTime(),
            pollAbort: new AbortController(),
            pollActive: true,
            connectedClients: new Set(),
            lastActivity: Date.now(),
            clientHeaders: {},
            lastEventCursor: state.lastEventCursor,
            currentPath: state.currentPath,
            windowClosedAt: null,
            closing: false,
            closeFallbackTimer: null,
            focusTarget: null,
        };
        sessions.set(state.agentSessionKey, entry);
        startPollLoop(entry);
        console.error(`[daemon] restored session ${state.agentSessionKey} (studio: ${state.sessionId})`);
    }
}
export function startServer(socketPath = SOCKET_PATH, pidFilePath = PID_FILE_PATH, metaPath, worktreeRoot) {
    for (const entry of sessions.values()) {
        clearCloseFallbackTimer(entry);
        entry.pollAbort.abort();
    }
    sessions.clear();
    boundSocketPath = socketPath;
    boundPidPath = pidFilePath;
    boundMetaPath = metaPath ?? null;
    boundWorktreeRoot = worktreeRoot ?? null;
    idleSince = null;
    daemonStartedAt = Date.now();
    buildFingerprint = computeBuildFingerprint();
    restoreSessionsFromDisk();
    fs.mkdirSync(path.dirname(socketPath), { recursive: true });
    try {
        fs.unlinkSync(socketPath);
    }
    catch { }
    const server = net.createServer(handleConnection);
    server.listen(socketPath, () => {
        fs.writeFileSync(pidFilePath, `${process.pid}\n`);
        // Dev daemons record a meta file so the zombie sweep can tell live from
        // orphaned. The prod singleton (no metaPath) writes none.
        if (metaPath && worktreeRoot) {
            const meta = {
                pid: process.pid,
                worktreeRoot,
                socketPath,
                pidPath: pidFilePath,
            };
            try {
                fs.writeFileSync(metaPath, `${JSON.stringify(meta)}\n`);
            }
            catch { }
        }
        console.error(`[daemon] listening on ${socketPath} (pid ${process.pid})`);
    });
    server.on("error", (err) => {
        console.error(`[daemon] server error: ${err.message}`);
        process.exit(1);
    });
    healthTimer = setInterval(reapClosedSessions, reapIntervalMs());
    process.setMaxListeners(process.getMaxListeners() + 3);
    process.on("SIGTERM", gracefulShutdown);
    process.on("SIGINT", gracefulShutdown);
    // Reap signal (worktree going away): close windows too, unlike SIGTERM which
    // stays window-preserving so a build-mismatch restart can reconnect to them.
    process.on("SIGHUP", reapShutdown);
    return server;
}
