// bitfab-plugin-lib/dist/daemon/server.js
import crypto from "node:crypto";
import fs from "node:fs";
import net from "node:net";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { readAllDaemonSessionStates, updateSessionStateForKey, writeSessionStateForKey, } from "../activeStudioSession.js";
import { createAgentSession, navigateStudioAndAwaitAck, pollAgentSessionEvents, } from "../agentSessionChannel.js";
import { openChromelessWindow } from "../browser.js";
import { buildStudioSessionUrl, resolveStudioInitialPath, } from "../studioUrl.js";
import { BROWSER_HEALTH_INTERVAL_MS, PID_FILE_PATH, SOCKET_PATH, serialize, } from "./protocol.js";
const BROWSER_INITIAL_GRACE_MS = 60_000;
const BROWSER_DISCONNECT_THRESHOLD_MS = 30_000;
const WINDOW_CLOSE_GRACE_MS = 10_000;
function computeBuildFingerprint() {
    try {
        const thisFile = fileURLToPath(import.meta.url);
        const content = fs.readFileSync(thisFile);
        return crypto
            .createHash("sha256")
            .update(content)
            .digest("hex")
            .slice(0, 16);
    }
    catch {
        return "unknown";
    }
}
const sessions = new Map();
let healthTimer = null;
let buildFingerprint = null;
function buildSessionState(entry) {
    return {
        sessionId: entry.studioSessionId,
        serviceUrl: entry.serviceUrl,
        windowPid: entry.browserPid,
        pollerPid: null,
        source: "daemon",
        openedAt: new Date(entry.browserLaunchedAt).toISOString(),
        currentPath: null,
        authenticated: true,
        windowState: "open",
        agentIntent: "open",
        lastBrowserPingAt: null,
        browserConnected: entry.browserConnected,
        lastEventAt: null,
        lastEventCursor: entry.lastEventCursor,
        agentSessionKey: entry.key,
    };
}
function persistSessionState(entry) {
    try {
        writeSessionStateForKey(entry.key, buildSessionState(entry));
    }
    catch { }
}
function send(socket, msg) {
    try {
        socket.write(serialize(msg));
    }
    catch {
        // socket already closed
    }
}
const STATE_CHANGING_EVENTS = new Set([
    "studio:session-started",
    "studio:window-closed",
    "studio:session-ended",
    "studio:not-member",
    "studio:authenticated",
]);
function broadcastEvent(entry, event) {
    const push = {
        id: event.id,
        event: event.type,
        sessionId: entry.studioSessionId,
        key: entry.key,
        data: event.data,
    };
    for (const client of entry.connectedClients) {
        send(client, push);
    }
    if (STATE_CHANGING_EVENTS.has(event.type)) {
        const updates = {
            lastEventAt: new Date().toISOString(),
            lastEventCursor: event.id || entry.lastEventCursor,
        };
        if (event.type === "studio:window-closed") {
            entry.browserConnected = false;
            entry.browserDisconnectedAt = Date.now();
            entry.windowClosedAt = Date.now();
            updates.browserConnected = false;
            updates.windowState = "closed";
        }
        else if (event.type === "studio:not-member") {
            updates.authenticated = false;
        }
        else if (event.type === "studio:authenticated") {
            const token = typeof event.data === "object" &&
                event.data !== null &&
                "token" in event.data
                ? event.data.token
                : undefined;
            if (typeof token === "string" && !entry.apiKey) {
                entry.apiKey = token;
                entry.pollAbort.abort();
                entry.pollAbort = new AbortController();
                startPollLoop(entry);
            }
            updates.authenticated = true;
        }
        else if (event.type === "studio:session-started") {
            entry.browserConnected = true;
            entry.browserEverConnected = true;
            entry.browserDisconnectedAt = null;
            entry.windowClosedAt = null;
            updates.browserConnected = true;
            updates.windowState = "open";
        }
        try {
            updateSessionStateForKey(entry.key, updates);
        }
        catch { }
    }
}
function startPollLoop(entry) {
    ;
    (async () => {
        if (!entry.lastEventCursor && entry.apiKey) {
            try {
                await createAgentSession({
                    serviceUrl: entry.serviceUrl,
                    apiKey: entry.apiKey,
                    sessionId: entry.studioSessionId,
                    extraHeaders: entry.clientHeaders,
                });
            }
            catch (err) {
                console.error(`[daemon] createAgentSession failed for ${entry.key}: ${err instanceof Error ? err.message : String(err)}`);
                return;
            }
        }
        await pollAgentSessionEvents({
            serviceUrl: entry.serviceUrl,
            apiKey: entry.apiKey || undefined,
            sessionId: entry.studioSessionId,
            startCursor: entry.lastEventCursor ?? undefined,
            abortSignal: entry.pollAbort.signal,
            onEvent: (event) => {
                entry.lastEventCursor = event.id;
                entry.lastActivity = Date.now();
                broadcastEvent(entry, event);
            },
            onBrowserStatus: (connected) => {
                const wasConnected = entry.browserConnected;
                entry.browserConnected = connected;
                if (connected) {
                    entry.browserEverConnected = true;
                    entry.browserDisconnectedAt = null;
                }
                else if (wasConnected && !connected) {
                    entry.browserDisconnectedAt = Date.now();
                }
                try {
                    updateSessionStateForKey(entry.key, {
                        browserConnected: connected,
                        lastBrowserPingAt: new Date().toISOString(),
                    });
                }
                catch { }
            },
            onError: (err) => {
                console.error(`[daemon] poll error for ${entry.key}: ${err.message}`);
            },
        });
    })().catch((err) => {
        if (!entry.pollAbort.signal.aborted) {
            console.error(`[daemon] poll terminated for ${entry.key}: ${err.message}`);
        }
    });
}
function launchBrowser(entry, initialPath) {
    const studioPath = resolveStudioInitialPath(initialPath);
    const url = buildStudioSessionUrl({
        serviceUrl: entry.serviceUrl,
        path: studioPath,
        sessionId: entry.studioSessionId,
    });
    entry.browserPid = openChromelessWindow(url);
    entry.browserLaunchedAt = Date.now();
    entry.browserConnected = false;
    entry.browserDisconnectedAt = null;
    try {
        updateSessionStateForKey(entry.key, {
            windowPid: entry.browserPid,
            browserConnected: false,
            windowState: "open",
            currentPath: studioPath,
        });
    }
    catch { }
}
async function handleOpen(socket, cmd) {
    const existing = sessions.get(cmd.key);
    if (existing) {
        if (existing.windowClosedAt != null) {
            existing.pollAbort.abort();
            sessions.delete(cmd.key);
        }
        else {
            existing.connectedClients.add(socket);
            existing.lastActivity = Date.now();
            if (cmd.apiKey && !existing.apiKey) {
                existing.apiKey = cmd.apiKey;
            }
            const RELAUNCH_COOLDOWN_MS = 10_000;
            if (!existing.browserConnected &&
                existing.browserEverConnected &&
                Date.now() - (existing.browserLaunchedAt ?? 0) > RELAUNCH_COOLDOWN_MS) {
                launchBrowser(existing);
            }
            send(socket, {
                ok: true,
                sessionId: existing.studioSessionId,
                browserConnected: existing.browserConnected,
                reconnected: true,
            });
            return;
        }
    }
    const studioSessionId = cmd.sessionId ?? crypto.randomUUID();
    const entry = {
        key: cmd.key,
        studioSessionId,
        serviceUrl: cmd.serviceUrl,
        apiKey: cmd.apiKey ?? "",
        browserPid: null,
        browserConnected: false,
        browserEverConnected: false,
        browserDisconnectedAt: null,
        browserLaunchedAt: Date.now(),
        pollAbort: new AbortController(),
        pollActive: true,
        connectedClients: new Set([socket]),
        lastActivity: Date.now(),
        clientHeaders: cmd.clientHeaders ?? {},
        lastEventCursor: null,
        windowClosedAt: null,
    };
    launchBrowser(entry, cmd.initialPath);
    sessions.set(cmd.key, entry);
    startPollLoop(entry);
    persistSessionState(entry);
    send(socket, {
        ok: true,
        sessionId: studioSessionId,
        browserConnected: false,
        reconnected: false,
    });
}
async function handleNavigate(socket, cmd) {
    const entry = sessions.get(cmd.key);
    if (!entry) {
        send(socket, { ok: false, error: `no session for key: ${cmd.key}` });
        return;
    }
    try {
        const result = await navigateStudioAndAwaitAck({
            serviceUrl: entry.serviceUrl,
            apiKey: entry.apiKey,
            sessionId: entry.studioSessionId,
        }, cmd.path);
        entry.lastActivity = Date.now();
        if (result.acked) {
            try {
                updateSessionStateForKey(cmd.key, {
                    currentPath: cmd.path,
                    windowState: "open",
                });
            }
            catch { }
        }
        send(socket, { ok: true, acked: result.acked, reason: result.reason });
    }
    catch (err) {
        send(socket, {
            ok: false,
            error: `navigate failed: ${err instanceof Error ? err.message : String(err)}`,
        });
    }
}
function handleClose(socket, cmd) {
    const entry = sessions.get(cmd.key);
    if (!entry) {
        send(socket, { ok: false, error: `no session for key: ${cmd.key}` });
        return;
    }
    entry.connectedClients.add(socket);
    try {
        updateSessionStateForKey(cmd.key, {
            agentIntent: "close",
        });
    }
    catch { }
    send(socket, { ok: true });
}
function handleDetach(socket, cmd) {
    const entry = sessions.get(cmd.key);
    if (!entry) {
        send(socket, { ok: true });
        return;
    }
    entry.connectedClients.delete(socket);
    entry.lastActivity = Date.now();
    send(socket, { ok: true });
}
function handleStatus(socket) {
    const infos = [];
    for (const [key, entry] of sessions) {
        infos.push({
            key,
            sessionId: entry.studioSessionId,
            browserPid: entry.browserPid,
            browserConnected: entry.browserConnected,
            connectedClients: entry.connectedClients.size,
            lastActivity: entry.lastActivity,
        });
    }
    send(socket, { ok: true, sessions: infos });
}
function handlePing(socket) {
    send(socket, { ok: true, build: buildFingerprint });
}
function dispatch(socket, cmd) {
    switch (cmd.cmd) {
        case "open":
            handleOpen(socket, cmd).catch((err) => {
                send(socket, {
                    ok: false,
                    error: `open failed: ${err instanceof Error ? err.message : String(err)}`,
                });
            });
            break;
        case "navigate":
            handleNavigate(socket, cmd).catch((err) => {
                send(socket, {
                    ok: false,
                    error: `navigate failed: ${err instanceof Error ? err.message : String(err)}`,
                });
            });
            break;
        case "close":
            handleClose(socket, cmd);
            break;
        case "detach":
            handleDetach(socket, cmd);
            break;
        case "status":
            handleStatus(socket);
            break;
        case "ping":
            handlePing(socket);
            break;
        default:
            send(socket, {
                ok: false,
                error: `unknown command: ${cmd.cmd}`,
            });
    }
}
function removeClientFromAllSessions(socket) {
    for (const entry of sessions.values()) {
        entry.connectedClients.delete(socket);
    }
}
function handleConnection(socket) {
    let buffer = "";
    socket.on("data", (chunk) => {
        buffer += chunk.toString();
        let newlineIdx = buffer.indexOf("\n");
        while (newlineIdx !== -1) {
            const line = buffer.slice(0, newlineIdx).trim();
            buffer = buffer.slice(newlineIdx + 1);
            if (line) {
                try {
                    const cmd = JSON.parse(line);
                    dispatch(socket, cmd);
                }
                catch {
                    send(socket, { ok: false, error: "invalid JSON" });
                }
            }
            newlineIdx = buffer.indexOf("\n");
        }
    });
    socket.on("close", () => {
        removeClientFromAllSessions(socket);
    });
    socket.on("error", () => {
        removeClientFromAllSessions(socket);
    });
}
function checkBrowserHealth() {
    const now = Date.now();
    for (const [key, entry] of sessions) {
        if (entry.windowClosedAt != null &&
            now - entry.windowClosedAt > WINDOW_CLOSE_GRACE_MS) {
            entry.pollAbort.abort();
            sessions.delete(key);
            continue;
        }
        if (entry.browserConnected) {
            continue;
        }
        if (entry.connectedClients.size === 0) {
            continue;
        }
        if (!entry.browserEverConnected) {
            if (now - entry.browserLaunchedAt > BROWSER_INITIAL_GRACE_MS) {
                console.error(`[daemon] browser never connected for ${entry.key}, relaunching`);
                launchBrowser(entry);
            }
            continue;
        }
        const disconnectedFor = entry.browserDisconnectedAt
            ? now - entry.browserDisconnectedAt
            : 0;
        if (disconnectedFor > BROWSER_DISCONNECT_THRESHOLD_MS) {
            console.error(`[daemon] browser disconnected for ${entry.key} (${Math.round(disconnectedFor / 1000)}s), relaunching`);
            launchBrowser(entry);
            entry.browserEverConnected = false;
        }
    }
}
function gracefulShutdown() {
    for (const entry of sessions.values()) {
        entry.pollAbort.abort();
        for (const client of entry.connectedClients) {
            client.destroy();
        }
    }
    sessions.clear();
    if (healthTimer) {
        clearInterval(healthTimer);
    }
    try {
        fs.unlinkSync(SOCKET_PATH);
    }
    catch { }
    try {
        fs.unlinkSync(PID_FILE_PATH);
    }
    catch { }
    process.exit(0);
}
function isProcessAlive(pid) {
    try {
        process.kill(pid, 0);
        return true;
    }
    catch (err) {
        return err.code === "EPERM";
    }
}
function restoreSessionsFromDisk() {
    const states = readAllDaemonSessionStates();
    for (const state of states) {
        if (!state.agentSessionKey) {
            continue;
        }
        if (sessions.has(state.agentSessionKey)) {
            continue;
        }
        if (state.windowState === "closed") {
            continue;
        }
        if (state.windowPid != null && !isProcessAlive(state.windowPid)) {
            continue;
        }
        const entry = {
            key: state.agentSessionKey,
            studioSessionId: state.sessionId,
            serviceUrl: state.serviceUrl,
            apiKey: "",
            browserPid: state.windowPid,
            browserConnected: state.browserConnected,
            browserEverConnected: state.browserConnected,
            browserDisconnectedAt: null,
            browserLaunchedAt: new Date(state.openedAt).getTime(),
            pollAbort: new AbortController(),
            pollActive: true,
            connectedClients: new Set(),
            lastActivity: Date.now(),
            clientHeaders: {},
            lastEventCursor: state.lastEventCursor,
            windowClosedAt: null,
        };
        sessions.set(state.agentSessionKey, entry);
        startPollLoop(entry);
        console.error(`[daemon] restored session ${state.agentSessionKey} (studio: ${state.sessionId})`);
    }
}
export function startServer(socketPath = SOCKET_PATH, pidFilePath = PID_FILE_PATH) {
    for (const entry of sessions.values()) {
        entry.pollAbort.abort();
    }
    sessions.clear();
    buildFingerprint = computeBuildFingerprint();
    restoreSessionsFromDisk();
    fs.mkdirSync(path.dirname(socketPath), { recursive: true });
    try {
        fs.unlinkSync(socketPath);
    }
    catch { }
    const server = net.createServer(handleConnection);
    server.listen(socketPath, () => {
        fs.writeFileSync(pidFilePath, `${process.pid}\n`);
        console.error(`[daemon] listening on ${socketPath} (pid ${process.pid})`);
    });
    server.on("error", (err) => {
        console.error(`[daemon] server error: ${err.message}`);
        process.exit(1);
    });
    healthTimer = setInterval(checkBrowserHealth, BROWSER_HEALTH_INTERVAL_MS);
    process.setMaxListeners(process.getMaxListeners() + 2);
    process.on("SIGTERM", gracefulShutdown);
    process.on("SIGINT", gracefulShutdown);
    return server;
}
