// bitfab-plugin-lib/dist/commands/clearStudioSession.js
import { clearActiveStudioSession, readActiveStudioSession, } from "../activeStudioSession.js";
import { closeStudioWindowsBySession } from "../browser.js";
import { emitJson } from "../output.js";
import { positionalArgs } from "../parseArgs.js";
import { resolveChannel } from "../studioChannel.js";
import { killPollerProcess, killStudioWindow } from "../studioTeardown.js";
/**
 * The reap-and-forget teardown for one Studio session, callable in-process.
 * Best-effort throughout: reaps the `--app` window and event poller, asks the
 * active channel to forget the session (so a daemon's in-memory entry can't
 * reconnect to a dead window), then force-clears the on-disk pointer. Returns
 * whether the pointer was actually cleared (false when it named a different
 * session). `runClearStudioSession` (the command) and `bitfab login --force`
 * both call this so the teardown lives in exactly one place.
 */
export async function clearStudioSessionById(sessionId, record) {
    // Reap BEFORE clearing: the kills resolve their pids from the record, which
    // is gone afterwards. Both helpers re-check the sessionId, so they no-op if
    // the record was re-bound (and the clear below then reports not-found).
    killStudioWindow(sessionId);
    // The pid kill is a no-op when the browser was already running (dead launcher
    // pid), so also best-effort close the window by URL match - unless the record
    // already shows this session closed, in which case nothing is left to close.
    const alreadyClosed = record?.sessionId === sessionId && record.windowState === "closed";
    if (!alreadyClosed) {
        closeStudioWindowsBySession(sessionId);
    }
    killPollerProcess(sessionId);
    // Forget the session in whichever channel backs Studio (the on-disk reaps
    // above don't reach the daemon's in-memory entry). Best-effort: never let a
    // daemon/channel hiccup block clearing the pointer below.
    try {
        const channel = await resolveChannel("", record?.serviceUrl ?? "");
        try {
            await channel.clearSession(sessionId);
        }
        finally {
            channel.destroy();
        }
    }
    catch { }
    return clearActiveStudioSession(sessionId, { force: true });
}
/**
 * Explicit recovery for a stale Studio pointer. `openStudioTo` gates (emits
 * `not-responding`) when the recorded window is believed live but unreachable
 * (crash, sleep, or a close no process witnessed); it never replaces it
 * automatically. After the user approves recovery, run this to remove the
 * pointer, then re-run `openStudioTo.js` to open a fresh window.
 *
 * Clearing means the user is abandoning the recorded session, so reap its
 * processes too (both best-effort, session-matched): the `--app` window (if
 * it was alive-but-unreachable, it would otherwise linger as a zombie next to
 * the fresh one) and the event poller still attached to the dead session.
 *
 * Clearing the on-disk pointer alone is not enough: in daemon mode the channel
 * holds its own in-memory SessionEntry that survives a crash/sleep (no
 * `window-closed` was witnessed), so the next open would reconnect to the dead
 * window and gate `not-responding` forever. We ask the active channel to forget
 * the session too (`clearSession`); the daemon drops its entry, the direct
 * channel no-ops (it keeps no cross-process state).
 *
 * With a `sessionId` arg, clears only that session's pointer; with none, clears
 * whatever is currently recorded for this scope. Always a deliberate (forced)
 * clear, but still matched on sessionId so a different session is never touched.
 */
export async function runClearStudioSession() {
    const ARGS = [
        {
            name: "sessionId",
            required: false,
            format: "uuid",
            description: "Studio session whose pointer to clear. Defaults to whatever session is currently recorded.",
        },
    ];
    const args = positionalArgs(ARGS, process.argv, {
        description: "Clear a stale Studio window pointer so a fresh window can be opened. Run after openStudioTo reports not-responding and the user approves recovery. Emits one JSON line: cleared, not-found, or no-active-studio.",
    });
    const record = readActiveStudioSession();
    const sessionId = args[0] ?? record?.sessionId ?? null;
    if (!sessionId) {
        emitJson({ event: "no-active-studio" });
        return;
    }
    // Report what actually happened: the pointer clear no-ops when the recorded
    // pointer names a different session, so emitting `cleared` unconditionally
    // would falsely claim success while a stale pointer remains.
    const cleared = await clearStudioSessionById(sessionId, record);
    emitJson({ event: cleared ? "cleared" : "not-found", sessionId });
}
