// bitfab-plugin-lib/dist/commands/closeStudio.js
import { clearActiveStudioSession, readActiveStudioSession, } from "../activeStudioSession.js";
import { closeStudio, fetchStreamTip, pollAgentSessionEvents, } from "../agentSessionChannel.js";
import { getConfig, hasCredentials } from "../config.js";
import { emitJson } from "../output.js";
import { positionalArgs } from "../parseArgs.js";
import { createWindowCloseArbiter, killPollerProcess, killStudioWindow, } from "../studioTeardown.js";
// How long to wait for the browser to confirm the window actually closed
// (`studio:window-closed`, the `beforeunload` beacon) before giving up. If it
// never arrives (the browser refused `window.close()` and the pid kill was a
// no-op), we leave the pointer in place rather than clear a window that may
// still be alive — the next open reuses it.
const CLOSE_CONFIRM_TIMEOUT_MS = 8_000;
// How long to give the web's own `window.close()` before falling back to
// SIGTERMing the spawned `--app` process.
const WINDOW_KILL_DELAY_MS = 2_500;
/**
 * Resolve true once a `studio:window-closed` survives the refresh grace
 * window (no `session-started` re-bind follows) — the shared arbiter in
 * `studioTeardown.ts` owns that disambiguation. Polls from `startCursor`
 * (the pre-close stream tip) so replayed historical events are ignored — only
 * a fresh teardown confirmation counts. `session-ended` / `session-closed`
 * are NOT confirmations: the web emits them before (or regardless of) the
 * `window.close()` attempt the browser can refuse.
 *
 * The overall deadline bounds how long we wait for a candidate
 * `window-closed`; a candidate already in its grace window when the deadline
 * fires is allowed to finish deciding (confirm on grace expiry, refute on a
 * reload's `session-started`).
 */
function awaitWindowClosed(client, startCursor) {
    return new Promise((resolve) => {
        const abortController = new AbortController();
        let timedOut = false;
        const finish = (closed) => {
            arbiter.cancel();
            clearTimeout(deadline);
            abortController.abort();
            resolve(closed);
        };
        const arbiter = createWindowCloseArbiter({
            onConfirmed: () => finish(true),
            onRefuted: () => {
                // The reload re-bound: that window-closed was a refresh, not a
                // close. Keep waiting (unless the deadline already passed).
                if (timedOut) {
                    finish(false);
                }
            },
        });
        const deadline = setTimeout(() => {
            timedOut = true;
            // Let a pending grace window decide; with no candidate, give up
            // without proof (the pointer is kept).
            if (!arbiter.hasPendingCandidate()) {
                finish(false);
            }
        }, CLOSE_CONFIRM_TIMEOUT_MS);
        pollAgentSessionEvents({
            ...client,
            startCursor,
            abortSignal: abortController.signal,
            onEvent: (event) => arbiter.handleEvent(event),
            onError: () => { },
        }).catch(() => {
            finish(false);
        });
    });
}
export async function runCloseStudio() {
    // Usage: closeStudio [message]
    // The session always comes from the on-disk pointer (the same single source
    // of truth `openStudioTo` uses), which makes the bare call safe to run
    // unconditionally at every flow exit (it no-ops when nothing is open).
    const ARGS = [
        {
            name: "message",
            required: false,
            description: "Optional farewell message shown in Studio before closing.",
        },
    ];
    const message = positionalArgs(ARGS, process.argv, {
        description: "Close the active Studio window (session read from the on-disk pointer) and clear its pointer. Safe to run when nothing is open (no-op).",
    })[0];
    const sessionId = readActiveStudioSession()?.sessionId ?? null;
    if (!sessionId) {
        emitJson({ event: "no-active-studio" });
        process.exit(0);
    }
    const config = getConfig();
    if (!hasCredentials() || !config.apiKey) {
        // Can't push the close event or confirm a teardown without credentials.
        // Still reap a lingering poller (it would just error-loop), but leave the
        // pointer and the window alone.
        killPollerProcess(sessionId);
        emitJson({ event: "close-skipped", sessionId, reason: "not-authenticated" });
        process.exit(0);
    }
    const client = {
        serviceUrl: config.serviceUrl,
        apiKey: config.apiKey,
        sessionId,
    };
    // Capture the stream tip first, then listen from there and send the close,
    // so we only react to the `window-closed` our own close produces — not one
    // replayed from history (which would clear a live tab's pointer).
    let closed;
    try {
        const tip = await fetchStreamTip(client);
        closed = awaitWindowClosed(client, tip);
        await closeStudio(client, message);
    }
    catch {
        // Session already gone (expired/unbound) — nothing to close. Leave the
        // pointer alone but reap a lingering poller; closing an already-closed
        // session is a no-op, not an error.
        killPollerProcess(sessionId);
        emitJson({ event: "close-skipped", sessionId, reason: "session-gone" });
        process.exit(0);
    }
    // Give the web's own `window.close()` a moment to be confirmed; if no proof
    // arrives, SIGTERM the spawned `--app` process as a second best-effort
    // close, then keep waiting for the confirmation it may produce.
    const quickConfirm = await Promise.race([
        closed,
        new Promise((resolve) => {
            setTimeout(() => resolve(null), WINDOW_KILL_DELAY_MS);
        }),
    ]);
    if (quickConfirm === null) {
        killStudioWindow(sessionId);
    }
    // Clear the pointer ONLY on proof the window is gone (`window-closed`). If
    // neither close attempt was confirmed within the timeout, leave the pointer
    // in place: the window may still be alive, and clearing it would let the
    // next open spawn a duplicate. `force` skips the pid guard (a different
    // process wrote the record) but still matches on sessionId.
    const confirmed = await closed;
    // Reap the background poller BEFORE clearing the pointer (the kill resolves
    // its pid from the record). On a confirmed close it would self-exit anyway;
    // on a refused close this is what stops it outliving the flow.
    killPollerProcess(sessionId);
    if (confirmed) {
        clearActiveStudioSession(sessionId, { force: true });
    }
    emitJson({
        event: confirmed ? "closed" : "close-unconfirmed",
        sessionId,
    });
    process.exit(0);
}
