// bitfab-plugin-lib/dist/commands/closeStudio.js
import { clearActiveStudioSession } from "../activeStudioSession.js";
import { closeStudio, pollAgentSessionEvents } from "../agentSessionChannel.js";
import { getConfig, hasCredentials } from "../config.js";
import { positionalArgs } from "../parseArgs.js";
// How long to wait for the browser to confirm the tab actually closed
// (`studio:session-closed`) before giving up. The web emits it from
// `closeStudioTab` and the `beforeunload` beacon, so it normally arrives within
// a second; if it never does (a refused `window.close()`), we leave the pointer
// in place rather than clear a window that may still be alive.
const CLOSE_CONFIRM_TIMEOUT_MS = 8_000;
/**
 * Read the current tail of the session's event stream and return the latest
 * event id (the "tip"), or undefined if the stream is empty/unreachable.
 * `awaitSessionClosed` starts polling AFTER this id so it only reacts to events
 * produced by our own close — never a `studio:session-closed` replayed from
 * history (e.g. a prior refresh's `beforeunload`), which would otherwise clear a
 * pointer to a still-live tab and let the next open spawn a duplicate.
 */
async function fetchStreamTip(client) {
    try {
        const url = new URL(`${client.serviceUrl}/api/studio/events`);
        url.searchParams.set("session", client.sessionId);
        const res = await fetch(url.toString(), {
            headers: { Authorization: `Bearer ${client.apiKey}` },
        });
        if (!res.ok) {
            return undefined;
        }
        const body = (await res.json());
        return body.events?.at(-1)?.id;
    }
    catch {
        return undefined;
    }
}
/**
 * Resolve true if a `studio:session-closed` arrives before the timeout. Polls
 * from `startCursor` (the pre-close stream tip) so replayed historical events
 * are ignored — only a fresh close confirmation counts.
 */
function awaitSessionClosed(client, startCursor) {
    return new Promise((resolve) => {
        const abortController = new AbortController();
        const timer = setTimeout(() => {
            abortController.abort();
            resolve(false);
        }, CLOSE_CONFIRM_TIMEOUT_MS);
        pollAgentSessionEvents({
            ...client,
            startCursor,
            abortSignal: abortController.signal,
            onEvent: (event) => {
                if (event.type === "studio:session-closed") {
                    clearTimeout(timer);
                    abortController.abort();
                    resolve(true);
                }
            },
            onError: () => { },
        }).catch(() => {
            clearTimeout(timer);
            resolve(false);
        });
    });
}
export async function runCloseStudio() {
    const ARGS = [
        { name: "sessionId", format: "uuid" },
        { name: "message", required: false },
    ];
    const args = positionalArgs(ARGS);
    const sessionId = args[0];
    const message = args[1];
    if (!hasCredentials()) {
        console.error("Not authenticated.");
        process.exit(1);
    }
    const config = getConfig();
    if (!config.apiKey) {
        console.error("No API key configured.");
        process.exit(1);
    }
    const client = {
        serviceUrl: config.serviceUrl,
        apiKey: config.apiKey,
        sessionId,
    };
    // Capture the stream tip first, then listen from there and send the close, so
    // we only react to the `session-closed` our own close produces — not one
    // replayed from history (which would clear a live tab's pointer).
    const tip = await fetchStreamTip(client);
    const closed = awaitSessionClosed(client, tip);
    await closeStudio(client, message);
    // Clear the pointer ONLY on proof the window is gone (`session-closed`). If
    // the browser refused to close the tab (no `session-closed` within the
    // timeout), leave the pointer in place: the window may still be alive, and
    // clearing it would let the next open spawn a duplicate. `force` skips the pid
    // guard (a different process wrote the record) but still matches on sessionId.
    if (await closed) {
        clearActiveStudioSession(sessionId, { force: true });
    }
    process.exit(0);
}
