// bitfab-plugin-lib/dist/commands/closeStudio.js
import { clearActiveStudioSession, readActiveStudioSession, } from "../activeStudioSession.js";
import { closeStudio, fetchStreamTip, pollAgentSessionEvents, } from "../agentSessionChannel.js";
import { getConfig, hasCredentials } from "../config.js";
import { positionalArgs } from "../parseArgs.js";
// How long to wait for the browser to confirm the window actually closed
// (`studio:window-closed`, the `beforeunload` beacon) before giving up. If it
// never arrives (the browser refused `window.close()` and the pid kill was a
// no-op), we leave the pointer in place rather than clear a window that may
// still be alive — the next open reuses it.
const CLOSE_CONFIRM_TIMEOUT_MS = 8_000;
// How long to give the web's own `window.close()` before falling back to
// SIGTERMing the spawned `--app` process.
const WINDOW_KILL_DELAY_MS = 2_500;
// A `window-closed` also fires on a page refresh (followed by a fresh
// `session-started` from the reload), so it only counts as proof after this
// grace period passes without a reload re-binding — the same disambiguation
// the openStudioTo event poller applies.
const WINDOW_CLOSE_GRACE_PERIOD_MS = 10_000;
/**
 * Resolve true once a `studio:window-closed` survives the refresh grace
 * window (no `session-started` re-bind follows). Polls from `startCursor`
 * (the pre-close stream tip) so replayed historical events are ignored — only
 * a fresh teardown confirmation counts. `session-ended` / `session-closed`
 * are NOT confirmations: the web emits them before (or regardless of) the
 * `window.close()` attempt the browser can refuse.
 *
 * The overall deadline bounds how long we wait for a candidate
 * `window-closed`; a candidate already in its grace window when the deadline
 * fires is allowed to finish deciding (confirm on grace expiry, refute on a
 * reload's `session-started`).
 */
function awaitWindowClosed(client, startCursor) {
    return new Promise((resolve) => {
        const abortController = new AbortController();
        let graceTimer = null;
        let timedOut = false;
        const finish = (closed) => {
            if (graceTimer) {
                clearTimeout(graceTimer);
                graceTimer = null;
            }
            clearTimeout(deadline);
            abortController.abort();
            resolve(closed);
        };
        const deadline = setTimeout(() => {
            timedOut = true;
            // Let a pending grace window decide; with no candidate, give up
            // without proof (the pointer is kept).
            if (!graceTimer) {
                finish(false);
            }
        }, CLOSE_CONFIRM_TIMEOUT_MS);
        pollAgentSessionEvents({
            ...client,
            startCursor,
            abortSignal: abortController.signal,
            onEvent: (event) => {
                if (event.type === "studio:window-closed") {
                    // Possibly a refresh: wait the grace window so a session-started
                    // from the reload can refute it.
                    if (graceTimer) {
                        clearTimeout(graceTimer);
                    }
                    graceTimer = setTimeout(() => {
                        graceTimer = null;
                        finish(true);
                    }, WINDOW_CLOSE_GRACE_PERIOD_MS);
                    return;
                }
                if (event.type === "studio:session-started" && graceTimer) {
                    // The reload re-bound: that window-closed was a refresh, not a
                    // close. Keep waiting (unless the deadline already passed).
                    clearTimeout(graceTimer);
                    graceTimer = null;
                    if (timedOut) {
                        finish(false);
                    }
                }
            },
            onError: () => { },
        }).catch(() => {
            finish(false);
        });
    });
}
/**
 * Best-effort kill of the spawned `--app` window process recorded in the
 * pointer. Genuinely best-effort: when the browser was already running at
 * spawn time, the recorded pid is a dead launcher that handed off to the
 * existing instance, so the kill is a no-op (the common case). Correctness
 * never depends on it — the pointer clears only on a confirmed
 * `studio:window-closed`.
 */
function killStudioWindow(sessionId) {
    const record = readActiveStudioSession();
    if (record?.sessionId !== sessionId || record.windowPid == null) {
        return;
    }
    try {
        process.kill(record.windowPid, "SIGTERM");
    }
    catch {
        // Already gone (or a dead launcher pid) — nothing to kill.
    }
}
export async function runCloseStudio() {
    const ARGS = [
        { name: "sessionId", format: "uuid" },
        { name: "message", required: false },
    ];
    const args = positionalArgs(ARGS);
    const sessionId = args[0];
    const message = args[1];
    if (!hasCredentials()) {
        console.error("Not authenticated.");
        process.exit(1);
    }
    const config = getConfig();
    if (!config.apiKey) {
        console.error("No API key configured.");
        process.exit(1);
    }
    const client = {
        serviceUrl: config.serviceUrl,
        apiKey: config.apiKey,
        sessionId,
    };
    // Capture the stream tip first, then listen from there and send the close,
    // so we only react to the `window-closed` our own close produces — not one
    // replayed from history (which would clear a live tab's pointer).
    const tip = await fetchStreamTip(client);
    const closed = awaitWindowClosed(client, tip);
    try {
        await closeStudio(client, message);
    }
    catch {
        // Session already gone (expired/unbound) — nothing to close. Leave the
        // pointer alone and exit quietly; closing an already-closed session is a
        // no-op, not an error.
        process.exit(0);
    }
    // Give the web's own `window.close()` a moment to be confirmed; if no proof
    // arrives, SIGTERM the spawned `--app` process as a second best-effort
    // close, then keep waiting for the confirmation it may produce.
    const quickConfirm = await Promise.race([
        closed,
        new Promise((resolve) => {
            setTimeout(() => resolve(null), WINDOW_KILL_DELAY_MS);
        }),
    ]);
    if (quickConfirm === null) {
        killStudioWindow(sessionId);
    }
    // Clear the pointer ONLY on proof the window is gone (`window-closed`). If
    // neither close attempt was confirmed within the timeout, leave the pointer
    // in place: the window may still be alive, and clearing it would let the
    // next open spawn a duplicate. `force` skips the pid guard (a different
    // process wrote the record) but still matches on sessionId.
    if (await closed) {
        clearActiveStudioSession(sessionId, { force: true });
    }
    process.exit(0);
}
