// bitfab-plugin-lib/dist/commands/detectCapabilities.js
import { detectReplayCapabilities, isReplayLanguage, } from "../replayCapabilities.js";
import { checkForSdkUpdates, checkForSdkUpdatesOffline, } from "../sdkUpdates.js";
/**
 * Build the per-SDK replay-capability record from a resolved update status.
 * Pure (no I/O).
 *
 * Capabilities and `versionResolved` are gated on `resolvedVersion` (from a
 * lockfile or the installed package), NOT `current` (which falls back to the
 * manifest's declared spec). A declared spec is a range/lower-bound, not a
 * confirmation of what's installed, so trusting it can over-report capabilities
 * the installed SDK lacks. When only a declared spec exists, the SDK isn't
 * confirmed installed, so capabilities stay unverified. `current`/`latest`
 * remain best-known values for the upgrade hint.
 */
export function buildReplayCapabilityRecord(status, language) {
    return {
        workspacePath: status.workspacePath,
        language,
        packageName: status.packageName,
        current: status.current,
        latest: status.latest,
        versionResolved: status.resolvedVersion !== null,
        updateAvailable: status.updateAvailable,
        renameFrom: status.renameFrom,
        manifestPath: status.manifestPath,
        lockfilePath: status.lockfilePath,
        capabilities: detectReplayCapabilities(language, status.resolvedVersion),
    };
}
function missingCapabilityLabels(caps) {
    const labels = [
        ["supportsCodeChanges", "code changes"],
        ["supportsExperimentGroups", "experiment groups"],
        ["supportsDatasetId", "dataset attribution"],
        ["supportsReplayTraceIds", "replay trace IDs"],
        ["supportsInputAdapters", "input adapters"],
    ];
    return labels.filter(([key]) => !caps[key]).map(([, label]) => label);
}
function printSummary(records) {
    for (const r of records) {
        const prefix = r.workspacePath && r.workspacePath !== "." ? `${r.workspacePath} ` : "";
        const version = r.current ?? "unknown";
        if (!r.versionResolved) {
            console.log(`  ${prefix}${r.language} (${r.packageName}) — couldn't confirm the installed version ` +
                "from a lockfile or installed package (a manifest version spec alone isn't a confirmed " +
                "install). Capabilities are unverified: install the SDK (or generate a lockfile / activate " +
                "the project's environment) and re-run, or check the installed package by hand.");
            continue;
        }
        const missing = missingCapabilityLabels(r.capabilities);
        const upgrade = r.updateAvailable ? ` (update available → ${r.latest})` : "";
        if (missing.length === 0) {
            console.log(`  ${prefix}${r.language} (${r.packageName}) v${version} — full replay capability${upgrade}`);
        }
        else {
            console.log(`  ${prefix}${r.language} (${r.packageName}) v${version} — missing: ${missing.join(", ")}${upgrade}`);
        }
    }
}
/**
 * Detect replay capabilities of the installed Bitfab SDK(s) in the current
 * directory, plus their version-update status. Emits a machine-readable
 * `<bitfab-replay-capabilities>` block (one JSON line per SDK) followed by a
 * human-readable summary, mirroring the `update` command's contract.
 *
 * Capabilities are gated on the resolved installed version against a capability
 * floor map (see replayCapabilities.ts): deterministic, uniform across
 * TypeScript/Python/Ruby, and free of package-manager-layout guessing.
 */
export async function runDetectCapabilities(cwd = process.cwd()) {
    const offline = checkForSdkUpdatesOffline(cwd).filter((s) => isReplayLanguage(s.language));
    if (offline.length === 0) {
        console.log("No Bitfab SDK with replay support (TypeScript, Python, or Ruby) detected in the current directory.");
        return;
    }
    let statuses;
    try {
        statuses = await checkForSdkUpdates(cwd);
    }
    catch {
        statuses = checkForSdkUpdatesOffline(cwd);
    }
    const records = [];
    for (const status of statuses) {
        if (!isReplayLanguage(status.language)) {
            continue;
        }
        records.push(buildReplayCapabilityRecord(status, status.language));
    }
    console.log("<bitfab-replay-capabilities>");
    for (const r of records) {
        console.log(JSON.stringify(r));
    }
    console.log("</bitfab-replay-capabilities>");
    console.log("\nReplay capability status:");
    printSummary(records);
}
