// bitfab-plugin-lib/dist/commands/loginForSession.js
import { writeActiveStudioSession } from "../activeStudioSession.js";
import { pollLoginEvents } from "../agentSessionChannel.js";
import { saveCredentials } from "../config.js";
import { openStudioTo } from "./openStudioTo.js";
const LOGIN_TIMEOUT_MS = 10 * 60 * 1000;
const SESSION_READY_TIMEOUT_MS = 30_000;
const SESSION_READY_POLL_MS = 500;
/**
 * Inline login for commands that open a Studio page while unauthenticated.
 * Opens sign-in with a redirect back to `targetPath`, polls for the token,
 * saves credentials, waits for the browser's event channel to be live
 * (`studio:session-started`), then writes the active studio session so the
 * caller can navigate the existing tab via `openStudioTo`.
 *
 * Returns the API key.
 */
export async function loginForSession(opts) {
    const separator = opts.targetPath.includes("?") ? "&" : "?";
    const redirectPath = `${opts.targetPath}${separator}session=${opts.sessionId}&pluginLogin=true`;
    const signInPath = `/studio/sign-in?redirect_url=${encodeURIComponent(redirectPath)}`;
    console.error("Opening Studio to sign in...");
    console.error("(If the browser didn't open, visit the Studio URL manually.)");
    await openStudioTo(signInPath, {
        apiKey: null,
        serviceUrl: opts.serviceUrl,
        sessionId: opts.sessionId,
    });
    const abortController = new AbortController();
    const timer = setTimeout(() => abortController.abort(), LOGIN_TIMEOUT_MS);
    const keepaliveTimer = setInterval(() => {
        process.stderr.write("");
    }, 30_000);
    try {
        const token = await pollLoginEvents({
            serviceUrl: opts.serviceUrl,
            sessionId: opts.sessionId,
            abortSignal: abortController.signal,
        });
        clearTimeout(timer);
        clearInterval(keepaliveTimer);
        saveCredentials(token);
        await waitForSessionReady({
            serviceUrl: opts.serviceUrl,
            apiKey: token,
            sessionId: opts.sessionId,
        });
        writeActiveStudioSession({
            sessionId: opts.sessionId,
            serviceUrl: opts.serviceUrl,
        });
        console.error("Authenticated.");
        return token;
    }
    catch (err) {
        clearTimeout(timer);
        clearInterval(keepaliveTimer);
        throw err;
    }
}
async function waitForSessionReady(opts) {
    const deadline = Date.now() + SESSION_READY_TIMEOUT_MS;
    let since = null;
    while (Date.now() < deadline) {
        try {
            const url = new URL(`${opts.serviceUrl}/api/studio/events`);
            url.searchParams.set("session", opts.sessionId);
            if (since) {
                url.searchParams.set("since", since);
            }
            const res = await fetch(url.toString(), {
                headers: { Authorization: `Bearer ${opts.apiKey}` },
            });
            if (res.ok) {
                const body = (await res.json());
                for (const event of body.events) {
                    since = event.id;
                    if (event.type === "studio:session-started") {
                        return;
                    }
                }
            }
        }
        catch {
            // transient network error, retry
        }
        await new Promise((resolve) => setTimeout(resolve, SESSION_READY_POLL_MS));
    }
    console.error("Warning: Studio session did not confirm readiness within 30s. Proceeding anyway.");
}
