// bitfab-plugin-lib/dist/commands/login.js
import crypto from "node:crypto";
import { reportHandoff } from "../analytics.js";
import { openChromelessWindow } from "../browser.js";
import { getConfig, saveCredentials } from "../config.js";
import { recordFocus } from "../frontmostApp.js";
const LOGIN_TIMEOUT_MS = 10 * 60 * 1000;
const POLL_INTERVAL_MS = 1500;
async function verifyToken(serviceUrl, token) {
    try {
        const res = await fetch(`${serviceUrl}/api/plugin/whoami`, {
            headers: { Authorization: `Bearer ${token}` },
        });
        if (!res.ok) {
            return null;
        }
        return (await res.json());
    }
    catch {
        return null;
    }
}
async function pollLoginStatus(serviceUrl, loginId, signal) {
    while (!signal.aborted) {
        try {
            const url = `${serviceUrl}/api/studio/login-status?loginId=${encodeURIComponent(loginId)}`;
            const res = await fetch(url, { signal });
            if (res.ok) {
                const body = (await res.json());
                if (body.status === "complete" && body.token) {
                    return body.token;
                }
            }
        }
        catch {
            if (signal.aborted) {
                throw new Error("Login polling aborted");
            }
        }
        await new Promise((resolve) => {
            const timer = setTimeout(resolve, POLL_INTERVAL_MS);
            signal.addEventListener("abort", () => {
                clearTimeout(timer);
                resolve();
            }, { once: true });
        });
    }
    throw new Error("Login polling aborted");
}
export async function runLogin(platform, pluginVersion, options) {
    const exitOnComplete = options?.exitOnComplete ?? true;
    const config = getConfig();
    const restoreFocus = recordFocus();
    const loginId = crypto.randomUUID();
    const redirectUrl = `/studio/close?pluginLogin=true&loginId=${loginId}`;
    const browserUrl = `${config.serviceUrl}/studio/sign-in?redirect_url=${encodeURIComponent(redirectUrl)}`;
    console.log("\nOpening Studio to sign in...");
    console.log(`  ${browserUrl}`);
    console.log("");
    console.log("(If the browser didn't open, visit the URL above manually.)");
    openChromelessWindow(browserUrl);
    const abortController = new AbortController();
    let loginTimer;
    const timeoutPromise = new Promise((_resolve, reject) => {
        loginTimer = setTimeout(() => {
            abortController.abort();
            reject(new Error("Authentication timed out after 10 minutes. Run login again."));
        }, LOGIN_TIMEOUT_MS);
    });
    // Keepalive heartbeat so agent shell tools don't classify this as idle
    const keepaliveTimer = setInterval(() => {
        process.stdout.write("");
    }, 30_000);
    try {
        const token = await Promise.race([
            pollLoginStatus(config.serviceUrl, loginId, abortController.signal),
            timeoutPromise,
        ]);
        if (loginTimer) {
            clearTimeout(loginTimer);
        }
        clearInterval(keepaliveTimer);
        saveCredentials(token);
        const who = await verifyToken(config.serviceUrl, token);
        const greeting = who?.user.email
            ? `Logged in as ${who.user.email}.`
            : "Authentication successful.";
        console.log(`\n${greeting}`);
        console.log("Bitfab MCP tools are now active. (via studio)");
        await reportHandoff(token, pluginVersion, platform, {
            flow: "auth.login",
            wonBy: "studio",
            loopbackAvailable: false,
            ticketAvailable: false,
        });
        restoreFocus();
        if (exitOnComplete) {
            process.exit(0);
        }
    }
    catch (err) {
        if (loginTimer) {
            clearTimeout(loginTimer);
        }
        clearInterval(keepaliveTimer);
        abortController.abort();
        if (exitOnComplete) {
            console.error(`\n${err.message}`);
            process.exit(1);
        }
        throw err;
    }
}
