// bitfab-plugin-lib/dist/commands/login.js
import { reportHandoff } from "../analytics.js";
import { spawnDetached } from "../browser.js";
import { getConfig, hasCredentials } from "../config.js";
import { openStudioTo, runLoginWindowMonitor, StudioNavigationError, } from "./openStudioTo.js";
// Re-invoke this exact CLI invocation as a DETACHED child, with
// `--drain <sessionId>` appended, so the child does nothing but run the login
// window monitor. This lets the parent login process exit the instant sign-in
// completes, while the child stays alive just long enough to witness the
// window's auto-close and clear the pointer. Returns false if the spawn could
// not even be attempted (e.g. a restrictive sandbox), so the caller falls back
// to monitoring in-process.
//
// We replay the WHOLE original argv (`process.argv.slice(1)`), not just the
// script path, so a dispatcher invocation (`node index.js login ...`) still
// routes to the login command in the child — re-execing only the script with
// `--drain` would feed `--drain` as the subcommand and never reach runLogin.
// runLogin checks `--drain` before anything else, so the replayed flags
// (`--force`, etc.) are harmless in the child.
function trySpawnDetachedMonitor(sessionId) {
    try {
        const pid = spawnDetached(process.execPath, [
            ...process.argv.slice(1),
            "--drain",
            sessionId,
        ]);
        return pid != null;
    }
    catch {
        return false;
    }
}
async function verifyToken(serviceUrl, token) {
    try {
        const res = await fetch(`${serviceUrl}/api/plugin/whoami`, {
            headers: { Authorization: `Bearer ${token}` },
        });
        if (!res.ok) {
            return null;
        }
        return (await res.json());
    }
    catch {
        return null;
    }
}
export async function runLogin(platform, pluginVersion, options) {
    const exitOnComplete = options?.exitOnComplete ?? true;
    // Detached-monitor mode: this process was spawned by a prior login invocation
    // (`--drain <sessionId>`) purely to watch the login window's auto-close and
    // clear the pointer. Do only that, then exit — none of the normal login flow.
    const drainIdx = process.argv.indexOf("--drain");
    if (drainIdx !== -1) {
        const sessionId = process.argv[drainIdx + 1];
        if (!sessionId) {
            // The re-exec always appends `--drain <sessionId>` together; a bare
            // `--drain` is malformed. Fail loudly rather than exiting 0 having
            // silently skipped the window cleanup (mirrors the CLI's handleDrainArg).
            throw new Error("--drain requires a sessionId argument");
        }
        await runLoginWindowMonitor(sessionId);
        if (exitOnComplete) {
            process.exit(0);
        }
        return;
    }
    const force = options?.force ?? process.argv.includes("--force");
    const config = getConfig();
    // Already-authenticated short-circuit: a re-login that's a no-op must not
    // open, navigate, or record ANYTHING (the old behavior opened a throwaway
    // window at the close page whose pointer went stale the moment it
    // self-closed, gating the next command). Only a VERIFIED key short-circuits:
    // a stale key falls through to a real login. `--force` skips the
    // short-circuit to deliberately re-mint the token (e.g. after switching
    // orgs in the browser).
    if (!force && hasCredentials() && config.apiKey) {
        const who = await verifyToken(config.serviceUrl, config.apiKey);
        if (who) {
            const identity = who.user.email ?? "your account";
            console.log(`\nAlready logged in as ${identity}. Run the login command with --force to re-authenticate.`);
            if (exitOnComplete) {
                process.exit(0);
            }
            return;
        }
    }
    console.log("\nOpening Studio to sign in...");
    try {
        // Open Studio for sign-in. No in-process monitor here: after the token we
        // hand the window's auto-close cleanup to a detached child so this process
        // can exit immediately. `forceLogin` takes the sign-in path even when a key
        // is on disk; `freshWindowPath` sends a fresh window to the self-closing
        // close page so it still auto-closes after sign-in (a reused existing
        // window lands on `/studio` instead). openStudioTo owns session +
        // pluginLogin injection and writes the active-session pointer on open.
        const result = await openStudioTo("/studio", {
            forceLogin: true,
            freshWindowPath: `/studio/close?autoClose=true&message=${encodeURIComponent("Login complete")}`,
            onLoginRequired: (_sessionId, signInUrl) => {
                console.log(`\nIf the browser didn't open, visit this URL manually:\n${signInUrl}`);
            },
        });
        const apiKey = getConfig().apiKey;
        if (apiKey) {
            await reportHandoff(apiKey, pluginVersion, platform, {
                flow: "auth.login",
                wonBy: "studio",
                loopbackAvailable: false,
                ticketAvailable: false,
            });
            const who = await verifyToken(config.serviceUrl, apiKey);
            const greeting = who?.user.email
                ? `Logged in as ${who.user.email}.`
                : "Authentication successful.";
            console.log(`\n${greeting}`);
            console.log("Bitfab MCP tools are now active. (via studio)");
        }
        // The pointer is written; now ensure something witnesses the window's
        // auto-close and clears it. Prefer a detached child so login returns
        // immediately; if that spawn isn't possible (e.g. a sandbox), run the
        // monitor in this process as a bounded fallback. Either way a refused/
        // parked close just leaves the valid pointer in place for reuse.
        if (!trySpawnDetachedMonitor(result.sessionId)) {
            await runLoginWindowMonitor(result.sessionId);
        }
        if (exitOnComplete) {
            process.exit(0);
        }
    }
    catch (err) {
        if (exitOnComplete) {
            if (err instanceof StudioNavigationError && err.staleSessionId) {
                // The recorded Studio window did not answer the sign-in navigation.
                console.error("\nA Studio window is recorded as open but is not responding. Close it (or run the clearStudioSession command), then try again.");
            }
            else {
                console.error(`\n${err.message}`);
            }
            process.exit(1);
        }
        throw err;
    }
}
