// bitfab-plugin-lib/dist/commands/login.js
import { reportHandoff } from "../analytics.js";
import { getConfig, hasCredentials } from "../config.js";
import { openStudioTo, StudioNavigationError } from "./openStudioTo.js";
async function verifyToken(serviceUrl, token) {
    try {
        const res = await fetch(`${serviceUrl}/api/plugin/whoami`, {
            headers: { Authorization: `Bearer ${token}` },
        });
        if (!res.ok) {
            return null;
        }
        return (await res.json());
    }
    catch {
        return null;
    }
}
export async function runLogin(platform, pluginVersion, options) {
    const exitOnComplete = options?.exitOnComplete ?? true;
    const force = options?.force ?? process.argv.includes("--force");
    const config = getConfig();
    // Already-authenticated short-circuit: a re-login that's a no-op must not
    // open, navigate, or record ANYTHING (the old behavior opened a throwaway
    // window at the close page whose pointer went stale the moment it
    // self-closed, gating the next command). Only a VERIFIED key short-circuits:
    // a stale key falls through to a real login. `--force` skips the
    // short-circuit to deliberately re-mint the token (e.g. after switching
    // orgs in the browser).
    if (!force && hasCredentials() && config.apiKey) {
        const who = await verifyToken(config.serviceUrl, config.apiKey);
        if (who) {
            const identity = who.user.email ?? "your account";
            console.log(`\nAlready logged in as ${identity}. Run the login command with --force to re-authenticate.`);
            if (exitOnComplete) {
                process.exit(0);
            }
            return;
        }
    }
    // Factual blocking note only: the setup flow intentionally runs login as a
    // bounded foreground call (10-minute timeout), so no "run in background"
    // guidance here, unlike the indefinite Studio commands.
    console.error("NOTE: this command blocks until sign-in completes in the browser (10-minute timeout).");
    console.log("\nOpening Studio to sign in...");
    try {
        // Reaching here means a login is genuinely needed (signed out, stale key,
        // or --force). `forceLogin` makes openStudioTo take the sign-in path even
        // when a key is on disk — same protocol either way: reuse the recorded
        // window when present, throwaway window when not.
        //
        // The post-login destination depends on which window does the signing
        // in, and openStudioTo decides that from ONE pointer read (pre-reading
        // the pointer here to pick a destination is a TOCTOU): a REUSED window
        // lands on `path` (Studio home — the close page would close or park the
        // very window the reuse protocol preserves), while a fresh THROWAWAY
        // window lands on `freshWindowPath` (the self-closing close page — its
        // only job was the login). openStudioTo also owns session + pluginLogin
        // injection, so neither path bakes those params in.
        const result = await openStudioTo("/studio", {
            forceLogin: true,
            freshWindowPath: `/studio/close?autoClose=true&message=${encodeURIComponent("Login complete")}`,
            onLoginRequired: (_sessionId, signInUrl) => {
                console.log(`\nIf the browser didn't open, visit this URL manually:\n${signInUrl}`);
            },
        });
        result.abort();
        const apiKey = getConfig().apiKey;
        if (apiKey) {
            await reportHandoff(apiKey, pluginVersion, platform, {
                flow: "auth.login",
                wonBy: "studio",
                loopbackAvailable: false,
                ticketAvailable: false,
            });
            const who = await verifyToken(config.serviceUrl, apiKey);
            const greeting = who?.user.email
                ? `Logged in as ${who.user.email}.`
                : "Authentication successful.";
            console.log(`\n${greeting}`);
            console.log("Bitfab MCP tools are now active. (via studio)");
        }
        if (exitOnComplete) {
            process.exit(0);
        }
    }
    catch (err) {
        if (exitOnComplete) {
            if (err instanceof StudioNavigationError && err.staleSessionId) {
                // The recorded Studio window did not answer the sign-in navigation.
                console.error("\nA Studio window is recorded as open but is not responding. Close it (or run the clearStudioSession command), then try again.");
            }
            else {
                console.error(`\n${err.message}`);
            }
            process.exit(1);
        }
        throw err;
    }
}
