// bitfab-plugin-lib/dist/commands/openTracePlan.js
import { readActiveStudioSession } from "../activeStudioSession.js";
import { isStudioBrowserConnected, navigateStudioAndAwaitAck, pollAgentSessionEvents, } from "../agentSessionChannel.js";
import { reportHandoff } from "../analytics.js";
import { runBrowserHandoff } from "../browserHandoff.js";
import { getConfig, hasCredentials } from "../config.js";
import { recordFocus } from "../frontmostApp.js";
const OPEN_TRACE_PLAN_TIMEOUT_MS = 30 * 60 * 1000;
const RESPONSE_HTML = `<html>
<body style="margin:0;background:#f8fafc;color:#0f172a;font-family:system-ui,-apple-system,sans-serif;display:flex;align-items:center;justify-content:center;height:100vh;overflow:hidden">
  <div style="text-align:center;max-width:400px">
    <h1 style="font-size:24px;font-weight:600;margin:0 0 16px 0">Bitfab</h1>
    <p style="margin:0;color:#059669">Trace plan recorded! You can close this window.</p>
    <button onclick="window.close()" style="margin-top:16px;padding:8px 16px;background:#0f172a;color:white;border:none;border-radius:8px;font-size:14px;font-weight:500;cursor:pointer">Close Window</button>
  </div>
</body>
</html>`;
function parseStatus(raw) {
    return raw === "confirmed" ? "confirmed" : "cancelled";
}
/**
 * Wait for the server to emit `tracePlan:confirmed` or `tracePlan:cancelled`
 * for `planId` on the given Studio session. Returns the user's choice when
 * one arrives. Returns:
 *   - "session-ended" if the Studio session ends mid-wait (user closed the
 *     tab — caller can fall back to a fresh chromeless window).
 *   - "timeout" if `opts.timeoutMs` elapses without a terminal event (e.g.
 *     user navigated Studio away from the trace plan page and never came
 *     back). Mirrors the loopback+ticket fallback's overall handoff
 *     timeout so the CLI never hangs indefinitely (Bugbot on #1662).
 */
async function awaitTracePlanDecision(opts) {
    const abortController = new AbortController();
    return new Promise((resolve) => {
        const timer = setTimeout(() => {
            resolve("timeout");
            abortController.abort();
        }, opts.timeoutMs);
        const settle = (decision) => {
            clearTimeout(timer);
            resolve(decision);
            abortController.abort();
        };
        pollAgentSessionEvents({
            serviceUrl: opts.serviceUrl,
            apiKey: opts.apiKey,
            sessionId: opts.sessionId,
            abortSignal: abortController.signal,
            onEvent: (event) => {
                if (event.type === "tracePlan:confirmed" &&
                    event.data.planId === opts.planId) {
                    settle("confirmed");
                    return;
                }
                if (event.type === "tracePlan:cancelled" &&
                    event.data.planId === opts.planId) {
                    settle("cancelled");
                    return;
                }
                if (event.type === "studio:session-ended") {
                    settle("session-ended");
                }
            },
            onError: () => {
                // Transient — pollAgentSessionEvents keeps going on its own.
            },
        }).catch(() => {
            if (abortController.signal.aborted) {
                return;
            }
            settle("session-ended");
        });
    });
}
/**
 * Open the trace plan confirmation page and block until the user confirms
 * or cancels.
 *
 * Three delivery paths, tried in order:
 *
 *   1. **Studio in-place**: when an active Studio session exists in this
 *      worktree (recorded by `openStudio.js`) and the browser tab is live
 *      (server heartbeat), navigate the existing tab to
 *      `/studio/trace-plan/<id>` and wait for the
 *      `tracePlan:confirmed|cancelled` events the server pushes on
 *      Confirm/Cancel. No second window pops up.
 *
 *   2. **Loopback HTTP callback**: fallback for fresh sessions or when
 *      Studio isn't around. Spawns a chromeless window and listens on a
 *      local port for the page to call back.
 *
 *   3. **Server-polled ticket**: parallel to loopback for environments
 *      that can't bind a local port (SSH, Docker, WSL, cloud IDEs).
 *
 *   (2) and (3) race together via `runBrowserHandoff` — same behavior as
 *   before this command became Studio-aware. (1) is gated on the shared
 *   navigate-ack contract in `navigateStudioAndAwaitAck`, so a stuck
 *   Studio falls back to (2)+(3) within ~12s rather than hanging on the
 *   full handoff timeout.
 */
export async function runOpenTracePlan(platform, pluginVersion) {
    const args = process.argv.slice(2);
    const planId = args[0];
    if (!planId) {
        console.error("Usage: openTracePlan <planId>");
        console.error("Plan id is required.");
        process.exit(1);
    }
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    const config = getConfig();
    // Path 1: route into the open Studio tab if we have one and the server
    // confirms it has an active SSE listener. The navigate-ack timeout
    // inside `navigateStudioAndAwaitAck` catches the case where the Studio
    // process is alive but the tab is stuck or closed without the
    // heartbeat having expired yet.
    const studio = readActiveStudioSession();
    if (studio && config.apiKey) {
        // Use the serviceUrl captured when openStudio launched, not the
        // user's current config.serviceUrl. The two can diverge if the user
        // runs /bitfab-plugin-url between openStudio and openTracePlan
        // (e.g. switching local vs prod), and pushing events to the wrong
        // server would silently skip the Studio path (Bugbot on #1662).
        const client = {
            serviceUrl: studio.serviceUrl,
            apiKey: config.apiKey,
            sessionId: studio.sessionId,
        };
        const connected = await isStudioBrowserConnected(client);
        if (connected) {
            const navigatePath = `/studio/trace-plan/${encodeURIComponent(planId)}`;
            const ack = await navigateStudioAndAwaitAck(client, navigatePath);
            if (ack.acked) {
                const decision = await awaitTracePlanDecision({
                    ...client,
                    planId,
                    timeoutMs: OPEN_TRACE_PLAN_TIMEOUT_MS,
                });
                if (decision === "confirmed" || decision === "cancelled") {
                    if (decision === "confirmed") {
                        console.log("Trace plan confirmed [via studio]");
                    }
                    else {
                        console.log("Trace plan cancelled [via studio]");
                    }
                    await reportHandoff(config.apiKey, pluginVersion, platform, {
                        flow: "tracePlan.confirm",
                        wonBy: "studio",
                        loopbackAvailable: false,
                        ticketAvailable: false,
                    });
                    process.exit(0);
                }
                if (decision === "timeout") {
                    // User navigated Studio away from the trace plan without
                    // confirming or cancelling, and never came back inside the
                    // 30-minute window. Exit non-zero rather than fall through to
                    // path 2+3 — popping a second window after a 30-minute wait
                    // would be more confusing than helpful, and the setup-modify
                    // flow handles non-zero exit via its inline fallback.
                    console.error(`Trace plan timed out after ${Math.round(OPEN_TRACE_PLAN_TIMEOUT_MS / 60_000)} minutes [via studio]. The user navigated away from the trace plan page without confirming.`);
                    process.exit(1);
                }
                // decision === "session-ended": Studio was closed mid-wait. Fall
                // through to path 2+3 so the user gets a fresh window to act on.
            }
            // ack.acked === false (timeout, session-ended, or push-failed):
            // fall through to the loopback+ticket path. No focus restore is
            // needed because we never raised a new window.
        }
    }
    // Path 2 + 3: existing browser-handoff race (loopback HTTP callback +
    // server-polled ticket). Spawns a new chromeless window.
    const restoreFocus = recordFocus();
    const loopbackRoute = {
        path: "/done",
        handle: (params, res, resolve) => {
            res.writeHead(200, { "Content-Type": "text/html" });
            res.end(RESPONSE_HTML);
            resolve({
                status: parseStatus(params.get("status")),
                isFetch: params.get("isFetch") === "true",
            });
        },
    };
    try {
        const { value, wonBy, loopbackAvailable, ticketAvailable } = await runBrowserHandoff({
            serviceUrl: config.serviceUrl,
            ticketKind: "tracePlan.confirm",
            ticketBearer: config.apiKey ?? undefined,
            timeoutMs: OPEN_TRACE_PLAN_TIMEOUT_MS,
            ticketTtlSeconds: OPEN_TRACE_PLAN_TIMEOUT_MS / 1000,
            buildUrl: ({ port, ticketId }) => {
                const url = new URL(`${config.serviceUrl}/trace-plan/${encodeURIComponent(planId)}`);
                if (port !== null) {
                    url.searchParams.set("callbackPort", String(port));
                }
                if (ticketId !== null) {
                    url.searchParams.set("ticket", ticketId);
                }
                return url.toString();
            },
            loopbackRoutes: [loopbackRoute],
        });
        const { status, isFetch } = value;
        if (status === "confirmed") {
            console.log(`Trace plan confirmed [via ${wonBy}]`);
        }
        else {
            console.log(`Trace plan cancelled [via ${wonBy}]`);
        }
        if (config.apiKey) {
            await reportHandoff(config.apiKey, pluginVersion, platform, {
                flow: "tracePlan.confirm",
                wonBy: wonBy,
                loopbackAvailable,
                ticketAvailable,
            });
        }
        if (isFetch) {
            restoreFocus();
        }
        process.exit(0);
    }
    catch (err) {
        console.error(err.message);
        process.exit(1);
    }
}
