// bitfab-plugin-lib/dist/commands/openTracePlan.js
import { navigateStudioAndAwaitAck, pollAgentSessionEvents, } from "../agentSessionChannel.js";
import { reportHandoff } from "../analytics.js";
import { getConfig, hasCredentials } from "../config.js";
import { recordFocus } from "../frontmostApp.js";
import { StudioNavigationError, withStudioSession, } from "../withStudioSession.js";
const OPEN_TRACE_PLAN_TIMEOUT_MS = 30 * 60 * 1000;
async function awaitTracePlanDecision(opts) {
    const abortController = new AbortController();
    return new Promise((resolve) => {
        const timer = setTimeout(() => {
            resolve("timeout");
            abortController.abort();
        }, opts.timeoutMs);
        const settle = (decision) => {
            clearTimeout(timer);
            resolve(decision);
            abortController.abort();
        };
        pollAgentSessionEvents({
            serviceUrl: opts.serviceUrl,
            apiKey: opts.apiKey,
            sessionId: opts.sessionId,
            abortSignal: abortController.signal,
            onEvent: (event) => {
                if (event.type === "tracePlan:confirmed" &&
                    event.data.planId === opts.planId) {
                    settle("confirmed");
                    return;
                }
                if (event.type === "tracePlan:cancelled" &&
                    event.data.planId === opts.planId) {
                    settle("cancelled");
                    return;
                }
                if (event.type === "studio:session-ended") {
                    settle("session-ended");
                }
            },
            onError: () => { },
        }).catch(() => {
            if (abortController.signal.aborted) {
                return;
            }
            settle("session-ended");
        });
    });
}
export async function runOpenTracePlan(platform, pluginVersion) {
    const args = process.argv.slice(2);
    const planId = args[0];
    if (!planId) {
        console.error("Usage: openTracePlan <planId>");
        console.error("Plan id is required.");
        process.exit(1);
    }
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    const config = getConfig();
    if (!config.apiKey) {
        console.error("API key missing in config.");
        process.exit(1);
    }
    const apiKey = config.apiKey;
    const restoreFocus = recordFocus();
    const targetPath = `/studio/trace-plan/${encodeURIComponent(planId)}`;
    let sessionId;
    try {
        const result = await withStudioSession({
            apiKey,
            serviceUrl: config.serviceUrl,
            targetPath,
        });
        sessionId = result.sessionId;
    }
    catch (err) {
        if (err instanceof StudioNavigationError) {
            console.error(`Studio navigation failed: ${err.message}`);
        }
        else {
            console.error(`Failed to open Studio: ${err.message}`);
        }
        process.exit(1);
    }
    let decision = await awaitTracePlanDecision({
        serviceUrl: config.serviceUrl,
        apiKey,
        sessionId,
        planId,
        timeoutMs: OPEN_TRACE_PLAN_TIMEOUT_MS,
    });
    if (decision === "session-ended") {
        decision = "cancelled";
    }
    if (decision === "timeout") {
        console.error(`Trace plan timed out after ${Math.round(OPEN_TRACE_PLAN_TIMEOUT_MS / 60_000)} minutes [via studio]. The user navigated away without confirming.`);
        process.exit(1);
    }
    if (decision === "confirmed") {
        console.log("Trace plan confirmed [via studio]");
    }
    else {
        console.log("Trace plan cancelled [via studio]");
    }
    await navigateStudioAndAwaitAck({ serviceUrl: config.serviceUrl, apiKey, sessionId }, "/studio", { ackTimeoutMs: 5_000 }).catch(() => { });
    await reportHandoff(apiKey, pluginVersion, platform, {
        flow: "tracePlan.confirm",
        wonBy: "studio",
        loopbackAvailable: false,
        ticketAvailable: false,
    });
    restoreFocus();
    process.exit(0);
}
