// bitfab-plugin-lib/dist/commands/openTracePlan.js
import { reportHandoff } from "../analytics.js";
import { getConfig, hasCredentials } from "../config.js";
import { blockingNote } from "../output.js";
import { positionalArgs } from "../parseArgs.js";
import { emitWindowOpened, handleStudioError, openStudioTo, } from "./openStudioTo.js";
const OPEN_TRACE_PLAN_TIMEOUT_MS = 30 * 60 * 1000;
const ARGS = [
    {
        name: "planId",
        format: "uuid",
        description: "Trace plan ID to open for review.",
    },
];
const BLOCKS_UNTIL = "the user confirms or cancels the trace plan in Studio (30-minute timeout)";
function emit(event) {
    process.stdout.write(`${JSON.stringify(event)}\n`);
}
export async function runOpenTracePlan(platform, pluginVersion) {
    const args = positionalArgs(ARGS, process.argv, {
        usageNote: blockingNote(BLOCKS_UNTIL),
        description: "Open a trace plan in Studio for the user to review and stream JSONL events to stdout until they confirm or cancel (or a 30-minute timeout elapses).",
    });
    const initialPlanId = args[0];
    const config = getConfig();
    const apiKey = hasCredentials() ? config.apiKey : null;
    const targetPath = `/studio/trace-plan/${encodeURIComponent(initialPlanId)}`;
    // openStudioTo resolves the key (from config or via inline login); capture it
    // for the reportHandoff call below, which runs after the try block.
    let resolvedApiKey = apiKey;
    let currentPlanId = initialPlanId;
    let decision = "timeout";
    let abort = () => { };
    const timer = setTimeout(() => {
        decision = "timeout";
        abort();
    }, OPEN_TRACE_PLAN_TIMEOUT_MS);
    try {
        const result = await openStudioTo(targetPath, {
            // This command WAITS for its own terminal events (return-to-agent /
            // session-ended / confirmed / cancelled) and exits. It polls as a
            // read-only reader - never claiming the monitor lock - so it works
            // whether or not a "start" monitor is already running for the session.
            monitor: "wait",
            onLoginRequired: (sessionId, signInUrl) => {
                emit({ event: "auth-required", sessionId, signInUrl });
            },
            onWindowOpened: (url) => {
                emitWindowOpened({ opened: true, url }, emit);
            },
            onAuthenticated: (sessionId) => {
                emit({ event: "authenticated", sessionId });
            },
            onEvent: (event) => {
                if (event.type === "tracePlan:created" &&
                    event.data.planId &&
                    event.data.planId !== currentPlanId) {
                    currentPlanId = event.data.planId;
                    return;
                }
                if (event.type === "tracePlan:confirmed" &&
                    event.data.planId === currentPlanId) {
                    clearTimeout(timer);
                    decision = "confirmed";
                    abort();
                    return;
                }
                if (event.type === "tracePlan:cancelled" &&
                    event.data.planId === currentPlanId) {
                    clearTimeout(timer);
                    decision = "cancelled";
                    abort();
                    return;
                }
                if (event.type === "studio:session-ended") {
                    clearTimeout(timer);
                    decision = "cancelled";
                    abort();
                }
            },
        });
        abort = result.abort;
        resolvedApiKey = result.apiKey;
        emit({
            event: "session-ready",
            sessionId: result.sessionId,
        });
        await result.done;
    }
    catch (err) {
        clearTimeout(timer);
        handleStudioError(err);
    }
    if (decision === "timeout") {
        emit({ event: "timeout", planId: currentPlanId });
        process.exit(1);
    }
    if (decision === "confirmed") {
        emit({ event: "confirmed", planId: currentPlanId });
    }
    else {
        emit({ event: "cancelled", planId: currentPlanId });
    }
    if (resolvedApiKey) {
        await reportHandoff(resolvedApiKey, pluginVersion, platform, {
            flow: "tracePlan.confirm",
            wonBy: "studio",
            loopbackAvailable: false,
            ticketAvailable: false,
        });
    }
    process.exit(0);
}
