// bitfab-plugin-lib/dist/commands/openExperiments.js
import { getConfig, hasCredentials } from "../config.js";
import { positionalArgs } from "../parseArgs.js";
import { openStudioTo, StudioNavigationError } from "./openStudioTo.js";
const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
export function parseTestRunIds(args) {
    const joined = args.join(",");
    const seen = new Set();
    for (const part of joined.split(",")) {
        const id = part.trim();
        if (id.length > 0 && UUID_PATTERN.test(id)) {
            seen.add(id.toLowerCase());
        }
    }
    return Array.from(seen);
}
export function buildExperimentsPath(testRunIds) {
    const params = new URLSearchParams();
    params.set("testRunIds", testRunIds.join(","));
    return `/studio/experiments?${params.toString()}`;
}
export function buildExperimentsPathForGroup(experimentGroupId) {
    const params = new URLSearchParams();
    params.set("experimentGroupId", experimentGroupId);
    return `/studio/experiments?${params.toString()}`;
}
const ARGS = [{ name: "testRunId", variadic: true }];
export async function runOpenExperiments(platform) {
    const args = positionalArgs(ARGS);
    const testRunIds = parseTestRunIds(args);
    if (testRunIds.length === 0) {
        console.error("Usage: openExperiments <testRunId>[,<testRunId>...] [<testRunId>...]");
        console.error("At least one valid UUID test-run id is required.");
        process.exit(1);
    }
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    const config = getConfig();
    if (!config.apiKey) {
        console.error("API key missing in config.");
        process.exit(1);
    }
    const targetPath = buildExperimentsPath(testRunIds);
    try {
        await openStudioTo(targetPath, {
            apiKey: config.apiKey,
            serviceUrl: config.serviceUrl,
        });
    }
    catch (err) {
        if (err instanceof StudioNavigationError) {
            console.error(`Studio navigation failed: ${err.message}`);
        }
        else {
            console.error(`Failed to open Studio: ${err.message}`);
        }
        process.exit(1);
    }
    console.log("Experiments viewer opened in Studio");
    process.exit(0);
}
