// bitfab-plugin-lib/dist/commands/openStudio.js
import crypto from "node:crypto";
import { clearActiveStudioSession, writeActiveStudioSession, } from "../activeStudioSession.js";
import { createAgentSession, pollAgentSessionEvents, } from "../agentSessionChannel.js";
import { openChromelessWindow } from "../browser.js";
import { clearStudioMapping, writeStudioMapping, } from "../chatSessions/arming.js";
import { getConfig, hasCredentials } from "../config.js";
import { NotAuthenticatedError } from "../errors.js";
import { recordFocus } from "../frontmostApp.js";
import { isValidStudioRoute } from "../studioRoutes.js";
const SESSION_CLOSE_GRACE_PERIOD_MS = 10_000;
function emitJson(event) {
    process.stdout.write(`${JSON.stringify(event)}\n`);
}
function ensureStudioPath(p) {
    if (!p.startsWith("/studio")) {
        return "/studio";
    }
    return p;
}
export async function openStudio({ serviceUrl, apiKey, sessionId, initialPath = "/studio", clientHeaders, reconnect, }) {
    if (!apiKey) {
        throw new NotAuthenticatedError();
    }
    initialPath = ensureStudioPath(initialPath);
    if (!isValidStudioRoute(initialPath)) {
        throw new Error(`Studio route not allowed: ${initialPath}`);
    }
    const restoreFocus = reconnect ? () => { } : recordFocus();
    await createAgentSession({
        serviceUrl,
        apiKey,
        sessionId,
        extraHeaders: clientHeaders,
    });
    writeActiveStudioSession({ sessionId, serviceUrl });
    if (!reconnect) {
        const separator = initialPath.includes("?") ? "&" : "?";
        const url = `${serviceUrl}${initialPath}${separator}session=${encodeURIComponent(sessionId)}`;
        openChromelessWindow(url);
    }
    const abortController = new AbortController();
    let sessionEndGraceTimer = null;
    pollAgentSessionEvents({
        serviceUrl,
        apiKey,
        sessionId,
        abortSignal: abortController.signal,
        onEvent: (event) => {
            if (event.type === "studio:return-to-agent") {
                emitJson({ event: "return-to-agent", sessionId });
                restoreFocus();
            }
            if (event.type === "studio:edit-with-agent") {
                emitJson({
                    event: "edit-with-agent",
                    sessionId,
                    datasetId: event.data.datasetId,
                });
                restoreFocus();
            }
            if (event.type === "studio:session-ended") {
                if (sessionEndGraceTimer) {
                    clearTimeout(sessionEndGraceTimer);
                    sessionEndGraceTimer = null;
                }
                emitJson({ event: "session-ended", sessionId });
                clearActiveStudioSession();
                restoreFocus();
                abortController.abort();
            }
            if (event.type === "studio:session-closed") {
                if (sessionEndGraceTimer) {
                    return;
                }
                sessionEndGraceTimer = setTimeout(() => {
                    sessionEndGraceTimer = null;
                    emitJson({ event: "session-ended", sessionId });
                    clearActiveStudioSession();
                    restoreFocus();
                    abortController.abort();
                }, SESSION_CLOSE_GRACE_PERIOD_MS);
            }
            if (event.type === "studio:session-started") {
                if (sessionEndGraceTimer) {
                    clearTimeout(sessionEndGraceTimer);
                    sessionEndGraceTimer = null;
                    console.error("Studio reconnected after page refresh");
                }
            }
            if (event.type === "studio:auth-required") {
                emitJson({ event: "auth-required", sessionId });
            }
            if (event.type === "studio:authenticated") {
                emitJson({ event: "authenticated", sessionId });
            }
            if (event.type === "studio:not-member") {
                emitJson({ event: "not-member", sessionId });
                clearActiveStudioSession();
                restoreFocus();
                abortController.abort();
            }
        },
        onError: (err) => {
            console.error(`studio agent-event poll: ${err.message}`);
        },
    }).catch((err) => {
        if (!abortController.signal.aborted) {
            console.error(`studio agent-event poll terminated: ${err.message}`);
        }
    });
    return {
        sessionId,
        abort: () => abortController.abort(),
    };
}
export async function runOpenStudio(platform) {
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    const config = getConfig();
    if (!config.apiKey) {
        console.error("No API key configured.");
        process.exit(1);
    }
    const args = process.argv.slice(2);
    const existingIdx = args.indexOf("--existing");
    const reconnect = existingIdx !== -1;
    let sessionId;
    let agentSessionId = null;
    let initialPath;
    if (reconnect) {
        const existingSessionId = args[existingIdx + 1];
        if (!existingSessionId) {
            console.error("Usage: openStudio --existing <sessionId> [agentSessionId]");
            process.exit(1);
        }
        sessionId = existingSessionId;
        agentSessionId = args[existingIdx + 2] ?? null;
        initialPath = "/studio";
        console.error(`Reconnecting to Studio session: ${sessionId}`);
    }
    else {
        initialPath = ensureStudioPath(args[0] ?? "/studio");
        if (!isValidStudioRoute(initialPath)) {
            console.error(`Studio route not allowed: ${initialPath}`);
            process.exit(1);
        }
        agentSessionId = args[1] ?? null;
        sessionId = crypto.randomUUID();
        const separator = initialPath.includes("?") ? "&" : "?";
        console.error(`Opening Studio: ${config.serviceUrl}${initialPath}${separator}session=${sessionId}`);
    }
    if (agentSessionId) {
        try {
            writeStudioMapping(agentSessionId, sessionId);
        }
        catch { }
    }
    const session = await openStudio({
        serviceUrl: config.serviceUrl,
        apiKey: config.apiKey,
        sessionId,
        initialPath,
        clientHeaders: {
            "x-bitfab-client-name": platform.cliBinary,
            "x-bitfab-client-display-name": platform.displayName,
        },
        reconnect,
    });
    emitJson({ event: reconnect ? "reconnected" : "started", sessionId });
    const shutdown = (exitCode) => {
        if (agentSessionId) {
            try {
                clearStudioMapping(agentSessionId);
            }
            catch { }
        }
        session.abort();
        clearActiveStudioSession();
        process.exit(exitCode);
    };
    process.on("SIGINT", () => shutdown(0));
    process.on("SIGTERM", () => shutdown(0));
    process.on("exit", () => {
        // `exit` is sync-only; rely on the synchronous fs.rmSync inside
        // clearActiveStudioSession. Safe to call even if abort() was already
        // invoked above — clear is idempotent.
        clearActiveStudioSession();
    });
}
