// bitfab-plugin-lib/dist/commands/openExperiments.js
import { blockingNote, emitJson } from "../output.js";
import { parseArgs, UUID_RE } from "../parseArgs.js";
import { handleStudioError, openStudioTo } from "./openStudioTo.js";
export function parseTestRunIds(args) {
    const joined = args.join(",");
    const seen = new Set();
    for (const part of joined.split(",")) {
        const id = part.trim();
        if (id.length > 0 && UUID_RE.test(id)) {
            seen.add(id.toLowerCase());
        }
    }
    return Array.from(seen);
}
function withTokens(params, opts) {
    if (opts?.showTokens) {
        params.set("tokens", "1");
        // Always emit the basis explicitly. The page's tokenType is sticky, so a
        // bare `?tokens=1` would leave a prior `uncached` view counting uncached
        // against an all-basis intent; pinning it on every tokens-bearing nav keeps
        // the lens matching the URL.
        params.set("tokenType", opts.tokenType ?? "all");
    }
    return `/studio/experiments?${params.toString()}`;
}
export function buildExperimentsPath(testRunIds, opts) {
    const params = new URLSearchParams();
    params.set("testRunIds", testRunIds.join(","));
    return withTokens(params, opts);
}
export function buildExperimentsPathForGroup(experimentGroupId, opts) {
    const params = new URLSearchParams();
    params.set("experimentGroupId", experimentGroupId);
    return withTokens(params, opts);
}
export function buildExperimentsPathForDataset(datasetId, opts) {
    const params = new URLSearchParams();
    params.set("datasetId", datasetId);
    return withTokens(params, opts);
}
const ARGS = [
    {
        name: "testRunId",
        variadic: true,
        description: "Experiment test-run ID(s) to display. UUIDs, space- or comma-separated.",
    },
];
const FLAGS = [
    {
        name: "show-tokens",
        boolean: true,
        description: "Show token-cost columns in the experiments view.",
    },
    // Switch the token-cost lens to the uncached basis. No-op without
    // --show-tokens; the lens defaults to all tokens otherwise.
    {
        name: "uncached",
        boolean: true,
        description: "Use the uncached token-cost basis (no-op without --show-tokens).",
    },
];
const BLOCKS_UNTIL = "the user returns to the agent or the Studio session ends";
/**
 * Open the experiments viewer in Studio and stay alive (holding the session's
 * heartbeat) until the user returns to the agent or the session ends. Stdout
 * is a JSON line stream:
 *
 *   {"event":"auth-required","sessionId":"..."}    // (logged-out only) sign-in opened
 *   {"event":"authenticated","sessionId":"..."}    // (logged-out only) sign-in complete
 *   {"event":"session-ready","sessionId":"..."}   // Studio is up
 *   {"event":"done"}                               // terminal (returned to agent)
 *   {"event":"ended"}                              // terminal (session ended)
 *
 * On a logged-out run the auth-required/authenticated pair precedes
 * session-ready; when already authenticated, session-ready is the first line.
 *
 * After a terminal event the process exits 0. This mirrors the poller-owning
 * shape of startDataset / openTracePlan so a session it opens always has a
 * live poller (heartbeat + event handling), instead of opening a tab that is
 * disconnected from birth.
 *
 * Free-form logs go to stderr so stdout stays a parseable JSONL stream.
 */
export async function runOpenExperiments(_platform) {
    const { positional, flags } = parseArgs({
        positional: ARGS,
        flags: FLAGS,
        usageNote: blockingNote(BLOCKS_UNTIL),
        description: "Open the experiments viewer in Studio and stream JSONL events to stdout (auth-required, authenticated, session-ready, then terminal done/ended) until the user returns to the agent or the session ends.",
    });
    const testRunIds = parseTestRunIds(positional);
    if (testRunIds.length === 0) {
        console.error("Usage: openExperiments <testRunId>[,<testRunId>...] [<testRunId>...] [--show-tokens] [--uncached]");
        console.error(blockingNote(BLOCKS_UNTIL));
        console.error("At least one valid UUID test-run id is required.");
        process.exit(1);
    }
    const targetPath = buildExperimentsPath(testRunIds, {
        showTokens: flags["show-tokens"] === true,
        tokenType: flags.uncached === true ? "uncached" : "all",
    });
    let abort = () => { };
    try {
        const result = await openStudioTo(targetPath, {
            // This command WAITS for its own terminal events (return-to-agent /
            // session-ended / confirmed / cancelled) and exits. It polls as a
            // read-only reader - never claiming the monitor lock - so it works
            // whether or not a "start" monitor is already running for the session.
            monitor: "wait",
            onLoginRequired: (sessionId, signInUrl) => {
                emitJson({ event: "auth-required", sessionId, signInUrl });
            },
            onAuthenticated: (sessionId) => {
                emitJson({ event: "authenticated", sessionId });
            },
            onEvent: (event) => {
                if (event.type === "studio:return-to-agent") {
                    emitJson({ event: "done" });
                    console.error("Experiments viewer returned to agent [via studio].");
                    abort();
                    return;
                }
                if (event.type === "studio:session-ended") {
                    emitJson({ event: "ended" });
                    console.error("Studio session ended.");
                    abort();
                }
            },
        });
        abort = result.abort;
        emitJson({ event: "session-ready", sessionId: result.sessionId });
        await result.done;
        process.exit(0);
    }
    catch (err) {
        handleStudioError(err);
    }
}
