// bitfab-plugin-lib/dist/commands/openStudioTo.js
import crypto from "node:crypto";
import { claimStudioMonitor, clearActiveStudioSession, readActiveStudioSession, releaseStudioMonitor, tryClaimStudioMonitor, writeActiveStudioSession, } from "../activeStudioSession.js";
import { fetchStreamTip, navigateStudioAndAwaitAck, pollAgentSessionEvents, pollLoginEvents, pollSessionAuthEvents, requestSessionReauth, } from "../agentSessionChannel.js";
import { openChromelessWindow } from "../browser.js";
import { getConfig, hasCredentials, saveCredentials } from "../config.js";
import { recordFocus } from "../frontmostApp.js";
import { blockingNote, emitJson, printBlockingNotice } from "../output.js";
import { isValidStudioRoute } from "../studioRoutes.js";
import { createWindowCloseArbiter } from "../studioTeardown.js";
import { createStudioSession } from "./createStudioSession.js";
import { waitForSessionReady } from "./sessionReady.js";
const LOGIN_TIMEOUT_MS = 10 * 60 * 1000;
// Printed once to stderr by `startEventPoller` whenever a command is about to
// block on the Studio event stream. Centralized at that chokepoint so EVERY
// blocking Studio command (present and future) warns agents to background it:
// they all funnel through `startEventPoller`, so a new blocking path cannot be
// added without inheriting the notice. The navigate-and-exit and login paths
// never start a poller, so they correctly stay silent.
const BLOCKING_NOTICE_UNTIL = "the Studio interaction completes or the session ends";
// How long the reused-window sign-in waits for proof the window is alive
// (the sign-in page's `studio:auth-required`, or `studio:authenticated`
// straight away when browser cookies are still valid). Longer than the
// authenticated branch's navigate-ack timeout because it includes a full
// page load of the sign-in page, not just a client-side route change.
const REAUTH_ACK_TIMEOUT_MS = 20_000;
// A no-ack only means "the window is gone" for these reasons. `blocked` means a
// LIVE tab actively refused the navigation (an app-level guard, e.g. unsaved
// changes) and `invalid-route` is a programming error — neither is a stale
// session, so they must NOT emit `not-responding` (which would prompt the user
// to clear + reopen and spawn a duplicate over a healthy tab). They surface as
// `open-failed` instead.
const UNREACHABLE_REASONS = new Set(["timeout", "push-failed", "window-closed"]);
function isUnreachable(reason) {
    return reason != null && UNREACHABLE_REASONS.has(reason);
}
export function handleStudioError(err) {
    if (err instanceof StudioNavigationError &&
        err.staleSessionId &&
        isUnreachable(err.reason)) {
        // The recorded session is believed live but unreachable (crash, sleep, or a
        // close no process witnessed). Emit the SAME structured gate event the
        // openStudioTo runner does, so every Studio-opening command participates in
        // the one gate protocol: the calling skill offers Try-again / clear-and-
        // reopen recovery uniformly instead of each command failing its own way.
        emitJson({ event: "not-responding", sessionId: err.staleSessionId });
    }
    else if (err instanceof StudioNavigationError) {
        // Blocked / invalid-route: a live tab refused, or a bad path — surface it,
        // do not offer clear-and-reopen.
        emitJson({ event: "open-failed", reason: err.blockedReason ?? err.message });
    }
    else {
        emitJson({
            event: "open-failed",
            reason: err instanceof Error ? err.message : String(err),
        });
    }
    process.exit(1);
}
export class StudioNavigationError extends Error {
    reason;
    blockedReason;
    staleSessionId;
    constructor(reason, blockedReason, staleSessionId) {
        super(blockedReason
            ? `Studio navigation failed (${reason}): ${blockedReason}`
            : `Studio navigation failed: ${reason}`);
        this.reason = reason;
        this.blockedReason = blockedReason;
        this.staleSessionId = staleSessionId;
        this.name = "StudioNavigationError";
    }
}
/**
 * Single entry point for ALL "show this path in Studio" operations.
 * Every command that needs to open or navigate Studio calls this function.
 *
 * Resolves `serviceUrl` and `apiKey` from the local config itself, so callers do
 * not thread credentials through. A null key (not authenticated) takes the
 * inline sign-in path.
 *
 * When `onEvent` is provided, starts a single event polling loop for the
 * session (whether existing or newly created). ALL events flow to the
 * callback. The caller filters to what it needs. No other poller should
 * be started by the caller.
 *
 * Authenticated. The on-disk pointer is the single source of truth:
 *   - present => navigate the existing window. If it acks, reuse it. If it does
 *     NOT ack (the tab is unreachable: crash, sleep, or a close no poller saw),
 *     throw StudioNavigationError so the caller gates (asks the user). The
 *     navigation already retries via ping-pong before reporting no-ack, so a
 *     no-ack means the tab was pinged twice and never answered. We never open a
 *     second window while a record exists.
 *   - absent => open a fresh Studio at the path and write the pointer.
 *
 * There is no `closedAt`/`endedAt` branching: the pointer is CLEARED on a
 * confirmed window teardown (`studio:window-closed`, after the refresh grace
 * period — in the event poller / `closeStudio`), so a closed window is simply
 * "absent" on the next open and a fresh window opens silently. An ENDED
 * session whose window never confirmed closing keeps its pointer: the window
 * is still alive (parked on the close page) and gets reused. Recovery from a
 * stale pointer is explicit (`clearStudioSession`), never automatic.
 *
 * Unauthenticated (no key):
 * The SAME single-window protocol, pointer first:
 *   - present => reuse the existing window for sign-in. The rich event
 *     channel is key-authenticated, so the navigation goes through the
 *     server's keyless reauth verb (sessionId-scoped): it pushes an
 *     `agent:navigate` to the sign-in page onto the session's stream, and
 *     the ack is awaited on the keyless auth-event poll. No ack => throw
 *     StudioNavigationError (gate), exactly like the authenticated branch.
 *     A push that cannot even be attempted (older server without the verb,
 *     network failure) gates the same way ("push-failed"). The pointer
 *     survives untouched (same sessionId, same window) through the whole
 *     re-auth round trip; nothing on this path ever clears it — recovery
 *     from a gate is the explicit clear + retry.
 *   - absent => opens Studio to sign-in with a redirect back to `path`.
 * Either way: polls for the auth token, saves credentials, waits for the
 * session to be live, then starts the event poller. Fires `onLoginRequired`
 * / `onAuthenticated` callbacks so the caller can emit events.
 */
export async function openStudioTo(path, opts = {}) {
    const config = getConfig();
    const serviceUrl = config.serviceUrl;
    const apiKey = !opts.forceLogin && hasCredentials() ? config.apiKey : null;
    const noop = () => { };
    const restoreFocus = recordFocus();
    if (!apiKey) {
        // Pointer first, same as the authenticated branch: a recorded window is
        // reused for sign-in, never duplicated. See the function doc for the
        // keyless reuse protocol and its fallback.
        const existing = readActiveStudioSession();
        if (existing) {
            const reauth = await requestSessionReauth({
                serviceUrl: existing.serviceUrl,
                sessionId: existing.sessionId,
                redirectPath: path,
            });
            if (!reauth.ok) {
                // The reauth push could not be attempted (older server without the
                // verb, or a network failure). Gate exactly like the authenticated
                // branch's failed navigate push: never open a fresh window over a
                // believed-live record, and never clear the pointer without proof —
                // recovery is the explicit clear + retry the gate offers. Note an
                // expired server-side session is NOT this case: the push succeeds
                // regardless (the server pushes without requiring the org binding)
                // and the login itself re-creates the binding.
                throw new StudioNavigationError("push-failed", undefined, existing.sessionId);
            }
            return loginViaExistingWindow({
                serviceUrl: existing.serviceUrl,
                sessionId: existing.sessionId,
                startCursor: reauth.pushedEventId,
                signInUrl: reauth.signInUrl,
                restoreFocus,
                opts,
            });
        }
        const sessionId = crypto.randomUUID();
        // `pluginLogin=true` on the redirect is load-bearing: it is what makes the
        // authenticated Studio shell POST /api/studio/login-complete, which is the
        // only emitter of a `studio:authenticated` event carrying the token that
        // `pollLoginEvents` waits for. The top-level `session=` lets the sign-in
        // page bind the session before redirecting. Drop either and login hangs
        // until LOGIN_TIMEOUT_MS.
        const freshPath = opts.freshWindowPath ?? path;
        const separator = freshPath.includes("?") ? "&" : "?";
        const redirectPath = `${freshPath}${separator}session=${encodeURIComponent(sessionId)}&pluginLogin=true`;
        const signInPath = `/studio/sign-in?redirect_url=${encodeURIComponent(redirectPath)}`;
        const signInUrl = `${serviceUrl}${signInPath}&session=${encodeURIComponent(sessionId)}`;
        const windowPid = openChromelessWindow(signInUrl);
        // Presentation belongs to the caller: login.js prints the URL for humans,
        // the JSONL commands carry signInUrl in the auth-required event. So don't
        // print it here (doing so double-prints for login.js).
        opts.onLoginRequired?.(sessionId, signInUrl);
        const abortController = new AbortController();
        const timer = setTimeout(() => abortController.abort(), LOGIN_TIMEOUT_MS);
        const keepalive = setInterval(() => process.stderr.write(""), 30_000);
        let loginApiKey;
        try {
            loginApiKey = await pollLoginEvents({
                serviceUrl,
                sessionId,
                abortSignal: abortController.signal,
            });
        }
        catch (err) {
            clearTimeout(timer);
            clearInterval(keepalive);
            throw err;
        }
        clearTimeout(timer);
        clearInterval(keepalive);
        saveCredentials(loginApiKey);
        opts.onAuthenticated?.(sessionId);
        if (opts.onEvent) {
            await waitForSessionReady({
                serviceUrl,
                apiKey: loginApiKey,
                sessionId,
            });
            writeActiveStudioSession({
                sessionId,
                serviceUrl,
                windowPid: windowPid ?? undefined,
            });
            const poller = startEventPoller(serviceUrl, loginApiKey, sessionId, opts.onEvent, restoreFocus, { claim: opts.monitor !== "wait" });
            return {
                sessionId,
                serviceUrl,
                apiKey: loginApiKey,
                opened: true,
                ...poller,
            };
        }
        restoreFocus();
        return {
            sessionId,
            serviceUrl,
            apiKey: loginApiKey,
            opened: true,
            abort: noop,
            done: Promise.resolve(),
        };
    }
    // The on-disk pointer is the single source of truth. Present => navigate the
    // existing window; a no-ack gates (throws) rather than spawning a duplicate.
    // Absent => open a fresh window. The pointer is cleared on a witnessed close
    // or deliberate end, so a closed/ended session reads as absent here.
    const existing = readActiveStudioSession();
    if (existing) {
        const client = {
            serviceUrl: existing.serviceUrl,
            apiKey,
            sessionId: existing.sessionId,
        };
        const ack = await navigateStudioAndAwaitAck(client, path);
        if (!ack.acked) {
            // The recorded tab did not answer (timeout after ping-pong, push failed,
            // session reported ended, or a live tab actively blocked the route). Gate
            // so the caller can surface it; never silently replace a believed-live
            // window. The pointer is preserved for an explicit clear + retry.
            throw new StudioNavigationError(ack.reason ?? "unknown", ack.blockedReason, existing.sessionId);
        }
        // Role on a reused window (mutually exclusive; see `monitor` opt):
        //   - "wait": a transient reader. Always polls its own events, claims
        //     nothing — so it neither blocks nor is blocked by a "start" monitor;
        //     both read the durable stream at once. It exits when its `onEvent`
        //     aborts on a terminal event.
        //   - "start": the singleton monitor. Polls only if it WINS the atomic
        //     claim (`tryClaimStudioMonitor`, an exclusive-create — two racers can
        //     never both win, no TOCTOU). On a loss (a live monitor already owns
        //     the session) it fire-and-exits, reusing that one; on a win (fresh, or
        //     the recorded monitor died) it takes over (auto-resume).
        //   - undefined: plain navigation — never polls; fire-and-exit.
        // `restoreFocus` runs now on the fire-and-exit path since it does not stay
        // alive to run it on a terminal event.
        if (opts.onEvent && opts.monitor === "wait") {
            const poller = startEventPoller(existing.serviceUrl, apiKey, existing.sessionId, opts.onEvent, restoreFocus, { claim: false });
            return {
                sessionId: existing.sessionId,
                serviceUrl: existing.serviceUrl,
                apiKey,
                opened: false,
                ...poller,
            };
        }
        if (opts.onEvent && opts.monitor === "start" && tryClaimStudioMonitor()) {
            const poller = startEventPoller(existing.serviceUrl, apiKey, existing.sessionId, opts.onEvent, restoreFocus, { claim: true });
            return {
                sessionId: existing.sessionId,
                serviceUrl: existing.serviceUrl,
                apiKey,
                opened: false,
                ...poller,
            };
        }
        restoreFocus();
        return {
            sessionId: existing.sessionId,
            serviceUrl: existing.serviceUrl,
            apiKey,
            opened: false,
            abort: noop,
            done: Promise.resolve(),
        };
    }
    const sessionId = crypto.randomUUID();
    const session = await createStudioSession({
        serviceUrl,
        apiKey,
        sessionId,
        initialPath: path,
        clientHeaders: opts.clientHeaders,
    });
    // A freshly opened window is always monitored (it must not be orphaned). It
    // claims the monitor lock UNLESS this is a "wait" caller — a transient reader
    // owns nothing even on a fresh open.
    const poller = opts.onEvent
        ? startEventPoller(session.serviceUrl, apiKey, session.sessionId, opts.onEvent, restoreFocus, { claim: opts.monitor !== "wait" })
        : { abort: noop, done: Promise.resolve() };
    return {
        sessionId: session.sessionId,
        serviceUrl: session.serviceUrl,
        apiKey,
        opened: true,
        ...poller,
    };
}
/**
 * Complete a sign-in inside the ALREADY-OPEN Studio window. The pointer
 * record stays untouched the whole way: same sessionId, same window, same
 * recorded windowPid — rewriting it here (as the fresh-window path does) is
 * exactly what used to orphan windows. Returns `opened: false`: this was a
 * navigation of an existing window, not a new one.
 */
async function loginViaExistingWindow(args) {
    const { serviceUrl, sessionId } = args;
    const noop = () => { };
    const keepalive = setInterval(() => process.stderr.write(""), 30_000);
    let loginApiKey;
    try {
        loginApiKey = await awaitReauthLogin({
            serviceUrl,
            sessionId,
            startCursor: args.startCursor,
            // Only fired once the sign-in page proved the window alive AND is
            // waiting on the user; a still-valid browser session re-auths silently
            // and the caller never surfaces a sign-in prompt.
            onAck: () => args.opts.onLoginRequired?.(sessionId, args.signInUrl),
        });
    }
    finally {
        clearInterval(keepalive);
    }
    saveCredentials(loginApiKey);
    args.opts.onAuthenticated?.(sessionId);
    // Mirror the authenticated reused-window branch exactly, since this reuses
    // the SAME recorded session: only poll when a monitor role was asked for, and
    // a "start" must WIN the atomic claim (`tryClaimStudioMonitor`) — if a live
    // monitor already owns this session, fire-and-exit and reuse it rather than
    // spawning a second poller and overwriting the lock. "wait" always polls as a
    // read-only reader; a plain navigation (no `monitor`) fire-and-exits.
    if (args.opts.onEvent &&
        (args.opts.monitor === "wait" ||
            (args.opts.monitor === "start" && tryClaimStudioMonitor()))) {
        await waitForSessionReady({
            serviceUrl,
            apiKey: loginApiKey,
            sessionId,
        });
        const poller = startEventPoller(serviceUrl, loginApiKey, sessionId, args.opts.onEvent, args.restoreFocus, { claim: args.opts.monitor !== "wait" });
        return {
            sessionId,
            serviceUrl,
            apiKey: loginApiKey,
            opened: false,
            ...poller,
        };
    }
    args.restoreFocus();
    return {
        sessionId,
        serviceUrl,
        apiKey: loginApiKey,
        opened: false,
        abort: noop,
        done: Promise.resolve(),
    };
}
/**
 * Wait on ONE keyless subscription for the login to complete in the existing
 * window. Resolves with the API token.
 *
 * Two-stage deadline on the same poll:
 *   - No auth event within REAUTH_ACK_TIMEOUT_MS => the window never loaded
 *     the sign-in page (dead, asleep, or parked outside the Studio shell —
 *     e.g. already on the sign-in page, which has no navigate handler):
 *     throw StudioNavigationError("timeout") so the caller gates exactly
 *     like the authenticated no-ack path. Never a duplicate window.
 *   - `studio:auth-required` => the sign-in page is up (it posts this on
 *     mount): ack, then the user gets the full login deadline.
 *   - `studio:authenticated` + token => done. May arrive WITHOUT a prior
 *     auth-required when the browser's Clerk session is still valid (the
 *     sign-in page redirects straight through): a silent re-auth.
 */
function awaitReauthLogin(args) {
    return new Promise((resolve, reject) => {
        const abortController = new AbortController();
        let settled = false;
        let acked = false;
        const settle = (fn) => {
            if (settled) {
                return;
            }
            settled = true;
            clearTimeout(ackTimer);
            clearTimeout(loginTimer);
            abortController.abort();
            fn();
        };
        const ackTimer = setTimeout(() => {
            settle(() => reject(new StudioNavigationError("timeout", undefined, args.sessionId)));
        }, REAUTH_ACK_TIMEOUT_MS);
        const loginTimer = setTimeout(() => {
            settle(() => reject(new Error("Login polling aborted")));
        }, LOGIN_TIMEOUT_MS);
        pollSessionAuthEvents({
            serviceUrl: args.serviceUrl,
            sessionId: args.sessionId,
            startCursor: args.startCursor,
            abortSignal: abortController.signal,
            onEvent: (event) => {
                if (event.type === "studio:auth-required" && !acked) {
                    acked = true;
                    clearTimeout(ackTimer);
                    args.onAck();
                    return;
                }
                if (event.type === "studio:authenticated") {
                    const token = event.data.token;
                    if (typeof token === "string") {
                        settle(() => resolve(token));
                    }
                }
            },
        }).catch((err) => {
            settle(() => reject(err instanceof Error ? err : new Error(String(err))));
        });
    });
}
/**
 * Starts the single event polling loop for a Studio session. Returns an
 * abort function.
 *
 * The pointer is removed ONLY on proof the window is gone: a
 * `studio:window-closed` event (the browser's `beforeunload` beacon — the one
 * signal nothing emits optimistically). It clears after a grace window so a
 * page refresh (window-closed, then `session-started` from the reload) cancels
 * the cleanup. The only other clears in the whole system are `closeStudio`
 * (which also waits for a confirmed window-closed) and the explicit
 * `clearStudioSession` recovery after a not-responding gate.
 *
 * `studio:session-ended` is the logical "end session" signal: it terminates
 * the command (forwarded once; commands key on it) but NEVER clears the
 * pointer — the web emits it before attempting `window.close()`, which the
 * browser can refuse, so the window may still be alive (parked on the close
 * page) and reusable by the next open. `studio:session-closed` (the close
 * flow completed) is informational and ignored here for the same reason.
 *
 * A window-closed with no preceding session-ended (manual tab close) still
 * terminates the command: after the grace expires it clears and forwards a
 * synthetic `session-ended` so the command's existing terminal logic runs.
 */
function startEventPoller(serviceUrl, apiKey, sessionId, onEvent, restoreFocus, { claim = true } = {}) {
    // Single warning point for ALL blocking Studio commands: anything that stays
    // alive on the event stream funnels through here, so emitting once at the
    // chokepoint means a new blocking command can never forget the notice.
    printBlockingNotice(BLOCKING_NOTICE_UNTIL);
    // A "start" monitor takes the lock and stamps its pid into the pointer, so
    // the `--monitor` dedupe gate (`tryClaimStudioMonitor`) sees a live monitor
    // and `closeStudio` (the SIGTERM-on-end) tracks the live poller — uniform
    // across every "start" entry point (open-new, fresh login, keyless reuse,
    // reused-window takeover; the takeover already holds the lock from its
    // winning `tryClaimStudioMonitor`, so this re-affirms it). A "wait" reader
    // passes `claim: false`: it owns nothing, so it neither blocks a "start"
    // monitor nor is treated as one.
    if (claim) {
        claimStudioMonitor();
    }
    const abortController = new AbortController();
    let endedNotified = false;
    // Drop the lock whenever THIS poller stops, by any path — `notifyEnded`, the
    // `not-member` early-out, or a caller-driven `abort` (e.g. the runner's
    // SIGTERM). Releasing only inside `notifyEnded` would leak the lock on those
    // other paths, blocking the next `--monitor` until the holder dies and the
    // lock is reclaimed as stale. Best-effort, idempotent (only removes our own
    // lock); a "wait" reader holds none, so this is a no-op for it.
    const releaseLock = () => {
        if (claim) {
            releaseStudioMonitor();
        }
    };
    // Forward the terminal `session-ended` to the command exactly once (it may
    // arrive both for real and synthesized from a window-closed) and stop the
    // loop. `baseEvent` seeds the synthetic event in the window-closed case.
    const notifyEnded = (baseEvent) => {
        if (!endedNotified) {
            endedNotified = true;
            restoreFocus();
            onEvent({ ...baseEvent, type: "studio:session-ended" });
        }
        releaseLock();
        abortController.abort();
    };
    // The shared arbiter owns the window-closed vs page-refresh disambiguation.
    // A confirmed teardown is the only thing that clears the pointer here.
    const arbiter = createWindowCloseArbiter({
        onConfirmed: (event) => {
            clearActiveStudioSession(sessionId);
            notifyEnded(event);
        },
        onRefuted: () => {
            console.error("Studio reconnected after page refresh");
        },
    });
    const done = (async () => {
        // Attach strictly AFTER the current stream tip: replayed history must not
        // drive the lifecycle. Reuse keeps ended-but-alive sessions around by
        // design, so the stream can legitimately contain an old `session-ended`
        // (or a previous run's terminal command events) that would instantly
        // terminate this command if replayed.
        const startCursor = await fetchStreamTip({ serviceUrl, apiKey, sessionId });
        if (abortController.signal.aborted) {
            return;
        }
        await pollAgentSessionEvents({
            serviceUrl,
            apiKey,
            sessionId,
            startCursor,
            abortSignal: abortController.signal,
            onEvent: (event) => {
                if (event.type === "studio:window-closed" ||
                    event.type === "studio:session-started") {
                    // Real teardown (`beforeunload`) vs page refresh: the arbiter holds
                    // the candidate through the grace window and clears the pointer only
                    // on a confirmed teardown (its onConfirmed above).
                    arbiter.handleEvent(event);
                    return;
                }
                if (event.type === "studio:session-ended") {
                    // Logical end only — NOT proof the window closed. Terminate the
                    // command but keep the pointer: if no window-closed ever arrives the
                    // window is still alive and the next open reuses it. The deliberate
                    // agent-driven close path (`closeStudio`) owns the confirmed-close
                    // clear.
                    //
                    // Only live page JS emits session-ended, so the window is alive right
                    // now — any PENDING grace candidate (armed by an earlier
                    // window-closed, i.e. a refresh teardown) is stale evidence and must
                    // not clear a live window's pointer. Cancel it like session-started
                    // does. The confirmed-deliberate-close order (session-ended THEN
                    // window-closed) is unaffected: its candidate is armed after this
                    // handler ran.
                    arbiter.cancel();
                    notifyEnded(event);
                    return;
                }
                if (event.type === "studio:session-closed") {
                    // Close flow completed — informational, never clears (the close
                    // attempt it reports can still be refused by the browser).
                    return;
                }
                if (event.type === "studio:not-member") {
                    restoreFocus();
                    onEvent(event);
                    releaseLock();
                    abortController.abort();
                    return;
                }
                if (event.type === "studio:return-to-agent" ||
                    event.type === "studio:edit-with-agent") {
                    restoreFocus();
                }
                onEvent(event);
            },
            onError: (err) => {
                console.error(`studio agent-event poll: ${err.message}`);
            },
        });
    })().catch((err) => {
        if (!abortController.signal.aborted) {
            console.error(`studio agent-event poll terminated: ${err.message}`);
        }
        // Release the monitor lock on an unexpected poller crash too, matching the
        // abort / not-member / notifyEnded stop paths — otherwise a throw after the
        // claim was taken leaks the lock until the next stale recovery (Bugbot).
        releaseLock();
    });
    return {
        abort: () => {
            releaseLock();
            abortController.abort();
        },
        done,
    };
}
/**
 * CLI runner. Parses args and calls openStudioTo.
 */
export async function runOpenStudioTo(platform) {
    const argv = process.argv.slice(2);
    let path = argv.find((a) => !a.startsWith("--"));
    // `--monitor` makes this invocation START the session's single durable
    // monitor (opening one if none is running). The backgrounded "open" step
    // passes it; plain navigations omit it and fire-and-exit. The CLI exposes
    // only "start"; the "wait" role is for in-process blocking commands.
    const monitor = argv.includes("--monitor") ? "start" : undefined;
    if (!path) {
        console.error("Usage: openStudioTo <path> [--monitor]");
        console.error(blockingNote(BLOCKING_NOTICE_UNTIL));
        process.exit(1);
    }
    if (!path.startsWith("/studio")) {
        path = "/studio";
    }
    if (!isValidStudioRoute(path)) {
        console.error(`Studio route not allowed: ${path}`);
        process.exit(1);
    }
    try {
        const result = await openStudioTo(path, {
            monitor,
            clientHeaders: {
                "x-bitfab-client-name": platform.cliBinary,
                "x-bitfab-client-display-name": platform.displayName,
            },
            onLoginRequired: (sessionId, signInUrl) => {
                emitJson({ event: "auth-required", sessionId, signInUrl });
            },
            onAuthenticated: (sessionId) => {
                emitJson({ event: "authenticated", sessionId });
            },
            onEvent: (event) => {
                const data = event.data;
                const payload = {
                    event: event.type.replace("studio:", ""),
                    sessionId: result.sessionId,
                };
                if (data.datasetId) {
                    payload.datasetId = data.datasetId;
                }
                if (data.path) {
                    payload.path = data.path;
                }
                if (data.search) {
                    payload.search = data.search;
                }
                if (data.token) {
                    payload.token = data.token;
                }
                emitJson(payload);
            },
        });
        if (result.opened) {
            emitJson({ event: "started", sessionId: result.sessionId });
        }
        else {
            emitJson({ event: "navigated", sessionId: result.sessionId, path });
        }
        const shutdown = () => {
            result.abort();
            process.exit(0);
        };
        process.on("SIGINT", shutdown);
        process.on("SIGTERM", shutdown);
        await result.done;
    }
    catch (err) {
        if (err instanceof StudioNavigationError &&
            err.staleSessionId &&
            isUnreachable(err.reason)) {
            emitJson({
                event: "not-responding",
                sessionId: err.staleSessionId,
                path,
            });
        }
        else if (err instanceof StudioNavigationError) {
            // Blocked / invalid-route: a live tab refused, or a bad path. Surface it
            // rather than prompting a clear-and-reopen that would duplicate a healthy
            // window.
            emitJson({
                event: "open-failed",
                sessionId: "none",
                reason: err.blockedReason ?? err.message,
            });
        }
        else {
            emitJson({
                event: "open-failed",
                sessionId: "none",
                reason: err.message,
            });
        }
        process.exit(1);
    }
}
