// bitfab-plugin-lib/dist/commands/openStudioTo.js
import crypto from "node:crypto";
import { updateActiveStudioSession } from "../activeStudioSession.js";
import { getConfig, hasCredentials, saveCredentials } from "../config.js";
import { StudioNavigationError } from "../errors.js";
import { recordFocus } from "../frontmostApp.js";
import { emitJson } from "../output.js";
import { resolveChannel } from "../studioChannel.js";
import { isValidStudioRoute } from "../studioRoutes.js";
export { StudioNavigationError } from "../errors.js";
const WINDOW_CLOSE_GRACE_PERIOD_MS = 10_000;
const LOGIN_TIMEOUT_MS = 10 * 60 * 1000;
const UNREACHABLE_REASONS = new Set(["timeout", "push-failed", "window-closed"]);
function isUnreachable(reason) {
    return reason != null && UNREACHABLE_REASONS.has(reason);
}
export function handleStudioError(err) {
    if (err instanceof StudioNavigationError &&
        err.staleSessionId &&
        isUnreachable(err.reason)) {
        emitJson({ event: "not-responding", sessionId: err.staleSessionId });
    }
    else if (err instanceof StudioNavigationError) {
        emitJson({ event: "open-failed", reason: err.blockedReason ?? err.message });
    }
    else {
        emitJson({
            event: "open-failed",
            reason: err instanceof Error ? err.message : String(err),
        });
    }
    process.exit(1);
}
function createEventSubscription(channel, sessionId, onEvent, restoreFocus) {
    let graceTimer = null;
    let endedNotified = false;
    const clearGraceTimer = () => {
        if (graceTimer) {
            clearTimeout(graceTimer);
            graceTimer = null;
        }
    };
    const notifyEnded = (baseEvent) => {
        if (!endedNotified) {
            endedNotified = true;
            restoreFocus();
            onEvent({ ...baseEvent, type: "studio:session-ended" });
        }
        clearGraceTimer();
        sub.abort();
    };
    const sub = channel.subscribe(sessionId, (event) => {
        if (event.type === "studio:window-closed") {
            clearGraceTimer();
            graceTimer = setTimeout(() => {
                graceTimer = null;
                updateActiveStudioSession({ windowState: "closed" });
                notifyEnded(event);
            }, WINDOW_CLOSE_GRACE_PERIOD_MS);
            return;
        }
        if (event.type === "studio:session-started") {
            if (graceTimer) {
                clearGraceTimer();
                console.error("Studio reconnected after page refresh");
            }
            return;
        }
        if (event.type === "studio:session-ended") {
            clearGraceTimer();
            notifyEnded(event);
            return;
        }
        if (event.type === "studio:session-closed") {
            return;
        }
        if (event.type === "studio:not-member") {
            restoreFocus();
            onEvent(event);
            sub.abort();
            return;
        }
        if (event.type === "studio:return-to-agent" ||
            event.type === "studio:edit-with-agent") {
            restoreFocus();
        }
        onEvent(event);
    }, (err) => {
        console.error(`studio event stream: ${err.message}`);
    });
    return {
        abort: () => {
            clearGraceTimer();
            sub.abort();
        },
        done: sub.done,
    };
}
export async function openStudioTo(path, opts = {}) {
    const config = getConfig();
    const serviceUrl = config.serviceUrl;
    const apiKey = !opts.forceLogin && hasCredentials() ? config.apiKey : null;
    const noop = () => { };
    const restoreFocus = recordFocus();
    const channel = await resolveChannel(apiKey ?? "", serviceUrl);
    if (!apiKey) {
        const sessionId = crypto.randomUUID();
        const destination = opts.freshWindowPath ?? path;
        const destSeparator = destination.includes("?") ? "&" : "?";
        const redirectPath = `${destination}${destSeparator}session=${encodeURIComponent(sessionId)}&pluginLogin=true`;
        const signInInitialPath = `/studio/sign-in?redirect_url=${encodeURIComponent(redirectPath)}&session=${encodeURIComponent(sessionId)}`;
        await channel.openOrNavigate(signInInitialPath, {
            clientHeaders: opts.clientHeaders,
            sessionId,
        });
        const signInUrl = `${serviceUrl}${signInInitialPath}`;
        opts.onLoginRequired?.(sessionId, signInUrl);
        const loginApiKey = await new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                sub.abort();
                reject(new Error("Login polling aborted"));
            }, LOGIN_TIMEOUT_MS);
            const keepalive = setInterval(() => process.stderr.write(""), 30_000);
            const cleanup = () => {
                clearTimeout(timer);
                clearInterval(keepalive);
            };
            const sub = channel.subscribe(sessionId, (event) => {
                if (event.type === "studio:authenticated" &&
                    typeof event.data?.token === "string") {
                    cleanup();
                    sub.abort();
                    resolve(event.data.token);
                }
            }, (err) => {
                cleanup();
                reject(err);
            });
        });
        saveCredentials(loginApiKey);
        opts.onAuthenticated?.(sessionId);
        channel.setApiKey(loginApiKey);
        if (opts.onEvent) {
            const subscription = createEventSubscription(channel, sessionId, opts.onEvent, restoreFocus);
            return {
                sessionId,
                serviceUrl,
                apiKey: loginApiKey,
                opened: true,
                ...subscription,
            };
        }
        restoreFocus();
        return {
            sessionId,
            serviceUrl,
            apiKey: loginApiKey,
            opened: true,
            abort: noop,
            done: Promise.resolve(),
        };
    }
    const { sessionId, opened } = await channel.openOrNavigate(path, {
        clientHeaders: opts.clientHeaders,
    });
    const subscription = opts.onEvent
        ? createEventSubscription(channel, sessionId, opts.onEvent, restoreFocus)
        : { abort: () => channel.destroy(), done: Promise.resolve() };
    return {
        sessionId,
        serviceUrl,
        apiKey,
        opened,
        ...subscription,
    };
}
export async function runOpenStudioTo(platform) {
    let path = process.argv.slice(2).find((a) => !a.startsWith("--"));
    if (!path) {
        console.error("Usage: openStudioTo <path>");
        process.exit(1);
    }
    if (!path.startsWith("/studio")) {
        path = "/studio";
    }
    if (!isValidStudioRoute(path)) {
        console.error(`Studio route not allowed: ${path}`);
        process.exit(1);
    }
    try {
        const result = await openStudioTo(path, {
            clientHeaders: {
                "x-bitfab-client-name": platform.cliBinary,
                "x-bitfab-client-display-name": platform.displayName,
            },
            onLoginRequired: (sessionId, signInUrl) => {
                emitJson({ event: "auth-required", sessionId, signInUrl });
            },
            onAuthenticated: (sessionId) => {
                emitJson({ event: "authenticated", sessionId });
            },
            onEvent: (event) => {
                const data = event.data;
                const payload = {
                    event: event.type.replace("studio:", ""),
                    sessionId: result.sessionId,
                };
                if (data.datasetId) {
                    payload.datasetId = data.datasetId;
                }
                if (data.path) {
                    payload.path = data.path;
                }
                if (data.search) {
                    payload.search = data.search;
                }
                if (data.token) {
                    payload.token = data.token;
                }
                emitJson(payload);
            },
        });
        if (result.opened) {
            emitJson({ event: "started", sessionId: result.sessionId });
        }
        else {
            emitJson({ event: "navigated", sessionId: result.sessionId, path });
        }
        const shutdown = () => {
            result.abort();
            process.exit(0);
        };
        process.on("SIGINT", shutdown);
        process.on("SIGTERM", shutdown);
        await result.done;
    }
    catch (err) {
        if (err instanceof StudioNavigationError &&
            err.staleSessionId &&
            isUnreachable(err.reason)) {
            emitJson({
                event: "not-responding",
                sessionId: err.staleSessionId,
                path,
            });
        }
        else if (err instanceof StudioNavigationError) {
            emitJson({
                event: "open-failed",
                sessionId: "none",
                reason: err.blockedReason ?? err.message,
            });
        }
        else {
            emitJson({
                event: "open-failed",
                sessionId: "none",
                reason: err.message,
            });
        }
        process.exit(1);
    }
}
