// bitfab-plugin-lib/dist/commands/openDatasetExperiments.js
import { blockingNote, emitJson } from "../output.js";
import { parseArgs } from "../parseArgs.js";
import { buildExperimentsPathForDataset } from "./openExperiments.js";
import { handleStudioError, openStudioTo } from "./openStudioTo.js";
const ARGS = [
    {
        name: "datasetId",
        format: "uuid",
        description: "Dataset whose experiments to display.",
    },
];
const FLAGS = [
    {
        name: "show-tokens",
        boolean: true,
        description: "Show token-usage columns in the experiments view.",
    },
];
const BLOCKS_UNTIL = "the user returns to the agent or the Studio session ends";
/**
 * Open the experiments viewer in Studio scoped to a single dataset and stay
 * alive (holding the session's heartbeat) until the user returns to the agent
 * or the session ends. The page lists every experiment (test run) that replayed
 * a trace belonging to the dataset, derived server-side from shared traces.
 *
 * Stdout is the same JSON line stream as openExperiments:
 *
 *   {"event":"auth-required","sessionId":"..."}    // (logged-out only) sign-in opened
 *   {"event":"authenticated","sessionId":"..."}    // (logged-out only) sign-in complete
 *   {"event":"session-ready","sessionId":"..."}   // Studio is up
 *   {"event":"done"}                               // terminal (returned to agent)
 *   {"event":"ended"}                              // terminal (session ended)
 *
 * Free-form logs go to stderr so stdout stays a parseable JSONL stream.
 */
export async function runOpenDatasetExperiments(_platform) {
    const { positional, flags } = parseArgs({
        positional: ARGS,
        flags: FLAGS,
        usageNote: blockingNote(BLOCKS_UNTIL),
        description: "Open the experiments viewer in Studio scoped to one dataset (lists every experiment that replayed a trace in it) and stream JSONL events to stdout until the user returns to the agent or the session ends.",
    });
    const datasetId = positional[0];
    const targetPath = buildExperimentsPathForDataset(datasetId, {
        showTokens: flags["show-tokens"] === true,
    });
    let abort = () => { };
    try {
        const result = await openStudioTo(targetPath, {
            // This command WAITS for its own terminal events (return-to-agent /
            // session-ended / confirmed / cancelled) and exits. It polls as a
            // read-only reader — never claiming the monitor lock — so it works
            // whether or not a "start" monitor is already running for the session.
            monitor: "wait",
            onLoginRequired: (sessionId, signInUrl) => {
                emitJson({ event: "auth-required", sessionId, signInUrl });
            },
            onAuthenticated: (sessionId) => {
                emitJson({ event: "authenticated", sessionId });
            },
            onEvent: (event) => {
                if (event.type === "studio:return-to-agent") {
                    emitJson({ event: "done" });
                    console.error("Experiments viewer returned to agent [via studio].");
                    abort();
                    return;
                }
                if (event.type === "studio:session-ended") {
                    emitJson({ event: "ended" });
                    console.error("Studio session ended.");
                    abort();
                }
            },
        });
        abort = result.abort;
        emitJson({ event: "session-ready", sessionId: result.sessionId });
        await result.done;
        process.exit(0);
    }
    catch (err) {
        handleStudioError(err);
    }
}
