// bitfab-plugin-lib/dist/commands/openStudioTo.js
import crypto from "node:crypto";
import { markActiveStudioSessionClosed, markActiveStudioSessionEnded, reactivateActiveStudioSession, readActiveStudioSession, writeActiveStudioSession, } from "../activeStudioSession.js";
import { navigateStudioAndAwaitAck, pollAgentSessionEvents, pollLoginEvents, } from "../agentSessionChannel.js";
import { openChromelessWindow } from "../browser.js";
import { getConfig, hasCredentials, saveCredentials } from "../config.js";
import { recordFocus } from "../frontmostApp.js";
import { emitJson } from "../output.js";
import { UUID_RE } from "../parseArgs.js";
import { isValidStudioRoute } from "../studioRoutes.js";
import { createStudioSession } from "./createStudioSession.js";
import { waitForSessionReady } from "./loginForSession.js";
const SESSION_CLOSE_GRACE_PERIOD_MS = 10_000;
const LOGIN_TIMEOUT_MS = 10 * 60 * 1000;
export function handleStudioError(err) {
    if (err instanceof StudioNavigationError) {
        console.error(`Studio navigation failed: ${err.message}`);
    }
    else {
        console.error(`Failed to open Studio: ${err.message}`);
    }
    process.exit(1);
}
export class StudioNavigationError extends Error {
    reason;
    blockedReason;
    staleSessionId;
    constructor(reason, blockedReason, staleSessionId) {
        super(blockedReason
            ? `Studio navigation failed (${reason}): ${blockedReason}`
            : `Studio navigation failed: ${reason}`);
        this.reason = reason;
        this.blockedReason = blockedReason;
        this.staleSessionId = staleSessionId;
        this.name = "StudioNavigationError";
    }
}
/**
 * Single entry point for ALL "show this path in Studio" operations.
 * Every command that needs to open or navigate Studio calls this function.
 *
 * When `onEvent` is provided, starts a single event polling loop for the
 * session (whether existing or newly created). ALL events flow to the
 * callback. The caller filters to what it needs. No other poller should
 * be started by the caller.
 *
 * Authenticated (apiKey provided). The on-disk record is never deleted; the open
 * path branches on how the previous session ended:
 * 1. Record has `closedAt` (the window is gone): skip the handshake and open a
 *    fresh Studio silently.
 * 2. Record is reachable (no `closedAt`) and responds: navigate and reuse it.
 * 3. Record is reachable but does not ack:
 *    - `endedAt` set (deliberately concluded): open a fresh Studio silently.
 *    - neither flag set (crash / sleep / unwitnessed close): throw
 *      StudioNavigationError so the caller asks the user (Try again / Open new).
 *      The record is preserved for the retry.
 * 4. No record (or forceNew): open a new Studio at the path.
 *
 * `closedAt` is stamped on a tab close, `endedAt` on a deliberate end (the
 * "return to agent & close" / `closeStudio`); a window that just went silent
 * gets neither, which is the only case that gates.
 *
 * Unauthenticated (no apiKey):
 * Opens Studio to sign-in with a redirect back to `path`. Polls for the
 * auth token, saves credentials, waits for the session to be live, then
 * starts the event poller. Fires `onLoginRequired` / `onAuthenticated`
 * callbacks so the caller can emit events.
 */
export async function openStudioTo(path, opts) {
    const sessionId = opts.sessionId ?? crypto.randomUUID();
    const noop = () => { };
    const restoreFocus = recordFocus();
    if (!opts.apiKey) {
        const targetWithSession = `${path}${path.includes("?") ? "&" : "?"}session=${encodeURIComponent(sessionId)}`;
        const signInUrl = `${opts.serviceUrl}/studio/sign-in?redirect_url=${encodeURIComponent(targetWithSession)}`;
        openChromelessWindow(signInUrl);
        opts.onLoginRequired?.(sessionId, signInUrl);
        const abortController = new AbortController();
        const timer = setTimeout(() => abortController.abort(), LOGIN_TIMEOUT_MS);
        const keepalive = setInterval(() => process.stderr.write(""), 30_000);
        let apiKey;
        try {
            apiKey = await pollLoginEvents({
                serviceUrl: opts.serviceUrl,
                sessionId,
                abortSignal: abortController.signal,
            });
        }
        catch (err) {
            clearTimeout(timer);
            clearInterval(keepalive);
            throw err;
        }
        clearTimeout(timer);
        clearInterval(keepalive);
        saveCredentials(apiKey);
        opts.onAuthenticated?.(sessionId);
        if (opts.onEvent) {
            await waitForSessionReady({
                serviceUrl: opts.serviceUrl,
                apiKey,
                sessionId,
            });
            writeActiveStudioSession({ sessionId, serviceUrl: opts.serviceUrl });
            const poller = startEventPoller(opts.serviceUrl, apiKey, sessionId, opts.onEvent, restoreFocus);
            return {
                sessionId,
                serviceUrl: opts.serviceUrl,
                opened: true,
                ...poller,
            };
        }
        restoreFocus();
        return {
            sessionId,
            serviceUrl: opts.serviceUrl,
            opened: true,
            abort: noop,
            done: Promise.resolve(),
        };
    }
    // An explicit sessionId is a reconnect request. If it matches the session we
    // already track locally on the same service, that tab should still be live, so
    // reach it first and reuse it instead of spawning a duplicate window. Only when
    // it does not match (stale or absent pointer, or a different configured
    // service) do we re-bind the id by opening a new window on the configured
    // service. With no explicit id we just navigate the local active session.
    // forceNew always starts fresh.
    //
    // A record marked `closedAt` means the window is gone, so don't try to reach
    // it: leave `reachable` null and open a fresh window below. A record with only
    // `endedAt` (or neither flag) is still worth a reconnect attempt. The record
    // is never deleted, only marked.
    const active = opts.forceNew ? null : readActiveStudioSession();
    const reachable = active && !active.closedAt ? active : null;
    const existing = opts.sessionId != null &&
        (reachable?.sessionId !== opts.sessionId ||
            reachable?.serviceUrl !== opts.serviceUrl)
        ? null
        : reachable;
    if (existing) {
        const client = {
            serviceUrl: existing.serviceUrl,
            apiKey: opts.apiKey,
            sessionId: existing.sessionId,
        };
        const ack = await navigateStudioAndAwaitAck(client, path);
        if (ack.acked) {
            // The window answered, so it is live again: drop any terminal marker so a
            // previously-ended record stops reporting the session as ended (otherwise
            // pushActivity keeps skipping a session we just reused).
            if (existing.endedAt != null || existing.closedAt != null) {
                reactivateActiveStudioSession(existing.sessionId);
            }
            const poller = opts.onEvent
                ? startEventPoller(existing.serviceUrl, opts.apiKey, existing.sessionId, opts.onEvent, restoreFocus)
                : { abort: noop, done: Promise.resolve() };
            return {
                sessionId: existing.sessionId,
                serviceUrl: existing.serviceUrl,
                opened: false,
                ...poller,
            };
        }
        // The reconnect attempt did not ack. Open a fresh window silently only when
        // the tab is genuinely GONE (no consumer, push failed, or no answer) AND
        // either the session was deliberately ended (`endedAt`) or this is an
        // explicit reconnect (the "rejoin session <id>" prompt already chose it, so
        // no second prompt). Everything else throws so the caller surfaces it:
        //  - neither flag set means an implicit navigation against a believed-live
        //    window: it may still be alive but unreachable, so ask (Try again / Open
        //    new) rather than silently replace it; the record is preserved, and
        //  - a `blocked` / `invalid-route` outcome means a LIVE tab actively refused
        //    this navigation (app-level guard, not the route whitelist): surface that
        //    rather than spawn a duplicate over a healthy window. This guard applies
        //    to the `endedAt` path too, not just explicit reconnects.
        const tabGone = ack.reason === "session-ended" ||
            ack.reason === "push-failed" ||
            ack.reason === "timeout";
        const canReopenSilently = tabGone && (existing.endedAt != null || opts.sessionId != null);
        if (!canReopenSilently) {
            throw new StudioNavigationError(ack.reason ?? "unknown", ack.blockedReason, existing.sessionId);
        }
    }
    const session = await createStudioSession({
        serviceUrl: opts.serviceUrl,
        apiKey: opts.apiKey,
        sessionId,
        initialPath: path,
        clientHeaders: opts.clientHeaders,
    });
    const poller = opts.onEvent
        ? startEventPoller(session.serviceUrl, opts.apiKey, session.sessionId, opts.onEvent, restoreFocus)
        : { abort: noop, done: Promise.resolve() };
    return {
        sessionId: session.sessionId,
        serviceUrl: session.serviceUrl,
        opened: true,
        ...poller,
    };
}
/**
 * Starts the single event polling loop for a Studio session. Returns an
 * abort function. Handles session-close grace period internally.
 */
function startEventPoller(serviceUrl, apiKey, sessionId, onEvent, restoreFocus) {
    const abortController = new AbortController();
    let sessionEndGraceTimer = null;
    const done = pollAgentSessionEvents({
        serviceUrl,
        apiKey,
        sessionId,
        abortSignal: abortController.signal,
        onEvent: (event) => {
            if (event.type === "studio:session-closed") {
                if (sessionEndGraceTimer) {
                    return;
                }
                sessionEndGraceTimer = setTimeout(() => {
                    sessionEndGraceTimer = null;
                    // The tab closed and did not reconnect within the grace window: the
                    // window is gone. Mark `closedAt` (not `endedAt` — this was not a
                    // deliberate end) so the next open skips the handshake and opens fresh.
                    markActiveStudioSessionClosed(sessionId);
                    restoreFocus();
                    onEvent({ ...event, type: "studio:session-ended" });
                    abortController.abort();
                }, SESSION_CLOSE_GRACE_PERIOD_MS);
                return;
            }
            if (event.type === "studio:session-started") {
                if (sessionEndGraceTimer) {
                    clearTimeout(sessionEndGraceTimer);
                    sessionEndGraceTimer = null;
                    console.error("Studio reconnected after page refresh");
                }
                return;
            }
            if (event.type === "studio:session-ended") {
                if (sessionEndGraceTimer) {
                    clearTimeout(sessionEndGraceTimer);
                    sessionEndGraceTimer = null;
                }
                // A real session-ended event is the user's deliberate "return to agent
                // & close": mark `endedAt` so the next open still tries to reconnect but,
                // if that fails, opens a fresh window silently instead of asking.
                markActiveStudioSessionEnded(sessionId);
                restoreFocus();
                onEvent(event);
                abortController.abort();
                return;
            }
            if (event.type === "studio:not-member") {
                restoreFocus();
                onEvent(event);
                abortController.abort();
                return;
            }
            if (event.type === "studio:return-to-agent" ||
                event.type === "studio:edit-with-agent") {
                restoreFocus();
            }
            onEvent(event);
        },
        onError: (err) => {
            console.error(`studio agent-event poll: ${err.message}`);
        },
    }).catch((err) => {
        if (!abortController.signal.aborted) {
            console.error(`studio agent-event poll terminated: ${err.message}`);
        }
    });
    return { abort: () => abortController.abort(), done };
}
/**
 * CLI runner. Parses args and calls openStudioTo.
 */
export async function runOpenStudioTo(platform) {
    const forceNew = process.argv.includes("--new");
    const raw = process.argv.slice(2).filter((a) => !a.startsWith("--"));
    // A UUID arg (in any position) names an existing Studio session to reconnect
    // to; the non-UUID arg is the route. This lets the disconnect popup's
    // "rejoin session <id>" prompt re-bind the exact session instead of minting a
    // new one.
    const reconnectSessionId = raw.find((a) => UUID_RE.test(a));
    let path = raw.find((a) => !UUID_RE.test(a));
    if (!path) {
        if (reconnectSessionId) {
            path = "/studio";
        }
        else {
            console.error("Usage: openStudioTo [--new] <path> [studioSessionId]");
            process.exit(1);
        }
    }
    if (!path.startsWith("/studio")) {
        path = "/studio";
    }
    if (!isValidStudioRoute(path)) {
        console.error(`Studio route not allowed: ${path}`);
        process.exit(1);
    }
    const config = getConfig();
    const apiKey = hasCredentials() ? config.apiKey : null;
    try {
        const result = await openStudioTo(path, {
            apiKey,
            serviceUrl: config.serviceUrl,
            forceNew,
            sessionId: reconnectSessionId,
            clientHeaders: {
                "x-bitfab-client-name": platform.cliBinary,
                "x-bitfab-client-display-name": platform.displayName,
            },
            onLoginRequired: (sessionId, signInUrl) => {
                emitJson({ event: "auth-required", sessionId, signInUrl });
            },
            onAuthenticated: (sessionId) => {
                emitJson({ event: "authenticated", sessionId });
            },
            onEvent: (event) => {
                const data = event.data;
                const payload = {
                    event: event.type.replace("studio:", ""),
                    sessionId: result.sessionId,
                };
                if (data.datasetId) {
                    payload.datasetId = data.datasetId;
                }
                if (data.path) {
                    payload.path = data.path;
                }
                if (data.search) {
                    payload.search = data.search;
                }
                if (data.token) {
                    payload.token = data.token;
                }
                emitJson(payload);
            },
        });
        if (result.opened) {
            emitJson({ event: "started", sessionId: result.sessionId });
        }
        else {
            emitJson({ event: "navigated", sessionId: result.sessionId, path });
        }
        const shutdown = () => {
            result.abort();
            process.exit(0);
        };
        process.on("SIGINT", shutdown);
        process.on("SIGTERM", shutdown);
        await result.done;
    }
    catch (err) {
        if (err instanceof StudioNavigationError && err.staleSessionId) {
            emitJson({
                event: "not-responding",
                sessionId: err.staleSessionId,
                path,
            });
        }
        else {
            emitJson({
                event: "open-failed",
                sessionId: "none",
                reason: err.message,
            });
        }
        process.exit(1);
    }
}
