// bitfab-plugin-lib/dist/commands/openExperiments.js
import { getConfig, hasCredentials } from "../config.js";
import { emitJson } from "../output.js";
import { positionalArgs, UUID_RE } from "../parseArgs.js";
import { handleStudioError, openStudioTo } from "./openStudioTo.js";
export function parseTestRunIds(args) {
    const joined = args.join(",");
    const seen = new Set();
    for (const part of joined.split(",")) {
        const id = part.trim();
        if (id.length > 0 && UUID_RE.test(id)) {
            seen.add(id.toLowerCase());
        }
    }
    return Array.from(seen);
}
export function buildExperimentsPath(testRunIds) {
    const params = new URLSearchParams();
    params.set("testRunIds", testRunIds.join(","));
    return `/studio/experiments?${params.toString()}`;
}
export function buildExperimentsPathForGroup(experimentGroupId) {
    const params = new URLSearchParams();
    params.set("experimentGroupId", experimentGroupId);
    return `/studio/experiments?${params.toString()}`;
}
const ARGS = [{ name: "testRunId", variadic: true }];
/**
 * Open the experiments viewer in Studio and stay alive (holding the session's
 * heartbeat) until the user returns to the agent or the session ends. Stdout
 * is a JSON line stream:
 *
 *   {"event":"auth-required","sessionId":"..."}    // (logged-out only) sign-in opened
 *   {"event":"authenticated","sessionId":"..."}    // (logged-out only) sign-in complete
 *   {"event":"session-ready","sessionId":"..."}   // Studio is up
 *   {"event":"done"}                               // terminal (returned to agent)
 *   {"event":"ended"}                              // terminal (session ended)
 *
 * On a logged-out run the auth-required/authenticated pair precedes
 * session-ready; when already authenticated, session-ready is the first line.
 *
 * After a terminal event the process exits 0. This mirrors the poller-owning
 * shape of startDataset / openTracePlan so a session it opens always has a
 * live poller (heartbeat + event handling), instead of opening a tab that is
 * disconnected from birth.
 *
 * Free-form logs go to stderr so stdout stays a parseable JSONL stream.
 */
export async function runOpenExperiments(_platform) {
    const args = positionalArgs(ARGS);
    const testRunIds = parseTestRunIds(args);
    if (testRunIds.length === 0) {
        console.error("Usage: openExperiments <testRunId>[,<testRunId>...] [<testRunId>...]");
        console.error("At least one valid UUID test-run id is required.");
        process.exit(1);
    }
    const config = getConfig();
    const apiKey = hasCredentials() ? config.apiKey : null;
    const serviceUrl = config.serviceUrl;
    const targetPath = buildExperimentsPath(testRunIds);
    let abort = () => { };
    try {
        const result = await openStudioTo(targetPath, {
            apiKey,
            serviceUrl,
            onLoginRequired: (sessionId, signInUrl) => {
                emitJson({ event: "auth-required", sessionId, signInUrl });
            },
            onAuthenticated: (sessionId) => {
                emitJson({ event: "authenticated", sessionId });
            },
            onEvent: (event) => {
                if (event.type === "studio:return-to-agent") {
                    emitJson({ event: "done" });
                    console.error("Experiments viewer returned to agent [via studio].");
                    abort();
                    return;
                }
                if (event.type === "studio:session-ended") {
                    emitJson({ event: "ended" });
                    console.error("Studio session ended.");
                    abort();
                }
            },
        });
        abort = result.abort;
        emitJson({ event: "session-ready", sessionId: result.sessionId });
        await result.done;
        process.exit(0);
    }
    catch (err) {
        handleStudioError(err);
    }
}
