// bitfab-plugin-lib/dist/commands/persistReplayLabels.js
import { readFileSync, unlinkSync } from "node:fs";
import { z } from "zod";
import { chunk } from "../arrayUtils.js";
import { hasCredentials } from "../config.js";
import { createMcpClient } from "../mcpClient.js";
import { positionalArgs } from "../parseArgs.js";
const BATCH_SIZE = 50;
const CONFIDENCE_LEVELS = [
    "VeryLow",
    "Low",
    "Medium",
    "High",
    "VeryHigh",
];
const verdictSchema = z.discriminatedUnion("kind", [
    z.object({
        kind: z.literal("verdict"),
        traceId: z.uuid(),
        label: z.boolean(),
        annotation: z.string().min(1),
        confidence: z.enum(CONFIDENCE_LEVELS).optional(),
    }),
    z.object({
        kind: z.literal("skip"),
        traceId: z.uuid(),
    }),
]);
const inputFileSchema = z.object({
    expectedTraceIds: z.array(z.uuid()).min(1),
    verdicts: z.array(z.unknown()).min(1),
});
/**
 * Pure: normalise a raw verdict entry into the discriminated form. Accepts
 * the agent-friendly `{ skip: true }` and `{ label, annotation, confidence? }`
 * shapes. `confidence` is optional and ignored on skip entries.
 */
export function parseVerdict(raw) {
    if (!raw || typeof raw !== "object") {
        return { error: "verdict entry must be an object" };
    }
    const obj = raw;
    const traceId = typeof obj.traceId === "string" ? obj.traceId : null;
    if (!traceId) {
        return { error: "verdict entry missing traceId" };
    }
    if (obj.skip === true) {
        return { kind: "skip", traceId };
    }
    if (typeof obj.label !== "boolean") {
        return { error: `verdict for ${traceId} missing boolean label` };
    }
    if (typeof obj.annotation !== "string" || obj.annotation.length === 0) {
        return { error: `verdict for ${traceId} missing annotation` };
    }
    let confidence;
    if (obj.confidence !== undefined) {
        if (!CONFIDENCE_LEVELS.includes(obj.confidence)) {
            return {
                error: `verdict for ${traceId} has invalid confidence (expected one of ${CONFIDENCE_LEVELS.join(", ")})`,
            };
        }
        confidence = obj.confidence;
    }
    return {
        kind: "verdict",
        traceId,
        label: obj.label,
        annotation: obj.annotation,
        ...(confidence !== undefined && { confidence }),
    };
}
/**
 * Pure: check the agent verdicted every expected replay trace. Returns the
 * gap (expected ∖ provided) so the caller can route the agent back to label
 * the missing IDs without a DB round trip.
 */
export function findMissingCoverage(expectedTraceIds, verdicts) {
    const seen = new Set(verdicts.map((v) => v.traceId));
    return expectedTraceIds.filter((id) => !seen.has(id));
}
/**
 * Build the MCP `update_agent_labels` payload from a verdict batch.
 * Exported for tests.
 */
export function toMcpLabels(verdicts) {
    return verdicts.map((v) => {
        if (v.kind === "skip") {
            return { traceId: v.traceId, skip: true };
        }
        return {
            traceId: v.traceId,
            label: v.label,
            annotation: v.annotation,
            ...(v.confidence !== undefined && { confidence: v.confidence }),
        };
    });
}
function isInvalidTraceIdResponse(text) {
    return (text.includes("None of the provided trace IDs belong to this organization") || text.includes("Skipped (not in this organization)"));
}
/**
 * Core orchestration - exposed so tests can drive it with a fake proxy.
 * Validates input, batches the MCP call, returns a structured result. Does
 * NOT touch the filesystem; the CLI wrapper handles read/unlink.
 */
export async function persistReplayLabels(opts) {
    const parsedFile = inputFileSchema.safeParse(opts.rawInput);
    if (!parsedFile.success) {
        return {
            status: "invalid-input",
            message: parsedFile.error.issues
                .map((i) => `${i.path.join(".") || "(root)"}: ${i.message}`)
                .join("; "),
        };
    }
    const verdicts = [];
    const seenTraceIds = new Set();
    for (const raw of parsedFile.data.verdicts) {
        const parsed = parseVerdict(raw);
        if ("error" in parsed) {
            return { status: "invalid-input", message: parsed.error };
        }
        if (seenTraceIds.has(parsed.traceId)) {
            return {
                status: "invalid-input",
                message: `duplicate verdict for traceId ${parsed.traceId}: each replay trace must have exactly one entry (verdict or skip)`,
            };
        }
        seenTraceIds.add(parsed.traceId);
        verdicts.push(parsed);
    }
    const missing = findMissingCoverage(parsedFile.data.expectedTraceIds, verdicts);
    if (missing.length > 0) {
        return {
            status: "missing-coverage",
            missingTraceIds: missing,
            providedTraceIds: verdicts.map((v) => v.traceId),
        };
    }
    const persisted = [];
    for (const batch of chunk(verdicts, BATCH_SIZE)) {
        const result = await opts.client.call("update_agent_labels", {
            labels: toMcpLabels(batch),
        });
        const text = result.content?.[0]?.text ?? "unknown MCP error";
        if (result.isError || isInvalidTraceIdResponse(text)) {
            return {
                status: "mcp-error",
                message: text,
                partialTraceIds: persisted,
            };
        }
        for (const v of batch) {
            persisted.push(v.traceId);
        }
    }
    let passCount = 0;
    let failCount = 0;
    let skipCount = 0;
    for (const v of verdicts) {
        if (v.kind === "skip") {
            skipCount += 1;
        }
        else if (v.label) {
            passCount += 1;
        }
        else {
            failCount += 1;
        }
    }
    return {
        status: "ok",
        persistedTraceIds: persisted,
        passCount,
        failCount,
        skipCount,
    };
}
/**
 * CLI entrypoint. Reads `<verdicts.json>`, runs the persist pipeline,
 * prints one JSON line, and exits. On `status: "ok"` the input file is
 * deleted so the next iteration starts clean; on any other status the file
 * is left in place so the agent can retry without losing its work.
 */
export async function runPersistReplayLabels(platform, pluginVersion) {
    const ARGS = [
        {
            name: "verdicts.json",
            format: "path",
            description: "Path to the replay verdicts JSON file to persist.",
        },
    ];
    const path = positionalArgs(ARGS, process.argv, {
        description: 'Persist replay pass/fail verdicts as agent labels. Prints one JSON result line; on status "ok" the input file is deleted, on any other status it is left in place so the run can be retried.',
    })[0];
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    let raw;
    try {
        raw = JSON.parse(readFileSync(path, "utf-8"));
    }
    catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        console.log(JSON.stringify({
            status: "invalid-input",
            message: `Failed to read or parse ${path}: ${message}`,
            path,
        }));
        process.exit(1);
    }
    const result = await persistReplayLabels({
        client: createMcpClient(platform, pluginVersion),
        rawInput: raw,
    });
    if (result.status === "ok") {
        try {
            unlinkSync(path);
        }
        catch {
            // Best-effort cleanup. Failure to unlink is not a script failure -
            // the labels persisted, which is what callers care about.
        }
        console.log(JSON.stringify({ ...result, path }));
        process.exit(0);
    }
    console.log(JSON.stringify({ ...result, path }));
    process.exit(1);
}
