// bitfab-plugin-lib/dist/commands/persistReplayLabels.js
import { readFileSync, unlinkSync } from "node:fs";
import { z } from "zod";
import { chunk } from "../arrayUtils.js";
import { hasCredentials } from "../config.js";
import { createMcpClient } from "../mcpClient.js";
import { positionalArgs } from "../parseArgs.js";
const BATCH_SIZE = 50;
const CONFIDENCE_LEVELS = [
    "VeryLow",
    "Low",
    "Medium",
    "High",
    "VeryHigh",
];
const lineageFileSchema = z
    .object({
    // The replay test run. Verdicts are keyed by the original-trace lineage, so
    // the server resolves each originalTraceId to its replay trace within this run.
    testRunId: z.uuid(),
    // Canonical; older item files carry the deprecated expectedSourceTraceIds.
    expectedOriginalTraceIds: z.array(z.uuid()).min(1).optional(),
    expectedSourceTraceIds: z.array(z.uuid()).min(1).optional(),
    verdicts: z.array(z.unknown()).min(1),
})
    .refine((d) => d.expectedOriginalTraceIds !== undefined ||
    d.expectedSourceTraceIds !== undefined, {
    message: "lineage file must carry expectedOriginalTraceIds (or the deprecated expectedSourceTraceIds)",
    path: ["expectedOriginalTraceIds"],
});
const directFileSchema = z.object({
    expectedTraceIds: z.array(z.uuid()).min(1),
    verdicts: z.array(z.unknown()).min(1),
});
/**
 * The verdict-entry keys read for each id kind, canonical first. Lineage mode
 * accepts `originalTraceId` or its deprecated `sourceTraceId` alias so older
 * item files still persist; direct mode reads only `traceId`.
 */
const ID_KIND_KEYS = {
    originalTraceId: ["originalTraceId", "sourceTraceId"],
    traceId: ["traceId"],
};
/**
 * Pure: normalise a raw verdict entry into the discriminated form. Reads the id
 * under `idKind` (`originalTraceId`, accepting the `sourceTraceId` alias, or
 * `traceId`). Accepts the agent-friendly `{ skip: true }` and
 * `{ label, annotation, confidence? }` shapes. `confidence` is optional and
 * ignored on skip entries.
 */
export function parseVerdict(raw, idKind) {
    if (!raw || typeof raw !== "object") {
        return { error: "verdict entry must be an object" };
    }
    const obj = raw;
    let id = null;
    for (const key of ID_KIND_KEYS[idKind]) {
        if (typeof obj[key] === "string") {
            id = obj[key];
            break;
        }
    }
    if (!id) {
        return { error: `verdict entry missing ${idKind}` };
    }
    if (obj.skip === true) {
        return { kind: "skip", id };
    }
    if (typeof obj.label !== "boolean") {
        return { error: `verdict for ${id} missing boolean label` };
    }
    if (typeof obj.annotation !== "string" || obj.annotation.length === 0) {
        return { error: `verdict for ${id} missing annotation` };
    }
    let confidence;
    if (obj.confidence !== undefined) {
        if (!CONFIDENCE_LEVELS.includes(obj.confidence)) {
            return {
                error: `verdict for ${id} has invalid confidence (expected one of ${CONFIDENCE_LEVELS.join(", ")})`,
            };
        }
        confidence = obj.confidence;
    }
    return {
        kind: "verdict",
        id,
        label: obj.label,
        annotation: obj.annotation,
        ...(confidence !== undefined && { confidence }),
    };
}
/**
 * Pure: check the agent verdicted every expected replay item. Returns the gap
 * (expected ∖ provided) so the caller can route the agent back to label the
 * missing items without a DB round trip.
 */
export function findMissingCoverage(expectedIds, verdicts) {
    const seen = new Set(verdicts.map((v) => v.id));
    return expectedIds.filter((id) => !seen.has(id));
}
/**
 * Build the MCP `update_agent_labels` payload from a verdict batch. Each label
 * targets its trace by `idKind`: the trace's own `traceId` (direct), or
 * `originalTraceId` (the server resolves it to the replay trace via lineage using
 * the run-level testRunId). Lineage labels also carry the deprecated
 * `sourceTraceId` alias with the same value: servers that predate the rename
 * strip the unknown `originalTraceId` key and would otherwise see a label with
 * no id at all, while current servers simply prefer the canonical key.
 * Exported for tests.
 */
export function toMcpLabels(verdicts, idKind) {
    const idFields = (id) => idKind === "originalTraceId"
        ? { originalTraceId: id, sourceTraceId: id }
        : { [idKind]: id };
    return verdicts.map((v) => {
        if (v.kind === "skip") {
            return { ...idFields(v.id), skip: true };
        }
        return {
            ...idFields(v.id),
            label: v.label,
            annotation: v.annotation,
            ...(v.confidence !== undefined && { confidence: v.confidence }),
        };
    });
}
function isInvalidTraceIdResponse(text) {
    return (text.includes("None of the provided trace IDs belong to this organization") ||
        text.includes("Skipped (not in this organization)") ||
        // Current servers say "for original trace(s)"; older servers said
        // "for source trace(s)". Match the common prefix so both are caught.
        text.includes("No replay trace found for"));
}
function expectedLabelState(verdict) {
    if (verdict.kind === "skip") {
        return "skipped";
    }
    return verdict.label ? "PASS" : "FAIL";
}
// update_agent_labels echoes one line per label keyed by the id the caller
// supplied (`originalTraceId` in lineage mode, `traceId` in direct mode), e.g.
// `- <id>: set; effective PASS`. Verification matches each line back to the
// verdict by that id, so the plugin never needs a server-generated replay id.
const EFFECTIVE_LABEL_LINE_RE = /^- ([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}): [^;]+; effective (PASS|FAIL|skipped|unlabeled)\b/;
export function parseUpdateAgentLabelsResult(mcpText) {
    const states = new Map();
    for (const line of mcpText.split(/\r?\n/)) {
        const match = EFFECTIVE_LABEL_LINE_RE.exec(line.trim());
        if (match) {
            states.set(match[1], match[2]);
        }
    }
    return states;
}
/**
 * Pure: confirm the MCP response actually persisted each verdict in the batch.
 * Reads the effective-label lines (keyed by the caller-supplied id) and compares
 * each against what the verdict asked for. On any gap it returns a
 * `verification-failed` result; otherwise it records the verified ids and
 * returns null.
 */
function verifyPersistedLabels(opts) {
    const states = parseUpdateAgentLabelsResult(opts.mcpText);
    const missingIds = [];
    const mismatchedIds = [];
    for (const verdict of opts.verdicts) {
        const expected = expectedLabelState(verdict);
        const actual = states.get(verdict.id);
        if (actual === undefined) {
            missingIds.push(verdict.id);
            mismatchedIds.push({ id: verdict.id, expected, actual: "missing" });
        }
        else if (actual !== expected) {
            mismatchedIds.push({ id: verdict.id, expected, actual });
        }
    }
    if (missingIds.length > 0 || mismatchedIds.length > 0) {
        return {
            status: "verification-failed",
            idKind: opts.idKind,
            message: "update_agent_labels returned text, but it did not contain the expected effective label lines.",
            verifiedIds: opts.verifiedIds,
            missingIds,
            mismatchedIds,
        };
    }
    for (const verdict of opts.verdicts) {
        opts.verifiedIds.push(verdict.id);
    }
    return null;
}
/**
 * Pure: pick the input mode (lineage vs direct) and pull out the id kind, the
 * expected coverage list, and the optional testRunId. Returns an error string
 * when the file matches neither shape, preferring the error from whichever
 * shape the file most resembles so the message is actionable.
 */
function resolveInputMode(rawInput) {
    const lineage = lineageFileSchema.safeParse(rawInput);
    if (lineage.success) {
        return {
            idKind: "originalTraceId",
            // Canonical, falling back to the deprecated alias for older item files.
            expectedIds: lineage.data.expectedOriginalTraceIds ??
                lineage.data.expectedSourceTraceIds ??
                [],
            testRunId: lineage.data.testRunId,
            verdicts: lineage.data.verdicts,
        };
    }
    const direct = directFileSchema.safeParse(rawInput);
    if (direct.success) {
        return {
            idKind: "traceId",
            expectedIds: direct.data.expectedTraceIds,
            verdicts: direct.data.verdicts,
        };
    }
    // Neither matched: surface the error from the shape the file looks like.
    const looksLineage = !!rawInput &&
        typeof rawInput === "object" &&
        ("expectedOriginalTraceIds" in rawInput ||
            "expectedSourceTraceIds" in rawInput ||
            "testRunId" in rawInput);
    const chosen = looksLineage ? lineage : direct;
    const message = chosen.success
        ? "input did not match the lineage (testRunId + expectedOriginalTraceIds) or direct (expectedTraceIds) shape"
        : chosen.error.issues
            .map((i) => `${i.path.join(".") || "(root)"}: ${i.message}`)
            .join("; ");
    return { error: message };
}
/**
 * Core orchestration - exposed so tests can drive it with a fake proxy.
 * Validates input, batches the MCP call, verifies each batch persisted, and
 * returns a structured result. Does NOT touch the filesystem; the CLI wrapper
 * handles read/unlink.
 */
export async function persistReplayLabels(opts) {
    const mode = resolveInputMode(opts.rawInput);
    if ("error" in mode) {
        return { status: "invalid-input", message: mode.error };
    }
    const { idKind, expectedIds, testRunId } = mode;
    const verdicts = [];
    const seenIds = new Set();
    for (const raw of mode.verdicts) {
        const parsed = parseVerdict(raw, idKind);
        if ("error" in parsed) {
            return { status: "invalid-input", message: parsed.error };
        }
        if (seenIds.has(parsed.id)) {
            return {
                status: "invalid-input",
                message: `duplicate verdict for ${idKind} ${parsed.id}: each replay item must have exactly one entry (verdict or skip)`,
            };
        }
        seenIds.add(parsed.id);
        verdicts.push(parsed);
    }
    const missing = findMissingCoverage(expectedIds, verdicts);
    if (missing.length > 0) {
        return {
            status: "missing-coverage",
            idKind,
            missingIds: missing,
            providedIds: verdicts.map((v) => v.id),
        };
    }
    const persisted = [];
    for (const batch of chunk(verdicts, BATCH_SIZE)) {
        const args = { labels: toMcpLabels(batch, idKind) };
        if (idKind === "originalTraceId" && testRunId !== undefined) {
            args.testRunId = testRunId;
        }
        const result = await opts.client.call("update_agent_labels", args);
        const text = result.content?.[0]?.text ?? "unknown MCP error";
        if (result.isError || isInvalidTraceIdResponse(text)) {
            return {
                status: "mcp-error",
                message: text,
                idKind,
                partialIds: persisted,
            };
        }
        const verificationFailure = verifyPersistedLabels({
            mcpText: text,
            verdicts: batch,
            idKind,
            verifiedIds: persisted,
        });
        if (verificationFailure) {
            return verificationFailure;
        }
    }
    let passCount = 0;
    let failCount = 0;
    let skipCount = 0;
    for (const v of verdicts) {
        if (v.kind === "skip") {
            skipCount += 1;
        }
        else if (v.label) {
            passCount += 1;
        }
        else {
            failCount += 1;
        }
    }
    return {
        status: "ok",
        idKind,
        persistedIds: persisted,
        passCount,
        failCount,
        skipCount,
    };
}
/**
 * CLI entrypoint. Reads `<verdicts.json>`, runs the persist pipeline,
 * prints one JSON line, and exits. On `status: "ok"` the input file is
 * deleted so the next iteration starts clean; on any other status the file
 * is left in place so the agent can retry without losing its work.
 */
export async function runPersistReplayLabels(platform, pluginVersion) {
    const ARGS = [
        {
            name: "verdicts.json",
            format: "path",
            description: "Path to the replay verdicts JSON file to persist.",
        },
    ];
    const path = positionalArgs(ARGS, process.argv, {
        description: 'Persist replay pass/fail verdicts as agent labels. Prints one JSON result line; on status "ok" the input file is deleted, on any other status it is left in place so the run can be retried.',
    })[0];
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    let raw;
    try {
        raw = JSON.parse(readFileSync(path, "utf-8"));
    }
    catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        console.log(JSON.stringify({
            status: "invalid-input",
            message: `Failed to read or parse ${path}: ${message}`,
            path,
        }));
        process.exit(1);
    }
    const result = await persistReplayLabels({
        client: createMcpClient(platform, pluginVersion),
        rawInput: raw,
    });
    if (result.status === "ok") {
        try {
            unlinkSync(path);
        }
        catch {
            // Best-effort cleanup. Failure to unlink is not a script failure -
            // the labels persisted, which is what callers care about.
        }
        console.log(JSON.stringify({ ...result, path }));
        process.exit(0);
    }
    console.log(JSON.stringify({ ...result, path }));
    process.exit(1);
}
