// bitfab-plugin-lib/dist/commands/pushActivity.js
import { spawn } from "node:child_process";
import { readActiveStudioSession } from "../activeStudioSession.js";
import { pushAgentSessionEvent } from "../agentSessionChannel.js";
import { getConfig, hasCredentials } from "../config.js";
import { printHelpIfRequested, UUID_RE } from "../parseArgs.js";
// Set on the detached worker copy of ourselves so it performs the blocking
// POST instead of spawning yet another child. See the fire-and-forget note in
// runPushActivity.
const WORKER_ENV_FLAG = "BITFAB_PUSH_ACTIVITY_WORKER";
/**
 * Activity events are pure telemetry for the Studio sidebar: nothing in any
 * flow reads the response. The agent runs this command in the foreground at
 * the start of nearly every phase, so a blocking POST puts the endpoint's
 * latency (cold serverless start, Redis round trips) directly on the agent's
 * critical path — observed at up to ~24s per call.
 *
 * Fire-and-forget: the parent resolves the session and credentials, re-execs
 * itself detached (own process group, stdio ignored, unref'd) to perform the
 * actual POST, then exits immediately. The agent's foreground call returns in
 * ~node-startup time regardless of how slow the endpoint is, and the detached
 * worker outlives us to complete (or silently drop) the write.
 */
export async function runPushActivity() {
    printHelpIfRequested({
        description: "Push an agent activity update to the live Studio window's activity feed. Exits 0 silently when no Studio session is active or the user is not authenticated.",
        positional: [
            {
                name: "sessionId",
                format: "uuid",
                required: false,
                description: "Target Studio session. Defaults to the active session when omitted.",
            },
            { name: "action", description: "One of: started, completed, failed." },
            {
                name: "displayName",
                description: "Human-readable activity name shown in Studio.",
            },
        ],
    });
    // Usage: pushActivity [<sessionId>] <action> <displayName>
    const raw = process.argv.slice(2).filter((a) => !a.startsWith("--"));
    let sessionId = null;
    let action;
    let displayName;
    if (raw.length >= 3 && UUID_RE.test(raw[0])) {
        sessionId = raw[0];
        action = raw[1];
        displayName = raw[2];
    }
    else if (raw.length >= 2) {
        action = raw[0];
        displayName = raw[1];
    }
    else {
        console.error("Usage: pushActivity [<sessionId>] <action> <displayName>");
        process.exit(1);
    }
    if (!["started", "completed", "failed"].includes(action)) {
        console.error("Action must be one of: started, completed, failed");
        process.exit(1);
    }
    if (!sessionId) {
        const active = readActiveStudioSession();
        // No live window to push to. A closed or deliberately-ended session has its
        // pointer cleared, so an absent record covers all of those cases.
        if (!active) {
            process.exit(0);
        }
        sessionId = active.sessionId;
    }
    if (!hasCredentials()) {
        process.exit(0);
    }
    const config = getConfig();
    if (!config.apiKey) {
        process.exit(0);
    }
    if (process.env[WORKER_ENV_FLAG] === "1") {
        // We are the detached worker: perform the blocking POST, swallow any
        // failure (telemetry must never surface an error), then exit.
        await pushAgentSessionEvent({ serviceUrl: config.serviceUrl, apiKey: config.apiKey, sessionId }, {
            type: "agent:activity",
            data: {
                action: action,
                tool: "assistant",
                displayName,
            },
        }).catch(() => { });
        process.exit(0);
    }
    // Parent path: re-exec ourselves detached with the session already resolved,
    // so the POST runs out-of-band and the agent's foreground call returns now.
    const child = spawn(process.execPath, [process.argv[1], sessionId, action, displayName], {
        detached: true,
        stdio: "ignore",
        env: { ...process.env, [WORKER_ENV_FLAG]: "1" },
    });
    child.unref();
    process.exit(0);
}
