// bitfab-plugin-lib/dist/commands/readTracesBatched.js
import { writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { chunk } from "../arrayUtils.js";
import { hasCredentials } from "../config.js";
import { createMcpClient } from "../mcpClient.js";
import { parseArgs, UUID_RE, } from "../parseArgs.js";
// The read_traces MCP tool caps each call at 10 IDs, both to bound the DB work
// per request and to keep each response small enough not to spill to disk. We
// keep that batch size and fan the batches out in parallel here rather than
// asking the agent to issue ceil(N/10) tool calls in a single turn (that
// choreography was unreliable in practice: it serialized in prod), so the
// deterministic loop lives in code.
const BATCH_SIZE = 10;
// Bound how many read_traces calls are in flight at once. Each call already
// batches its own DB queries (HVT-2308), so a handful of concurrent calls is
// safe; the cap just keeps a very large ID set from opening too many
// connections at once.
const DEFAULT_CONCURRENCY = 6;
const SCOPES = ["summary", "full"];
// Same separator mcpService.readTraces uses between traces, so concatenating
// per-batch responses reproduces one seamless stream.
const TRACE_SEPARATOR = "\n\n---\n\n";
/**
 * Pure: validate + dedupe the trace IDs, preserving first-seen order. A
 * duplicate ID would waste a slot in its batch and read the same trace twice,
 * so we collapse them here rather than at the MCP boundary.
 */
export function normalizeTraceIds(traceIds) {
    if (traceIds.length === 0) {
        return { ok: false, message: "No trace IDs provided." };
    }
    const seen = new Set();
    const ids = [];
    for (const id of traceIds) {
        if (!UUID_RE.test(id)) {
            return { ok: false, message: `Invalid trace ID (not a UUID): ${id}` };
        }
        if (seen.has(id)) {
            continue;
        }
        seen.add(id);
        ids.push(id);
    }
    return { ok: true, ids };
}
/**
 * Run `tasks` with at most `limit` in flight at once, preserving input order
 * in the returned array. Used so a large ID set doesn't open one connection
 * per batch all at once.
 */
async function mapWithConcurrency(items, limit, fn) {
    const results = new Array(items.length);
    let cursor = 0;
    const workers = Array.from({ length: Math.min(limit, items.length) }, async () => {
        while (cursor < items.length) {
            const index = cursor;
            cursor += 1;
            results[index] = await fn(items[index], index);
        }
    });
    await Promise.all(workers);
    return results;
}
/**
 * Core orchestration, exposed so tests can drive it with a fake proxy.
 * Validates input, fans the batches out in parallel, and concatenates the
 * per-batch text in input order. Does NOT touch the filesystem; the CLI
 * wrapper handles writing the result file.
 */
export async function readTracesBatched(opts) {
    const normalized = normalizeTraceIds(opts.traceIds);
    if (!normalized.ok) {
        return { status: "invalid-input", message: normalized.message };
    }
    const batches = chunk(normalized.ids, BATCH_SIZE);
    // Warm the MCP session once before fanning out. Otherwise the parallel
    // call()s each see a null sessionId and race to initialize, opening several
    // throwaway sessions. ensureSession is idempotent and call() re-checks, so a
    // failure here is non-fatal: the per-batch calls surface it.
    try {
        await opts.client.ensureSession();
    }
    catch {
        // Non-fatal: call() re-runs ensureSession and reports its own error.
    }
    const outcomes = await mapWithConcurrency(batches, opts.concurrency ?? DEFAULT_CONCURRENCY, async (batch) => {
        const result = await opts.client.call("read_traces", {
            traceIds: batch,
            scope: opts.scope,
        });
        if (result.isError) {
            return {
                ok: false,
                message: result.content?.[0]?.text ?? "unknown MCP error",
            };
        }
        return { ok: true, text: result.content?.[0]?.text ?? "" };
    });
    const firstFailure = outcomes.findIndex((o) => !o.ok);
    if (firstFailure !== -1) {
        const failed = outcomes[firstFailure];
        return {
            status: "mcp-error",
            message: failed.ok ? "unknown MCP error" : failed.message,
            failedBatchIndex: firstFailure,
        };
    }
    const text = outcomes
        .map((o) => (o.ok ? o.text : ""))
        .filter((t) => t.length > 0)
        .join(TRACE_SEPARATOR);
    return {
        status: "ok",
        text,
        traceCount: normalized.ids.length,
        batchCount: batches.length,
    };
}
/**
 * CLI entrypoint. Reads trace IDs (variadic positional) and an optional
 * `--scope`, fans the read_traces calls out in parallel batches of 10, writes
 * the concatenated trace text to a temp file, and prints one JSON line with the
 * file path so the agent can Read it. Large `full`-scope payloads stay out of
 * the tool-call channel this way.
 */
export async function runReadTracesBatched(platform, pluginVersion) {
    const POSITIONAL = [
        {
            name: "trace-id",
            format: "uuid",
            variadic: true,
            description: "One or more trace IDs (UUIDs) to read.",
        },
    ];
    const FLAGS = [
        {
            name: "scope",
            format: "string",
            default: "summary",
            description: 'Span detail level: "summary" (capped) or "full" (untruncated).',
        },
    ];
    const { positional, flags } = parseArgs({
        positional: POSITIONAL,
        flags: FLAGS,
        description: "Read many traces at once: fans read_traces out in parallel batches of 10, writes the concatenated result to a temp file, and prints {status, traceCount, batchCount, outputFile} as JSON.",
    });
    const scope = flags.scope;
    if (scope !== "summary" && scope !== "full") {
        console.log(JSON.stringify({
            status: "invalid-input",
            message: `Invalid --scope "${scope}": expected "summary" or "full".`,
        }));
        process.exit(1);
    }
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    const result = await readTracesBatched({
        client: createMcpClient(platform, pluginVersion),
        traceIds: positional,
        scope,
    });
    if (result.status !== "ok") {
        console.log(JSON.stringify(result));
        process.exit(1);
    }
    const outputFile = join(tmpdir(), `bitfab-read-traces-${process.pid}-${Date.now()}.txt`);
    try {
        writeFileSync(outputFile, result.text, "utf-8");
    }
    catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        console.log(JSON.stringify({
            status: "write-error",
            message: `Failed to write output file: ${message}`,
        }));
        process.exit(1);
    }
    console.log(JSON.stringify({
        status: "ok",
        traceCount: result.traceCount,
        batchCount: result.batchCount,
        outputFile,
    }));
    process.exit(0);
}
