// bitfab-plugin-lib/dist/commands/readReplayStatus.js
import { getConfig, hasCredentials } from "../config.js";
import { emitJson } from "../output.js";
import { parseArgs } from "../parseArgs.js";
const POSITIONAL = [
    {
        name: "testRunId",
        format: "uuid",
        description: "Replay test run ID to poll for current status.",
    },
];
export async function fetchReplayStatus(opts) {
    const baseUrl = opts.serviceUrl.replace(/\/$/, "");
    let response;
    try {
        response = await fetch(`${baseUrl}/api/sdk/replay/status`, {
            method: "POST",
            headers: {
                Authorization: `Bearer ${opts.apiKey}`,
                "Content-Type": "application/json",
            },
            body: JSON.stringify({ testRunId: opts.testRunId }),
        });
    }
    catch (err) {
        return {
            status: "error",
            message: err instanceof Error ? err.message : String(err),
        };
    }
    let payload;
    try {
        payload = await response.json();
    }
    catch {
        payload = null;
    }
    if (!response.ok) {
        const message = payload &&
            typeof payload === "object" &&
            typeof payload.error === "string"
            ? payload.error
            : `HTTP ${response.status}`;
        return { status: "error", message };
    }
    const data = payload;
    const testRunStatus = typeof data.status === "string" ? data.status : null;
    const traceIds = data && typeof data.traceIds === "object" && data.traceIds !== null
        ? data.traceIds
        : {};
    const traceCount = typeof data.traceCount === "number" && Number.isFinite(data.traceCount)
        ? data.traceCount
        : Object.keys(traceIds).length;
    return {
        status: "ok",
        testRunId: opts.testRunId,
        testRunStatus,
        traceIds,
        traceCount,
        mappedCount: Object.keys(traceIds).length,
    };
}
export async function runReadReplayStatus(platform) {
    const { positional } = parseArgs({
        positional: POSITIONAL,
        description: "Read the current replay test run status, including local-replay-trace-id to server-trace-id mapping. Safe to poll while replay is still running.",
    });
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    const config = getConfig();
    if (!config.apiKey) {
        console.error("API key missing in config.");
        process.exit(1);
    }
    const result = await fetchReplayStatus({
        serviceUrl: config.serviceUrl,
        apiKey: config.apiKey,
        testRunId: positional[0],
    });
    emitJson(result);
    process.exit(result.status === "ok" ? 0 : 1);
}
