// bitfab-plugin-lib/dist/commands/reapDaemon.js
import fs from "node:fs";
import net from "node:net";
import { devDaemonPaths, resolveDevWorktreeRoot, sweepZombieDaemons, } from "../daemon/index.js";
import { serialize } from "../daemon/protocol.js";
import { emitJson } from "../output.js";
import { positionalArgs } from "../parseArgs.js";
const SHUTDOWN_TIMEOUT_MS = 4_000;
// Grace after SIGHUP before SIGKILL. The daemon's SIGHUP handler closes its
// Studio windows via synchronous AppleScript (osascript, ~4s timeout per
// Chromium app), so this must be generous enough not to abort an in-flight
// window teardown; SIGKILL is only the backstop for a truly wedged daemon.
const REAP_KILL_GRACE_MS = 12_000;
function sleep(ms) {
    return new Promise((r) => setTimeout(r, ms));
}
function processIsAlive(pid) {
    try {
        process.kill(pid, 0);
        return true;
    }
    catch (err) {
        return err.code !== "ESRCH";
    }
}
function readPid(pidPath) {
    try {
        const pid = Number.parseInt(fs.readFileSync(pidPath, "utf-8").trim(), 10);
        return Number.isNaN(pid) ? null : pid;
    }
    catch {
        return null;
    }
}
/**
 * Connect to a running daemon WITHOUT ensuring/spawning one (DaemonClient.connect
 * would spawn — the opposite of reaping) and ask it to shut down. Resolves true
 * once it acks OR the connection closes after we sent the command (the daemon
 * exits fast on shutdown, so `close` can beat `data` — treating that as failure
 * would wrongly trigger fallbackKill against an already-reaping daemon). Resolves
 * false only when nothing was listening (connect error or timeout).
 */
function rawShutdown(socketPath) {
    return new Promise((resolve) => {
        let settled = false;
        let sent = false;
        const finish = (value) => {
            if (settled) {
                return;
            }
            settled = true;
            clearTimeout(timeout);
            try {
                socket.destroy();
            }
            catch { }
            resolve(value);
        };
        const timeout = setTimeout(() => finish(false), SHUTDOWN_TIMEOUT_MS);
        const socket = net.createConnection(socketPath, () => {
            sent = true;
            socket.write(serialize({ cmd: "shutdown" }));
        });
        socket.on("data", () => finish(true));
        // A close OR error after the command was sent means the daemon accepted it
        // and exited (it tears the connection down on shutdown). Only a failure
        // before we sent anything (connect refused) is a real "nothing listening".
        socket.on("close", () => finish(sent));
        socket.on("error", () => finish(sent));
    });
}
async function fallbackKill(paths) {
    const pid = readPid(paths.pidPath);
    if (pid !== null && processIsAlive(pid)) {
        try {
            // SIGHUP (not SIGTERM): a daemon still alive here closes its Studio
            // windows before exiting, instead of orphaning them.
            process.kill(pid, "SIGHUP");
        }
        catch { }
        const deadline = Date.now() + REAP_KILL_GRACE_MS;
        while (processIsAlive(pid) && Date.now() < deadline) {
            await sleep(50);
        }
        if (processIsAlive(pid)) {
            try {
                process.kill(pid, "SIGKILL");
            }
            catch { }
        }
    }
    for (const p of [
        paths.socketPath,
        paths.pidPath,
        `${paths.socketPath}.lock`,
        paths.metaPath,
    ]) {
        if (!p) {
            continue;
        }
        try {
            fs.unlinkSync(p);
        }
        catch { }
    }
}
/**
 * Reap this worktree's dev Studio daemon: close the windows it owns and stop it,
 * then sweep any zombie daemons (dead process, or worktree gone). Targets only
 * the `dev-<worktreeHash>` daemon for the cwd's worktree — never the prod
 * singleton, never another worktree. Idempotent: a no-op if nothing is there.
 * Run on worktree close / merge / teardown.
 */
export async function runReapDaemon() {
    positionalArgs([], process.argv, {
        description: "Reap this worktree's dev Studio daemon (close its windows and stop it) and sweep zombie daemons. Targets only the per-worktree dev daemon, never the shared prod daemon. Idempotent. Emits one JSON line.",
    });
    // Resolve the same way resolveDaemonPaths does (module dir, with the prod-path
    // guard), so the socket we reap is exactly the one the daemon bound.
    const worktreeRoot = resolveDevWorktreeRoot();
    let reaped = false;
    if (worktreeRoot) {
        const paths = devDaemonPaths(worktreeRoot);
        const present = fs.existsSync(paths.socketPath) || fs.existsSync(paths.pidPath);
        if (present) {
            const acked = await rawShutdown(paths.socketPath);
            if (!acked) {
                await fallbackKill(paths);
            }
            reaped = true;
        }
    }
    // Also clear daemons from worktrees that vanished without a reap.
    sweepZombieDaemons();
    emitJson({
        event: "daemon-reaped",
        worktreeRoot: worktreeRoot ?? null,
        reaped,
    });
}
