// bitfab-plugin-lib/dist/commands/replayProgress.js
import { spawn } from "node:child_process";
import { randomUUID } from "node:crypto";
import { appendFileSync, createWriteStream, existsSync, mkdirSync, readFileSync, renameSync, rmSync, statSync, writeFileSync, } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { createInterface } from "node:readline";
import { printHelpIfRequested } from "../parseArgs.js";
import { formatHeartbeat, formatNoProgressNotice, formatReplayCommandFailure, formatReplayHeader, formatReplayResultMissing, formatReplaySummary, formatTraceLine, } from "../replayProgressView.js";
/**
 * How long to go without a per-trace line before emitting a liveness heartbeat,
 * in milliseconds. A single replayed trace (an LLM call) can run tens of seconds;
 * without this the run would print nothing in that gap and look frozen. Overridable
 * via env for tests (which can't wait the real interval).
 */
const HEARTBEAT_MS = Number(process.env.BITFAB_REPLAY_HEARTBEAT_MS) || 12_000;
/**
 * Prefix the SDK reporter writes to the replay's stderr, one line per trace:
 *
 *   @@bitfab:progress {"completed":3,"total":10,"succeeded":3,"errored":0,"item":{"traceId":"...","error":null}}
 *
 * stderr (not an extra fd) because progress has to survive the launcher hops
 * the agent uses (`npx tsx`, `python`, `bundle exec`), which don't forward
 * extra fds; stderr/stdout are inherited through every hop.
 */
const PROGRESS_SENTINEL = "@@bitfab:progress ";
/**
 * Every line this command prints to stdout is also appended here, fresh per run,
 * so the user can `tail -f` it in a separate terminal for a truly live, never-
 * collapsing view. Claude Code's tool card collapses output and can't stream; a
 * tail is the one surface that shows every line the instant it is written.
 */
const DEFAULT_LOG = ".bitfab/replay-progress.log";
const DEFAULT_EVENTS_LOG = ".bitfab/replay-events.jsonl";
const DEFAULT_ITEMS_DIR = ".bitfab/replay-items";
/**
 * Where the full ReplayResult JSON is written (and read back from) when `--out`
 * is not passed. Kept for backwards compatibility with replay scripts that
 * predate the run-dir event log: scripts that print human output to stdout and
 * write the result file themselves still resolve (via readExistingResult), and
 * old consumers that read this path still find it. The events.jsonl `complete`
 * row is the modern source; this file is the compatibility shim.
 */
const DEFAULT_OUT = "scripts/replay-result.json";
const DEFAULT_REPLAYS_DIR = ".bitfab/replays";
function hasOwn(obj, key) {
    return Object.hasOwn(obj, key);
}
function parseEvent(json) {
    try {
        const obj = JSON.parse(json);
        if (typeof obj.completed === "number" &&
            typeof obj.total === "number" &&
            typeof obj.succeeded === "number" &&
            typeof obj.errored === "number") {
            const event = {
                completed: obj.completed,
                total: obj.total,
                succeeded: obj.succeeded,
                errored: obj.errored,
            };
            const testRunId = obj.testRunId ?? obj.test_run_id;
            if (typeof testRunId === "string") {
                event.testRunId = testRunId;
            }
            const item = obj.item;
            if (item && typeof item === "object") {
                const rec = item;
                // Accept camelCase (TS) and snake_case (Python/Ruby).
                const traceId = rec.traceId ?? rec.trace_id;
                const replayTraceId = rec.replayTraceId ?? rec.replay_trace_id;
                const durationMs = rec.durationMs ?? rec.duration_ms;
                event.item = {
                    traceId: typeof traceId === "string" ? traceId : null,
                    replayTraceId: typeof replayTraceId === "string" ? replayTraceId : null,
                    error: typeof rec.error === "string" ? rec.error : null,
                    durationMs: typeof durationMs === "number" ? durationMs : null,
                    ...(hasOwn(rec, "input") ? { input: rec.input } : {}),
                    ...(hasOwn(rec, "result") ? { result: rec.result } : {}),
                    ...(hasOwn(rec, "originalOutput")
                        ? { originalOutput: rec.originalOutput }
                        : {}),
                    ...(hasOwn(rec, "original_output")
                        ? { originalOutput: rec.original_output }
                        : {}),
                    ...(hasOwn(rec, "tokens") ? { tokens: rec.tokens } : {}),
                    ...(hasOwn(rec, "model") ? { model: rec.model } : {}),
                    ...(hasOwn(rec, "dbSnapshotRef")
                        ? { dbSnapshotRef: rec.dbSnapshotRef }
                        : {}),
                    ...(hasOwn(rec, "db_snapshot_ref")
                        ? { dbSnapshotRef: rec.db_snapshot_ref }
                        : {}),
                };
            }
            return event;
        }
    }
    catch {
        // A progress UI must never crash the run: drop malformed lines.
    }
    return null;
}
/**
 * Run a replay and emit one ready-to-relay line per trace to stdout while it
 * runs, so a coding agent can background this command and simply show each new
 * line to the user (no parsing, no cadence decisions on the agent).
 *
 * The replay's own stdout (the ReplayResult JSON) is captured and appended as a
 * final `complete` event in the JSONL event log. This command's stdout carries
 * ONLY the progress lines (a header, one line per trace as it settles with its
 * pass/fail glyph + `n/total` + duration, and a final summary). Exits with the
 * replay's exit code.
 */
export async function runReplayProgress() {
    printHelpIfRequested({
        description: "Run a replay and emit one ready-to-relay line per trace to stdout for a coding agent to relay. " +
            "Background this command: it prints a header, one line per trace as it finishes (pass/fail glyph, " +
            "running n/total, and that trace's duration), and a final summary, and appends progress plus the final " +
            "ReplayResult JSON to the JSONL event log. Everything after `--` is the replay command, run as-is.",
        positional: [
            {
                name: "command",
                variadic: true,
                description: "The replay command and its arguments, after a `--` separator, e.g. " +
                    "`replayProgress --label <key> -- npx tsx scripts/replay.ts <key> --dataset-id <id>`.",
            },
        ],
        flags: [
            { name: "label", description: "Function key shown in the header." },
            {
                name: "total",
                format: "number",
                description: "Total trace count for the header, when known up front.",
            },
            {
                name: "out",
                format: "path",
                description: "Optional compatibility path where the full ReplayResult JSON should also be written.",
            },
            {
                name: "log",
                format: "path",
                description: `Append every progress line here too, fresh per run, for \`tail -f\` (default <run-dir>/progress.log).`,
            },
            {
                name: "events-log",
                format: "path",
                description: `Append normalized replay events here as JSONL, fresh per run, for automation (default <run-dir>/events.jsonl).`,
            },
            {
                name: "items-dir",
                format: "path",
                description: `Write full settled replay item payloads here, fresh per run (default <run-dir>/items).`,
            },
            {
                name: "run-dir",
                format: "path",
                description: `Write run artifacts under this directory (default ${DEFAULT_REPLAYS_DIR}/<unique-run-id>).`,
            },
        ],
    });
    const argv = process.argv.slice(2);
    const sepIndex = argv.indexOf("--");
    const head = sepIndex >= 0 ? argv.slice(0, sepIndex) : [];
    const command = sepIndex >= 0 ? argv.slice(sepIndex + 1) : argv;
    let label;
    let total;
    let out = DEFAULT_OUT;
    let logPath;
    let eventsLogPath;
    let itemsDir;
    let runDir;
    for (let i = 0; i < head.length; i++) {
        if (head[i] === "--label") {
            label = head[++i];
        }
        else if (head[i] === "--total") {
            const n = Number(head[++i]);
            total = Number.isFinite(n) && n > 0 ? n : undefined;
        }
        else if (head[i] === "--out") {
            out = head[++i] ?? DEFAULT_OUT;
        }
        else if (head[i] === "--log") {
            logPath = head[++i] ?? DEFAULT_LOG;
        }
        else if (head[i] === "--events-log") {
            eventsLogPath = head[++i] ?? DEFAULT_EVENTS_LOG;
        }
        else if (head[i] === "--items-dir") {
            itemsDir = head[++i] ?? DEFAULT_ITEMS_DIR;
        }
        else if (head[i] === "--run-dir") {
            runDir = head[++i];
        }
    }
    if (command.length === 0) {
        console.error("Usage: replayProgress [--label <key>] [--total N] [--out <path>] [--log <path>] [--events-log <path>] [--items-dir <path>] -- <command> [args...]\n" +
            "Everything after -- is the replay command to run.");
        process.exit(1);
    }
    const [cmd, ...args] = command;
    const startedAt = Date.now();
    const artifacts = resolveReplayArtifacts({
        runDir,
        logPath,
        eventsLogPath,
        itemsDir,
    });
    logPath = artifacts.logPath;
    eventsLogPath = artifacts.eventsLogPath;
    itemsDir = artifacts.itemsDir;
    writeRunMetadata(artifacts, { label, startedAt, status: "running" });
    ensureReplayArtifactsExcluded(process.cwd());
    // Tee every line to a log file (fresh per run) so the user can `tail -f` it for
    // a live, uncollapsing view the tool card can't give. Best-effort: a logging
    // failure must never crash or stall the replay, so any error drops the stream.
    let logStream = null;
    try {
        const target = resolve(logPath);
        mkdirSync(dirname(target), { recursive: true });
        logStream = createWriteStream(target, { flags: "w" });
        logStream.on("error", () => {
            logStream = null;
        });
    }
    catch {
        logStream = null;
    }
    let eventsLogStream = null;
    let progressItemFileCount = 0;
    try {
        const target = resolve(eventsLogPath);
        mkdirSync(dirname(target), { recursive: true });
        eventsLogStream = createWriteStream(target, { flags: "w" });
        eventsLogStream.on("error", () => {
            eventsLogStream = null;
        });
    }
    catch {
        eventsLogStream = null;
    }
    const resolvedItemsDir = resolve(itemsDir);
    try {
        rmSync(resolvedItemsDir, { recursive: true, force: true });
        mkdirSync(resolvedItemsDir, { recursive: true });
    }
    catch {
        // Item files are best-effort: progress lines and final status should still
        // stream even if the item directory is not writable.
    }
    // Track when the last real line printed, so the heartbeat fires exactly one
    // interval after the run goes quiet (not up to two). Every line (header,
    // per-trace, heartbeat, summary) goes through emitLine so it prints, tees to
    // the log, and stamps the clock.
    let lastLineAt = Date.now();
    function emitLine(line) {
        process.stdout.write(`${line}\n`);
        logStream?.write(`${line}\n`);
        lastLineAt = Date.now();
    }
    // Header first, so the user knows a replay started before the first trace.
    emitLine(formatReplayHeader(label, total));
    clearPreviousResult(out);
    // stdin inherited; stdout + stderr piped so we capture the result JSON and
    // parse the progress lines. Our OWN stdout carries only progress lines.
    const child = spawn(cmd, args, { stdio: ["inherit", "pipe", "pipe"] });
    let resultBuf = "";
    child.stdout?.on("data", (d) => {
        resultBuf += d;
    });
    let lastEvent = null;
    let testRunId = null;
    function noteTestRunId(id) {
        if (!id || testRunId === id) {
            return;
        }
        testRunId = id;
        writeRunMetadata(artifacts, {
            label,
            startedAt,
            status: "running",
            testRunId,
        });
        writeTestRunPointer(artifacts, testRunId);
    }
    // Sum of each trace's own duration, for the average in the summary.
    let sumDurationMs = 0;
    if (child.stderr) {
        const rl = createInterface({ input: child.stderr });
        rl.on("line", (line) => {
            if (!line.startsWith(PROGRESS_SENTINEL)) {
                // Forward the replay's own stderr (its errors and human summary) instead
                // of swallowing it. Progress lines go to OUR stdout, so the child's
                // stderr stays on stderr: wrapping a replay is then never worse than
                // running it bare, which is the whole point of capturing this stream.
                process.stderr.write(`${line}\n`);
                return;
            }
            const event = parseEvent(line.slice(PROGRESS_SENTINEL.length));
            if (!event) {
                return;
            }
            noteTestRunId(event.testRunId);
            lastEvent = event;
            const d = event.item?.durationMs;
            if (typeof d === "number" && Number.isFinite(d) && d >= 0) {
                sumDurationMs += d;
            }
            const externalized = externalizeProgressEvent(event, resolvedItemsDir, progressItemFileCount);
            if (externalized.itemPath) {
                progressItemFileCount += 1;
            }
            eventsLogStream?.write(`${JSON.stringify({ type: "progress", ...externalized.event })}\n`);
            // One self-contained line per trace as it settles.
            emitLine(formatTraceLine(event));
        });
    }
    // Liveness heartbeat: print one once a full interval has passed with no real
    // line, then re-arm so it repeats while still quiet (elapsed climbing). The tick
    // runs finer than the interval so the first heartbeat lands close to one full
    // interval after the last line, not up to two. unref so it never keeps the
    // process alive on its own.
    const heartbeat = setInterval(() => {
        if (Date.now() - lastLineAt < HEARTBEAT_MS) {
            return;
        }
        emitLine(formatHeartbeat(lastEvent?.completed ?? 0, total ?? lastEvent?.total ?? 0, Date.now() - startedAt));
    }, Math.min(HEARTBEAT_MS, 2_000));
    heartbeat.unref?.();
    // Set process.exitCode and let the process exit naturally once stdout drains.
    // Calling process.exit() here would race the flush: under an agent's Bash tool
    // stdout is a PIPE (not a TTY), so its writes are async, and a hard exit drops
    // every buffered progress line. The child has fully closed by now and the
    // readline interface ends with it, so nothing keeps the loop alive.
    await new Promise((resolve) => {
        child.on("error", (err) => {
            clearInterval(heartbeat);
            console.error(`replayProgress: failed to run ${cmd}: ${err.message}`);
            logStream?.end();
            eventsLogStream?.end();
            process.exitCode = 1;
            resolve();
        });
        child.on("close", (code, signal) => {
            clearInterval(heartbeat);
            const result = parseReplayResult(out, resultBuf);
            noteTestRunId(extractTestRunId(result.result));
            const resultAvailable = result.ok;
            if (result.result !== undefined) {
                const completeItemRefs = externalizeCompleteResult(result.result, resolvedItemsDir);
                eventsLogStream?.write(`${JSON.stringify({
                    type: "complete",
                    result: buildCompleteSummary(result.result),
                    items: completeItemRefs,
                })}\n`);
            }
            const failed = signal != null || (code ?? 0) !== 0;
            const resultLocation = out ?? eventsLogPath;
            if (lastEvent && failed) {
                emitLine(formatReplayCommandFailure({
                    code: code ?? null,
                    signal: signal ?? null,
                    out: resultLocation,
                }));
            }
            else if (lastEvent && !resultAvailable) {
                emitLine(formatReplayResultMissing(resultLocation));
            }
            else if (lastEvent) {
                emitLine(formatReplaySummary({
                    completed: lastEvent.completed,
                    errored: lastEvent.errored,
                    elapsedMs: Date.now() - startedAt,
                    sumDurationMs,
                }));
            }
            else {
                // No trace ever settled. Always close with a clear line instead of going
                // silent, since "no per-trace lines" otherwise reads as a hang: say
                // whether the replay ran-but-reported-nothing (too-old SDK / unwired
                // onProgress) or failed outright.
                emitLine(formatNoProgressNotice({
                    failed,
                    code: code ?? null,
                    signal: signal ?? null,
                    out: resultLocation,
                }));
            }
            logStream?.end();
            eventsLogStream?.end();
            process.exitCode = signal || !resultAvailable ? 1 : (code ?? 0);
            writeRunMetadata(artifacts, {
                label,
                startedAt,
                finishedAt: Date.now(),
                status: process.exitCode === 0 ? "complete" : "failed",
                testRunId,
            });
            resolve();
        });
    });
}
function resolveReplayArtifacts(input) {
    const usesLegacyDefaults = !input.runDir &&
        (input.logPath !== undefined ||
            input.eventsLogPath !== undefined ||
            input.itemsDir !== undefined);
    const runId = makeRunId();
    const resolvedRunDir = resolve(input.runDir ??
        (usesLegacyDefaults
            ? DEFAULT_REPLAYS_DIR
            : join(DEFAULT_REPLAYS_DIR, runId)));
    return {
        runId,
        runDir: resolvedRunDir,
        logPath: resolve(input.logPath ??
            (usesLegacyDefaults
                ? DEFAULT_LOG
                : join(resolvedRunDir, "progress.log"))),
        eventsLogPath: resolve(input.eventsLogPath ??
            (usesLegacyDefaults
                ? DEFAULT_EVENTS_LOG
                : join(resolvedRunDir, "events.jsonl"))),
        itemsDir: resolve(input.itemsDir ??
            (usesLegacyDefaults
                ? DEFAULT_ITEMS_DIR
                : join(resolvedRunDir, "items"))),
        replaysRoot: usesLegacyDefaults
            ? null
            : input.runDir
                ? dirname(resolvedRunDir)
                : resolve(DEFAULT_REPLAYS_DIR),
    };
}
function makeRunId() {
    const timestamp = new Date().toISOString().replace(/[^0-9A-Za-z]/g, "");
    return `${timestamp}-${process.pid}-${randomUUID().slice(0, 8)}`;
}
function writeRunMetadata(artifacts, metadata) {
    try {
        mkdirSync(artifacts.runDir, { recursive: true });
        writeFileSync(join(artifacts.runDir, "run.json"), `${JSON.stringify({
            runId: artifacts.runId,
            label: metadata.label ?? null,
            status: metadata.status,
            testRunId: metadata.testRunId ?? null,
            startedAt: new Date(metadata.startedAt).toISOString(),
            finishedAt: metadata.finishedAt
                ? new Date(metadata.finishedAt).toISOString()
                : null,
            eventsLogPath: artifacts.eventsLogPath,
            itemsDir: artifacts.itemsDir,
            logPath: artifacts.logPath,
        }, null, 2)}\n`);
    }
    catch {
        // Best-effort metadata must never block the replay.
    }
}
function writeTestRunPointer(artifacts, testRunId) {
    if (!artifacts.replaysRoot) {
        return;
    }
    try {
        const dir = join(artifacts.replaysRoot, "by-test-run", sanitizeFilename(testRunId));
        mkdirSync(dir, { recursive: true });
        writeFileSync(join(dir, "run.json"), `${JSON.stringify({
            testRunId,
            runId: artifacts.runId,
            runDir: artifacts.runDir,
            eventsLogPath: artifacts.eventsLogPath,
            itemsDir: artifacts.itemsDir,
            logPath: artifacts.logPath,
        }, null, 2)}\n`);
    }
    catch {
        // Best-effort lookup metadata; the run directory itself is authoritative.
    }
}
function extractTestRunId(result) {
    if (!result || typeof result !== "object") {
        return null;
    }
    const rec = result;
    if (typeof rec.testRunId === "string") {
        return rec.testRunId;
    }
    if (typeof rec.test_run_id === "string") {
        return rec.test_run_id;
    }
    return null;
}
function ensureReplayArtifactsExcluded(startDir) {
    const gitDir = findGitDir(startDir);
    if (!gitDir) {
        return;
    }
    const excludePath = join(gitDir, "info", "exclude");
    try {
        const existing = existsSync(excludePath)
            ? readFileSync(excludePath, "utf8")
            : "";
        const patterns = [
            ".bitfab/replays/",
            ".bitfab/replay-events.jsonl",
            ".bitfab/replay-progress.log",
            ".bitfab/replay-items/",
            "scripts/replay-result.json",
        ];
        const missing = patterns.filter((pattern) => !existing.includes(pattern));
        if (missing.length === 0) {
            return;
        }
        mkdirSync(dirname(excludePath), { recursive: true });
        appendFileSync(excludePath, `${existing.endsWith("\n") || existing.length === 0 ? "" : "\n"}# Bitfab replay runtime artifacts\n${missing.join("\n")}\n`);
    }
    catch {
        // Best-effort: ignore metadata should never block a replay.
    }
}
function findGitDir(startDir) {
    let current = resolve(startDir);
    while (true) {
        const candidate = join(current, ".git");
        try {
            if (existsSync(candidate)) {
                const stat = statSync(candidate);
                if (stat.isDirectory()) {
                    return candidate;
                }
                const contents = readFileSync(candidate, "utf8");
                const match = /^gitdir:\s*(.+)$/m.exec(contents);
                if (match?.[1]) {
                    return resolve(current, match[1].trim());
                }
            }
        }
        catch {
            return null;
        }
        const parent = dirname(current);
        if (parent === current) {
            return null;
        }
        current = parent;
    }
}
function clearPreviousResult(out) {
    if (!out) {
        return;
    }
    try {
        rmSync(resolve(out), { force: true });
    }
    catch {
        // Best-effort: a stale result should not block the replay command itself.
    }
}
/**
 * Parse the replay's captured stdout (the ReplayResult JSON) for the final event,
 * but ONLY when that stdout actually parses as JSON. A replay that prints human
 * text or env-loader / banner noise to stdout would otherwise clobber a clean
 * result file the script wrote itself; in that case leave the script's file
 * untouched and warn, so malformed stdout is never silently treated as a result.
 */
function parseReplayResult(out, body) {
    if (body.trim().length === 0) {
        return readExistingResult(out);
    }
    let result;
    try {
        result = JSON.parse(body);
    }
    catch {
        process.stderr.write(`replayProgress: captured stdout is not valid JSON${out ? `; left ${out} as the replay script wrote it` : ""}. ` +
            "The replay script should print its ReplayResult JSON to stdout and human output to stderr.\n");
        return readExistingResult(out);
    }
    writeCompatibilityResult(out, body);
    return { ok: true, result };
}
function writeCompatibilityResult(out, body) {
    if (!out) {
        return;
    }
    try {
        const target = resolve(out);
        mkdirSync(dirname(target), { recursive: true });
        writeFileSync(target, body);
    }
    catch {
        // Best-effort: the event log still carries the parsed result.
    }
}
function readExistingResult(out) {
    if (!out) {
        return { ok: false };
    }
    try {
        return { ok: true, result: JSON.parse(readFileSync(out, "utf8")) };
    }
    catch {
        return { ok: false };
    }
}
function externalizeProgressEvent(event, itemsDir, index) {
    if (!event.item) {
        return { event };
    }
    const item = event.item;
    const hasPayload = hasValue(item.input) ||
        hasValue(item.result) ||
        hasValue(item.originalOutput) ||
        hasValue(item.tokens) ||
        hasValue(item.model) ||
        hasValue(item.dbSnapshotRef);
    if (!hasPayload) {
        return { event };
    }
    const key = item.replayTraceId ?? item.traceId ?? `progress-${index}`;
    const itemPath = writeItemFile(itemsDir, key, {
        testRunId: event.testRunId ?? null,
        sourceTraceId: item.traceId ?? null,
        replayTraceId: item.replayTraceId ?? null,
        input: item.input,
        result: item.result,
        originalOutput: item.originalOutput,
        error: item.error ?? null,
        durationMs: item.durationMs ?? null,
        tokens: item.tokens,
        model: item.model,
        dbSnapshotRef: item.dbSnapshotRef,
    });
    if (!itemPath) {
        return { event };
    }
    return {
        itemPath,
        event: {
            ...event,
            item: {
                traceId: item.traceId,
                replayTraceId: item.replayTraceId,
                error: item.error,
                durationMs: item.durationMs,
                itemPath,
            },
        },
    };
}
function externalizeCompleteResult(result, itemsDir) {
    if (!result || typeof result !== "object") {
        return [];
    }
    const rec = result;
    if (!Array.isArray(rec.items)) {
        return [];
    }
    const refs = [];
    for (let i = 0; i < rec.items.length; i++) {
        const rawItem = rec.items[i];
        if (!rawItem || typeof rawItem !== "object") {
            continue;
        }
        const item = rawItem;
        const serverTraceId = typeof item.traceId === "string"
            ? item.traceId
            : typeof item.trace_id === "string"
                ? item.trace_id
                : null;
        const durationMs = item.durationMs ?? item.duration_ms;
        const itemPath = writeItemFile(itemsDir, serverTraceId ?? `complete-${i}`, {
            testRunId: typeof rec.testRunId === "string"
                ? rec.testRunId
                : typeof rec.test_run_id === "string"
                    ? rec.test_run_id
                    : null,
            serverTraceId,
            input: item.input,
            result: item.result,
            originalOutput: item.originalOutput ?? item.original_output,
            error: item.error ?? null,
            durationMs: typeof durationMs === "number" ? durationMs : null,
            tokens: item.tokens,
            model: item.model,
            dbSnapshotRef: item.dbSnapshotRef ?? item.db_snapshot_ref,
        });
        if (!itemPath) {
            continue;
        }
        refs.push({
            serverTraceId,
            itemPath,
            error: typeof item.error === "string" ? item.error : null,
            durationMs: typeof durationMs === "number" ? durationMs : null,
        });
    }
    return refs;
}
function buildCompleteSummary(result) {
    if (!result || typeof result !== "object") {
        return {};
    }
    const rec = result;
    const items = Array.isArray(rec.items) ? rec.items : [];
    return {
        ...(typeof rec.testRunId === "string" ? { testRunId: rec.testRunId } : {}),
        ...(typeof rec.test_run_id === "string"
            ? { testRunId: rec.test_run_id }
            : {}),
        ...(typeof rec.testRunUrl === "string"
            ? { testRunUrl: rec.testRunUrl }
            : {}),
        ...(typeof rec.test_run_url === "string"
            ? { testRunUrl: rec.test_run_url }
            : {}),
        itemCount: items.length,
    };
}
function writeItemFile(itemsDir, key, payload) {
    try {
        mkdirSync(itemsDir, { recursive: true });
        const filename = `${sanitizeFilename(key)}.json`;
        const target = join(itemsDir, filename);
        const tmp = `${target}.tmp-${process.pid}-${Date.now()}`;
        writeFileSync(tmp, `${JSON.stringify(payload)}\n`);
        renameSync(tmp, target);
        return target;
    }
    catch {
        return null;
    }
}
function sanitizeFilename(value) {
    const cleaned = value.replace(/[^A-Za-z0-9._-]/g, "_");
    return cleaned.length > 0 ? cleaned : "item";
}
function hasValue(value) {
    return value !== undefined;
}
