// bitfab-plugin-lib/dist/commands/replayProgress.js
import { spawn } from "node:child_process";
import { createWriteStream, existsSync, mkdirSync, readFileSync, rmSync, writeFileSync, } from "node:fs";
import { dirname, resolve } from "node:path";
import { createInterface } from "node:readline";
import { printHelpIfRequested } from "../parseArgs.js";
import { formatHeartbeat, formatNoProgressNotice, formatReplayCommandFailure, formatReplayHeader, formatReplayResultMissing, formatReplaySummary, formatTraceLine, } from "../replayProgressView.js";
/**
 * How long to go without a per-trace line before emitting a liveness heartbeat,
 * in milliseconds. A single replayed trace (an LLM call) can run tens of seconds;
 * without this the run would print nothing in that gap and look frozen. Overridable
 * via env for tests (which can't wait the real interval).
 */
const HEARTBEAT_MS = Number(process.env.BITFAB_REPLAY_HEARTBEAT_MS) || 12_000;
/**
 * Prefix the SDK reporter writes to the replay's stderr, one line per trace:
 *
 *   @@bitfab:progress {"completed":3,"total":10,"succeeded":3,"errored":0,"item":{"traceId":"...","error":null}}
 *
 * stderr (not an extra fd) because progress has to survive the launcher hops
 * the agent uses (`npx tsx`, `python`, `bundle exec`), which don't forward
 * extra fds; stderr/stdout are inherited through every hop.
 */
const PROGRESS_SENTINEL = "@@bitfab:progress ";
const DEFAULT_OUT = "scripts/replay-result.json";
/**
 * Every line this command prints to stdout is also appended here, fresh per run,
 * so the user can `tail -f` it in a separate terminal for a truly live, never-
 * collapsing view. Claude Code's tool card collapses output and can't stream; a
 * tail is the one surface that shows every line the instant it is written.
 */
const DEFAULT_LOG = ".bitfab/replay-progress.log";
function parseEvent(json) {
    try {
        const obj = JSON.parse(json);
        if (typeof obj.completed === "number" &&
            typeof obj.total === "number" &&
            typeof obj.succeeded === "number" &&
            typeof obj.errored === "number") {
            const event = {
                completed: obj.completed,
                total: obj.total,
                succeeded: obj.succeeded,
                errored: obj.errored,
            };
            const item = obj.item;
            if (item && typeof item === "object") {
                const rec = item;
                // Accept camelCase (TS) and snake_case (Python/Ruby).
                const traceId = rec.traceId ?? rec.trace_id;
                const durationMs = rec.durationMs ?? rec.duration_ms;
                event.item = {
                    traceId: typeof traceId === "string" ? traceId : null,
                    error: typeof rec.error === "string" ? rec.error : null,
                    durationMs: typeof durationMs === "number" ? durationMs : null,
                };
            }
            return event;
        }
    }
    catch {
        // A progress UI must never crash the run: drop malformed lines.
    }
    return null;
}
/**
 * Run a replay and emit one ready-to-relay line per trace to stdout while it
 * runs, so a coding agent can background this command and simply show each new
 * line to the user (no parsing, no cadence decisions on the agent).
 *
 * The replay's own stdout (the ReplayResult JSON) is captured and written to the
 * `--out` file (default `scripts/replay-result.json`) for the agent to read in
 * full, and this command's stdout carries ONLY the progress lines (a header, one
 * line per trace as it settles with its pass/fail glyph + `n/total` + duration,
 * and a final summary). Exits with the replay's exit code.
 */
export async function runReplayProgress() {
    printHelpIfRequested({
        description: "Run a replay and emit one ready-to-relay line per trace to stdout for a coding agent to relay. " +
            "Background this command: it prints a header, one line per trace as it finishes (pass/fail glyph, " +
            "running n/total, and that trace's duration), and a final summary, and writes the replay's full " +
            "ReplayResult JSON to the --out file. Everything after `--` is the replay command, run as-is.",
        positional: [
            {
                name: "command",
                variadic: true,
                description: "The replay command and its arguments, after a `--` separator, e.g. " +
                    "`replayProgress --label <key> -- npx tsx scripts/replay.ts <key> --dataset-id <id>`.",
            },
        ],
        flags: [
            { name: "label", description: "Function key shown in the header." },
            {
                name: "total",
                format: "number",
                description: "Total trace count for the header, when known up front.",
            },
            {
                name: "out",
                format: "path",
                description: `Where to write the full ReplayResult JSON (default ${DEFAULT_OUT}).`,
            },
            {
                name: "log",
                format: "path",
                description: `Append every progress line here too, fresh per run, for \`tail -f\` (default ${DEFAULT_LOG}).`,
            },
        ],
    });
    const argv = process.argv.slice(2);
    const sepIndex = argv.indexOf("--");
    const head = sepIndex >= 0 ? argv.slice(0, sepIndex) : [];
    const command = sepIndex >= 0 ? argv.slice(sepIndex + 1) : argv;
    let label;
    let total;
    let out = DEFAULT_OUT;
    let logPath = DEFAULT_LOG;
    for (let i = 0; i < head.length; i++) {
        if (head[i] === "--label") {
            label = head[++i];
        }
        else if (head[i] === "--total") {
            const n = Number(head[++i]);
            total = Number.isFinite(n) && n > 0 ? n : undefined;
        }
        else if (head[i] === "--out") {
            out = head[++i] ?? DEFAULT_OUT;
        }
        else if (head[i] === "--log") {
            logPath = head[++i] ?? DEFAULT_LOG;
        }
    }
    if (command.length === 0) {
        console.error("Usage: replayProgress [--label <key>] [--total N] [--out <path>] [--log <path>] -- <command> [args...]\n" +
            "Everything after -- is the replay command to run.");
        process.exit(1);
    }
    const [cmd, ...args] = command;
    const startedAt = Date.now();
    // Tee every line to a log file (fresh per run) so the user can `tail -f` it for
    // a live, uncollapsing view the tool card can't give. Best-effort: a logging
    // failure must never crash or stall the replay, so any error drops the stream.
    let logStream = null;
    try {
        const target = resolve(logPath);
        mkdirSync(dirname(target), { recursive: true });
        logStream = createWriteStream(target, { flags: "w" });
        logStream.on("error", () => {
            logStream = null;
        });
    }
    catch {
        logStream = null;
    }
    // Track when the last real line printed, so the heartbeat fires exactly one
    // interval after the run goes quiet (not up to two). Every line (header,
    // per-trace, heartbeat, summary) goes through emitLine so it prints, tees to
    // the log, and stamps the clock.
    let lastLineAt = Date.now();
    function emitLine(line) {
        process.stdout.write(`${line}\n`);
        logStream?.write(`${line}\n`);
        lastLineAt = Date.now();
    }
    // Header first, so the user knows a replay started before the first trace.
    emitLine(formatReplayHeader(label, total));
    clearPreviousResult(out);
    // stdin inherited; stdout + stderr piped so we capture the result JSON and
    // parse the progress lines. Our OWN stdout carries only progress lines.
    const child = spawn(cmd, args, { stdio: ["inherit", "pipe", "pipe"] });
    let resultBuf = "";
    child.stdout?.on("data", (d) => {
        resultBuf += d;
    });
    let lastEvent = null;
    // Sum of each trace's own duration, for the average in the summary.
    let sumDurationMs = 0;
    if (child.stderr) {
        const rl = createInterface({ input: child.stderr });
        rl.on("line", (line) => {
            if (!line.startsWith(PROGRESS_SENTINEL)) {
                // Forward the replay's own stderr (its errors and human summary) instead
                // of swallowing it. Progress lines go to OUR stdout, so the child's
                // stderr stays on stderr: wrapping a replay is then never worse than
                // running it bare, which is the whole point of capturing this stream.
                process.stderr.write(`${line}\n`);
                return;
            }
            const event = parseEvent(line.slice(PROGRESS_SENTINEL.length));
            if (!event) {
                return;
            }
            lastEvent = event;
            const d = event.item?.durationMs;
            if (typeof d === "number" && Number.isFinite(d) && d >= 0) {
                sumDurationMs += d;
            }
            // One self-contained line per trace as it settles.
            emitLine(formatTraceLine(event));
        });
    }
    // Liveness heartbeat: print one once a full interval has passed with no real
    // line, then re-arm so it repeats while still quiet (elapsed climbing). The tick
    // runs finer than the interval so the first heartbeat lands close to one full
    // interval after the last line, not up to two. unref so it never keeps the
    // process alive on its own.
    const heartbeat = setInterval(() => {
        if (Date.now() - lastLineAt < HEARTBEAT_MS) {
            return;
        }
        emitLine(formatHeartbeat(lastEvent?.completed ?? 0, total ?? lastEvent?.total ?? 0, Date.now() - startedAt));
    }, Math.min(HEARTBEAT_MS, 2_000));
    heartbeat.unref?.();
    // Set process.exitCode and let the process exit naturally once stdout drains.
    // Calling process.exit() here would race the flush: under an agent's Bash tool
    // stdout is a PIPE (not a TTY), so its writes are async, and a hard exit drops
    // every buffered progress line. The child has fully closed by now and the
    // readline interface ends with it, so nothing keeps the loop alive.
    await new Promise((resolve) => {
        child.on("error", (err) => {
            clearInterval(heartbeat);
            console.error(`replayProgress: failed to run ${cmd}: ${err.message}`);
            logStream?.end();
            process.exitCode = 1;
            resolve();
        });
        child.on("close", (code, signal) => {
            clearInterval(heartbeat);
            const resultAvailable = writeResult(out, resultBuf);
            const failed = signal != null || (code ?? 0) !== 0;
            if (lastEvent && failed) {
                emitLine(formatReplayCommandFailure({
                    code: code ?? null,
                    signal: signal ?? null,
                    out,
                }));
            }
            else if (lastEvent && !resultAvailable) {
                emitLine(formatReplayResultMissing(out));
            }
            else if (lastEvent) {
                emitLine(formatReplaySummary({
                    completed: lastEvent.completed,
                    errored: lastEvent.errored,
                    elapsedMs: Date.now() - startedAt,
                    sumDurationMs,
                }));
            }
            else {
                // No trace ever settled. Always close with a clear line instead of going
                // silent, since "no per-trace lines" otherwise reads as a hang: say
                // whether the replay ran-but-reported-nothing (too-old SDK / unwired
                // onProgress) or failed outright.
                emitLine(formatNoProgressNotice({
                    failed,
                    code: code ?? null,
                    signal: signal ?? null,
                    out,
                }));
            }
            logStream?.end();
            process.exitCode = signal || !resultAvailable ? 1 : (code ?? 0);
            resolve();
        });
    });
}
function clearPreviousResult(out) {
    try {
        rmSync(resolve(out), { force: true });
    }
    catch {
        // Best-effort: a stale result should not block the replay command itself.
    }
}
/**
 * Persist the replay's captured stdout (the ReplayResult JSON) for the agent, but
 * ONLY when that stdout actually parses as JSON. A replay that prints human text
 * or env-loader / banner noise to stdout would otherwise clobber a clean result
 * file the script wrote itself; in that case leave the script's file untouched and
 * warn, so a malformed stdout is never silently written over a good `--out`.
 */
function writeResult(out, body) {
    if (body.trim().length === 0) {
        return existingResultIsJson(out);
    }
    try {
        JSON.parse(body);
    }
    catch {
        process.stderr.write(`replayProgress: captured stdout is not valid JSON; left ${out} as the replay script wrote it. ` +
            "The replay script should print its ReplayResult JSON to stdout and human output to stderr.\n");
        return existingResultIsJson(out);
    }
    try {
        const target = resolve(out);
        mkdirSync(dirname(target), { recursive: true });
        writeFileSync(target, body);
        return true;
    }
    catch {
        // Best-effort: the agent can still parse the JSON tail of the run's output.
        return existingResultIsJson(out);
    }
}
function existingResultIsJson(out) {
    try {
        if (!existsSync(out)) {
            return false;
        }
        JSON.parse(readFileSync(out, "utf8"));
        return true;
    }
    catch {
        return false;
    }
}
