// bitfab-plugin-lib/dist/commands/startDataset.js
import { reportHandoff } from "../analytics.js";
import { requireAuth } from "../config.js";
import { emitJson } from "../output.js";
import { positionalArgs } from "../parseArgs.js";
import { handleStudioError, openStudioTo } from "./openStudioTo.js";
/**
 * Open the per-dataset review page in Studio and stay alive until the user
 * clicks Done (or the session ends). Stdout is a JSON line stream:
 *
 *   {"event":"modify","datasetId":"...","ts":"..."}     // non-terminal
 *   {"event":"saved","status":"saved"}                  // terminal
 *   {"event":"cancelled","status":"cancelled"}          // terminal
 *
 * After a terminal event the process exits 0. The calling agent reads stdout
 * incrementally, reacting to each modify event by doing add/remove work.
 *
 * Free-form logs go to stderr so stdout stays a parseable JSONL stream.
 */
const ARGS = [
    { name: "functionKey" },
    { name: "datasetId", format: "uuid" },
];
export async function runStartDataset(platform, pluginVersion) {
    const args = positionalArgs(ARGS);
    const functionKey = args[0];
    const datasetId = args[1];
    const { apiKey, serviceUrl } = requireAuth(platform.loginHint);
    const targetPath = `/studio/trace-functions/${encodeURIComponent(functionKey)}/datasets/${encodeURIComponent(datasetId)}`;
    let abort = () => { };
    try {
        const result = await openStudioTo(targetPath, {
            apiKey,
            serviceUrl,
            onEvent: (event) => {
                if (event.type === "studio:edit-with-agent") {
                    emitJson({
                        event: "modify",
                        datasetId: event.data.datasetId,
                        ts: new Date().toISOString(),
                    });
                    return;
                }
                if (event.type === "studio:return-to-agent") {
                    emitJson({ event: "saved", status: "saved" });
                    console.error("Labeling complete [via studio].");
                    abort();
                    return;
                }
                if (event.type === "studio:session-ended") {
                    emitJson({ event: "cancelled", status: "cancelled" });
                    console.error("Labeling ended without saving.");
                    abort();
                }
            },
        });
        abort = result.abort;
        await result.done;
        await reportHandoff(apiKey, pluginVersion, platform, {
            flow: "dataset.start",
            wonBy: "studio",
            loopbackAvailable: false,
            ticketAvailable: false,
        });
        process.exit(0);
    }
    catch (err) {
        handleStudioError(err);
    }
}
