// bitfab-plugin-lib/dist/commands/startTemplatePreview.js
import crypto from "node:crypto";
import { clearActivePreviewSession, writeActivePreviewSession, } from "../activePreviewSession.js";
import { reportHandoff } from "../analytics.js";
import { getConfig, hasCredentials } from "../config.js";
import { emitJson } from "../output.js";
import { positionalArgs } from "../parseArgs.js";
import { loginForSession } from "./loginForSession.js";
import { handleStudioError, openStudioTo } from "./openStudioTo.js";
export function buildTemplatePreviewUrl(serviceUrl, functionKey) {
    return `${serviceUrl}/studio/trace-functions/${encodeURIComponent(functionKey)}/template-preview`;
}
const ARGS = [{ name: "functionKey" }];
export async function runStartTemplatePreview(platform, pluginVersion) {
    const args = positionalArgs(ARGS);
    const functionKey = args[0];
    const config = getConfig();
    let apiKey = hasCredentials() ? config.apiKey : null;
    const serviceUrl = config.serviceUrl;
    const targetPath = `/studio/trace-functions/${encodeURIComponent(functionKey)}/template-preview`;
    if (!apiKey) {
        apiKey = await loginForSession({
            targetPath,
            serviceUrl,
            sessionId: crypto.randomUUID(),
        });
    }
    let abort = () => { };
    try {
        const result = await openStudioTo(targetPath, {
            apiKey,
            serviceUrl,
            onEvent: (event) => {
                if (event.type === "studio:element-clicked") {
                    emitJson({
                        event: "click",
                        ts: new Date().toISOString(),
                        ...event.data,
                    });
                    return;
                }
                if (event.type === "focusChanged") {
                    emitJson({
                        event: "focus",
                        ts: new Date().toISOString(),
                        ...event.data,
                    });
                    return;
                }
                if (event.type === "studio:return-to-agent" ||
                    event.type === "studio:session-ended") {
                    console.log("Template preview closed [via studio].");
                    abort();
                }
            },
        });
        abort = result.abort;
        writeActivePreviewSession({
            sessionId: result.sessionId,
            serviceUrl,
        });
        const cleanupActiveSession = () => {
            try {
                clearActivePreviewSession();
            }
            catch { }
        };
        process.on("exit", cleanupActiveSession);
        const onSignal = () => {
            cleanupActiveSession();
            process.exit(0);
        };
        process.on("SIGINT", onSignal);
        process.on("SIGTERM", onSignal);
        process.on("SIGHUP", onSignal);
        await result.done;
        await reportHandoff(apiKey, pluginVersion, platform, {
            flow: "template.preview",
            wonBy: "studio",
            loopbackAvailable: false,
            ticketAvailable: false,
        });
        process.exit(0);
    }
    catch (err) {
        handleStudioError(err);
    }
}
