// bitfab-plugin-lib/dist/commands/startTemplatePreview.js
import { clearActivePreviewSession, writeActivePreviewSession, } from "../activePreviewSession.js";
import { reportHandoff } from "../analytics.js";
import { getConfig, hasCredentials } from "../config.js";
import { positionalArgs } from "../parseArgs.js";
import { openStudioTo, StudioNavigationError } from "./openStudioTo.js";
function emitJson(obj) {
    process.stdout.write(`${JSON.stringify(obj)}\n`);
}
export function buildTemplatePreviewUrl(serviceUrl, functionKey) {
    return `${serviceUrl}/studio/trace-functions/${encodeURIComponent(functionKey)}/template-preview`;
}
const ARGS = [{ name: "functionKey" }];
export async function runStartTemplatePreview(platform, pluginVersion) {
    const args = positionalArgs(ARGS);
    const functionKey = args[0];
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    const config = getConfig();
    if (!config.apiKey) {
        console.error("API key missing in config.");
        process.exit(1);
    }
    const apiKey = config.apiKey;
    const targetPath = `/studio/trace-functions/${encodeURIComponent(functionKey)}/template-preview`;
    let abort = () => { };
    try {
        const result = await openStudioTo(targetPath, {
            apiKey,
            serviceUrl: config.serviceUrl,
            onEvent: (event) => {
                if (event.type === "studio:element-clicked") {
                    emitJson({
                        event: "click",
                        ts: new Date().toISOString(),
                        ...event.data,
                    });
                    return;
                }
                if (event.type === "focusChanged") {
                    emitJson({
                        event: "focus",
                        ts: new Date().toISOString(),
                        ...event.data,
                    });
                    return;
                }
                if (event.type === "studio:return-to-agent" ||
                    event.type === "studio:session-ended") {
                    console.log("Template preview closed [via studio].");
                    abort();
                }
            },
        });
        abort = result.abort;
        writeActivePreviewSession({
            sessionId: result.sessionId,
            serviceUrl: config.serviceUrl,
        });
        const cleanupActiveSession = () => {
            try {
                clearActivePreviewSession();
            }
            catch { }
        };
        process.on("exit", cleanupActiveSession);
        const onSignal = () => {
            cleanupActiveSession();
            process.exit(0);
        };
        process.on("SIGINT", onSignal);
        process.on("SIGTERM", onSignal);
        process.on("SIGHUP", onSignal);
        await result.done;
        await reportHandoff(apiKey, pluginVersion, platform, {
            flow: "template.preview",
            wonBy: "studio",
            loopbackAvailable: false,
            ticketAvailable: false,
        });
        process.exit(0);
    }
    catch (err) {
        if (err instanceof StudioNavigationError) {
            console.error(`Studio navigation failed: ${err.message}`);
        }
        else {
            console.error(`Failed to open Studio: ${err.message}`);
        }
        process.exit(1);
    }
}
