// bitfab-plugin-lib/dist/commands/switchOrg.js
import { clearActiveStudioSession, readActiveStudioSession, } from "../activeStudioSession.js";
import { requireAuth, saveCredentials } from "../config.js";
import { emitJson } from "../output.js";
import { positionalArgs } from "../parseArgs.js";
import { killPollerProcess, killStudioWindow } from "../studioTeardown.js";
async function requestSwitch(serviceUrl, apiKey, clerkOrganizationId) {
    const res = await fetch(`${serviceUrl}/api/plugin/organizations/switch`, {
        method: "POST",
        headers: {
            Authorization: `Bearer ${apiKey}`,
            "Content-Type": "application/json",
        },
        body: JSON.stringify({ clerkOrganizationId }),
    });
    if (!res.ok) {
        throw new Error(`Failed to switch organization (HTTP ${res.status})`);
    }
    return (await res.json());
}
/**
 * Drop the active Studio pointer (and reap its window + poller) so the next
 * Studio-opening command starts a FRESH session bound to the newly active org.
 * A recorded session is bound in Redis to the previous org and can't rebind;
 * reusing it would show the old org's data. The browser only actually flips
 * when that fresh session's StudioOrgGate runs the client-side `setActive`.
 */
function resetActiveStudio() {
    const active = readActiveStudioSession();
    if (!active) {
        return;
    }
    killStudioWindow(active.sessionId);
    killPollerProcess(active.sessionId);
    clearActiveStudioSession(active.sessionId, { force: true });
}
const ARGS = [{ name: "clerkOrganizationId", required: false }];
/**
 * Switch the plugin's active organization to `<clerkOrganizationId>`:
 *
 *   switchOrg <clerkOrganizationId>
 *     {"event":"switched","status":"switched"|"already-aligned","clerkOrganizationId":"...","organizationName":"...","apiKey":"..."}
 *     {"event":"not-member","clerkOrganizationId":"..."}   // user isn't in that org
 *
 * Discovering which org to pass is the `list_organizations` MCP tool's job (the
 * agent reads it directly); this command only performs the switch. On success
 * the new org's API key is written to credentials.json AND echoed back in the
 * `apiKey` field so the caller can sync the app's runtime `BITFAB_API_KEY` to
 * the switched org: `get_bitfab_api_key` resolves a `BITFAB_API_KEY` process
 * env override ahead of credentials.json, so it can lag a just-completed switch;
 * this field is the authoritative switched-org key. The active Studio pointer is
 * cleared. The live browser does NOT flip on its own;
 * it follows on the next Studio open (a fresh session whose StudioOrgGate runs
 * the client-side `setActive`) or via the in-app org switcher.
 *
 * Free-form logs go to stderr; stdout stays a parseable JSONL stream.
 */
export async function runSwitchOrg(platform) {
    const { apiKey, serviceUrl } = requireAuth(platform.loginHint);
    const target = positionalArgs(ARGS)[0] ?? null;
    if (!target) {
        console.error("Usage: switchOrg <clerkOrganizationId>");
        console.error("Discover org ids with the list_organizations MCP tool, then pass one here.");
        process.exit(1);
    }
    try {
        const result = await requestSwitch(serviceUrl, apiKey, target);
        if (result.status === "not-member") {
            emitJson({ event: "not-member", clerkOrganizationId: target });
            return;
        }
        // Replace the local credential with the (re-resolved) org key.
        saveCredentials(result.apiKey);
        // Only reset Studio when the org actually changed. On `already-aligned` the
        // active Studio session is already bound to the right org, so tearing it
        // down would needlessly close the user's in-progress Studio work.
        if (result.status === "switched") {
            resetActiveStudio();
        }
        emitJson({
            event: "switched",
            status: result.status,
            clerkOrganizationId: result.clerkOrganizationId,
            organizationName: result.organizationName,
            apiKey: result.apiKey,
        });
    }
    catch (err) {
        emitJson({
            event: "error",
            reason: err instanceof Error ? err.message : String(err),
        });
        process.exit(1);
    }
}
