// bitfab-plugin-lib/dist/commands/waitForTrace.js
import { hasCredentials } from "../config.js";
import { createMcpClient } from "../mcpClient.js";
import { parseArgs } from "../parseArgs.js";
const DEFAULT_POLL_INTERVAL_MS = 10_000;
const DEFAULT_TIMEOUT_MS = 10 * 60 * 1000;
const TRACE_ID_REGEX = /ID:\s+([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})/i;
const POSITIONAL = [
    {
        name: "traceFunctionKey",
        description: "Trace function key to watch for its first trace.",
    },
];
const FLAGS = [
    {
        name: "timeout-ms",
        format: "number",
        default: DEFAULT_TIMEOUT_MS,
        description: "How long to keep polling before giving up.",
    },
    {
        name: "poll-interval-ms",
        format: "number",
        default: DEFAULT_POLL_INTERVAL_MS,
        description: "Delay between polls.",
    },
];
/**
 * Pure function - extracts the first trace UUID from the plaintext response
 * returned by the server's `search_traces` tool. Returns null when no match
 * exists (e.g. "No traces found matching the filter criteria.").
 */
export function extractFirstTraceId(text) {
    const match = text.match(TRACE_ID_REGEX);
    return match?.[1] ?? null;
}
export function parseWaitForTraceArgs(argv) {
    const { positional, flags } = parseArgs({
        positional: POSITIONAL,
        flags: FLAGS,
        argv: ["node", "waitForTrace.js", ...argv],
        description: "Block until the first trace for a function arrives, then print one JSON line ({status, traceId, url, waitedMs}) and exit. Status is found, timeout, interrupted, or error.",
    });
    return {
        traceFunctionKey: positional[0],
        timeoutMs: flags["timeout-ms"],
        pollIntervalMs: flags["poll-interval-ms"],
    };
}
/**
 * Core polling loop - repeatedly calls `search_traces` against the given
 * MCP proxy until the first trace arrives, the timeout elapses, or the
 * caller signals interrupt. Exported so tests can drive it with a fake
 * proxy + config instead of real creds + HTTP.
 */
export async function pollForFirstTrace(opts) {
    const startMs = Date.now();
    const isInterrupted = opts.isInterrupted ?? (() => false);
    while (!isInterrupted() && Date.now() - startMs < opts.timeoutMs) {
        const result = await opts.client.call("search_traces", {
            traceFunctionKey: opts.traceFunctionKey,
            limit: 1,
        });
        const text = result.content?.[0]?.text ?? "";
        if (result.isError) {
            return { status: "error", message: text || "search_traces error" };
        }
        const traceId = extractFirstTraceId(text);
        if (traceId) {
            const serviceUrl = opts.client.serviceUrl.replace(/\/$/, "");
            return {
                status: "found",
                traceId,
                url: `${serviceUrl}/traces/${traceId}`,
                waitedMs: Date.now() - startMs,
            };
        }
        await sleep(opts.pollIntervalMs, isInterrupted);
    }
    return {
        status: isInterrupted() ? "interrupted" : "timeout",
        waitedMs: Date.now() - startMs,
    };
}
/**
 * CLI entrypoint. Parses argv, checks creds, invokes `pollForFirstTrace`,
 * prints one JSON line to stdout, and exits. Exists so setup skills can
 * block on the first trace via a single long-timeout `Bash` call instead
 * of a 40-iteration agent-side poll loop.
 */
export async function runWaitForTrace(platform, pluginVersion) {
    const args = parseWaitForTraceArgs(process.argv.slice(2));
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    let interrupted = false;
    const onSignal = () => {
        interrupted = true;
    };
    process.on("SIGINT", onSignal);
    process.on("SIGTERM", onSignal);
    const result = await pollForFirstTrace({
        client: createMcpClient(platform, pluginVersion),
        traceFunctionKey: args.traceFunctionKey,
        timeoutMs: args.timeoutMs,
        pollIntervalMs: args.pollIntervalMs,
        isInterrupted: () => interrupted,
    });
    if (result.status === "error") {
        console.error(result.message);
        process.exit(1);
    }
    console.log(JSON.stringify(result));
    process.exit(0);
}
function sleep(ms, isInterrupted) {
    return new Promise((resolve) => {
        const interval = 200;
        let elapsed = 0;
        const timer = setInterval(() => {
            elapsed += interval;
            if (isInterrupted() || elapsed >= ms) {
                clearInterval(timer);
                resolve();
            }
        }, interval);
    });
}
