// bitfab-plugin-lib/dist/activePreviewSession.js
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
/**
 * Process-local handshake between `startTemplatePreview` and the plugin's
 * MCP server, both of which run as separate Node processes.
 * `startTemplatePreview` writes the active preview session id (and the
 * service URL it bound to) here on launch; the MCP server reads it inside
 * the `update_template` wrapper so it can push `editStarted` / `editEnded`
 * agent events to the right session.
 *
 * Single file = single active preview at a time. Concurrent previews
 * overwrite the file; the most recent one wins. Stale files (writer
 * crashed) are detected at read time via the recorded pid.
 */
const FILE_PATH = path.join(os.homedir(), ".config", "bitfab", "active-preview-session.json");
function processIsAlive(pid) {
    try {
        process.kill(pid, 0);
        return true;
    }
    catch (err) {
        // ESRCH = process not found; EPERM = exists but we can't signal it (still
        // alive). Anything else: treat as alive so we don't accidentally clear a
        // legitimate session.
        return err.code !== "ESRCH";
    }
}
export function writeActivePreviewSession(session) {
    fs.mkdirSync(path.dirname(FILE_PATH), { recursive: true });
    const payload = {
        ...session,
        pid: process.pid,
        startedAt: new Date().toISOString(),
    };
    const tmp = `${FILE_PATH}.${process.pid}.tmp`;
    fs.writeFileSync(tmp, `${JSON.stringify(payload, null, 2)}\n`);
    fs.renameSync(tmp, FILE_PATH);
}
/**
 * Clear the file only if the recorded pid matches the current process. This
 * prevents a slow-exiting old preview from wiping out a newer one's record.
 */
export function clearActivePreviewSession() {
    try {
        const current = readActivePreviewSessionRaw();
        if (current && current.pid !== process.pid) {
            return;
        }
        fs.rmSync(FILE_PATH, { force: true });
    }
    catch {
        // File already gone or unreadable — nothing to do.
    }
}
function readActivePreviewSessionRaw() {
    try {
        const raw = fs.readFileSync(FILE_PATH, "utf-8");
        const parsed = JSON.parse(raw);
        if (typeof parsed.sessionId === "string" &&
            typeof parsed.serviceUrl === "string" &&
            typeof parsed.pid === "number" &&
            typeof parsed.startedAt === "string") {
            return parsed;
        }
        return null;
    }
    catch {
        return null;
    }
}
/**
 * Return the active preview session, or null if no preview is open. Treats
 * the file as stale when the recorded pid is no longer running.
 */
export function readActivePreviewSession() {
    const raw = readActivePreviewSessionRaw();
    if (!raw) {
        return null;
    }
    if (!processIsAlive(raw.pid)) {
        return null;
    }
    return raw;
}
