// bitfab-plugin-lib/dist/activeStudioSession.js
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { getAgentSessionId } from "./getAgentSessionId.js";
function shortHash(input) {
    return crypto.createHash("sha256").update(input).digest("hex").slice(0, 16);
}
function sessionScope() {
    return shortHash(getAgentSessionId() ?? process.cwd());
}
function filePath() {
    return path.join(os.homedir(), ".config", "bitfab", `active-studio-session.${sessionScope()}.json`);
}
export function writeActiveStudioSession(session) {
    const target = filePath();
    fs.mkdirSync(path.dirname(target), { recursive: true });
    const payload = {
        ...session,
        pid: process.pid,
        startedAt: new Date().toISOString(),
    };
    const tmp = `${target}.${process.pid}.tmp`;
    fs.writeFileSync(tmp, `${JSON.stringify(payload, null, 2)}\n`);
    fs.renameSync(tmp, target);
}
/**
 * Clear the file only if the recorded pid matches this process. Prevents a
 * slow-exiting old Studio from wiping a newer one's record.
 */
export function clearActiveStudioSession() {
    try {
        const current = readActiveStudioSessionRaw();
        if (current && current.pid !== process.pid) {
            return;
        }
        fs.rmSync(filePath(), { force: true });
    }
    catch {
        // File already gone or unreadable — nothing to do.
    }
}
function readActiveStudioSessionRaw() {
    try {
        const raw = fs.readFileSync(filePath(), "utf-8");
        const parsed = JSON.parse(raw);
        if (typeof parsed.sessionId === "string" &&
            typeof parsed.serviceUrl === "string" &&
            typeof parsed.pid === "number" &&
            typeof parsed.startedAt === "string") {
            return parsed;
        }
        return null;
    }
    catch {
        return null;
    }
}
/**
 * Return the active Studio session for this worktree, or null if none is
 * recorded. Does NOT check whether the writer process is still alive:
 * the browser tab can outlive the CLI process that created it (e.g.
 * `openTracePlan.js` exits on confirm/cancel but the Studio tab stays
 * open for the next iteration). The real liveness check happens at the
 * navigation layer (`navigateStudioAndAwaitAck` in `openStudioTo`),
 * which clears the file if the tab is unresponsive.
 */
export function readActiveStudioSession() {
    return readActiveStudioSessionRaw();
}
