// bitfab-plugin-lib/dist/activeStudioSession.js
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
function cwdHash() {
    return crypto
        .createHash("sha256")
        .update(process.cwd())
        .digest("hex")
        .slice(0, 16);
}
function filePath() {
    return path.join(os.homedir(), ".config", "bitfab", `active-studio-session.${cwdHash()}.json`);
}
function processIsAlive(pid) {
    try {
        process.kill(pid, 0);
        return true;
    }
    catch (err) {
        // ESRCH = process not found; EPERM = exists but we can't signal it.
        // Treat anything other than ESRCH as alive so we don't accidentally
        // clear a legitimate session.
        return err.code !== "ESRCH";
    }
}
export function writeActiveStudioSession(session) {
    const target = filePath();
    fs.mkdirSync(path.dirname(target), { recursive: true });
    const payload = {
        ...session,
        pid: process.pid,
        startedAt: new Date().toISOString(),
    };
    const tmp = `${target}.${process.pid}.tmp`;
    fs.writeFileSync(tmp, `${JSON.stringify(payload, null, 2)}\n`);
    fs.renameSync(tmp, target);
}
/**
 * Clear the file only if the recorded pid matches this process. Prevents a
 * slow-exiting old Studio from wiping a newer one's record.
 */
export function clearActiveStudioSession() {
    try {
        const current = readActiveStudioSessionRaw();
        if (current && current.pid !== process.pid) {
            return;
        }
        fs.rmSync(filePath(), { force: true });
    }
    catch {
        // File already gone or unreadable — nothing to do.
    }
}
function readActiveStudioSessionRaw() {
    try {
        const raw = fs.readFileSync(filePath(), "utf-8");
        const parsed = JSON.parse(raw);
        if (typeof parsed.sessionId === "string" &&
            typeof parsed.serviceUrl === "string" &&
            typeof parsed.pid === "number" &&
            typeof parsed.startedAt === "string") {
            return parsed;
        }
        return null;
    }
    catch {
        return null;
    }
}
/**
 * Return the active Studio session for this worktree, or null if none is
 * open. Treats the file as stale when the recorded pid is no longer alive.
 * Liveness of the actual browser tab is a separate, server-side check
 * (`browserConnected` on `GET /api/studio/events`) — callers that need to
 * route into the tab should run that check before pushing events.
 */
export function readActiveStudioSession() {
    const raw = readActiveStudioSessionRaw();
    if (!raw) {
        return null;
    }
    if (!processIsAlive(raw.pid)) {
        return null;
    }
    return raw;
}
