// bitfab-plugin-lib/dist/activeStudioSession.js
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { getAgentSessionId } from "./getAgentSessionId.js";
function shortHash(input) {
    return crypto.createHash("sha256").update(input).digest("hex").slice(0, 16);
}
function sessionScope() {
    return shortHash(getAgentSessionId() ?? process.cwd());
}
function filePath() {
    return path.join(os.homedir(), ".config", "bitfab", `active-studio-session.${sessionScope()}.json`);
}
/** Atomically persist a full record (tmp write + rename). */
function writeRecord(record) {
    const target = filePath();
    fs.mkdirSync(path.dirname(target), { recursive: true });
    const tmp = `${target}.${process.pid}.tmp`;
    fs.writeFileSync(tmp, `${JSON.stringify(record, null, 2)}\n`);
    fs.renameSync(tmp, target);
}
export function writeActiveStudioSession(session) {
    writeRecord({
        ...session,
        pid: process.pid,
        startedAt: new Date().toISOString(),
    });
}
/**
 * Whether a process is still running. `process.kill(pid, 0)` sends no signal but
 * throws `ESRCH` if the pid is gone (`EPERM` means it exists under another user,
 * i.e. still alive).
 */
function isProcessAlive(pid) {
    try {
        process.kill(pid, 0);
        return true;
    }
    catch (err) {
        return err.code === "EPERM";
    }
}
/**
 * Remove the active-session pointer. The window is gone, so the next
 * `openStudioTo` opens a fresh one instead of trying to reach a dead tab.
 *
 * No-op unless the record still names `sessionId` (we never clear a different
 * session's pointer). The event poller passes no `force`, so it additionally
 * defers to a newer owner: it only clears when THIS process wrote the record or
 * the writer is no longer alive, which stops a stale poller from deleting a
 * pointer a newer process just re-bound. A deliberate clear (`closeStudio`, the
 * `clearStudioSession` recovery command) passes `force: true` to skip that pid
 * guard while still matching on `sessionId`.
 */
export function clearActiveStudioSession(sessionId, opts = {}) {
    try {
        const current = readActiveStudioSessionRaw();
        if (!current || current.sessionId !== sessionId) {
            return false;
        }
        if (!opts.force &&
            current.pid !== process.pid &&
            isProcessAlive(current.pid)) {
            return false;
        }
        fs.unlinkSync(filePath());
        return true;
    }
    catch {
        // File already gone or unreadable — nothing to do.
        return false;
    }
}
function readActiveStudioSessionRaw() {
    try {
        const raw = fs.readFileSync(filePath(), "utf-8");
        const parsed = JSON.parse(raw);
        // Migration: records written by the old stamped-marker lifecycle carried
        // `closedAt`/`endedAt` to mean "the window is gone/ended". The present/absent
        // model has no such fields, so a record still carrying one is a stale legacy
        // pointer to a dead window — treat it as absent so the next open is fresh
        // rather than gating on an unreachable tab.
        if (parsed.closedAt != null || parsed.endedAt != null) {
            return null;
        }
        if (typeof parsed.sessionId === "string" &&
            typeof parsed.serviceUrl === "string" &&
            typeof parsed.pid === "number" &&
            typeof parsed.startedAt === "string") {
            return {
                sessionId: parsed.sessionId,
                serviceUrl: parsed.serviceUrl,
                pid: parsed.pid,
                startedAt: parsed.startedAt,
            };
        }
        return null;
    }
    catch {
        return null;
    }
}
/**
 * Return the active Studio session for this scope, or null if none is recorded.
 * Does NOT check whether the writer process is still alive: the browser tab can
 * outlive the CLI process that created it (e.g. `openTracePlan.js` exits on
 * confirm/cancel but the Studio tab stays open for the next iteration). The real
 * liveness check happens at the navigation layer (`navigateStudioAndAwaitAck` in
 * `openStudioTo`); when that fails, the record is cleared (on a witnessed close)
 * or left for the user to clear (on an unwitnessed disappearance).
 */
export function readActiveStudioSession() {
    return readActiveStudioSessionRaw();
}
