// bitfab-plugin-lib/dist/agentSessionChannel.js
import { NotAuthenticatedError } from "./errors.js";
import { isValidStudioRoute } from "./studioRoutes.js";
const DEFAULT_POLL_INTERVAL_MS = 1500;
export async function createAgentSession(opts) {
    if (!opts.apiKey) {
        throw new NotAuthenticatedError();
    }
    const url = `${opts.serviceUrl}/api/studio/sessions`;
    const res = await fetch(url, {
        method: "POST",
        headers: {
            Authorization: `Bearer ${opts.apiKey}`,
            "Content-Type": "application/json",
            ...opts.extraHeaders,
        },
        body: JSON.stringify({ sessionId: opts.sessionId }),
    });
    if (!res.ok) {
        if (res.status === 401) {
            throw new NotAuthenticatedError("Session expired or invalid. Run the login command to re-authenticate.");
        }
        throw new Error(`createAgentSession failed (${res.status}): ${await res.text()}`);
    }
    return (await res.json());
}
/**
 * Push an event into a session's event stream. Use this to send commands
 * (e.g. "agent:navigate") to the browser tab paired with `sessionId`.
 * Event types and data shapes come from the typed `AgentEventDataMap`
 * union.
 */
export async function pushAgentSessionEvent(client, event) {
    const url = `${client.serviceUrl}/api/studio/events`;
    const res = await fetch(url, {
        method: "POST",
        headers: {
            Authorization: `Bearer ${client.apiKey}`,
            "Content-Type": "application/json",
        },
        body: JSON.stringify({
            session: client.sessionId,
            type: event.type,
            data: event.data,
        }),
    });
    if (!res.ok) {
        throw new Error(`pushAgentSessionEvent failed (${res.status}): ${await res.text()}`);
    }
    return (await res.json());
}
/**
 * Continuously poll a session's event stream. Each poll request is short-
 * lived (Vercel-friendly) and asks for events strictly after `since`. The
 * cursor advances after each successful poll.
 *
 * The loop runs until `abortSignal` fires. Network errors are surfaced via
 * `onError` and the loop continues; transient failures should not terminate
 * the channel.
 */
export async function pollAgentSessionEvents(opts) {
    const interval = opts.intervalMs ?? DEFAULT_POLL_INTERVAL_MS;
    let since = opts.startCursor ?? null;
    while (!opts.abortSignal.aborted) {
        try {
            const url = new URL(`${opts.serviceUrl}/api/studio/events`);
            url.searchParams.set("session", opts.sessionId);
            if (since) {
                url.searchParams.set("since", since);
            }
            const res = await fetch(url.toString(), {
                headers: { Authorization: `Bearer ${opts.apiKey}` },
                signal: opts.abortSignal,
            });
            if (res.ok) {
                const body = (await res.json());
                for (const event of body.events) {
                    opts.onEvent(event);
                    since = event.id;
                }
                if (opts.onBrowserStatus &&
                    typeof body.browserConnected === "boolean") {
                    opts.onBrowserStatus(body.browserConnected);
                }
            }
            else if (opts.onError) {
                opts.onError(new Error(`agent-session poll failed (${res.status})`));
            }
        }
        catch (err) {
            if (opts.abortSignal.aborted) {
                return;
            }
            if (opts.onError) {
                opts.onError(err instanceof Error ? err : new Error(String(err)));
            }
        }
        await new Promise((resolve) => {
            const onAbort = () => {
                clearTimeout(timer);
                resolve();
            };
            const timer = setTimeout(() => {
                opts.abortSignal.removeEventListener("abort", onAbort);
                resolve();
            }, interval);
            opts.abortSignal.addEventListener("abort", onAbort, { once: true });
        });
    }
}
/**
 * Poll the unauthenticated sessions/poll endpoint for auth-related events.
 * Resolves with the API token when a `studio:authenticated` event carrying
 * a `token` field is received. Used during login before the plugin has
 * credentials.
 */
export async function pollLoginEvents(opts) {
    const interval = opts.intervalMs ?? DEFAULT_POLL_INTERVAL_MS;
    let since = null;
    while (!opts.abortSignal.aborted) {
        try {
            const url = new URL(`${opts.serviceUrl}/api/studio/sessions/poll`);
            url.searchParams.set("session", opts.sessionId);
            if (since) {
                url.searchParams.set("since", since);
            }
            const res = await fetch(url.toString(), {
                signal: opts.abortSignal,
            });
            if (res.ok) {
                const body = (await res.json());
                for (const event of body.events) {
                    since = event.id;
                    if (event.type === "studio:authenticated" &&
                        typeof event.data.token === "string") {
                        return event.data.token;
                    }
                }
            }
        }
        catch {
            if (opts.abortSignal.aborted) {
                break;
            }
        }
        await new Promise((resolve) => {
            const onAbort = () => {
                clearTimeout(timer);
                resolve();
            };
            const timer = setTimeout(() => {
                opts.abortSignal.removeEventListener("abort", onAbort);
                resolve();
            }, interval);
            opts.abortSignal.addEventListener("abort", onAbort, { once: true });
        });
    }
    throw new Error("Login polling aborted");
}
export async function closeStudio(client, message) {
    return pushAgentSessionEvent(client, {
        type: "agent:close-studio",
        data: message ? { message } : {},
    });
}
export async function navigateStudio(client, path) {
    if (!isValidStudioRoute(path)) {
        throw new Error(`Studio route not allowed: ${path}`);
    }
    return pushAgentSessionEvent(client, {
        type: "agent:navigate",
        data: { path },
    });
}
const STUDIO_NAVIGATE_ACK_TIMEOUT_MS = 12_000;
export async function navigateStudioAndAwaitAck(client, path, opts = {}) {
    const ackTimeoutMs = opts.ackTimeoutMs ?? STUDIO_NAVIGATE_ACK_TIMEOUT_MS;
    const abortController = new AbortController();
    let timer = null;
    const result = new Promise((resolve) => {
        timer = setTimeout(() => {
            resolve({ acked: false, reason: "timeout" });
            abortController.abort();
        }, ackTimeoutMs);
        const pathOnly = path.split("?")[0];
        pollAgentSessionEvents({
            serviceUrl: client.serviceUrl,
            apiKey: client.apiKey,
            sessionId: client.sessionId,
            abortSignal: abortController.signal,
            onEvent: (event) => {
                if (event.type === "studio:navigated" && event.data.path === pathOnly) {
                    if (timer) {
                        clearTimeout(timer);
                    }
                    resolve({ acked: true });
                    abortController.abort();
                    return;
                }
                if (event.type === "studio:navigation-blocked" &&
                    event.data.path === pathOnly) {
                    if (timer) {
                        clearTimeout(timer);
                    }
                    resolve({
                        acked: false,
                        reason: "blocked",
                        blockedReason: event.data.reason,
                    });
                    abortController.abort();
                    return;
                }
                if (event.type === "studio:session-ended" ||
                    event.type === "studio:session-closed") {
                    if (timer) {
                        clearTimeout(timer);
                    }
                    resolve({ acked: false, reason: "session-ended" });
                    abortController.abort();
                }
            },
            onError: () => { },
        }).catch(() => {
            if (abortController.signal.aborted) {
                return;
            }
            if (timer) {
                clearTimeout(timer);
            }
            resolve({ acked: false, reason: "timeout" });
        });
    });
    if (!isValidStudioRoute(path)) {
        if (timer) {
            clearTimeout(timer);
        }
        abortController.abort();
        return { acked: false, reason: "invalid-route" };
    }
    try {
        await navigateStudio(client, path);
    }
    catch {
        if (timer) {
            clearTimeout(timer);
        }
        abortController.abort();
        return { acked: false, reason: "push-failed" };
    }
    return result;
}
export async function isStudioBrowserConnected(client, opts = {}) {
    const timeoutMs = opts.timeoutMs ?? 2_000;
    try {
        const url = new URL(`${client.serviceUrl}/api/studio/events`);
        url.searchParams.set("session", client.sessionId);
        url.searchParams.set("limit", "1");
        const res = await fetch(url.toString(), {
            headers: { Authorization: `Bearer ${client.apiKey}` },
            signal: AbortSignal.timeout(timeoutMs),
        });
        if (!res.ok) {
            return false;
        }
        const body = (await res.json());
        return body.browserConnected === true;
    }
    catch {
        return false;
    }
}
export async function getStudioState(client) {
    const url = new URL(`${client.serviceUrl}/api/studio/state`);
    url.searchParams.set("session", client.sessionId);
    const res = await fetch(url.toString(), {
        headers: { Authorization: `Bearer ${client.apiKey}` },
    });
    if (!res.ok) {
        throw new Error(`getStudioState failed (${res.status}): ${await res.text()}`);
    }
    return (await res.json());
}
