// bitfab-plugin-lib/dist/config.js
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
export const DEFAULT_SERVICE_URL = "https://bitfab.ai";
export const DSM_NOT_STARTED_URL = "dsm:not-started";
const GLOBAL_CONFIG_DIR = path.join(os.homedir(), ".config", "bitfab");
const GLOBAL_CONFIG_FILE = path.join(GLOBAL_CONFIG_DIR, "config.json");
const GLOBAL_CREDENTIALS_FILE = path.join(GLOBAL_CONFIG_DIR, "credentials.json");
const PROJECT_CONFIG_RELATIVE = path.join(".bitfab", "config.local.json");
const PROJECT_CREDENTIALS_RELATIVE = path.join(".bitfab", "credentials.local.json");
function projectSearchStartDirs() {
    const rawDirs = [
        process.cwd(),
        process.env.BITFAB_PROJECT_ROOT,
        process.env.BITFAB_WORKSPACE_ROOT,
    ];
    const dirs = [];
    const seen = new Set();
    for (const rawDir of rawDirs) {
        if (typeof rawDir !== "string" || !path.isAbsolute(rawDir)) {
            continue;
        }
        const dir = path.resolve(rawDir);
        if (seen.has(dir)) {
            continue;
        }
        seen.add(dir);
        dirs.push(dir);
    }
    return dirs;
}
function readJsonFile(filePath) {
    try {
        const content = fs.readFileSync(filePath, "utf-8");
        return JSON.parse(content);
    }
    catch {
        return null;
    }
}
function findProjectFiles() {
    for (const startDir of projectSearchStartDirs()) {
        let dir = startDir;
        let foundConfigFile = null;
        let foundCredentialsFile = null;
        while (true) {
            const configFile = path.join(dir, PROJECT_CONFIG_RELATIVE);
            const credentialsFile = path.join(dir, PROJECT_CREDENTIALS_RELATIVE);
            if (!foundConfigFile && fs.existsSync(configFile)) {
                foundConfigFile = configFile;
            }
            if (!foundCredentialsFile && fs.existsSync(credentialsFile)) {
                foundCredentialsFile = credentialsFile;
            }
            const parent = path.dirname(dir);
            if (parent === dir) {
                break;
            }
            dir = parent;
        }
        if (foundConfigFile || foundCredentialsFile) {
            return {
                configFile: foundConfigFile,
                credentialsFile: foundCredentialsFile,
            };
        }
    }
    return null;
}
function getProjectConfigData() {
    const filePath = findProjectFiles()?.configFile;
    if (!filePath) {
        return null;
    }
    return readJsonFile(filePath);
}
function getProjectCredentialsData() {
    const filePath = findProjectFiles()?.credentialsFile;
    if (!filePath) {
        return null;
    }
    // File exists: if it can't be parsed, treat as empty (preserve isolation)
    return readJsonFile(filePath) ?? {};
}
function getConfigData() {
    return readJsonFile(GLOBAL_CONFIG_FILE) ?? {};
}
function getCredentialsData() {
    return readJsonFile(GLOBAL_CREDENTIALS_FILE) ?? {};
}
function getServiceUrl() {
    if (process.env.BITFAB_SERVICE_URL) {
        return process.env.BITFAB_SERVICE_URL;
    }
    const project = getProjectConfigData();
    if (project && typeof project.serviceUrl === "string") {
        return project.serviceUrl;
    }
    const config = getConfigData();
    return typeof config.serviceUrl === "string"
        ? config.serviceUrl
        : DEFAULT_SERVICE_URL;
}
function getApiKey() {
    if (process.env.BITFAB_API_KEY) {
        return process.env.BITFAB_API_KEY;
    }
    // Project-local credentials override global when present
    const projectCreds = getProjectCredentialsData();
    if (projectCreds !== null) {
        return typeof projectCreds.apiKey === "string" ? projectCreds.apiKey : null;
    }
    const creds = getCredentialsData();
    return typeof creds.apiKey === "string" ? creds.apiKey : null;
}
function getVerbose() {
    if (process.env.BITFAB_VERBOSE === "true" ||
        process.env.BITFAB_VERBOSE === "1") {
        return true;
    }
    if (process.env.BITFAB_VERBOSE === "false" ||
        process.env.BITFAB_VERBOSE === "0") {
        return false;
    }
    const config = getConfigData();
    return config.verbose === true;
}
function getDebug() {
    if (process.env.BITFAB_DEBUG === "true" || process.env.BITFAB_DEBUG === "1") {
        return true;
    }
    const config = getConfigData();
    return config.debug === true;
}
function getSessionLogConsentValue() {
    const config = getConfigData();
    if (config.sessionLogConsent === true) {
        return true;
    }
    if (config.sessionLogConsent === false) {
        return false;
    }
    return null;
}
export function getSessionLogConsent() {
    return getSessionLogConsentValue();
}
export function setSessionLogConsent(value) {
    fs.mkdirSync(GLOBAL_CONFIG_DIR, { recursive: true });
    const existing = readJsonFile(GLOBAL_CONFIG_FILE) ?? {};
    existing.sessionLogConsent = value;
    fs.writeFileSync(GLOBAL_CONFIG_FILE, `${JSON.stringify(existing, null, 2)}\n`);
}
export function getConfig() {
    return {
        serviceUrl: getServiceUrl(),
        apiKey: getApiKey(),
        verbose: getVerbose(),
        debug: getDebug(),
        sessionLogConsent: getSessionLogConsentValue(),
    };
}
export function isProductionServiceUrl(serviceUrl) {
    try {
        const url = new URL(serviceUrl);
        return (url.protocol === "https:" &&
            url.hostname === "bitfab.ai" &&
            (url.pathname === "" || url.pathname === "/"));
    }
    catch {
        return false;
    }
}
export function formatNonProductionEndpointSuffix(serviceUrl) {
    if (isProductionServiceUrl(serviceUrl)) {
        return "";
    }
    try {
        const url = new URL(serviceUrl);
        if (!url.host) {
            return ` @ ${serviceUrl}`;
        }
        return ` @ ${url.host}`;
    }
    catch {
        return ` @ ${serviceUrl}`;
    }
}
export function saveCredentials(apiKey) {
    const projectFile = findProjectFiles()?.credentialsFile;
    if (projectFile) {
        fs.writeFileSync(projectFile, `${JSON.stringify({ apiKey }, null, 2)}\n`);
        return;
    }
    fs.mkdirSync(GLOBAL_CONFIG_DIR, { recursive: true });
    fs.writeFileSync(GLOBAL_CREDENTIALS_FILE, `${JSON.stringify({ apiKey }, null, 2)}\n`);
}
export function deleteCredentials() {
    try {
        fs.unlinkSync(GLOBAL_CREDENTIALS_FILE);
    }
    catch {
        // already gone
    }
}
export function hasCredentials() {
    return getApiKey() !== null;
}
export function requireAuth(loginHint) {
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${loginHint} first.`);
        process.exit(1);
    }
    const config = getConfig();
    if (!config.apiKey) {
        console.error("API key missing in config.");
        process.exit(1);
    }
    return { apiKey: config.apiKey, serviceUrl: config.serviceUrl };
}
export function getOrCreateInstallId() {
    const config = getConfigData();
    if (typeof config.installId === "string" && config.installId.length > 0) {
        return config.installId;
    }
    const installId = crypto.randomUUID();
    try {
        fs.mkdirSync(GLOBAL_CONFIG_DIR, { recursive: true });
        fs.writeFileSync(GLOBAL_CONFIG_FILE, `${JSON.stringify({ ...config, installId }, null, 2)}\n`);
    }
    catch {
        // best-effort: if we can't persist, just return the ephemeral ID
    }
    return installId;
}
export { GLOBAL_CREDENTIALS_FILE as CREDENTIALS_FILE };
