// bitfab-plugin-lib/dist/config.js
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
const DEFAULT_SERVICE_URL = "https://bitfab.ai";
const GLOBAL_CONFIG_DIR = path.join(os.homedir(), ".config", "bitfab");
const GLOBAL_CONFIG_FILE = path.join(GLOBAL_CONFIG_DIR, "config.json");
const GLOBAL_CREDENTIALS_FILE = path.join(GLOBAL_CONFIG_DIR, "credentials.json");
const PROJECT_CONFIG_RELATIVE = path.join(".bitfab", "config.local.json");
const PROJECT_CREDENTIALS_RELATIVE = path.join(".bitfab", "credentials.local.json");
function readJsonFile(filePath) {
    try {
        const content = fs.readFileSync(filePath, "utf-8");
        return JSON.parse(content);
    }
    catch {
        return null;
    }
}
function findProjectFile(relativePath, startDir = process.cwd()) {
    let dir = startDir;
    while (true) {
        const candidate = path.join(dir, relativePath);
        if (fs.existsSync(candidate)) {
            return candidate;
        }
        const parent = path.dirname(dir);
        if (parent === dir) {
            return null;
        }
        dir = parent;
    }
}
function getProjectConfigData() {
    const filePath = findProjectFile(PROJECT_CONFIG_RELATIVE);
    if (!filePath) {
        return null;
    }
    return readJsonFile(filePath);
}
function getProjectCredentialsData() {
    const filePath = findProjectFile(PROJECT_CREDENTIALS_RELATIVE);
    if (!filePath) {
        return null;
    }
    // File exists: if it can't be parsed, treat as empty (preserve isolation)
    return readJsonFile(filePath) ?? {};
}
function getConfigData() {
    return readJsonFile(GLOBAL_CONFIG_FILE) ?? {};
}
function getCredentialsData() {
    return readJsonFile(GLOBAL_CREDENTIALS_FILE) ?? {};
}
function getServiceUrl() {
    if (process.env.BITFAB_SERVICE_URL) {
        return process.env.BITFAB_SERVICE_URL;
    }
    const project = getProjectConfigData();
    if (project && typeof project.serviceUrl === "string") {
        return project.serviceUrl;
    }
    const config = getConfigData();
    return typeof config.serviceUrl === "string"
        ? config.serviceUrl
        : DEFAULT_SERVICE_URL;
}
function getApiKey() {
    if (process.env.BITFAB_API_KEY) {
        return process.env.BITFAB_API_KEY;
    }
    // Project-local credentials override global when present
    const projectCreds = getProjectCredentialsData();
    if (projectCreds !== null) {
        return typeof projectCreds.apiKey === "string" ? projectCreds.apiKey : null;
    }
    const creds = getCredentialsData();
    return typeof creds.apiKey === "string" ? creds.apiKey : null;
}
function getVerbose() {
    if (process.env.BITFAB_VERBOSE === "true" ||
        process.env.BITFAB_VERBOSE === "1") {
        return true;
    }
    if (process.env.BITFAB_VERBOSE === "false" ||
        process.env.BITFAB_VERBOSE === "0") {
        return false;
    }
    const config = getConfigData();
    return config.verbose === true;
}
function getDebug() {
    if (process.env.BITFAB_DEBUG === "true" || process.env.BITFAB_DEBUG === "1") {
        return true;
    }
    const config = getConfigData();
    return config.debug === true;
}
function getSessionLogConsentValue() {
    const config = getConfigData();
    if (config.sessionLogConsent === true) {
        return true;
    }
    if (config.sessionLogConsent === false) {
        return false;
    }
    return null;
}
export function getSessionLogConsent() {
    return getSessionLogConsentValue();
}
export function setSessionLogConsent(value) {
    fs.mkdirSync(GLOBAL_CONFIG_DIR, { recursive: true });
    const existing = readJsonFile(GLOBAL_CONFIG_FILE) ?? {};
    existing.sessionLogConsent = value;
    fs.writeFileSync(GLOBAL_CONFIG_FILE, `${JSON.stringify(existing, null, 2)}\n`);
}
export function getConfig() {
    return {
        serviceUrl: getServiceUrl(),
        apiKey: getApiKey(),
        verbose: getVerbose(),
        debug: getDebug(),
        sessionLogConsent: getSessionLogConsentValue(),
    };
}
export function saveCredentials(apiKey) {
    const projectFile = findProjectFile(PROJECT_CREDENTIALS_RELATIVE);
    if (projectFile) {
        fs.writeFileSync(projectFile, `${JSON.stringify({ apiKey }, null, 2)}\n`);
        return;
    }
    fs.mkdirSync(GLOBAL_CONFIG_DIR, { recursive: true });
    fs.writeFileSync(GLOBAL_CREDENTIALS_FILE, `${JSON.stringify({ apiKey }, null, 2)}\n`);
}
export function deleteCredentials() {
    try {
        fs.unlinkSync(GLOBAL_CREDENTIALS_FILE);
    }
    catch {
        // already gone
    }
}
export function hasCredentials() {
    return getApiKey() !== null;
}
export function requireAuth(loginHint) {
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${loginHint} first.`);
        process.exit(1);
    }
    const config = getConfig();
    if (!config.apiKey) {
        console.error("API key missing in config.");
        process.exit(1);
    }
    return { apiKey: config.apiKey, serviceUrl: config.serviceUrl };
}
export function getOrCreateInstallId() {
    const config = getConfigData();
    if (typeof config.installId === "string" && config.installId.length > 0) {
        return config.installId;
    }
    const installId = crypto.randomUUID();
    try {
        fs.mkdirSync(GLOBAL_CONFIG_DIR, { recursive: true });
        fs.writeFileSync(GLOBAL_CONFIG_FILE, `${JSON.stringify({ ...config, installId }, null, 2)}\n`);
    }
    catch {
        // best-effort: if we can't persist, just return the ephemeral ID
    }
    return installId;
}
export { GLOBAL_CREDENTIALS_FILE as CREDENTIALS_FILE };
