// bitfab-plugin-lib/dist/errors.js
export class NotAuthenticatedError extends Error {
    constructor(message) {
        super(message ??
            "Not authenticated. Run the login command to connect your Bitfab account.");
        this.name = "NotAuthenticatedError";
    }
}
/**
 * The daemon cannot run in this context (e.g. a bundled standalone CLI that
 * ships no daemon entry point). This is an expected, quiet fallback to the
 * direct channel, not an error worth surfacing to the user.
 */
export class DaemonUnavailableError extends Error {
    constructor(message) {
        super(message);
        this.name = "DaemonUnavailableError";
    }
}
/**
 * The daemon accepted the socket connection and answered `ping`, but never
 * responded to an `open` within the timeout: a wedged daemon. The client reaps
 * and respawns the daemon, then retries the open once.
 */
export class DaemonOpenTimeoutError extends Error {
    constructor(message = "daemon did not respond to open") {
        super(message);
        this.name = "DaemonOpenTimeoutError";
    }
}
export class StudioNavigationError extends Error {
    reason;
    blockedReason;
    staleSessionId;
    constructor(reason, blockedReason, staleSessionId) {
        super(blockedReason
            ? `Studio navigation failed (${reason}): ${blockedReason}`
            : `Studio navigation failed: ${reason}`);
        this.reason = reason;
        this.blockedReason = blockedReason;
        this.staleSessionId = staleSessionId;
        this.name = "StudioNavigationError";
    }
}
