// bitfab-plugin-lib/dist/frontmostApp.js
import { execFileSync, spawn } from "node:child_process";
import os from "node:os";
// Cross-platform helpers for handing focus back to the terminal/editor after a
// browser handoff finishes. Uses execFileSync/spawn (no shell) so nothing
// writes to the parent process's stdout, which matters for commands whose
// stdout is a parseable JSONL stream.
//
// macOS: AppleScript (app name)
// Linux: xdotool (X11 window ID; no-ops on Wayland or if xdotool is absent)
// Windows: user32.dll P/Invoke via PowerShell (HWND as int64)
function getFrontmostApp() {
    const platform = os.platform();
    try {
        if (platform === "darwin") {
            return execFileSync("osascript", [
                "-e",
                'tell application "System Events" to get name of first process whose frontmost is true',
            ], { stdio: "pipe", encoding: "utf-8" }).trim();
        }
        if (platform === "linux") {
            return execFileSync("xdotool", ["getactivewindow"], {
                stdio: "pipe",
                encoding: "utf-8",
            }).trim();
        }
        if (platform === "win32") {
            return execFileSync("powershell.exe", [
                "-NoProfile",
                "-Command",
                `Add-Type -MemberDefinition '[DllImport("user32.dll")] public static extern IntPtr GetForegroundWindow();' -Name WinFocus -Namespace Bitfab; [Bitfab.WinFocus]::GetForegroundWindow().ToInt64()`,
            ], { stdio: "pipe", encoding: "utf-8", timeout: 5000 }).trim();
        }
    }
    catch {
        return null;
    }
    return null;
}
// Silently swallow ENOENT when a binary (osascript, xdotool, powershell) is
// not found. Without this, spawn emits an unhandled 'error' event.
function spawnQuiet(cmd, args) {
    spawn(cmd, [...args], { stdio: "ignore" }).on("error", () => { });
}
function focusApp(identifier) {
    if (!identifier) {
        return;
    }
    const platform = os.platform();
    if (platform === "darwin") {
        spawnQuiet("osascript", [
            "-e",
            `tell application "${identifier}" to activate`,
        ]);
        return;
    }
    if (platform === "linux") {
        spawnQuiet("xdotool", ["windowactivate", identifier]);
        return;
    }
    if (platform === "win32") {
        if (!/^\d+$/.test(identifier)) {
            return;
        }
        // Simulate an Alt keypress before SetForegroundWindow. Windows only grants
        // foreground rights to processes that received the last input event; without
        // this, SetForegroundWindow just flashes the taskbar icon.
        const memberDef = [
            '[DllImport("user32.dll")] [return: MarshalAs(UnmanagedType.Bool)] public static extern bool SetForegroundWindow(IntPtr h);',
            '[DllImport("user32.dll")] public static extern void keybd_event(byte bVk, byte bScan, uint dwFlags, UIntPtr dwExtraInfo);',
        ].join(" ");
        const script = [
            `Add-Type -MemberDefinition '${memberDef}' -Name WinFocus -Namespace Bitfab;`,
            `[Bitfab.WinFocus]::keybd_event(0x12, 0, 1, [UIntPtr]::Zero);`,
            `[Bitfab.WinFocus]::SetForegroundWindow([IntPtr]::new(${identifier}));`,
            `[Bitfab.WinFocus]::keybd_event(0x12, 0, 3, [UIntPtr]::Zero)`,
        ].join(" ");
        spawnQuiet("powershell.exe", ["-NoProfile", "-Command", script]);
    }
}
export function recordFocus() {
    const app = getFrontmostApp();
    return () => focusApp(app);
}
