// bitfab-plugin-lib/dist/frontmostApp.js
import { execFileSync, spawn } from "node:child_process";
import os from "node:os";
function findAncestorApp() {
    try {
        let pid = process.ppid;
        while (pid > 1) {
            const output = execFileSync("ps", ["-o", "ppid=,comm=", "-p", String(pid)], { stdio: "pipe", encoding: "utf-8" }).trim();
            const match = output.match(/^\s*(\d+)\s+(.+)$/);
            if (!match) {
                break;
            }
            const parentPid = parseInt(match[1], 10);
            const comm = match[2].trim();
            const appMatch = comm.match(/\/([^/]+)\.app\//);
            if (appMatch) {
                const bundleName = appMatch[1];
                const executableName = comm.split("/").pop() ?? bundleName;
                return { bundleName, executableName };
            }
            pid = parentPid;
        }
    }
    catch {
        // Process tree walking failed
    }
    return null;
}
function focusAncestorApp(app, itermSessionId) {
    if (app.bundleName === "iTerm" || app.executableName === "iTerm2") {
        focusItermSession(itermSessionId);
        return;
    }
    if (/["\\]/.test(app.bundleName)) {
        return;
    }
    spawnQuiet("osascript", [
        "-e",
        `tell application "${app.bundleName}" to activate`,
    ]);
}
function focusItermSession(sessionId) {
    if (!sessionId || /["\\]/.test(sessionId)) {
        spawnQuiet("osascript", ["-e", 'tell application "iTerm2" to activate']);
        return;
    }
    // ITERM_SESSION_ID format: "w0t0p0:UUID". Walk all sessions to find and
    // raise the matching window+tab. activate runs first so iTerm comes to
    // front even if the session loop finds nothing.
    const script = [
        'tell application "iTerm2"',
        "  activate",
        "  repeat with w in windows",
        "    repeat with t in tabs of w",
        "      repeat with s in sessions of t",
        `        if id of s is "${sessionId}" then`,
        "          select t",
        "          set frontmost of w to true",
        "          return",
        "        end if",
        "      end repeat",
        "    end repeat",
        "  end repeat",
        "end tell",
    ].join("\n");
    spawnQuiet("osascript", ["-e", script]);
}
function getFrontmostApp() {
    const platform = os.platform();
    try {
        if (platform === "darwin") {
            return execFileSync("osascript", [
                "-e",
                'tell application "System Events" to get name of first process whose frontmost is true',
            ], { stdio: "pipe", encoding: "utf-8" }).trim();
        }
        if (platform === "linux") {
            return execFileSync("xdotool", ["getactivewindow"], {
                stdio: "pipe",
                encoding: "utf-8",
            }).trim();
        }
        if (platform === "win32") {
            return execFileSync("powershell.exe", [
                "-NoProfile",
                "-Command",
                `Add-Type -MemberDefinition '[DllImport("user32.dll")] public static extern IntPtr GetForegroundWindow();' -Name WinFocus -Namespace Bitfab; [Bitfab.WinFocus]::GetForegroundWindow().ToInt64()`,
            ], { stdio: "pipe", encoding: "utf-8", timeout: 5000 }).trim();
        }
    }
    catch {
        return null;
    }
    return null;
}
// Silently swallow ENOENT when a binary (osascript, xdotool, powershell) is
// not found. Without this, spawn emits an unhandled 'error' event.
function spawnQuiet(cmd, args) {
    spawn(cmd, [...args], { stdio: "ignore" }).on("error", () => { });
}
function focusApp(identifier) {
    if (!identifier) {
        return;
    }
    const platform = os.platform();
    if (platform === "darwin") {
        spawnQuiet("osascript", [
            "-e",
            `tell application "${identifier}" to activate`,
        ]);
        return;
    }
    if (platform === "linux") {
        spawnQuiet("xdotool", ["windowactivate", identifier]);
        return;
    }
    if (platform === "win32") {
        if (!/^\d+$/.test(identifier)) {
            return;
        }
        // Simulate an Alt keypress before SetForegroundWindow. Windows only grants
        // foreground rights to processes that received the last input event; without
        // this, SetForegroundWindow just flashes the taskbar icon.
        const memberDef = [
            '[DllImport("user32.dll")] [return: MarshalAs(UnmanagedType.Bool)] public static extern bool SetForegroundWindow(IntPtr h);',
            '[DllImport("user32.dll")] public static extern void keybd_event(byte bVk, byte bScan, uint dwFlags, UIntPtr dwExtraInfo);',
        ].join(" ");
        const script = [
            `Add-Type -MemberDefinition '${memberDef}' -Name WinFocus -Namespace Bitfab;`,
            `[Bitfab.WinFocus]::keybd_event(0x12, 0, 1, [UIntPtr]::Zero);`,
            `[Bitfab.WinFocus]::SetForegroundWindow([IntPtr]::new(${identifier}));`,
            `[Bitfab.WinFocus]::keybd_event(0x12, 0, 3, [UIntPtr]::Zero)`,
        ].join(" ");
        spawnQuiet("powershell.exe", ["-NoProfile", "-Command", script]);
    }
}
export function recordFocus() {
    const platform = os.platform();
    if (platform === "linux") {
        const windowId = process.env.WINDOWID;
        if (windowId) {
            return () => spawnQuiet("xdotool", ["windowactivate", windowId]);
        }
    }
    if (platform === "darwin") {
        const app = findAncestorApp();
        if (app) {
            const itermSessionId = process.env.ITERM_SESSION_ID ?? null;
            return () => focusAncestorApp(app, itermSessionId);
        }
    }
    const frontmost = getFrontmostApp();
    return () => focusApp(frontmost);
}
