// bitfab-plugin-lib/dist/index.js
export { closeStudio, getStudioState, navigateStudio, pushAgentSessionEvent, sendPing, sendPong, } from "./agentSessionChannel.js";
export { formatDevBuildBannerLines, formatDevBuildStatusLine, formatRelativeTime, getDevBuildSummary, hashSrcDir, isDevBuildActive, readBuildInfo, } from "./buildInfo.js";
export { captureFromHook } from "./chatSessions/index.js";
export { buildBitfabClientHeaders, buildBitfabRequestHeaders, } from "./clientHeaders.js";
export { runClearStudioSession } from "./commands/clearStudioSession.js";
export { runCloseStudio } from "./commands/closeStudio.js";
export { createStudioSession } from "./commands/createStudioSession.js";
export { buildReplayCapabilityRecord, runDetectCapabilities, } from "./commands/detectCapabilities.js";
export { runLogin } from "./commands/login.js";
export { runLogout } from "./commands/logout.js";
export { runOpenDatasetExperiments } from "./commands/openDatasetExperiments.js";
export { runOpenExperiments } from "./commands/openExperiments.js";
export { handleStudioError, openStudioTo, runOpenStudioTo, StudioNavigationError, } from "./commands/openStudioTo.js";
export { runOpenTracePlan } from "./commands/openTracePlan.js";
export { runPersistReplayLabels } from "./commands/persistReplayLabels.js";
export { runPushActivity } from "./commands/pushActivity.js";
export { runReadReplayStatus } from "./commands/readReplayStatus.js";
export { runReadTracesBatched } from "./commands/readTracesBatched.js";
export { runReapDaemon } from "./commands/reapDaemon.js";
export { runReplayProgress } from "./commands/replayProgress.js";
export { parseSessionLogConsentArgs, runSessionLogConsent, } from "./commands/sessionLogConsent.js";
export { runStartDataset } from "./commands/startDataset.js";
export { runStartTemplatePreview } from "./commands/startTemplatePreview.js";
export { runStatus } from "./commands/status.js";
export { runSwitchOrg } from "./commands/switchOrg.js";
export { runTrackSkill } from "./commands/trackSkill.js";
export { parseUpdateMode, runUpdate } from "./commands/update.js";
export { runWaitForTrace } from "./commands/waitForTrace.js";
export { getConfig, getSessionLogConsent, hasCredentials, requireAuth, setSessionLogConsent, } from "./config.js";
export { NotAuthenticatedError } from "./errors.js";
export { readHookStdin, runCaptureHook } from "./hooks/captureHook.js";
export { runSessionStart } from "./hooks/sessionStart.js";
export { extractSkillFromPrompt, runUserPromptSubmit, } from "./hooks/userPromptSubmit.js";
export { getCurrentVersion, getInstalledSdks } from "./installedSdk.js";
export { detectClaudeInstallScopes } from "./installScope.js";
export { createMcpClient } from "./mcpClient.js";
export { startMcpServer } from "./mcpServer.js";
export { emitJson } from "./output.js";
export { detectReplayCapabilities, isReplayLanguage, isVersionAtLeast, REPLAY_CAPABILITY_MIN_VERSIONS, } from "./replayCapabilities.js";
export { formatHeartbeat, formatNoProgressNotice, formatReplayHeader, formatReplaySummary, formatTraceLine, } from "./replayProgressView.js";
export { checkForSdkUpdates, checkForSdkUpdatesOffline, formatSdkUpdateLine, getBakedSdkVersion, getLatestSdkVersion, isSdkVersionNewer, } from "./sdkUpdates.js";
export { BITFAB_SKILLS } from "./skills.js";
export { isValidStudioRoute, STUDIO_ROUTE_PATTERNS, } from "./studioRoutes.js";
export { checkForUpdate, detectLegacyInstall, formatStaleMessage, formatUpdateMessage, legacyMigrationMessage, } from "./updates.js";
export { detectWorkspaces } from "./workspaces.js";
