// bitfab-plugin-lib/dist/installedSdk.js
import fs from "node:fs";
import path from "node:path";
import * as semver from "semver";
import { detectWorkspaces } from "./workspaces.js";
const TS_PACKAGE = "@bitfab/sdk";
const TS_PACKAGE_LEGACY = "bitfab";
const PY_PACKAGE = "bitfab-py";
const RB_PACKAGE = "bitfab";
const GO_MODULE = "github.com/Project-White-Rabbit/bitfab-go";
function readFileSafe(filePath) {
    try {
        return fs.readFileSync(filePath, "utf-8");
    }
    catch {
        return null;
    }
}
function stripRange(raw) {
    // Coerce any declared spec or recorded version to a clean semver string via
    // node-semver: "^0.20.0" → "0.20.0", "~> 0.10" → "0.10.0", ">=0.18" →
    // "0.18.0", "0.13.5" → "0.13.5", preserving pre-release ("^0.11.0-beta.1" →
    // "0.11.0-beta.1"). Replaces a hand-rolled regex that had corrupted
    // pre-release tags ("-beta.1" → "0.110.1").
    return semver.coerce(raw, { includePrerelease: true })?.version ?? "";
}
function readTsDeclared(cwd) {
    const manifestPath = path.join(cwd, "package.json");
    const content = readFileSafe(manifestPath);
    if (!content) {
        return null;
    }
    try {
        const pkg = JSON.parse(content);
        const raw = pkg.dependencies?.[TS_PACKAGE] ??
            pkg.devDependencies?.[TS_PACKAGE] ??
            pkg.peerDependencies?.[TS_PACKAGE];
        const legacyRaw = pkg.dependencies?.[TS_PACKAGE_LEGACY] ??
            pkg.devDependencies?.[TS_PACKAGE_LEGACY] ??
            pkg.peerDependencies?.[TS_PACKAGE_LEGACY];
        if (raw) {
            return {
                version: stripRange(raw) || null,
                path: manifestPath,
                legacy: false,
                hasLegacy: legacyRaw !== undefined,
            };
        }
        if (legacyRaw) {
            return {
                version: stripRange(legacyRaw) || null,
                path: manifestPath,
                legacy: true,
                hasLegacy: true,
            };
        }
        return null;
    }
    catch {
        return null;
    }
}
function readTsFromPnpm(cwd) {
    const p = path.join(cwd, "pnpm-lock.yaml");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    const match = content.match(/['"/\s](?:@bitfab\/sdk|bitfab)@(\d+\.\d+\.\d+(?:[-+][\w.]+)?)/);
    return match ? { version: match[1], path: p } : null;
}
function readTsFromNpm(cwd) {
    const p = path.join(cwd, "package-lock.json");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    try {
        const pkg = JSON.parse(content);
        for (const name of [TS_PACKAGE, TS_PACKAGE_LEGACY]) {
            const v = pkg.packages?.[`node_modules/${name}`]?.version ??
                pkg.dependencies?.[name]?.version;
            if (v) {
                return { version: v, path: p };
            }
        }
        return null;
    }
    catch {
        return null;
    }
}
function readTsFromYarn(cwd) {
    const p = path.join(cwd, "yarn.lock");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    // Classic (v1): `version "x"` under a quoted descriptor key.
    const classic = content.match(/^(?:"?(?:@bitfab\/sdk|bitfab)@[^",:\n]+"?(?:,\s*"?(?:@bitfab\/sdk|bitfab)@[^",:\n]+"?)*):\s*\n(?:[^\n]*\n)*?\s+version\s+"([^"]+)"/m);
    if (classic) {
        return { version: classic[1], path: p };
    }
    // Berry (v2+) is YAML and records the exact resolved version on a
    // `resolution:` line via the `@npm:` protocol, e.g.
    //   resolution: "@bitfab/sdk@npm:0.15.0"
    const berry = content.match(/^\s*resolution:\s*"(?:@bitfab\/sdk|bitfab)@npm:([0-9][\w.\-+]*)"/m);
    return berry ? { version: berry[1], path: p } : null;
}
function readTsFromBun(cwd) {
    const p = path.join(cwd, "bun.lock");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    const match = content.match(/"(?:@bitfab\/sdk|bitfab)@(\d+\.\d+\.\d+(?:[-+][\w.]+)?)"/);
    return match ? { version: match[1], path: p } : null;
}
function readTsFromNodeModules(cwd) {
    for (const name of [TS_PACKAGE, TS_PACKAGE_LEGACY]) {
        const p = path.join(cwd, "node_modules", name, "package.json");
        const content = readFileSafe(p);
        if (!content) {
            continue;
        }
        try {
            const pkg = JSON.parse(content);
            if (pkg.version) {
                return { version: pkg.version, path: p };
            }
        }
        catch { }
    }
    return null;
}
function readTsResolved(dir, monorepoRoot) {
    const local = readTsFromPnpm(dir) ??
        readTsFromNpm(dir) ??
        readTsFromYarn(dir) ??
        readTsFromBun(dir) ??
        readTsFromNodeModules(dir);
    if (local || dir === monorepoRoot) {
        return local;
    }
    return (readTsFromPnpm(monorepoRoot) ??
        readTsFromNpm(monorepoRoot) ??
        readTsFromYarn(monorepoRoot) ??
        readTsFromBun(monorepoRoot) ??
        readTsFromNodeModules(monorepoRoot));
}
function readPyDeclared(cwd) {
    const pyproject = path.join(cwd, "pyproject.toml");
    const pypContent = readFileSafe(pyproject);
    if (pypContent?.includes(PY_PACKAGE)) {
        const m = pypContent.match(new RegExp(`${PY_PACKAGE}\\s*[=~><^]*\\s*["']?([0-9][^"'\\s,\\]]*)?`));
        return {
            version: m?.[1] ? stripRange(m[1]) || null : null,
            path: pyproject,
        };
    }
    const reqPath = path.join(cwd, "requirements.txt");
    const reqContent = readFileSafe(reqPath);
    if (reqContent) {
        for (const rawLine of reqContent.split("\n")) {
            const line = rawLine.trim();
            if (!line || line.startsWith("#")) {
                continue;
            }
            const m = line.match(new RegExp(`^${PY_PACKAGE}\\s*([=~><!]+)\\s*([0-9][^\\s;]*)`));
            if (m) {
                return { version: stripRange(m[2]) || null, path: reqPath };
            }
            if (line === PY_PACKAGE) {
                return { version: null, path: reqPath };
            }
        }
    }
    return null;
}
function readPyTomlLockVersion(lockPath) {
    const content = readFileSafe(lockPath);
    if (!content) {
        return null;
    }
    const blocks = content.split(/^\[\[package\]\]\s*$/m);
    for (const block of blocks) {
        if (new RegExp(`^\\s*name\\s*=\\s*["']${PY_PACKAGE}["']\\s*$`, "m").test(block)) {
            const m = block.match(/^\s*version\s*=\s*["']([^"']+)["']/m);
            if (m) {
                return { version: m[1], path: lockPath };
            }
        }
    }
    return null;
}
function readPyFromPipfileLock(cwd) {
    const p = path.join(cwd, "Pipfile.lock");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    try {
        const data = JSON.parse(content);
        const raw = data.default?.[PY_PACKAGE]?.version ?? data.develop?.[PY_PACKAGE]?.version;
        return raw ? { version: stripRange(raw) || null, path: p } : null;
    }
    catch {
        return null;
    }
}
// In-project virtualenv directories worth probing for an installed copy.
const PY_VENV_DIRS = [".venv", "venv", "env"];
// Installed dist-info dir, e.g. `bitfab_py-0.20.0.dist-info` (PyPI normalizes
// `bitfab-py` to `bitfab_py`); the version lives in the directory name.
const PY_DIST_INFO_RE = /^bitfab[_-]py-(\d[\w.\-+]*)\.dist-info$/;
/**
 * Last-resort Python resolution: read the actually-installed version from an
 * in-project virtualenv's `.dist-info` directory. Covers the genuine gap where
 * a project declares the dep loosely and ships no lockfile (common with uv,
 * which puts `.venv` in the project). Best-effort: only finds in-project venvs,
 * not centralized poetry/conda envs (those still fall through to unresolved).
 */
function readPyFromVenv(dir) {
    for (const venv of PY_VENV_DIRS) {
        const sitePackages = [];
        const libDir = path.join(dir, venv, "lib");
        try {
            for (const entry of fs.readdirSync(libDir)) {
                if (entry.startsWith("python")) {
                    sitePackages.push(path.join(libDir, entry, "site-packages"));
                }
            }
        }
        catch { }
        // Windows layout.
        sitePackages.push(path.join(dir, venv, "Lib", "site-packages"));
        for (const site of sitePackages) {
            try {
                for (const entry of fs.readdirSync(site)) {
                    const m = entry.match(PY_DIST_INFO_RE);
                    if (m) {
                        return { version: m[1], path: path.join(site, entry) };
                    }
                }
            }
            catch { }
        }
    }
    return null;
}
function readPyResolvedAt(d) {
    return (readPyTomlLockVersion(path.join(d, "uv.lock")) ??
        readPyTomlLockVersion(path.join(d, "poetry.lock")) ??
        readPyTomlLockVersion(path.join(d, "pdm.lock")) ??
        readPyFromPipfileLock(d) ??
        readPyFromVenv(d));
}
function readPyResolved(dir, monorepoRoot) {
    const local = readPyResolvedAt(dir);
    if (local || dir === monorepoRoot) {
        return local;
    }
    return readPyResolvedAt(monorepoRoot);
}
function readRbDeclared(cwd) {
    const gemfilePath = path.join(cwd, "Gemfile");
    const content = readFileSafe(gemfilePath);
    if (!content) {
        return null;
    }
    const m = content.match(new RegExp(`gem\\s+["']${RB_PACKAGE}["'](?:\\s*,\\s*["']([^"']+)["'])?`));
    if (!m) {
        return null;
    }
    return {
        version: m[1] ? stripRange(m[1]) || null : null,
        path: gemfilePath,
    };
}
function readRbResolved(cwd) {
    const p = path.join(cwd, "Gemfile.lock");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    const m = content.match(new RegExp(`^\\s{4}${RB_PACKAGE}\\s+\\(([^)]+)\\)`, "m"));
    return m ? { version: stripRange(m[1]) || null, path: p } : null;
}
function readGoResolved(cwd) {
    const p = path.join(cwd, "go.mod");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    const m = content.match(new RegExp(`${GO_MODULE.replace(/\./g, "\\.")}\\s+v(\\d[\\w.\\-+]*)`));
    return m ? { version: m[1], path: p } : null;
}
function detectAtDir(dir, monorepoRoot, workspacePath) {
    const result = [];
    const tsDeclared = readTsDeclared(dir);
    const tsResolved = tsDeclared ? readTsResolved(dir, monorepoRoot) : null;
    if (tsDeclared) {
        result.push({
            language: "typescript",
            packageName: TS_PACKAGE,
            declaredVersion: tsDeclared.version,
            resolvedVersion: tsResolved?.version ?? null,
            manifestPath: tsDeclared.path,
            lockfilePath: tsResolved?.path ?? null,
            workspacePath,
            renameFrom: tsDeclared.legacy ? TS_PACKAGE_LEGACY : null,
            deprecated: tsDeclared.hasLegacy || undefined,
        });
    }
    const pyDeclared = readPyDeclared(dir);
    const pyResolved = pyDeclared ? readPyResolved(dir, monorepoRoot) : null;
    if (pyDeclared) {
        result.push({
            language: "python",
            packageName: PY_PACKAGE,
            declaredVersion: pyDeclared.version,
            resolvedVersion: pyResolved?.version ?? null,
            manifestPath: pyDeclared.path,
            lockfilePath: pyResolved?.path ?? null,
            workspacePath,
            renameFrom: null,
        });
    }
    const rbDeclared = readRbDeclared(dir);
    const rbResolved = rbDeclared ? readRbResolved(dir) : null;
    if (rbDeclared) {
        result.push({
            language: "ruby",
            packageName: RB_PACKAGE,
            declaredVersion: rbDeclared.version,
            resolvedVersion: rbResolved?.version ?? null,
            manifestPath: rbDeclared.path,
            lockfilePath: rbResolved?.path ?? null,
            workspacePath,
            renameFrom: null,
        });
    }
    const goResolved = readGoResolved(dir);
    if (goResolved) {
        result.push({
            language: "go",
            packageName: GO_MODULE,
            declaredVersion: null,
            resolvedVersion: goResolved.version,
            manifestPath: null,
            lockfilePath: goResolved.path,
            workspacePath,
            renameFrom: null,
        });
    }
    return result;
}
export function getInstalledSdks(rootDir) {
    const workspaces = detectWorkspaces(rootDir);
    const dirs = [
        { dir: rootDir, workspacePath: "." },
        ...workspaces.map((d) => ({
            dir: d,
            workspacePath: path.relative(rootDir, d) || ".",
        })),
    ];
    const seen = new Set();
    const results = [];
    for (const { dir, workspacePath } of dirs) {
        for (const sdk of detectAtDir(dir, rootDir, workspacePath)) {
            const key = `${workspacePath}\0${sdk.language}`;
            if (seen.has(key)) {
                continue;
            }
            seen.add(key);
            results.push(sdk);
        }
    }
    return results;
}
export function getCurrentVersion(sdk) {
    return sdk.resolvedVersion ?? sdk.declaredVersion;
}
