// bitfab-plugin-lib/dist/mcpServer.js
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";
import { readActivePreviewSession } from "./activePreviewSession.js";
import { pushAgentSessionEvent } from "./agentSessionChannel.js";
import { writeAgentRunId } from "./chatSessions/arming.js";
import { McpProxy } from "./mcpProxy.js";
import { resolveCodingAgentDisplayName } from "./resolveCodingAgentDisplayName.js";
const DATE_PATTERN = /^\d{4}-\d{2}-\d{2}$/;
// read_trace_labels returns only label + annotation + approved per trace (no
// span content), so its payload is bounded and one call can cover a whole
// dataset; read_traces stays capped at 10 because full span content is large.
// Kept in sync with bitfab-web/src/app/mcp/route.ts.
const READ_TRACE_LABELS_MAX_IDS = 100;
// Several MCP clients (Claude Code, Claude Desktop, some SSE clients)
// stringify non-string tool arguments before sending them in `tools/call` -
// numbers, booleans, and arrays all arrive as raw strings even though the
// advertised JSON Schema declares the right types. This is a known upstream
// bug tracked in:
//   - anthropics/claude-code#3966
//   - anthropics/claude-code#5504
//   - anthropics/claude-code#11359
//   - anthropics/claude-code#26094
// Most production MCP servers (Notion, Azure DevOps, FastMCP, ...) defend
// themselves with z.coerce / z.preprocess - we do the same. The advertised
// schema stays strict so the LLM still sees the right types. Use it as
// `z.preprocess(parseJsonString, z.number()...)`. Malformed JSON is left
// as-is so zod can emit a clear "Invalid arguments" error.
//
// This helper is duplicated in bitfab-web/src/app/mcp/route.ts so each MCP
// server in the chain defends itself independently - the plugin handles
// args from stdio clients, the server handles args from any direct HTTP
// callers. Keep them in sync.
function parseJsonString(val) {
    if (typeof val !== "string") {
        return val;
    }
    try {
        return JSON.parse(val);
    }
    catch {
        return val;
    }
}
// Structured shape for create_trace_plan's `tree`, mirroring bitfab-web's
// tracePlanTreeSchema (src/app/lib/domains/tracePlan.ts). It exists so the
// tool's advertised JSON Schema shows the agent the per-node `analysis` field
// (classification + kind) as a real property instead of burying it in prose:
// with `z.unknown()` the agent never saw `analysis` and shipped plans with every
// node unclassified (LLM calls included). Lenient on purpose (optional
// `analysis`, `.passthrough()` for fields like `fields` we don't restate): the
// web server does authoritative validation and rejects a fully-unclassified
// plan, so this schema only has to make `analysis` visible without
// over-rejecting otherwise-valid trees. Keep in sync with bitfab-web.
// Pre-redesign classification names, mapped onto the current taxonomy so a plan
// carrying legacy labels (e.g. an old stored plan replayed through the tool) is
// accepted here too, matching bitfab-web's own leniency.
const LEGACY_CLASSIFICATION = {
    mock_boundary: "model_call",
    live_side_effect: "side_effect",
};
const traceNodeAnalysisShape = z
    .object({
    classification: z.preprocess((value) => typeof value === "string"
        ? (LEGACY_CLASSIFICATION[value] ?? value)
        : value, z.enum(["pure", "model_call", "external_read", "side_effect"])),
    mockable: z.boolean().optional(),
    unmockableReason: z.string().min(1).optional(),
    inputSerializable: z.boolean().optional(),
    outputSerializable: z.boolean().optional(),
    innerCall: z.object({ name: z.string() }).optional(),
    sideEffectKind: z
        .enum([
        "db_write",
        "http_outbound",
        "queue",
        "email",
        "filesystem",
        "vector_write",
    ])
        .optional(),
    readKind: z
        .enum([
        "db_read",
        "http_read",
        "vector_search",
        "cache_read",
        "filesystem_read",
    ])
        .optional(),
})
    .passthrough();
const traceNodeShape = z
    .object({
    id: z.string(),
    name: z.string(),
    kind: z.enum(["manual", "auto", "pure"]),
    file: z.string(),
    line: z.number(),
    signature: z.string().nullable(),
    parentId: z.string().nullable(),
    childIds: z.array(z.string()),
    framework: z.string().optional(),
    sampleInput: z.unknown().optional(),
    sampleOutput: z.unknown().optional(),
    analysis: traceNodeAnalysisShape.optional(),
    // Observed current state of the code, NOT a recommendation. Surfaced as an
    // explicit field (rather than left to .passthrough()) for the same reason as
    // `analysis`: so the tool's JSON Schema shows the analyzer these exist and it
    // actually populates them. `alreadyTraced`: an existing span already wraps
    // this node (a hand-written withSpan/@span/bitfab_span, or a framework
    // auto-capture). `alreadyMocked`: that existing span is already tagged to
    // serve its recorded output on replay. Only analyze-repo sets these, when it
    // plans an already-instrumented workflow, so the delta between what's captured
    // now and what it recommends is recorded on the plan for later analysis. This
    // is data tracking only - nothing in the UI renders differently. Omit on
    // greenfield.
    alreadyTraced: z.boolean().optional(),
    alreadyMocked: z.boolean().optional(),
})
    .passthrough();
const tracePlanTreeShape = z.object({
    rootId: z.string(),
    nodes: z.record(z.string(), traceNodeShape),
});
export function createMcpServer(platform, getConfig, pluginVersion) {
    const proxy = new McpProxy(platform, pluginVersion);
    function proxyToolCall(toolName, args) {
        const agentRunId = process.env.BITFAB_AGENT_RUN_ID;
        const hasAgentRunId = typeof agentRunId === "string" && agentRunId.trim() !== "";
        if (toolName === "create_trace_plan" && hasAgentRunId) {
            const agentSessionId = process.env.CLAUDE_CODE_SESSION_ID;
            if (agentSessionId) {
                writeAgentRunId(agentSessionId, agentRunId);
            }
        }
        const nextArgs = toolName === "create_trace_plan" && hasAgentRunId
            ? { ...args, source: args.source ?? "analyze_repo", agentRunId }
            : args;
        return proxy.toolCall(getConfig(), toolName, nextArgs);
    }
    // Push an event to the active preview session, if one is open. Returns
    // a promise the caller can await to guarantee ordering (e.g. so an
    // editStarted lands on the bus before the work that follows kicks off).
    // Failures are swallowed: the agent event channel is non-critical and
    // the user already has the SSE `template:updated` re-render path as the
    // real success signal.
    async function pushToActivePreview(event) {
        const active = readActivePreviewSession();
        if (!active) {
            return;
        }
        const config = getConfig();
        if (!config.apiKey) {
            return;
        }
        try {
            await pushAgentSessionEvent({
                serviceUrl: active.serviceUrl,
                apiKey: config.apiKey,
                sessionId: active.sessionId,
            }, event);
        }
        catch {
            // Non-critical: agent-event channel may not be reachable; the
            // `template:updated` SSE path is the real success signal.
        }
    }
    const server = new McpServer({
        name: "Bitfab",
        version: pluginVersion,
    });
    server.registerTool("get_bitfab_api_key", {
        description: "Get your Bitfab API key for SDK initialization. Returns the plaintext key to use in environment variables. The SDK setup reference (install commands, initialization, instrumentation patterns) lives at https://docs.bitfab.ai; fetch pages as plain markdown by appending .md (TypeScript: /typescript-sdk.md, Python: /python-sdk.md, Ruby: /ruby-sdk.md, Go: /go-sdk.md), or fetch https://docs.bitfab.ai/llms.txt for the full page index. Read the relevant page when setting up.",
        inputSchema: {},
    }, async () => proxyToolCall("get_bitfab_api_key", {}));
    server.registerTool("get_api_key_context", {
        description: "Returns which Bitfab org the plugin MCP reads from and writes to (it can differ from the project's BITFAB_API_KEY and from the org open in the user's Studio tab). Call before the first plugin MCP write of a session, and whenever the user says data you just wrote isn't visible in their Studio - the `Organization:` field is the answer to 'which org should I switch the Studio to?'.",
        inputSchema: {},
    }, async () => proxyToolCall("get_api_key_context", {}));
    server.registerTool("list_organizations", {
        description: "Lists the Bitfab organizations the signed-in user belongs to, marking the one the plugin currently reads and writes ([current]). Use this to answer 'which org can I switch to?' or before switching the plugin's active org. Each entry includes the org id to pass to the switchOrg plugin command (which performs the switch and replaces the local API key).",
        inputSchema: {},
    }, async () => proxyToolCall("list_organizations", {}));
    server.registerTool("get_database_connection_status", {
        description: "Check whether this organization has connected a database for per-trace snapshot replay (DB branching). Returns one of: connected (ready - replay can branch the database to each trace's capture time), checking (provisioning in progress, poll again in ~15s), failed (the saved connection could not be provisioned), or none (no database connected yet). Use this during db-branching setup to wait until the customer's database is connected before enabling dbSnapshot.",
        inputSchema: {},
    }, async () => proxyToolCall("get_database_connection_status", {}));
    server.registerTool("list_trace_functions", {
        description: "List all traced functions in your organization. Use this to discover which functions are instrumented with the Bitfab SDK.",
        inputSchema: {},
    }, async () => proxyToolCall("list_trace_functions", {}));
    server.registerTool("search_traces", {
        description: "Search and filter traces for a traced function. Returns matching traces with IDs, status, timestamps, total duration (ms), total token usage (in/out/cached/total), model, and output preview (all when available). Replay traces (from test runs) carry the same totals. Use read_traces for per-span duration, tokens, time-to-first-token, and model. Supports keyword search, date range, status filter, and regex matching. Use drillDown to refine previous results.",
        inputSchema: {
            traceFunctionKey: z
                .string()
                .describe("The trace function key to search"),
            searchQuery: z.string().optional().describe("Full-text keyword search"),
            status: z
                .enum(["pending", "completed", "failed"])
                .optional()
                .describe("Filter by trace status"),
            createdAfter: z
                .string()
                .regex(DATE_PATTERN, "Must be YYYY-MM-DD format")
                .optional()
                .describe("Only traces after this date (YYYY-MM-DD)"),
            createdBefore: z
                .string()
                .regex(DATE_PATTERN, "Must be YYYY-MM-DD format")
                .optional()
                .describe("Only traces before this date (YYYY-MM-DD)"),
            traceFilterRegex: z
                .string()
                .optional()
                .describe("Regex to filter trace output"),
            limit: z
                .preprocess(parseJsonString, z.number().min(1).max(50))
                .optional()
                .describe("Max traces to return (default 10, max 50)"),
            drillDown: z
                .preprocess(parseJsonString, z.boolean())
                .optional()
                .describe("If true, merge with previous search filters to narrow results"),
            traceIds: z
                .preprocess(parseJsonString, z.array(z.uuid()))
                .optional()
                .describe("Specific trace IDs to fetch"),
            environment: z
                .string()
                .optional()
                .describe("Filter by environment (e.g., production, staging, development, preview). Omit to search across all environments."),
            labelSource: z
                .enum(["human", "agent"])
                .optional()
                .describe('Filter by who authored the label: "human" for human-authored labels only, "agent" for agent-authored labels only. Omit to include any source. This is a pure authorship filter - for trust/validation, use the separate `validated` parameter.'),
            validated: z
                .preprocess(parseJsonString, z.boolean())
                .optional()
                .describe('If true, only return traces with a validated label - a label that is either human-authored OR an agent-authored label that a human has approved (`approvedAt` is set). Use this when you want trustworthy/reviewed labels. Combine with `labelSource: "agent"` to get only approved-agent labels (excluding pure-human labels). Omit (or pass false) to include unapproved labels too.'),
            labelResult: z
                .preprocess(parseJsonString, z.boolean())
                .optional()
                .describe("Filter by label result: true for passing labels, false for failing labels"),
            testRunId: z
                .uuid()
                .optional()
                .describe("Filter to traces from a specific experiment (test run). Automatically sets includeReplays to true."),
            datasetId: z
                .uuid()
                .optional()
                .describe("Filter to traces belonging to a specific dataset. Combine with other filters (status, labelResult, etc.) to narrow within the dataset."),
            hasDbSnapshot: z
                .preprocess(parseJsonString, z.boolean())
                .optional()
                .describe("Filter by whether the trace captured a database snapshot reference. true returns only traces that can be replayed against a pinned historical DB branch (the 'Snapshot captured' traces); false returns only traces with no snapshot. Omit to include both."),
        },
    }, async (args) => proxyToolCall("search_traces", args));
    server.registerTool("read_traces", {
        description: 'Read one or more traces by ID. Includes inputs, outputs, status, environment tag, trace-level duration (ms), token usage (in/out/cached/total), and model (when available). Use these to compare latency and cost between a trace and its replay. By default returns truncated span details (input, output, reasoning, context, errors) plus per-span duration, tokens, time-to-first-token, and model when available; set scope="full" for the complete untruncated detail of a handful of traces. To load only labels + annotations for many traces at once (no span content), use read_trace_labels instead.',
        inputSchema: {
            traceIds: z
                .preprocess(parseJsonString, z.array(z.uuid()).min(1).max(10))
                .describe("Trace IDs to read (1-10)"),
            scope: z
                .enum(["summary", "full"])
                .optional()
                .default("summary")
                .describe('Level of span detail. "summary" (default) renders the full span structure but caps each span and trace-level field so the response stays small enough not to spill to disk: prefer it when scanning or identifying candidate traces. "full" renders every field untruncated (up to 10k chars each) with no per-span ceiling: use it when you need complete detail on a few traces.'),
        },
    }, async ({ traceIds, scope }) => proxyToolCall("read_traces", { traceIds, scope }));
    server.registerTool("read_trace_labels", {
        description: `Read just the labels for one or more traces by ID: returns each trace's pass/fail verdict, its annotation (the reviewer's reasoning), and whether the label is human-validated (a human authored or approved it). No span content, inputs, or outputs are loaded, so the response is small and one call accepts up to ${READ_TRACE_LABELS_MAX_IDS} IDs. Use this to load a whole dataset's verdicts into context in a single call (e.g. when building or confirming a dataset). When you need a trace's actual inputs/outputs/spans, use read_traces instead.`,
        inputSchema: {
            traceIds: z
                .preprocess(parseJsonString, z.array(z.uuid()).min(1).max(READ_TRACE_LABELS_MAX_IDS))
                .describe(`Trace IDs to read labels for (1-${READ_TRACE_LABELS_MAX_IDS})`),
        },
    }, async ({ traceIds }) => proxyToolCall("read_trace_labels", { traceIds }));
    server.registerTool("read_span_field", {
        description: "Fetch the COMPLETE, untruncated value of a single span field. read_traces truncates large span fields and prints a note telling you to call this when you need the full text. Pass the traceId, the spanId (the `[ID: ...]` shown for that span in read_traces output), and the field to retrieve. Returns up to 100k chars by default; pass a higher maxChars to read more of an extremely large field. Prefer this over re-reading the whole trace when you only need one large field in full.",
        inputSchema: {
            traceId: z.uuid().describe("The trace ID the span belongs to"),
            spanId: z
                .uuid()
                .describe("The span ID to read, taken from the `[ID: ...]` shown for the span in read_traces output"),
            field: z
                .enum([
                "input",
                "output",
                "reasoning",
                "content",
                "errors",
                "contexts",
            ])
                .describe("Which span field to return in full: input, output, reasoning, content, errors, or contexts"),
            maxChars: z
                .preprocess(parseJsonString, z.number().int().positive())
                .optional()
                .describe("Maximum characters to return (default 100000). Raise it to read further into an extremely large field."),
        },
    }, async ({ traceId, spanId, field, maxChars }) => proxyToolCall("read_span_field", { traceId, spanId, field, maxChars }));
    server.registerTool("update_agent_labels", {
        description: "Set, skip, or archive the agent's pass/fail verdict on one or more traces (`labelSource=\"agent\"`). Use this AFTER you have read the traces with read_traces and decided yourself whether each one looks like a pass, a fail, or genuinely cannot be judged. To set a verdict, pass `label` (true=PASS, false=FAIL) and `annotation` (your reasoning, shown to the human reviewer in the labeling UI). Optionally pass `confidence` (`VeryLow|Low|Medium|High|VeryHigh`) to record how confident you are - surfaced in the labeling UI so reviewers can prioritize low-confidence verdicts. To explicitly skip a trace you cannot decide on (instead of leaving it unlabeled), pass `skip: true` and omit label/annotation - this records an explicit skip so coverage checks know you intentionally did not verdict it. To clear a previously-set agent verdict (e.g., you changed your mind or labeled the wrong trace), pass `archive: true` and omit label/annotation. New verdicts start unapproved; once a human approves one in the UI, the label joins the validated dataset (queryable via `search_traces` with `validated: true`). Archiving is non-destructive - the row is hidden from all reads but kept for audit, and you can immediately re-label the trace from scratch. For replay results, key each label by the replay item's `originalTraceId` (the original trace it was replayed from; `sourceTraceId` is accepted as a deprecated alias) plus the top-level `testRunId` instead of a `traceId`: the server resolves it to the replay trace via lineage, so you never need a server-generated replay trace id. Returns an agent-readable summary with one parseable effective label line per updated trace, keyed by the id you supplied (`originalTraceId` for replay verdicts, otherwise `traceId`), so command callers can verify persistence.",
        inputSchema: {
            testRunId: z
                .uuid()
                .optional()
                .describe("The replay test run id. Required when any label targets a trace by originalTraceId (replay verdicts); ignored otherwise."),
            labels: z
                .preprocess(parseJsonString, z
                .array(z.object({
                traceId: z
                    .uuid()
                    .optional()
                    .describe("The trace ID to update. Provide this OR originalTraceId, not both."),
                originalTraceId: z
                    .uuid()
                    .optional()
                    .describe("For replay verdicts: the original trace ID the replay item was replayed from (the replay item's originalTraceId). The server resolves it to the replay trace via lineage, so you never need a server replay trace id. Requires the top-level testRunId. Provide this OR traceId, not both."),
                sourceTraceId: z
                    .uuid()
                    .optional()
                    .meta({ deprecated: true })
                    .describe("Deprecated alias for originalTraceId. Prefer originalTraceId; accepted for back-compat."),
                archive: z
                    .boolean()
                    .optional()
                    .describe("If true, archive the existing agent label for this trace and ignore label/annotation/skip/confidence. Default false."),
                skip: z
                    .boolean()
                    .optional()
                    .describe("If true, record an explicit skip (cannot judge) and ignore label/annotation/confidence. Default false."),
                label: z
                    .boolean()
                    .optional()
                    .describe("true for PASS, false for FAIL. Required when archive and skip are both false/omitted."),
                annotation: z
                    .string()
                    .min(1)
                    .optional()
                    .describe("Your reasoning for the verdict - shown to the human reviewer in the labeling UI. Required when archive and skip are both false/omitted."),
                confidence: z
                    .enum(["VeryLow", "Low", "Medium", "High", "VeryHigh"])
                    .optional()
                    .describe("How confident you are in this verdict: VeryLow / Low / Medium / High / VeryHigh. Surfaced as a chip in the labeling UI; low-confidence labels are prioritized for human review. Optional; only meaningful when paired with a label."),
            }))
                .min(1)
                .max(50))
                .describe("One update per trace (1-50)"),
        },
    }, async ({ testRunId, labels }) => proxyToolCall("update_agent_labels", {
        testRunId,
        // Mirror originalTraceId onto the deprecated sourceTraceId alias:
        // servers that predate the rename strip the unknown canonical key and
        // would otherwise see a label with no id; current servers prefer the
        // canonical key and ignore the alias.
        labels: labels.map((l) => l.originalTraceId !== undefined && l.sourceTraceId === undefined
            ? { ...l, sourceTraceId: l.originalTraceId }
            : l),
    }));
    server.registerTool("set_human_labels", {
        description: 'Record one or more human-authored pass/fail verdicts that are VALIDATED immediately (`labelSource="human"`, no approval step). Unlike update_agent_labels (which writes agent suggestions that start unapproved), labels set here join the validated dataset the instant they are written and satisfy `search_traces` with `validated: true`. Pass `label` (true=PASS, false=FAIL), `annotation` (the reasoning behind the verdict), and optionally `confidence` (`VeryLow|Low|Medium|High|VeryHigh`). USE ONLY when a human has explicitly decided the verdict, for example saving a known production bug as a test case. Do NOT use this for the agent\'s own first-pass guesses on traces awaiting human review; use update_agent_labels for those so they go through the normal approve/edit loop.',
        inputSchema: {
            labels: z
                .preprocess(parseJsonString, z
                .array(z.object({
                traceId: z.uuid().describe("The trace ID to label"),
                label: z
                    .boolean()
                    .describe("true for PASS, false for FAIL. Required."),
                annotation: z
                    .string()
                    .min(1)
                    .describe("The reasoning behind the verdict. Required."),
                confidence: z
                    .enum(["VeryLow", "Low", "Medium", "High", "VeryHigh"])
                    .optional()
                    .describe("How confident the verdict is: VeryLow / Low / Medium / High / VeryHigh. Optional."),
            }))
                .min(1)
                .max(50))
                .describe("One verdict per trace (1-50)"),
        },
    }, async ({ labels }) => proxyToolCall("set_human_labels", { labels }));
    server.registerTool("create_dataset", {
        description: "Create a labeled dataset for a traced function. Datasets are named buckets of traces (e.g. 'Hallucination failures', 'Edge cases for grader X') that humans review and that experiments replay against. Pass traceFunctionKey, a short human-readable name, and an optional description. Returns the new dataset id; use it with add_traces_to_dataset to populate.",
        inputSchema: {
            traceFunctionKey: z
                .string()
                .min(1)
                .describe("The trace function key the dataset is scoped to (must match traces you'll add)"),
            name: z
                .string()
                .min(1)
                .describe("Short human-readable name shown in the labeling UI (e.g. 'Hallucination failures')"),
            description: z
                .string()
                .optional()
                .describe("Optional longer description of what's in the dataset"),
        },
    }, async ({ traceFunctionKey, name, description }) => proxyToolCall("create_dataset", {
        traceFunctionKey,
        name,
        description,
    }));
    server.registerTool("list_datasets", {
        description: "List all datasets for a traced function, with id, name, description, and current trace count for each. Call this before add_traces_to_dataset / remove_traces_from_dataset to discover dataset ids, or before create_dataset to avoid duplicates.",
        inputSchema: {
            traceFunctionKey: z
                .string()
                .min(1)
                .describe("The trace function key to list datasets for"),
        },
    }, async ({ traceFunctionKey }) => proxyToolCall("list_datasets", { traceFunctionKey }));
    server.registerTool("add_traces_to_dataset", {
        description: "Add one or more traces to a dataset. Idempotent: re-adding a trace already in the dataset is a no-op. Traces are validated against the dataset's traceFunctionKey; ids that don't belong to this organization, or to the dataset's trace function, are silently skipped (the response reports how many were added vs skipped).",
        inputSchema: {
            datasetId: z
                .uuid()
                .describe("The dataset id (from create_dataset or list_datasets)"),
            traceIds: z
                .preprocess(parseJsonString, z.array(z.uuid()).min(1).max(100))
                .describe("Trace ids to add (1-100)"),
        },
    }, async ({ datasetId, traceIds }) => proxyToolCall("add_traces_to_dataset", { datasetId, traceIds }));
    server.registerTool("remove_traces_from_dataset", {
        description: "Remove one or more traces from a dataset. Trace ids that aren't currently in the dataset are silently ignored (the response reports how many were actually removed). The traces themselves are not deleted; only their membership in this dataset.",
        inputSchema: {
            datasetId: z.uuid().describe("The dataset id"),
            traceIds: z
                .preprocess(parseJsonString, z.array(z.uuid()).min(1).max(100))
                .describe("Trace ids to remove (1-100)"),
        },
    }, async ({ datasetId, traceIds }) => proxyToolCall("remove_traces_from_dataset", { datasetId, traceIds }));
    server.registerTool("list_experiments", {
        description: "List experiments (replay test runs) for a traced function, with status, pass/fail totals, and delta (fixed/regressed/still-passing/still-failing) for each. Call this to see past experiment results, or at the start of experiment mode to understand what's been tried before.",
        inputSchema: {
            traceFunctionKey: z
                .string()
                .min(1)
                .describe("The trace function key to list experiments for"),
            limit: z
                .preprocess(parseJsonString, z.number().min(1).max(50))
                .optional()
                .describe("Max experiments to return (default 20, max 50)"),
        },
    }, async ({ traceFunctionKey, limit }) => proxyToolCall("list_experiments", { traceFunctionKey, limit }));
    server.registerTool("get_experiment_traces", {
        description: "Get individual trace results for an experiment (test run), including each replay trace's verdict (fixed, regressed, still-passing, still-failing, unpaired) by comparing against the original trace's label, plus token usage (input, output, cached, total) for the replay and the paired original so you can reason about cost and cache-read deltas. Use after list_experiments to drill into a specific experiment's results.",
        inputSchema: {
            testRunId: z
                .uuid()
                .describe("The experiment (test run) ID to get traces for"),
        },
    }, async ({ testRunId }) => proxyToolCall("get_experiment_traces", { testRunId }));
    server.registerTool("get_replay_status", {
        description: "Read the current replay test run status, including local replay trace ID to server replay trace ID mapping. Safe to poll while replay is still running.",
        inputSchema: {
            testRunId: z
                .uuid()
                .describe("The replay test run ID to get status for"),
        },
    }, async ({ testRunId }) => proxyToolCall("get_replay_status", { testRunId }));
    server.registerTool("get_template_reference", {
        description: "Read the agent-facing reference for Bitfab span templates: the engine (Nunjucks), the render-context schema (variables bound at render time), the composition model (no `{% extends %}` - assembler injects into base.njk), the registered custom filters and tests, common patterns from the live default templates, and error behavior. **Call this once per session before your first `get_template` or `update_template`** - without it, edits typically use the wrong syntax dialect, the wrong variable names, or break the layout. Read-only, no organization data, safe to cache.",
        inputSchema: {},
    }, async () => proxyToolCall("get_template_reference", {}));
    server.registerTool("get_template", {
        description: "Read the rendering template for a span type. Templates control how a span's input/output renders in the Bitfab UI. **Pass `traceFunctionKey` to scope to a specific trace function** (this is the normal mode for the `templates` skill). Resolution: per-key row → org-global override → file default. The response labels the source so you know whether you're editing a per-key row, seeding from the org-global, or starting from the file default. Without `traceFunctionKey`, returns the org-global override. Span types: llm, agent, function, guardrail, handoff, custom. Always call this before update_template so your edit starts from the live content. **If you haven't called `get_template_reference` this session, call it first** - it documents the Nunjucks engine, available variables, and base-template composition model, which you'll need to write a correct edit.",
        inputSchema: {
            spanType: z
                .enum(["llm", "agent", "function", "guardrail", "handoff", "custom"])
                .describe("Span type whose template to read (llm | agent | function | guardrail | handoff | custom)"),
            traceFunctionKey: z
                .string()
                .min(1)
                .optional()
                .describe("Trace function key to scope the template to. When set, returns the per-key row if present; otherwise seeds from the org-global override; otherwise the file default. Omit to read the org-global template."),
        },
    }, async ({ spanType, traceFunctionKey }) => proxyToolCall("get_template", { spanType, traceFunctionKey }));
    server.registerTool("update_template", {
        description: "Upsert the rendering template for a span type. **Pass `traceFunctionKey` to scope the override to a single trace function** (this is the normal mode for the `templates` skill); the per-key row is created on first save and applied at render time only for spans on traces of that function. Without `traceFunctionKey`, edits the org-global override (applies when no per-key row matches). Updates are in-place: no version bump, no row juggling. The `template` body is the full template source you want stored, passed verbatim - do NOT wrap in `<![CDATA[ ... ]]>`, triple-backtick code fences, or any other escape construct; the value is stored exactly as received and a stray wrapper will silently break rendering (e.g. an opening `<![CDATA[` swallows the next `<style>` tag). Optional `name` and `description` are metadata shown in the templates UI. Span types: llm, agent, function, guardrail, handoff, custom. Always call get_template first so you're editing the current content rather than overwriting unrelated changes. **If you haven't called `get_template_reference` this session, call it first** - edits that don't match the documented Nunjucks variable shape will silently render empty in production.",
        inputSchema: {
            spanType: z
                .enum(["llm", "agent", "function", "guardrail", "handoff", "custom"])
                .describe("Span type whose template to upsert (llm | agent | function | guardrail | handoff | custom)"),
            traceFunctionKey: z
                .string()
                .min(1)
                .optional()
                .describe("Trace function key to scope the override to. When set, upserts a per-key row that the renderer prefers over the org-global for spans on traces of this function. Omit to edit the org-global template."),
            template: z
                .string()
                .min(1)
                .describe("Full template source to store, verbatim (replaces existing content). Pass as a raw string - no CDATA wrapper, no triple-backtick code fence, no escape construct."),
            name: z
                .string()
                .min(1)
                .optional()
                .describe("Optional new name. Defaults to 'Default {spanType} template' on first creation."),
            description: z
                .string()
                .optional()
                .describe("Optional description. Pass an empty string to clear an existing description."),
        },
    }, async ({ spanType, traceFunctionKey, template, name, description }) => {
        // Await the editStarted push so the banner / card animation hit the
        // browser BEFORE the save completes. Without this, a fast DB write
        // can finish before the SSE delivery, and the editEnded event lands
        // on the bus before editStarted - the user sees the "Saved" flash
        // with no preceding "editing" state.
        await pushToActivePreview({
            type: "editStarted",
            data: {
                spanType,
                traceFunctionKey: traceFunctionKey ?? null,
                agentLabel: resolveCodingAgentDisplayName(),
            },
        });
        try {
            const result = await proxyToolCall("update_template", {
                spanType,
                traceFunctionKey,
                template,
                name,
                description,
            });
            await pushToActivePreview({
                type: "editEnded",
                data: {
                    spanType,
                    traceFunctionKey: traceFunctionKey ?? null,
                    success: true,
                    summary: traceFunctionKey
                        ? `Updated ${spanType} template for ${traceFunctionKey}`
                        : `Updated org-global ${spanType} template`,
                },
            });
            return result;
        }
        catch (err) {
            await pushToActivePreview({
                type: "editEnded",
                data: {
                    spanType,
                    traceFunctionKey: traceFunctionKey ?? null,
                    success: false,
                    summary: err.message ?? "update_template failed",
                },
            });
            throw err;
        }
    });
    server.registerTool("create_trace_plan", {
        description: "Post a tracing instrumentation plan and get back a URL the user opens to confirm or adjust. Send the full call tree (root + children + framework spans + ~10 surrounding callers above the root and ~10 surrounding callees below each leaf as `pure` context nodes), initial recommended captured node ids, replay `analysis` on every node, and any sample input/output stored on the relevant nodes. Pass `traceFunctionKey` so future Modify cycles for this same key can bootstrap from this plan via get_trace_plan. Returns a URL the user opens in the browser; after they confirm or abort there, call get_trace_plan to read the final captured set and act on it.",
        inputSchema: {
            language: z
                .enum(["python", "typescript", "ruby", "go"])
                .describe("Source language of the user's code"),
            // Structured (not z.unknown()) so the tool's JSON Schema shows the agent
            // the per-node `analysis` field instead of burying it in prose, which is
            // how plans arrived with every node unclassified.
            tree: z
                .preprocess(parseJsonString, tracePlanTreeShape)
                .describe("TracePlanTree: { rootId, nodes: { [id]: TraceNode } }. Each TraceNode has id, name, kind ('manual' | 'auto' | 'pure'), file, line, signature, parentId, childIds, plus optional framework, fields, sampleInput, sampleOutput. Include ~10 surrounding callers above the root and ~10 surrounding callees below each leaf as `pure` (uncaptured) context nodes so the user can see what's adjacent in the codebase when they edit in the UI. Set `analysis` on every node, including uncaptured context nodes, describing WHAT THAT NODE DOES; classify by its own body: `analysis` is { classification: 'pure' | 'model_call' | 'external_read' | 'side_effect', mockable?: boolean, unmockableReason?: string, inputSerializable?: boolean, outputSerializable?: boolean, innerCall?: { name }, sideEffectKind?: 'db_write' | 'http_outbound' | 'queue' | 'email' | 'filesystem' | 'vector_write', readKind?: 'db_read' | 'http_read' | 'vector_search' | 'cache_read' | 'filesystem_read' }. Set `mockable` mechanically from the node's `kind`: `kind: 'manual'` (a hand-written `withSpan`/`@span`) is mockable, omit `mockable`; `kind: 'auto'` (captured by a framework handler/processor/stream/collector) gets `mockable: false` + `unmockableReason`, the ONE exception being Vercel AI SDK model spans (its `wrapLanguageModel` middleware routes the call through `withSpan`), which are mockable, omit; the root gets omit (never mockable). Rationale: mocking returns a span's recorded output instead of running the call, which only works when the call goes through a `withSpan` wrapper; `auto` framework spans are observed, not wrapped. Key hazard: an `auto` `external_read`/`side_effect` (a framework tool hitting a DB/HTTP) left mockable promises a mock replay can't deliver; its fix is a manual `withSpan` around that call or a db-snapshot, never a mock. Serializability is two raw facts, distinct from `mockable`: set `outputSerializable: false` when the recorded OUTPUT does not round-trip through serialization, and set `inputSerializable: false` when the recorded INPUT does not (its arguments hold a DB client, an open stream, a callback, or a class instance with no JSON form); omit either when it serializes (the default). Two rules follow. (1) Do NOT choose a root whose input is not serializable: replay re-runs the root against its recorded input, so a non-serializable-input root is not replayable, promote the root to a caller that takes a serializable input instead. (2) Do NOT mock a node whose output is not serializable (replay has no recorded value to return); the server enforces this by forcing any `outputSerializable: false` node unmockable, so you don't need to also set `mockable` for that reason. Use `model_call` only for the span that IS the model call (the actual LLM invocation, e.g. an auto-captured `ChatOpenAI` leaf); they re-run live with the user's changes, that is the point of replay. A framework wrapper or orchestrator whose model call is represented by a child node (a LangChain `chain.invoke`, a LangGraph node, the root function that just calls model-call children) is `pure`, NOT `model_call`; classify each node by its own body excluding work in child nodes, never bubble a child's `model_call` up to its parent. Use `external_read` for reads of external mutable state (DB SELECT, GET, vector search), set `readKind`. Use `side_effect` for writes/mutations that fire for real on replay (DB write, outbound POST/PUT/DELETE, email, queue, charge, file/vector write), set `sideEffectKind`. Prefer the smallest external boundary: if a broad read/write wrapper contains parsing, ranking, prompt construction, model calls, or other live code, classify/wrap the lower DB/HTTP/write call instead; the server suppresses default mocks for broad external parents with live descendants. Use `pure` for deterministic local compute. The server derives the replay disposition (`mockOnReplay`) and the validation summary from this classification and mockability, so do NOT send them. Mockable `external_read` and `side_effect` nodes default to mocked replay, except broad external parents with live descendants; `model_call` and `pure` nodes default to live replay."),
            capturedNodeIds: z
                .preprocess(parseJsonString, z.array(z.string()))
                .describe("Initial recommended captured set. Must reference ids present in tree.nodes. Must form a connected sub-tree (selecting any descendant implies its ancestors). Surrounding `pure` context nodes are not captured by default."),
            stats: z
                .preprocess(parseJsonString, z.unknown())
                .optional()
                .describe("Optional sample-run stats: { durationMs?, tokens?, cost? }"),
            traceFunctionKey: z
                .string()
                .min(1)
                .optional()
                .describe("Optional trace function key (the string passed to getFunction / get_function / bitfab_function / WithFunctionName). Persisted on the plan so Modify cycles can fetch this plan as the `before` tree via get_trace_plan with the same key."),
            source: z
                .enum(["interactive", "analyze_repo"])
                .optional()
                .describe("How this plan was produced. Omit (defaults to 'interactive') for a normal Instrument/Modify cycle the user confirms in the browser. Set 'analyze_repo' ONLY from the non-interactive `analyze-repo` batch mode, so these auto-drafted, unconfirmed plans are marked distinctly from interactively-confirmed ones."),
        },
    }, async ({ language, tree, capturedNodeIds, stats, traceFunctionKey, source, }) => proxyToolCall("create_trace_plan", {
        language,
        tree,
        capturedNodeIds,
        stats,
        traceFunctionKey,
        source,
    }));
    server.registerTool("confirm_trace_plan", {
        description: "Confirm a trace plan WITHOUT the browser, for the inline 'Continue' path where the user reviewed the ASCII plan in chat and chose not to open Studio. Persists the plan as the latest confirmed plan for its traceFunctionKey (so later get_trace_plan by key, `setup view`, and `setup modify` find it) and returns the final captured set plus per-node replay/mock decisions, exactly like get_trace_plan. Studio's Confirm button does this for the browser path; call this only for Continue. Pass the recommended capturedNodeIds; omit mockOnReplayByNodeId to accept the analysis-derived replay decisions (pass it only to override specific nodes the way a Studio toggle would).",
        inputSchema: {
            planId: z
                .string()
                .describe("The trace plan id returned by create_trace_plan."),
            capturedNodeIds: z
                .preprocess(parseJsonString, z.array(z.string()).min(1))
                .describe("The captured set to confirm. Must reference ids present in the plan's tree and form a single connected sub-tree with one entry point (its top-most captured span). The entry point does NOT have to be the plan's tree root: untracing the tree root and rooting at a captured descendant is honored and persisted as-is."),
            mockOnReplayByNodeId: z
                .preprocess(parseJsonString, z.record(z.string(), z.boolean()))
                .optional()
                .describe("Optional per-node overrides of the replay/mock decision (the equivalent of a Studio toggle). Omit to accept the server's analysis-derived defaults. The replay entry point (the top-most captured span) always re-runs live and cannot be mocked."),
        },
    }, async ({ planId, capturedNodeIds, mockOnReplayByNodeId }) => proxyToolCall("confirm_trace_plan", {
        planId,
        capturedNodeIds,
        mockOnReplayByNodeId,
    }));
    server.registerTool("get_trace_plan", {
        description: "Read a trace plan. Two modes: pass `planId` to read a specific plan (use this after create_trace_plan + browser confirmation to learn whether the user confirmed, cancelled, or is still pending); or pass `traceFunctionKey` to fetch the latest *confirmed* plan for that key (use this at the start of a Modify cycle to bootstrap the `before` tree from the prior plan, so you don't have to re-derive it from the code). Returns the full plan including the tree as JSON.",
        inputSchema: {
            planId: z
                .string()
                .optional()
                .describe("The trace plan id returned by create_trace_plan."),
            traceFunctionKey: z
                .string()
                .optional()
                .describe("Trace function key. Returns the latest confirmed plan for this key in the caller's organization, or a 'no prior plan' message if none exists."),
        },
    }, async ({ planId, traceFunctionKey }) => proxyToolCall("get_trace_plan", { planId, traceFunctionKey }));
    return { server, proxy };
}
export async function startMcpServer(platform, getConfig, pluginVersion) {
    const { server } = createMcpServer(platform, getConfig, pluginVersion);
    const transport = new StdioServerTransport();
    await server.connect(transport);
}
