// bitfab-plugin-lib/dist/parseArgs.js
export const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
function formatPositionalLabel(spec) {
    const tag = spec.format && spec.format !== "string" ? `:${spec.format}` : "";
    if (spec.variadic) {
        return spec.required !== false
            ? `<${spec.name}${tag}> [<${spec.name}${tag}>...]`
            : `[<${spec.name}${tag}>...]`;
    }
    return spec.required !== false
        ? `<${spec.name}${tag}>`
        : `[${spec.name}${tag}]`;
}
function formatFlagLabel(spec) {
    if (spec.boolean) {
        return `[--${spec.name}]`;
    }
    const tag = spec.format && spec.format !== "string" ? `:${spec.format}` : "";
    const valueHint = spec.format === "number" ? " N" : ` <${spec.name}${tag}>`;
    const defaultHint = spec.default !== undefined ? ` (default: ${spec.default})` : "";
    return `[--${spec.name}${valueHint}${defaultHint}]`;
}
function validateFormat(value, format) {
    if (format === "uuid" && !UUID_RE.test(value)) {
        return `"${value}" is not a valid UUID`;
    }
    if (format === "number") {
        const n = Number(value);
        if (!Number.isFinite(n) || n <= 0) {
            return `"${value}" is not a valid positive number`;
        }
    }
    return null;
}
export function buildUsage(command, positional, flags) {
    const parts = [
        `Usage: ${command}`,
        ...positional.map(formatPositionalLabel),
        ...(flags ?? []).map(formatFlagLabel),
    ];
    return parts.join(" ");
}
/**
 * Full help block: usage line, command description, then per-argument and
 * per-flag descriptions. Printed on -h / --help.
 */
export function buildHelp(command, positional, flags, description) {
    const lines = [buildUsage(command, positional, flags)];
    if (description) {
        lines.push("", description);
    }
    const describedArgs = positional.filter((s) => s.description);
    const flagSpecs = flags ?? [];
    const labelWidth = Math.max(0, ...describedArgs.map((s) => s.name.length), ...flagSpecs.map((s) => s.name.length + 2));
    if (describedArgs.length > 0) {
        lines.push("", "Arguments:");
        for (const spec of describedArgs) {
            lines.push(`  ${spec.name.padEnd(labelWidth)}  ${spec.description}`);
        }
    }
    if (flagSpecs.length > 0) {
        lines.push("", "Flags:");
        for (const spec of flagSpecs) {
            const detail = [
                spec.description,
                spec.default !== undefined ? `(default: ${spec.default})` : undefined,
            ]
                .filter(Boolean)
                .join(" ");
            lines.push(`  ${`--${spec.name}`.padEnd(labelWidth)}  ${detail}`.trimEnd());
        }
    }
    return lines.join("\n");
}
function helpRequested(raw) {
    return raw.includes("-h") || raw.includes("--help");
}
/**
 * Help support for commands that hand-roll their own argv parsing or take no
 * args at all. Prints the help block and exits 0 when -h / --help is present;
 * otherwise a no-op. Commands using `parseArgs` get this behavior built in.
 */
export function printHelpIfRequested(opts) {
    const argv = opts.argv ?? process.argv;
    if (!helpRequested(argv.slice(2))) {
        return;
    }
    console.log(buildHelp(extractCommandName(argv), opts.positional ?? [], opts.flags, opts.description));
    process.exit(0);
}
function extractCommandName(argv) {
    return argv[1]?.replace(/.*\//, "").replace(/\.[jt]s$/, "") ?? "command";
}
function fail(commandName, positional, flags, message, usageNote) {
    console.error(buildUsage(commandName, positional, flags));
    if (usageNote) {
        console.error(usageNote);
    }
    console.error(message);
    process.exit(1);
}
export function parseArgs(opts) {
    const argv = opts.argv ?? process.argv;
    const raw = argv.slice(2);
    const commandName = extractCommandName(argv);
    const flagSpecs = opts.flags ?? [];
    const failWith = (message) => fail(commandName, opts.positional, opts.flags, message, opts.usageNote);
    // -h/--help handling is shared: delegate to the same helper hand-rolled
    // entrypoints use, so the print-and-exit logic lives in exactly one place.
    printHelpIfRequested({
        description: opts.description ?? "",
        positional: opts.positional,
        flags: opts.flags,
        argv,
    });
    const positionalTokens = [];
    const parsedFlags = {};
    for (const spec of flagSpecs) {
        if (spec.default !== undefined) {
            parsedFlags[spec.name] = spec.default;
        }
    }
    for (let i = 0; i < raw.length; i++) {
        const token = raw[i];
        if (!token.startsWith("--")) {
            positionalTokens.push(token);
            continue;
        }
        const spec = flagSpecs.find((f) => `--${f.name}` === token);
        if (!spec) {
            failWith(`Unknown flag: ${token}`);
            continue;
        }
        if (spec.boolean) {
            parsedFlags[spec.name] = true;
            continue;
        }
        const value = raw[i + 1];
        if (value === undefined || value.startsWith("--")) {
            failWith(`Flag ${token} requires a value.`);
            continue;
        }
        const formatErr = validateFormat(value, spec.format);
        if (formatErr) {
            failWith(`Flag "${spec.name}": ${formatErr}`);
        }
        parsedFlags[spec.name] = spec.format === "number" ? Number(value) : value;
        i += 1;
    }
    const hasVariadic = opts.positional.some((s) => s.variadic);
    const requiredCount = opts.positional.filter((s) => s.required !== false && !s.variadic).length;
    const requiredVariadicMin = opts.positional.some((s) => s.variadic && s.required !== false)
        ? 1
        : 0;
    const minArgs = requiredCount + requiredVariadicMin;
    const maxArgs = hasVariadic ? Infinity : opts.positional.length;
    if (positionalTokens.length < minArgs) {
        failWith(`Expected at least ${minArgs} argument${minArgs === 1 ? "" : "s"}, got ${positionalTokens.length}.`);
    }
    if (positionalTokens.length > maxArgs) {
        failWith(`Expected at most ${maxArgs} argument${maxArgs === 1 ? "" : "s"}, got ${positionalTokens.length}.`);
    }
    for (let j = 0; j < positionalTokens.length; j++) {
        const spec = j < opts.positional.length ? opts.positional[j] : opts.positional.at(-1);
        const err = validateFormat(positionalTokens[j], spec.format);
        if (err) {
            failWith(`Argument "${spec.name}": ${err}`);
        }
    }
    return { positional: positionalTokens, flags: parsedFlags };
}
/**
 * Convenience wrapper for commands that only use positional args (no flags).
 * Rejects any --flag tokens with a clear error.
 */
export function positionalArgs(specs, argv = process.argv, opts) {
    return parseArgs({
        positional: specs,
        argv,
        usageNote: opts?.usageNote,
        description: opts?.description,
    }).positional;
}
