// bitfab-plugin-lib/dist/parseArgs.js
const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
function formatPositionalLabel(spec) {
    const tag = spec.format && spec.format !== "string" ? `:${spec.format}` : "";
    if (spec.variadic) {
        return spec.required !== false
            ? `<${spec.name}${tag}> [<${spec.name}${tag}>...]`
            : `[<${spec.name}${tag}>...]`;
    }
    return spec.required !== false
        ? `<${spec.name}${tag}>`
        : `[${spec.name}${tag}]`;
}
function formatFlagLabel(spec) {
    const tag = spec.format && spec.format !== "string" ? `:${spec.format}` : "";
    const valueHint = spec.format === "number" ? " N" : ` <${spec.name}${tag}>`;
    const defaultHint = spec.default !== undefined ? ` (default: ${spec.default})` : "";
    return `[--${spec.name}${valueHint}${defaultHint}]`;
}
function validateFormat(value, format) {
    if (format === "uuid" && !UUID_RE.test(value)) {
        return `"${value}" is not a valid UUID`;
    }
    if (format === "number") {
        const n = Number(value);
        if (!Number.isFinite(n) || n <= 0) {
            return `"${value}" is not a valid positive number`;
        }
    }
    return null;
}
export function buildUsage(command, positional, flags) {
    const parts = [
        `Usage: ${command}`,
        ...positional.map(formatPositionalLabel),
        ...(flags ?? []).map(formatFlagLabel),
    ];
    return parts.join(" ");
}
function extractCommandName(argv) {
    return argv[1]?.replace(/.*\//, "").replace(/\.[jt]s$/, "") ?? "command";
}
function fail(commandName, positional, flags, message) {
    console.error(buildUsage(commandName, positional, flags));
    console.error(message);
    process.exit(1);
}
export function parseArgs(opts) {
    const argv = opts.argv ?? process.argv;
    const raw = argv.slice(2);
    const commandName = extractCommandName(argv);
    const flagSpecs = opts.flags ?? [];
    const positionalTokens = [];
    const parsedFlags = {};
    for (const spec of flagSpecs) {
        if (spec.default !== undefined) {
            parsedFlags[spec.name] = spec.default;
        }
    }
    for (let i = 0; i < raw.length; i++) {
        const token = raw[i];
        if (!token.startsWith("--")) {
            positionalTokens.push(token);
            continue;
        }
        const spec = flagSpecs.find((f) => `--${f.name}` === token);
        if (!spec) {
            fail(commandName, opts.positional, opts.flags, `Unknown flag: ${token}`);
            continue;
        }
        const value = raw[i + 1];
        if (value === undefined || value.startsWith("--")) {
            fail(commandName, opts.positional, opts.flags, `Flag ${token} requires a value.`);
            continue;
        }
        const formatErr = validateFormat(value, spec.format);
        if (formatErr) {
            fail(commandName, opts.positional, opts.flags, `Flag "${spec.name}": ${formatErr}`);
        }
        parsedFlags[spec.name] = spec.format === "number" ? Number(value) : value;
        i += 1;
    }
    const hasVariadic = opts.positional.some((s) => s.variadic);
    const requiredCount = opts.positional.filter((s) => s.required !== false && !s.variadic).length;
    const requiredVariadicMin = opts.positional.some((s) => s.variadic && s.required !== false)
        ? 1
        : 0;
    const minArgs = requiredCount + requiredVariadicMin;
    const maxArgs = hasVariadic ? Infinity : opts.positional.length;
    if (positionalTokens.length < minArgs) {
        fail(commandName, opts.positional, opts.flags, `Expected at least ${minArgs} argument${minArgs === 1 ? "" : "s"}, got ${positionalTokens.length}.`);
    }
    if (positionalTokens.length > maxArgs) {
        fail(commandName, opts.positional, opts.flags, `Expected at most ${maxArgs} argument${maxArgs === 1 ? "" : "s"}, got ${positionalTokens.length}.`);
    }
    for (let j = 0; j < positionalTokens.length; j++) {
        const spec = j < opts.positional.length ? opts.positional[j] : opts.positional.at(-1);
        const err = validateFormat(positionalTokens[j], spec.format);
        if (err) {
            fail(commandName, opts.positional, opts.flags, `Argument "${spec.name}": ${err}`);
        }
    }
    return { positional: positionalTokens, flags: parsedFlags };
}
/**
 * Convenience wrapper for commands that only use positional args (no flags).
 * Rejects any --flag tokens with a clear error.
 */
export function positionalArgs(specs, argv = process.argv) {
    return parseArgs({ positional: specs, argv }).positional;
}
