// bitfab-plugin-lib/dist/replayProgressView.js
/**
 * Line formatting for the `replayProgress` command.
 *
 * The replay can't show live progress in a coding-agent pane (no streaming,
 * output collapses), so the agent runs `replayProgress` in the background and
 * relays the lines it prints into chat. To keep the agent's job trivial and
 * deterministic, ALL of "what to say" lives here: the command turns the SDK's
 * per-trace `@@bitfab:progress` events into one self-contained line per trace as
 * it finishes (a pass/fail glyph, the running `n/total`, that trace's duration,
 * and the error reason when it failed), bracketed by a header and a closing
 * summary. The agent never decides what or when to print, it just shows each new
 * line.
 *
 * Pure functions so the formatting is deterministic and unit-testable; the
 * command owns the output stream.
 */
/** Seconds with one decimal (`0.9s`), or "" when the duration is unknown. */
function formatDuration(ms) {
    if (ms == null || !Number.isFinite(ms) || ms < 0) {
        return "";
    }
    return `${(ms / 1000).toFixed(1)}s`;
}
function errorSummary(error) {
    if (!error) {
        return "errored";
    }
    for (const raw of error.split("\n")) {
        const line = raw.trim();
        if (line.length > 0) {
            return line;
        }
    }
    return "errored";
}
/** The start banner: what is being replayed, and how many traces if known. */
export function formatReplayHeader(label, total) {
    const count = total && total > 0 ? `${total} trace${total === 1 ? "" : "s"}` : "traces";
    return label && label.length > 0 ? `▶ ${label} · ${count}` : `▶ ${count}`;
}
/**
 * One self-contained line for the trace that just finished:
 *
 *   ✓ 5/12 · 1.1s
 *   ✗ 4/12 · 0.2s · missing OPENAI_API_KEY
 *
 * The glyph is this trace's pass/fail, `n/total` is overall progress, the time is
 * this trace's own duration (omitted if unknown), and the reason rides along only
 * on failure.
 */
export function formatTraceLine(e) {
    const completed = Math.max(0, Math.trunc(e.completed));
    const total = Math.max(completed, Math.trunc(Math.max(0, e.total)));
    const failed = Boolean(e.item?.error);
    const parts = [`${failed ? "✗" : "✓"} ${completed}/${total}`];
    const time = formatDuration(e.item?.durationMs);
    if (time) {
        parts.push(time);
    }
    if (failed) {
        parts.push(errorSummary(e.item?.error));
    }
    return `  ${parts.join(" · ")}`;
}
/**
 * A liveness line emitted when no trace has settled for a while, so a slow trace
 * (an LLM call can take tens of seconds) never makes the run look frozen. Decoupled
 * from completion: a timer prints this on a quiet interval, then real per-trace
 * lines resume. Whole-second elapsed, since it is about "still alive", not precision.
 */
export function formatHeartbeat(completed, total, elapsedMs) {
    const c = Math.max(0, Math.trunc(completed));
    const t = Math.max(0, Math.trunc(total));
    const elapsed = Math.max(0, Math.round(Math.max(0, elapsedMs) / 1000));
    const count = t > 0 ? `${c}/${t} running` : "running";
    return `  … ${count} · ${elapsed}s elapsed`;
}
/** The closing summary: counts, wall time, and average per-trace time. */
export function formatReplaySummary(s) {
    const completed = Math.max(0, Math.trunc(s.completed));
    const errored = Math.min(Math.max(0, Math.trunc(s.errored)), completed);
    const ok = Math.max(0, completed - errored);
    const elapsed = (Math.max(0, s.elapsedMs) / 1000).toFixed(1);
    const parts = [`${errored === 0 ? "✓" : "⚠"} done`, `${ok} ok`];
    if (errored > 0) {
        parts.push(`${errored} failed`);
    }
    parts.push(`${elapsed}s`);
    if (s.sumDurationMs != null &&
        Number.isFinite(s.sumDurationMs) &&
        completed > 0) {
        const avg = (s.sumDurationMs / completed / 1000).toFixed(1);
        parts.push(`avg ${avg}s`);
    }
    return `  ${parts.join(" · ")}`;
}
export function formatReplayCommandFailure(opts) {
    const how = opts.signal ? `signal ${opts.signal}` : `exit ${opts.code ?? 1}`;
    return `  ✗ failed (${how}) · replay command exited non-zero · check ${opts.out}`;
}
export function formatReplayResultMissing(out) {
    return `  ✗ failed · replay finished without ReplayResult JSON · check ${out}`;
}
/**
 * The closing line when the replay settled NO traces (zero `@@bitfab:progress`
 * events). Without it the run would just end after a few heartbeats, and "no
 * per-trace lines" is ambiguous: it looks the same as a hang. Two cases:
 *  - it ran (exit 0) but reported nothing: its `@bitfab/sdk` predates progress
 *    reporting, or the replay script never wired `onProgress`.
 *  - it failed before reporting anything: point at the captured output.
 *
 * Deliberately version-agnostic: the wrapper only spawns a command, it can't read
 * the replay's SDK version, and the reporter's minimum version isn't released yet.
 */
export function formatNoProgressNotice(opts) {
    if (!opts.failed) {
        return "  ⚠ done · replay finished but reported no progress · its @bitfab/sdk predates progress reporting, or onProgress isn't wired";
    }
    const how = opts.signal ? `signal ${opts.signal}` : `exit ${opts.code ?? 1}`;
    return `  ✗ failed (${how}) · no progress reported · check ${opts.out}`;
}
