// bitfab-plugin-lib/dist/sdkUpdates.js
import * as semver from "semver";
import { BAKED_SDK_VERSIONS } from "./bakedSdkVersions.js";
import { getCurrentVersion, getInstalledSdks, } from "./installedSdk.js";
const TS_PACKAGE = "@bitfab/sdk";
const PY_PACKAGE = "bitfab-py";
const RB_PACKAGE = "bitfab";
const GO_MODULE = "github.com/Project-White-Rabbit/bitfab-go";
async function fetchJson(url) {
    try {
        const response = await fetch(url, { signal: AbortSignal.timeout(3000) });
        if (!response.ok) {
            return null;
        }
        return await response.json();
    }
    catch {
        return null;
    }
}
function encodeGoModulePath(modulePath) {
    return modulePath.replace(/[A-Z]/g, (ch) => `!${ch.toLowerCase()}`);
}
export async function getLatestSdkVersion(language) {
    if (language === "typescript") {
        const data = (await fetchJson(`https://registry.npmjs.org/${TS_PACKAGE}/latest`));
        return data?.version ?? null;
    }
    if (language === "python") {
        const data = (await fetchJson(`https://pypi.org/pypi/${PY_PACKAGE}/json`));
        return data?.info?.version ?? null;
    }
    if (language === "ruby") {
        const data = (await fetchJson(`https://rubygems.org/api/v1/gems/${RB_PACKAGE}.json`));
        return data?.version ?? null;
    }
    if (language === "go") {
        const data = (await fetchJson(`https://proxy.golang.org/${encodeGoModulePath(GO_MODULE)}/@latest`));
        return data?.Version ? data.Version.replace(/^v/, "") : null;
    }
    return null;
}
export function getBakedSdkVersion(language) {
    return BAKED_SDK_VERSIONS[language];
}
export function isSdkVersionNewer(latest, current) {
    if (!current) {
        return false;
    }
    // node-semver handles numeric, pre-release, and build-metadata precedence
    // (0.11.0 > 0.11.0-beta.1; build metadata is ignored). coerce tolerates the
    // loose version strings we extract from lockfiles/manifests.
    const l = semver.coerce(latest, { includePrerelease: true });
    const c = semver.coerce(current, { includePrerelease: true });
    if (!l || !c) {
        return false;
    }
    return semver.gt(l, c);
}
function toStatus(sdk, latest, source, remoteCheckFailed) {
    const current = getCurrentVersion(sdk);
    const versionNewer = latest !== null && isSdkVersionNewer(latest, current);
    return {
        language: sdk.language,
        packageName: sdk.packageName,
        declaredVersion: sdk.declaredVersion,
        resolvedVersion: sdk.resolvedVersion,
        current,
        latest,
        latestSource: source,
        remoteCheckFailed,
        updateAvailable: versionNewer || sdk.renameFrom !== null,
        manifestPath: sdk.manifestPath,
        lockfilePath: sdk.lockfilePath,
        workspacePath: sdk.workspacePath,
        renameFrom: sdk.renameFrom,
        deprecated: sdk.deprecated ?? false,
    };
}
/**
 * Synchronous, offline SDK update check. Compares the resolved version
 * (from lockfiles) against the version that was baked into the plugin at
 * build time. No network, no process spawning — safe to call on every
 * session start.
 */
export function checkForSdkUpdatesOffline(cwd) {
    const installed = getInstalledSdks(cwd);
    if (installed.length === 0) {
        return [];
    }
    return installed.map((sdk) => toStatus(sdk, getBakedSdkVersion(sdk.language), "baked", false));
}
/**
 * Full SDK update check: offline first, then a remote registry fetch.
 * - If the remote fetch returns a version, it replaces the baked value.
 * - If the remote fetch fails, remoteCheckFailed is set so the caller
 *   can punt to the agent (which can run the package manager's native
 *   outdated command to determine the latest).
 */
export async function checkForSdkUpdates(cwd) {
    const offline = checkForSdkUpdatesOffline(cwd);
    if (offline.length === 0) {
        return [];
    }
    return Promise.all(offline.map(async (status) => {
        const installed = {
            language: status.language,
            packageName: status.packageName,
            declaredVersion: status.declaredVersion,
            resolvedVersion: status.resolvedVersion,
            manifestPath: status.manifestPath,
            lockfilePath: status.lockfilePath,
            workspacePath: status.workspacePath,
            renameFrom: status.renameFrom,
            deprecated: status.deprecated || undefined,
        };
        const remote = await getLatestSdkVersion(status.language);
        if (remote === null) {
            return toStatus(installed, status.latest, "baked", true);
        }
        return toStatus(installed, remote, "remote", false);
    }));
}
export function formatSdkUpdateLine(status) {
    const current = status.current ?? "unknown";
    const prefix = status.workspacePath && status.workspacePath !== "."
        ? `${status.workspacePath} `
        : "";
    const rename = status.renameFrom
        ? ` (rename: ${status.renameFrom} → ${status.packageName})`
        : "";
    return `${prefix}${status.language} (${status.packageName}) ${current} → ${status.latest}${rename}`;
}
