// bitfab-plugin-lib/dist/studioChannel.js
import crypto from "node:crypto";
import { readActiveStudioSession, writeActiveStudioSession, } from "./activeStudioSession.js";
import { closeStudio as closeStudioHttp, fetchStreamTip, navigateStudioAndAwaitAck, pollAgentSessionEvents, } from "./agentSessionChannel.js";
import { resolveAgentSessionKey } from "./agentSessionKey.js";
import { openChromelessWindow } from "./browser.js";
import { createStudioSession } from "./commands/createStudioSession.js";
import { DaemonClient } from "./daemon/index.js";
import { DaemonUnavailableError, StudioNavigationError } from "./errors.js";
import { isValidStudioRoute } from "./studioRoutes.js";
export class DirectChannel {
    apiKey;
    serviceUrl;
    constructor(apiKey, serviceUrl) {
        this.apiKey = apiKey;
        this.serviceUrl = serviceUrl;
    }
    setApiKey(apiKey) {
        this.apiKey = apiKey;
    }
    async openOrNavigate(path, opts) {
        // Keyless sign-in flow (login): there is no API key yet, so the
        // authenticated createStudioSession path would throw NotAuthenticated.
        // Mirror the daemon's keyless open (server.ts handleOpen) by launching
        // the browser straight at the session-scoped URL; the event stream is
        // polled anonymously via the session-scope header until the token lands.
        if (!this.apiKey) {
            const sessionId = opts?.sessionId ?? crypto.randomUUID();
            const initialPath = path.startsWith("/studio") ? path : "/studio";
            if (!isValidStudioRoute(initialPath)) {
                throw new Error(`Studio route not allowed: ${initialPath}`);
            }
            const url = initialPath.includes("session=")
                ? `${this.serviceUrl}${initialPath}`
                : `${this.serviceUrl}${initialPath}${initialPath.includes("?") ? "&" : "?"}session=${encodeURIComponent(sessionId)}`;
            const windowPid = openChromelessWindow(url);
            writeActiveStudioSession({
                sessionId,
                serviceUrl: this.serviceUrl,
                windowPid,
                currentPath: initialPath,
            });
            return { sessionId, opened: true };
        }
        const existing = readActiveStudioSession();
        if (existing && existing.windowState !== "closed") {
            const ack = await navigateStudioAndAwaitAck({
                serviceUrl: existing.serviceUrl,
                apiKey: this.apiKey,
                sessionId: existing.sessionId,
            }, path);
            if (!ack.acked) {
                throw new StudioNavigationError(ack.reason ?? "unknown", ack.blockedReason, existing.sessionId);
            }
            return { sessionId: existing.sessionId, opened: false };
        }
        const sessionId = opts?.sessionId ?? crypto.randomUUID();
        const session = await createStudioSession({
            serviceUrl: this.serviceUrl,
            apiKey: this.apiKey,
            sessionId,
            initialPath: path,
            clientHeaders: opts?.clientHeaders,
        });
        return { sessionId: session.sessionId, opened: true };
    }
    async close(sessionId, message) {
        await closeStudioHttp({ serviceUrl: this.serviceUrl, apiKey: this.apiKey, sessionId }, message);
    }
    subscribe(sessionId, onEvent, onError) {
        const abortController = new AbortController();
        const done = (async () => {
            const startCursor = await fetchStreamTip({
                serviceUrl: this.serviceUrl,
                apiKey: this.apiKey,
                sessionId,
            });
            if (abortController.signal.aborted) {
                return;
            }
            await pollAgentSessionEvents({
                serviceUrl: this.serviceUrl,
                apiKey: this.apiKey,
                sessionId,
                startCursor,
                abortSignal: abortController.signal,
                onEvent,
                onError,
            });
        })().catch((err) => {
            if (!abortController.signal.aborted) {
                onError(err instanceof Error ? err : new Error(String(err)));
            }
        });
        return { abort: () => abortController.abort(), done };
    }
    destroy() { }
}
class DaemonChannel {
    client;
    key;
    apiKey;
    serviceUrl;
    doneResolve = null;
    constructor(client, key, apiKey, serviceUrl) {
        this.client = client;
        this.key = key;
        this.apiKey = apiKey;
        this.serviceUrl = serviceUrl;
    }
    static async connect(apiKey, serviceUrl) {
        const client = new DaemonClient();
        await client.connect();
        const key = resolveAgentSessionKey();
        return new DaemonChannel(client, key, apiKey, serviceUrl);
    }
    setApiKey(apiKey) {
        this.apiKey = apiKey;
    }
    async openOrNavigate(path, opts) {
        const openRes = await this.client.open({
            key: this.key,
            serviceUrl: this.serviceUrl,
            apiKey: this.apiKey,
            sessionId: opts?.sessionId,
            initialPath: path,
            clientHeaders: opts?.clientHeaders,
        });
        if (openRes.reconnected) {
            const navRes = await this.client.navigate(this.key, path);
            if (!navRes.acked) {
                this.client.destroy();
                throw new StudioNavigationError(navRes.reason ?? "unknown", undefined, openRes.sessionId);
            }
            writeActiveStudioSession({
                sessionId: openRes.sessionId,
                serviceUrl: this.serviceUrl,
                source: "daemon",
                currentPath: path,
            });
            return { sessionId: openRes.sessionId, opened: false };
        }
        writeActiveStudioSession({
            sessionId: openRes.sessionId,
            serviceUrl: this.serviceUrl,
            source: "daemon",
            currentPath: path,
        });
        return { sessionId: openRes.sessionId, opened: true };
    }
    async close(sessionId, message) {
        await Promise.all([
            closeStudioHttp({ serviceUrl: this.serviceUrl, apiKey: this.apiKey, sessionId }, message),
            this.client.close(this.key).catch(() => { }),
        ]);
    }
    subscribe(_sessionId, onEvent, _onError) {
        const done = new Promise((resolve) => {
            this.doneResolve = resolve;
        });
        this.client.onEvent((push) => {
            const event = {
                id: "",
                type: push.event,
                data: push.data,
            };
            onEvent(event);
        });
        return {
            abort: () => this.destroy(),
            done,
        };
    }
    destroy() {
        this.client.detach(this.key).catch(() => { });
        this.client.destroy();
        if (this.doneResolve) {
            this.doneResolve();
            this.doneResolve = null;
        }
    }
}
function isDaemonEnabled() {
    return process.env.BITFAB_STUDIO_DAEMON !== "0";
}
function withDirectFallback(primary, apiKey, serviceUrl) {
    let active = primary;
    return {
        async openOrNavigate(path, opts) {
            try {
                return await active.openOrNavigate(path, opts);
            }
            catch (err) {
                if (err instanceof StudioNavigationError || active !== primary) {
                    throw err;
                }
                console.error(`daemon failed, falling back to direct: ${err instanceof Error ? err.message : String(err)}`);
                active.destroy();
                active = new DirectChannel(apiKey, serviceUrl);
                return active.openOrNavigate(path, opts);
            }
        },
        async close(sessionId, message) {
            await active.close(sessionId, message);
        },
        subscribe(sessionId, onEvent, onError) {
            return active.subscribe(sessionId, onEvent, onError);
        },
        setApiKey(key) {
            active.setApiKey(key);
        },
        destroy() {
            active.destroy();
        },
    };
}
export async function resolveChannel(apiKey, serviceUrl) {
    if (isDaemonEnabled()) {
        try {
            const daemon = await DaemonChannel.connect(apiKey, serviceUrl);
            return withDirectFallback(daemon, apiKey, serviceUrl);
        }
        catch (err) {
            if (err instanceof StudioNavigationError) {
                throw err;
            }
            // A daemon that simply isn't packaged in this build (e.g. the bundled
            // standalone CLI) is an expected, quiet fallback. Only surface the
            // noisier failure modes (spawn timeout, socket errors, etc.).
            if (!(err instanceof DaemonUnavailableError)) {
                console.error(`daemon unavailable, falling back to direct: ${err instanceof Error ? err.message : String(err)}`);
            }
        }
    }
    return new DirectChannel(apiKey, serviceUrl);
}
