// bitfab-plugin-lib/dist/studioTeardown.js
import { readActiveStudioSession } from "./activeStudioSession.js";
// A `window-closed` also fires on a page refresh (followed by a fresh
// `session-started` from the reload), so it only counts as proof of teardown
// after this grace period passes without a reload re-binding.
export const WINDOW_CLOSE_GRACE_PERIOD_MS = 10_000;
/**
 * Shared disambiguation for `studio:window-closed`: it fires both on real
 * teardown and as the first half of a page refresh (which re-binds with a
 * `session-started` moments later). The arbiter holds each candidate for the
 * grace period, refutes it if a `session-started` arrives, and confirms it
 * (with the original event) if the grace expires untouched. A newer
 * window-closed restarts the grace. Both consumers of the close signal feed
 * their event streams through this: the `openStudioTo` event poller (clears
 * the pointer + synthesizes the terminal `session-ended`) and `closeStudio`'s
 * confirmation wait (resolves the close as proven).
 */
export function createWindowCloseArbiter(handlers) {
    let graceTimer = null;
    const cancel = () => {
        if (graceTimer) {
            clearTimeout(graceTimer);
            graceTimer = null;
        }
    };
    return {
        handleEvent(event) {
            if (event.type === "studio:window-closed") {
                cancel();
                graceTimer = setTimeout(() => {
                    graceTimer = null;
                    handlers.onConfirmed(event);
                }, WINDOW_CLOSE_GRACE_PERIOD_MS);
                return;
            }
            if (event.type === "studio:session-started" && graceTimer) {
                cancel();
                handlers.onRefuted?.();
            }
        },
        hasPendingCandidate: () => graceTimer != null,
        cancel,
    };
}
/**
 * Best-effort process reaping shared by the two ways a Studio session ends:
 * the cooperative close (`closeStudio`, window reachable, clears only on a
 * confirmed `window-closed`) and the unilateral clear (`clearStudioSession`,
 * window unreachable, user approved abandoning it). Both want the same
 * cleanup: no zombie window and no orphaned event poller outliving the flow.
 *
 * Every kill re-reads the pointer at call time and matches on `sessionId`, so
 * a record re-bound by a newer session between resolution and the kill is
 * never touched. Correctness never depends on a kill succeeding.
 */
/**
 * Best-effort kill of the spawned `--app` window process recorded in the
 * pointer. Genuinely best-effort: when the browser was already running at
 * spawn time, the recorded pid is a dead launcher that handed off to the
 * existing instance, so the kill is a no-op (the common case). The pointer
 * clears only on a confirmed `studio:window-closed` (close path) or an
 * explicit user-approved clear (recovery path), never because of this kill.
 */
export function killStudioWindow(sessionId) {
    const record = readActiveStudioSession();
    if (record?.sessionId !== sessionId || record.windowPid == null) {
        return;
    }
    try {
        process.kill(record.windowPid, "SIGTERM");
    }
    catch {
        // Already gone (or a dead launcher pid); nothing to kill.
    }
}
/**
 * Best-effort kill of the process that wrote the pointer; in flow runs that
 * is the long-running `openStudioTo.js` event poller. The poller self-exits
 * ~10s after a CONFIRMED close (it sees the same `window-closed`), so this
 * only matters when the window never confirms (refused close, or an
 * unreachable session being cleared) and the poller would otherwise outlive
 * the flow as an orphan. It catches SIGTERM and exits cleanly. Skipped when
 * the record names a different session or ourselves; a reused/dead pid makes
 * the kill a no-op.
 */
export function killPollerProcess(sessionId) {
    const record = readActiveStudioSession();
    if (record?.sessionId !== sessionId || record.pid === process.pid) {
        return;
    }
    try {
        process.kill(record.pid, "SIGTERM");
    }
    catch {
        // Already gone; nothing to kill.
    }
}
