// bitfab-plugin-lib/dist/studioEventLog.js
import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
// The daemon is the durable buffer for agent-session events. Every event it
// polls is appended here as one JSON line, independent of whether any client is
// connected, so the agent's Monitor can attach to this file at any time and
// replay from a cursor without ever missing an event.
const SESSIONS_DIR = path.join(os.homedir(), ".config", "bitfab", "sessions");
function shortHash(input) {
    return crypto.createHash("sha256").update(input).digest("hex").slice(0, 16);
}
export function eventLogPathForKey(key) {
    return path.join(SESSIONS_DIR, `${shortHash(key)}.events.jsonl`);
}
// Map a raw agent event to the line the file carries: the `studio:` prefix
// stripped and ALL of the event's data fields (so every consumer, e.g.
// openTracePlan's planId or the template preview's element-clicked fields, can
// read what it needs from the one bus). The auth token is the sole exclusion so
// a secret never lands in this plaintext file (the daemon / in-process writer
// saves credentials directly when it sees `studio:authenticated`).
export function studioEventToLogLine(event, sessionId) {
    const data = (event.data ?? {});
    const line = {
        event: event.type.replace("studio:", ""),
        sessionId,
        ...data,
    };
    delete line.token;
    return line;
}
export function appendStudioEvent(key, line) {
    try {
        const target = eventLogPathForKey(key);
        fs.mkdirSync(path.dirname(target), { recursive: true });
        fs.appendFileSync(target, `${JSON.stringify(line)}\n`);
    }
    catch {
        // best-effort; a failed append must never take down the daemon poll loop
    }
}
// Reset the log for a fresh session. Called when a new (non-reconnect) session
// is created and on close/clear, so a resumed run never sees a prior session's
// tail.
export function truncateStudioEventLog(key) {
    try {
        fs.rmSync(eventLogPathForKey(key), { force: true });
    }
    catch { }
}
// Read lines from a 1-based offset (the cursor the flow tracks as <NEXT_LINE>).
// Returns the parsed lines and the next cursor to resume from.
export function readStudioEventLogTail(key, fromLine = 1) {
    try {
        const raw = fs.readFileSync(eventLogPathForKey(key), "utf-8");
        const all = raw.split("\n").filter((l) => l.length > 0);
        const start = Math.max(0, fromLine - 1);
        const slice = all.slice(start);
        const lines = [];
        for (const l of slice) {
            try {
                lines.push(JSON.parse(l));
            }
            catch {
                // skip a torn line (concurrent append); the next read will pick it up
            }
        }
        return { lines, nextLine: all.length + 1 };
    }
    catch {
        return { lines: [], nextLine: fromLine };
    }
}
// Block reading the event file, delivering each new line to `onLine`, until the
// abort signal fires. This is how every consumer (the agent's tail equivalent,
// blocking commands) reads the single bus: it polls the file from a cursor
// rather than holding its own poll against the backend. `onLine` may trigger the
// abort itself (e.g. on a terminal event); no line is delivered after abort.
export function tailStudioEvents(key, onLine, opts) {
    const pollMs = opts.pollMs ?? 200;
    let cursor = opts.fromLine ?? 1;
    return new Promise((resolve) => {
        if (opts.abortSignal.aborted) {
            resolve();
            return;
        }
        const tick = () => {
            const { lines, nextLine } = readStudioEventLogTail(key, cursor);
            cursor = nextLine;
            for (const line of lines) {
                if (opts.abortSignal.aborted) {
                    return;
                }
                onLine(line);
            }
        };
        const timer = setInterval(() => {
            if (!opts.abortSignal.aborted) {
                tick();
            }
        }, pollMs);
        opts.abortSignal.addEventListener("abort", () => {
            clearInterval(timer);
            resolve();
        }, { once: true });
        tick();
    });
}
