// bitfab-plugin-lib/dist/studioChannel.js
import crypto from "node:crypto";
import { readActiveStudioSession, writeActiveStudioSession, } from "./activeStudioSession.js";
import { closeStudio as closeStudioHttp, fetchStreamTip, navigateStudioAndAwaitAck, pollAgentSessionEvents, } from "./agentSessionChannel.js";
import { resolveAgentSessionKey } from "./agentSessionKey.js";
import { closeStudioWindowsBySession, openChromelessWindow } from "./browser.js";
import { createStudioSession } from "./commands/createStudioSession.js";
import { DaemonClient } from "./daemon/index.js";
import { DaemonUnavailableError, StudioNavigationError } from "./errors.js";
import { buildStudioSessionUrl, resolveStudioInitialPath } from "./studioUrl.js";
/**
 * Per-invocation, in-process Studio channel. URL/route building and event
 * shape are shared with the daemon (see `studioUrl.ts` and `DaemonPush.id`) so
 * the two never drift. The asymmetry is deliberate, NOT missing parity: this
 * channel intentionally has no dead-browser relaunch and no live
 * `browserConnected` / `windowState` tracking. Those are persistence concerns
 * that only matter to the long-lived daemon serving many commands; a one-shot
 * direct open (a CLI login or a daemon-less fallback) has no window to keep
 * warm and nothing to recover, so don't port that machinery here.
 */
export class DirectChannel {
    apiKey;
    serviceUrl;
    constructor(apiKey, serviceUrl) {
        this.apiKey = apiKey;
        this.serviceUrl = serviceUrl;
    }
    setApiKey(apiKey) {
        this.apiKey = apiKey;
    }
    async openOrNavigate(path, opts) {
        // Keyless sign-in flow (login): there is no API key yet, so the
        // authenticated createStudioSession path would throw NotAuthenticated.
        // Mirror the daemon's keyless open (server.ts handleOpen) by launching
        // the browser straight at the session-scoped URL; the event stream is
        // polled anonymously via the session-scope header until the token lands.
        if (!this.apiKey) {
            const sessionId = opts?.sessionId ?? crypto.randomUUID();
            const initialPath = resolveStudioInitialPath(path);
            const url = buildStudioSessionUrl({
                serviceUrl: this.serviceUrl,
                path: initialPath,
                sessionId,
            });
            const windowPid = openChromelessWindow(url);
            writeActiveStudioSession({
                sessionId,
                serviceUrl: this.serviceUrl,
                windowPid,
                currentPath: initialPath,
            });
            return { sessionId, opened: true };
        }
        const existing = readActiveStudioSession();
        if (existing && existing.windowState !== "closed") {
            const ack = await navigateStudioAndAwaitAck({
                serviceUrl: existing.serviceUrl,
                apiKey: this.apiKey,
                sessionId: existing.sessionId,
            }, path);
            if (!ack.acked) {
                throw new StudioNavigationError(ack.reason ?? "unknown", ack.blockedReason, existing.sessionId);
            }
            return { sessionId: existing.sessionId, opened: false };
        }
        const sessionId = opts?.sessionId ?? crypto.randomUUID();
        const session = await createStudioSession({
            serviceUrl: this.serviceUrl,
            apiKey: this.apiKey,
            sessionId,
            initialPath: path,
            clientHeaders: opts?.clientHeaders,
        });
        return { sessionId: session.sessionId, opened: true };
    }
    async close(sessionId, message) {
        await closeStudioHttp({ serviceUrl: this.serviceUrl, apiKey: this.apiKey, sessionId }, message);
    }
    forceCloseWindow(sessionId) {
        // No daemon to do it server-side: run the AppleScript close ourselves. This
        // is what actually closes a window parked inside an already-running Chrome
        // (the inline pid kill in closeStudio handles the rare standalone case).
        // Best-effort; never throws.
        closeStudioWindowsBySession(sessionId);
    }
    async clearSession(_sessionId) {
        // No-op: a per-invocation channel holds no session state beyond the on-disk
        // pointer, which the caller clears, and the window/poller reap also lives in
        // the caller (runClearStudioSession), mode-independent. There is no
        // out-of-process registry to forget here.
    }
    subscribe(sessionId, onEvent, onError) {
        const abortController = new AbortController();
        const done = (async () => {
            const startCursor = await fetchStreamTip({
                serviceUrl: this.serviceUrl,
                apiKey: this.apiKey,
                sessionId,
            });
            if (abortController.signal.aborted) {
                return;
            }
            await pollAgentSessionEvents({
                serviceUrl: this.serviceUrl,
                apiKey: this.apiKey,
                sessionId,
                startCursor,
                abortSignal: abortController.signal,
                onEvent,
                onError,
            });
        })().catch((err) => {
            if (!abortController.signal.aborted) {
                onError(err instanceof Error ? err : new Error(String(err)));
            }
        });
        return { abort: () => abortController.abort(), done };
    }
    destroy() { }
}
class DaemonChannel {
    client;
    key;
    apiKey;
    serviceUrl;
    doneResolve = null;
    constructor(client, key, apiKey, serviceUrl) {
        this.client = client;
        this.key = key;
        this.apiKey = apiKey;
        this.serviceUrl = serviceUrl;
    }
    static async connect(apiKey, serviceUrl) {
        const client = new DaemonClient();
        await client.connect();
        const key = resolveAgentSessionKey();
        return new DaemonChannel(client, key, apiKey, serviceUrl);
    }
    setApiKey(apiKey) {
        this.apiKey = apiKey;
    }
    async openOrNavigate(path, opts) {
        const openRes = await this.client.open({
            key: this.key,
            serviceUrl: this.serviceUrl,
            apiKey: this.apiKey,
            sessionId: opts?.sessionId,
            initialPath: path,
            clientHeaders: opts?.clientHeaders,
        });
        if (openRes.reconnected) {
            const navRes = await this.client.navigate(this.key, path);
            if (!navRes.acked) {
                this.client.destroy();
                throw new StudioNavigationError(navRes.reason ?? "unknown", undefined, openRes.sessionId);
            }
            writeActiveStudioSession({
                sessionId: openRes.sessionId,
                serviceUrl: this.serviceUrl,
                source: "daemon",
                currentPath: path,
            });
            return { sessionId: openRes.sessionId, opened: false };
        }
        writeActiveStudioSession({
            sessionId: openRes.sessionId,
            serviceUrl: this.serviceUrl,
            source: "daemon",
            currentPath: path,
        });
        return { sessionId: openRes.sessionId, opened: true };
    }
    async close(sessionId, message) {
        await Promise.all([
            closeStudioHttp({ serviceUrl: this.serviceUrl, apiKey: this.apiKey, sessionId }, message),
            this.client.close(this.key).catch(() => { }),
        ]);
    }
    forceCloseWindow(_sessionId) {
        // No-op: the daemon already arms its own force-close fallback when it
        // receives the close command (see daemon/server.ts handleClose). Doing it
        // here too would just duplicate the AppleScript pass.
    }
    async clearSession(sessionId) {
        // Drop the daemon's in-memory SessionEntry for this session so the next open
        // starts fresh instead of reconnecting to the dead window. State only: the
        // window reap is the caller's (runClearStudioSession's) job, kept off the
        // daemon's synchronous handler. Best-effort: a daemon that no longer knows
        // this session (already gone, or re-bound under this key) reports cleared:0.
        await this.client.clear(sessionId).catch(() => { });
    }
    subscribe(_sessionId, onEvent, onError) {
        const done = new Promise((resolve) => {
            this.doneResolve = resolve;
        });
        this.client.onEvent((push) => {
            const event = {
                id: push.id,
                type: push.event,
                data: push.data,
            };
            onEvent(event);
        });
        // Mirror DirectChannel: surface a lost event stream to the caller. For the
        // daemon path that means losing the socket to the daemon (poll errors are
        // logged daemon-side); without this the caller would silently hang.
        this.client.onDisconnect((err) => {
            onError(err);
        });
        return {
            abort: () => this.destroy(),
            done,
        };
    }
    destroy() {
        this.client.detach(this.key).catch(() => { });
        this.client.destroy();
        if (this.doneResolve) {
            this.doneResolve();
            this.doneResolve = null;
        }
    }
}
function isDaemonEnabled() {
    return process.env.BITFAB_STUDIO_DAEMON !== "0";
}
function withDirectFallback(primary, apiKey, serviceUrl) {
    let active = primary;
    return {
        async openOrNavigate(path, opts) {
            try {
                return await active.openOrNavigate(path, opts);
            }
            catch (err) {
                if (err instanceof StudioNavigationError || active !== primary) {
                    throw err;
                }
                console.error(`daemon failed, falling back to direct: ${err instanceof Error ? err.message : String(err)}`);
                active.destroy();
                active = new DirectChannel(apiKey, serviceUrl);
                return active.openOrNavigate(path, opts);
            }
        },
        async close(sessionId, message) {
            await active.close(sessionId, message);
        },
        forceCloseWindow(sessionId) {
            active.forceCloseWindow(sessionId);
        },
        async clearSession(sessionId) {
            await active.clearSession(sessionId);
        },
        subscribe(sessionId, onEvent, onError) {
            return active.subscribe(sessionId, onEvent, onError);
        },
        setApiKey(key) {
            active.setApiKey(key);
        },
        destroy() {
            active.destroy();
        },
    };
}
export async function resolveChannel(apiKey, serviceUrl) {
    if (isDaemonEnabled()) {
        try {
            const daemon = await DaemonChannel.connect(apiKey, serviceUrl);
            return withDirectFallback(daemon, apiKey, serviceUrl);
        }
        catch (err) {
            if (err instanceof StudioNavigationError) {
                throw err;
            }
            // A daemon that simply isn't packaged in this build (e.g. the bundled
            // standalone CLI) is an expected, quiet fallback. Only surface the
            // noisier failure modes (spawn timeout, socket errors, etc.).
            if (!(err instanceof DaemonUnavailableError)) {
                console.error(`daemon unavailable, falling back to direct: ${err instanceof Error ? err.message : String(err)}`);
            }
        }
    }
    return new DirectChannel(apiKey, serviceUrl);
}
