// bitfab-plugin-lib/dist/updates.js
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
async function getLatestVersion(platform) {
    try {
        const response = await fetch(`https://raw.githubusercontent.com/${platform.repo}/main/${platform.remotePackageJsonPath}`, { signal: AbortSignal.timeout(3000) });
        if (!response.ok) {
            return null;
        }
        const pkg = (await response.json());
        return pkg.version;
    }
    catch {
        return null;
    }
}
/**
 * Fetch the gitSha of the latest published build from the public repo's
 * `dist/buildInfo.json`. The path is derived from `remotePackageJsonPath`
 * (flat for Claude/Cursor → "dist/buildInfo.json"; subdir for Codex →
 * "plugins/bitfab/dist/buildInfo.json"). Returns null if the file is absent
 * (older releases predate the build stamp) or unreachable.
 */
async function getLatestBuildSha(platform) {
    const buildInfoPath = platform.remotePackageJsonPath.replace(/package\.json$/, "dist/buildInfo.json");
    try {
        const response = await fetch(`https://raw.githubusercontent.com/${platform.repo}/main/${buildInfoPath}`, { signal: AbortSignal.timeout(3000) });
        if (!response.ok) {
            return null;
        }
        const info = (await response.json());
        return typeof info.gitSha === "string" ? info.gitSha : null;
    }
    catch {
        return null;
    }
}
function isNewer(latest, current) {
    const [lMajor, lMinor, lPatch] = latest.split(".").map(Number);
    const [cMajor, cMinor, cPatch] = current.split(".").map(Number);
    if (lMajor !== cMajor) {
        return lMajor > cMajor;
    }
    if (lMinor !== cMinor) {
        return lMinor > cMinor;
    }
    return lPatch > cPatch;
}
function isAutoUpdateEnabled(platform) {
    try {
        const marketplacesPath = path.join(os.homedir(), ".claude", "plugins", "known_marketplaces.json");
        const content = fs.readFileSync(marketplacesPath, "utf-8");
        const data = JSON.parse(content);
        const byName = data[platform.marketplaceName];
        if (byName) {
            return byName.autoUpdate === true;
        }
        for (const marketplace of Object.values(data)) {
            if (marketplace.source?.repo === platform.repo) {
                return marketplace.autoUpdate === true;
            }
        }
    }
    catch {
        // File doesn't exist or can't be parsed
    }
    return false;
}
export async function checkForUpdate(currentVersion, platform, localBuildSha) {
    const latest = await getLatestVersion(platform);
    const versionNewer = latest !== null && isNewer(latest, currentVersion);
    // Same-version staleness: only worth a network round-trip when the version
    // check came back equal (not newer, not older) and we actually know the
    // local build's commit to compare against.
    let staleSameVersion = false;
    if (!versionNewer && latest === currentVersion && localBuildSha) {
        const latestBuildSha = await getLatestBuildSha(platform);
        staleSameVersion =
            latestBuildSha !== null && latestBuildSha !== localBuildSha;
    }
    return {
        current: currentVersion,
        latest,
        updateAvailable: versionNewer || staleSameVersion,
        autoUpdateEnabled: platform.supportsAutoUpdate
            ? isAutoUpdateEnabled(platform)
            : false,
        staleSameVersion,
    };
}
export function formatUpdateMessage(latest, platform) {
    return `v${latest} available — run ${platform.updateHint} to update`;
}
/**
 * Message for the stale-same-version case: the version string is unchanged but
 * the published build moved on. There's no higher version to name, so point the
 * user at the update/reinstall path to refresh the cached build.
 */
export function formatStaleMessage(platform) {
    return `cached build is behind the published release — run ${platform.updateHint} to refresh`;
}
/**
 * Detect whether the plugin is still installed from a legacy custom marketplace
 * (e.g. "bitfab@bitfab") rather than the official marketplace.
 */
export function detectLegacyInstall(_platform) {
    return false;
}
/**
 * Build the migration message shown when a legacy custom-marketplace install is detected.
 */
export function legacyMigrationMessage(platform, prefix = "[Bitfab]") {
    const legacyMarketplace = platform.legacyMarketplaceKey.split("@")[1];
    return [
        `${prefix} You're running from the old "${legacyMarketplace}" marketplace. Migrate to the official marketplace:`,
        ``,
        `  /plugin marketplace remove ${legacyMarketplace}`,
        `  /plugin install ${platform.pluginName}@${platform.marketplaceName}`,
        ``,
        `  Then restart ${platform.displayName}.`,
    ].join("\n");
}
