// bitfab-plugin-lib/dist/withStudioSession.js
import crypto from "node:crypto";
import { readActiveStudioSession } from "./activeStudioSession.js";
import { navigateStudioAndAwaitAck } from "./agentSessionChannel.js";
import { openStudio } from "./commands/openStudio.js";
import { NotAuthenticatedError } from "./errors.js";
export class StudioNavigationError extends Error {
    reason;
    blockedReason;
    constructor(reason, blockedReason) {
        super(blockedReason
            ? `Studio navigation failed (${reason}): ${blockedReason}`
            : `Studio navigation failed: ${reason}`);
        this.reason = reason;
        this.blockedReason = blockedReason;
        this.name = "StudioNavigationError";
    }
}
export async function withStudioSession(opts) {
    if (!opts.apiKey) {
        throw new NotAuthenticatedError();
    }
    const existing = readActiveStudioSession();
    if (existing) {
        const client = {
            serviceUrl: existing.serviceUrl,
            apiKey: opts.apiKey,
            sessionId: existing.sessionId,
        };
        const ack = await navigateStudioAndAwaitAck(client, opts.targetPath);
        if (ack.acked) {
            return { sessionId: existing.sessionId, serviceUrl: existing.serviceUrl };
        }
        throw new StudioNavigationError(ack.reason ?? "unknown", ack.blockedReason);
    }
    const sessionId = crypto.randomUUID();
    const session = await openStudio({
        serviceUrl: opts.serviceUrl,
        apiKey: opts.apiKey,
        sessionId,
        initialPath: opts.targetPath,
        clientHeaders: opts.clientHeaders,
    });
    return { sessionId: session.sessionId, serviceUrl: opts.serviceUrl };
}
